// rclnodejs/generated/nav2_msgs/nav2_msgs__action__ComputeRoute_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const ComputeRoute_GoalRefStruct = StructType({
  start_id: primitiveTypes.uint16,
  start: PoseStampedWrapper.refObjectType,
  goal_id: primitiveTypes.uint16,
  goal: PoseStampedWrapper.refObjectType,
  use_start: primitiveTypes.bool,
  use_poses: primitiveTypes.bool,
});
const ComputeRoute_GoalRefArray = ArrayType(ComputeRoute_GoalRefStruct);
const ComputeRoute_GoalRefStructArray = StructType({
  data: ComputeRoute_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ComputeRoute_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ComputeRoute_GoalRefStruct(other._refObject.toObject());
      this._start_idIntialized = true;
      this._wrapperFields.start =  new PoseStampedWrapper(other._wrapperFields.start);
      this._goal_idIntialized = true;
      this._wrapperFields.goal =  new PoseStampedWrapper(other._wrapperFields.goal);
      this._use_startIntialized = true;
      this._use_posesIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ComputeRoute_GoalRefStruct();
    this._start_idIntialized = false;
    this._wrapperFields.start = new PoseStampedWrapper();
    this._goal_idIntialized = false;
    this._wrapperFields.goal = new PoseStampedWrapper();
    this._use_startIntialized = false;
    this._use_posesIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ComputeRoute_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ComputeRoute_GoalArrayWrapper;
  }
  static get ArrayType() {
    return ComputeRoute_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return ComputeRoute_GoalRefStructArray
  }
  static get refObjectType() {
    return ComputeRoute_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._start_idIntialized) {
        throw new TypeError('Invalid argument: start_id in ComputeRoute_Goal');
      }
      if (!this._goal_idIntialized) {
        throw new TypeError('Invalid argument: goal_id in ComputeRoute_Goal');
      }
      if (!this._use_startIntialized) {
        throw new TypeError('Invalid argument: use_start in ComputeRoute_Goal');
      }
      if (!this._use_posesIntialized) {
        throw new TypeError('Invalid argument: use_poses in ComputeRoute_Goal');
      }
    }
    this._wrapperFields.start.freeze(own, checkConsistency);
    this._refObject.start = this._wrapperFields.start.refObject;
    this._wrapperFields.goal.freeze(own, checkConsistency);
    this._refObject.goal = this._wrapperFields.goal.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._start_idIntialized = true;
    this._refObject.start_id = refObject.start_id;
    this._wrapperFields.start.copyRefObject(refObject.start);
    this._goal_idIntialized = true;
    this._refObject.goal_id = refObject.goal_id;
    this._wrapperFields.goal.copyRefObject(refObject.goal);
    this._use_startIntialized = true;
    this._refObject.use_start = refObject.use_start;
    this._use_posesIntialized = true;
    this._refObject.use_poses = refObject.use_poses;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseStampedWrapper.freeStruct(refObject.start);
    PoseStampedWrapper.freeStruct(refObject.goal);
  }
  static destoryRawROS(msg) {
    ComputeRoute_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'ComputeRoute_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get start_id() {
    if (!this._start_idIntialized) {
      return undefined;
    }
    return this._refObject.start_id;
  }
  set start_id(value) {
    this._start_idIntialized = true;
    this._refObject.start_id = value;
  }
  get start() {
    return this._wrapperFields.start;
  }
  set start(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.start.copy(value);
    } else {
      this._wrapperFields.start.copy(new PoseStampedWrapper(value));
    }
  }
  get goal_id() {
    if (!this._goal_idIntialized) {
      return undefined;
    }
    return this._refObject.goal_id;
  }
  set goal_id(value) {
    this._goal_idIntialized = true;
    this._refObject.goal_id = value;
  }
  get goal() {
    return this._wrapperFields.goal;
  }
  set goal(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.goal.copy(value);
    } else {
      this._wrapperFields.goal.copy(new PoseStampedWrapper(value));
    }
  }
  get use_start() {
    if (!this._use_startIntialized) {
      return undefined;
    }
    return this._refObject.use_start;
  }
  set use_start(value) {
    this._use_startIntialized = true;
    this._refObject.use_start = value;
  }
  get use_poses() {
    if (!this._use_posesIntialized) {
      return undefined;
    }
    return this._refObject.use_poses;
  }
  set use_poses(value) {
    this._use_posesIntialized = true;
    this._refObject.use_poses = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ComputeRoute_GoalRefStruct(refObject.toObject());
    this._start_idIntialized = true;
    this._wrapperFields.start.copyRefObject(this._refObject.start);
    this._goal_idIntialized = true;
    this._wrapperFields.goal.copyRefObject(this._refObject.goal);
    this._use_startIntialized = true;
    this._use_posesIntialized = true;
  }
  copy(other) {
    this._refObject = new ComputeRoute_GoalRefStruct(other._refObject.toObject());
    this._start_idIntialized = true;
    this._wrapperFields.start.copy(other._wrapperFields.start);
    this._goal_idIntialized = true;
    this._wrapperFields.goal.copy(other._wrapperFields.goal);
    this._use_startIntialized = true;
    this._use_posesIntialized = true;
  }
  static get classType() {
    return ComputeRoute_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"start_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goal_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"goal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"use_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"use_poses","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"ComputeRoute_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ComputeRoute_Goal"};
  }
  hasMember(name) {
    let memberNames = ["start_id","start","goal_id","goal","use_start","use_poses"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ComputeRoute_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ComputeRoute_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ComputeRoute_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputeRoute_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputeRoute_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ComputeRoute_GoalRefArray(size);
    this._refObject = new ComputeRoute_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ComputeRoute_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ComputeRoute_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ComputeRoute_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ComputeRoute_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ComputeRoute_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ComputeRoute_GoalArrayWrapper;
  }
}
module.exports = ComputeRoute_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "start_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "goal_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "goal",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "use_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "use_poses",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "ComputeRoute_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ComputeRoute_Goal"
}
*/
