// rclnodejs/generated/nav2_msgs/nav2_msgs__srv__SetRouteGraph_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SetRouteGraph_RequestRefStruct = StructType({
  graph_filepath: primitiveTypes.string,
});
const SetRouteGraph_RequestRefArray = ArrayType(SetRouteGraph_RequestRefStruct);
const SetRouteGraph_RequestRefStructArray = StructType({
  data: SetRouteGraph_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetRouteGraph_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetRouteGraph_RequestRefStruct(other._refObject.toObject());
      this._graph_filepathIntialized = true;
      this._wrapperFields.graph_filepath =  new StringWrapper(other._wrapperFields.graph_filepath);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetRouteGraph_RequestRefStruct();
    this._graph_filepathIntialized = false;
    this._wrapperFields.graph_filepath = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetRouteGraph_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetRouteGraph_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetRouteGraph_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetRouteGraph_RequestRefStructArray
  }
  static get refObjectType() {
    return SetRouteGraph_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._graph_filepathIntialized) {
        throw new TypeError('Invalid argument: graph_filepath in SetRouteGraph_Request');
      }
    }
    if (own) {
      this._wrapperFields.graph_filepath.freeze(own, checkConsistency);
    }
    this._refObject.graph_filepath = this._wrapperFields.graph_filepath.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._graph_filepathIntialized = true;
    this._wrapperFields.graph_filepath.data = refObject.graph_filepath.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.graph_filepath);
  }
  static destoryRawROS(msg) {
    SetRouteGraph_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'srv', interfaceName: 'SetRouteGraph_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get graph_filepath() {
    if (!this._graph_filepathIntialized) {
      return undefined;
    }
    return this._wrapperFields.graph_filepath.data;
  }
  set graph_filepath(value) {
    this._graph_filepathIntialized = true;
    this._wrapperFields.graph_filepath.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetRouteGraph_RequestRefStruct(refObject.toObject());
    this._graph_filepathIntialized = true;
    this._wrapperFields.graph_filepath.copyRefObject(this._refObject.graph_filepath);
  }
  copy(other) {
    this._refObject = new SetRouteGraph_RequestRefStruct(other._refObject.toObject());
    this._graph_filepathIntialized = true;
    this._wrapperFields.graph_filepath.copy(other._wrapperFields.graph_filepath);
  }
  static get classType() {
    return SetRouteGraph_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"graph_filepath","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"SetRouteGraph_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetRouteGraph_Request"};
  }
  hasMember(name) {
    let memberNames = ["graph_filepath"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetRouteGraph_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetRouteGraph_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetRouteGraph_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetRouteGraph_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetRouteGraph_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetRouteGraph_RequestRefArray(size);
    this._refObject = new SetRouteGraph_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetRouteGraph_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetRouteGraph_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetRouteGraph_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetRouteGraph_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetRouteGraph_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetRouteGraph_RequestArrayWrapper;
  }
}
module.exports = SetRouteGraph_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "graph_filepath",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "SetRouteGraph_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetRouteGraph_Request"
}
*/
