// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__EdgeCost.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const EdgeCostRefStruct = StructType({
  edgeid: primitiveTypes.uint16,
  cost: primitiveTypes.float32,
});
const EdgeCostRefArray = ArrayType(EdgeCostRefStruct);
const EdgeCostRefStructArray = StructType({
  data: EdgeCostRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EdgeCostWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EdgeCostRefStruct(other._refObject.toObject());
      this._edgeidIntialized = true;
      this._costIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EdgeCostRefStruct();
    this._edgeidIntialized = false;
    this._costIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EdgeCostWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EdgeCostArrayWrapper;
  }
  static get ArrayType() {
    return EdgeCostArrayWrapper;
  }
  static get refObjectArrayType() {
    return EdgeCostRefStructArray
  }
  static get refObjectType() {
    return EdgeCostRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._edgeidIntialized) {
        throw new TypeError('Invalid argument: edgeid in EdgeCost');
      }
      if (!this._costIntialized) {
        throw new TypeError('Invalid argument: cost in EdgeCost');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._edgeidIntialized = true;
    this._refObject.edgeid = refObject.edgeid;
    this._costIntialized = true;
    this._refObject.cost = refObject.cost;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    EdgeCostWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'EdgeCost'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get edgeid() {
    if (!this._edgeidIntialized) {
      return undefined;
    }
    return this._refObject.edgeid;
  }
  set edgeid(value) {
    this._edgeidIntialized = true;
    this._refObject.edgeid = value;
  }
  get cost() {
    if (!this._costIntialized) {
      return undefined;
    }
    return this._refObject.cost;
  }
  set cost(value) {
    this._costIntialized = true;
    this._refObject.cost = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EdgeCostRefStruct(refObject.toObject());
    this._edgeidIntialized = true;
    this._costIntialized = true;
  }
  copy(other) {
    this._refObject = new EdgeCostRefStruct(other._refObject.toObject());
    this._edgeidIntialized = true;
    this._costIntialized = true;
  }
  static get classType() {
    return EdgeCostWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"edgeid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cost","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"EdgeCost","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EdgeCost"};
  }
  hasMember(name) {
    let memberNames = ["edgeid","cost"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EdgeCostArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EdgeCostWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EdgeCostWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EdgeCostArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EdgeCostArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EdgeCostRefArray(size);
    this._refObject = new EdgeCostRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EdgeCostWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EdgeCostArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EdgeCostArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EdgeCostWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EdgeCostWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EdgeCostArrayWrapper;
  }
}
module.exports = EdgeCostWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "edgeid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cost",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "EdgeCost",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EdgeCost"
}
*/
