// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__RouteEdge.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PointWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point.js');
const RouteEdgeRefStruct = StructType({
  edgeid: primitiveTypes.uint16,
  start: PointWrapper.refObjectType,
  end: PointWrapper.refObjectType,
});
const RouteEdgeRefArray = ArrayType(RouteEdgeRefStruct);
const RouteEdgeRefStructArray = StructType({
  data: RouteEdgeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RouteEdgeWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RouteEdgeRefStruct(other._refObject.toObject());
      this._edgeidIntialized = true;
      this._wrapperFields.start =  new PointWrapper(other._wrapperFields.start);
      this._wrapperFields.end =  new PointWrapper(other._wrapperFields.end);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RouteEdgeRefStruct();
    this._edgeidIntialized = false;
    this._wrapperFields.start = new PointWrapper();
    this._wrapperFields.end = new PointWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new RouteEdgeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RouteEdgeArrayWrapper;
  }
  static get ArrayType() {
    return RouteEdgeArrayWrapper;
  }
  static get refObjectArrayType() {
    return RouteEdgeRefStructArray
  }
  static get refObjectType() {
    return RouteEdgeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._edgeidIntialized) {
        throw new TypeError('Invalid argument: edgeid in RouteEdge');
      }
    }
    this._wrapperFields.start.freeze(own, checkConsistency);
    this._refObject.start = this._wrapperFields.start.refObject;
    this._wrapperFields.end.freeze(own, checkConsistency);
    this._refObject.end = this._wrapperFields.end.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._edgeidIntialized = true;
    this._refObject.edgeid = refObject.edgeid;
    this._wrapperFields.start.copyRefObject(refObject.start);
    this._wrapperFields.end.copyRefObject(refObject.end);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PointWrapper.freeStruct(refObject.start);
    PointWrapper.freeStruct(refObject.end);
  }
  static destoryRawROS(msg) {
    RouteEdgeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'RouteEdge'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get edgeid() {
    if (!this._edgeidIntialized) {
      return undefined;
    }
    return this._refObject.edgeid;
  }
  set edgeid(value) {
    this._edgeidIntialized = true;
    this._refObject.edgeid = value;
  }
  get start() {
    return this._wrapperFields.start;
  }
  set start(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.start.copy(value);
    } else {
      this._wrapperFields.start.copy(new PointWrapper(value));
    }
  }
  get end() {
    return this._wrapperFields.end;
  }
  set end(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.end.copy(value);
    } else {
      this._wrapperFields.end.copy(new PointWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new RouteEdgeRefStruct(refObject.toObject());
    this._edgeidIntialized = true;
    this._wrapperFields.start.copyRefObject(this._refObject.start);
    this._wrapperFields.end.copyRefObject(this._refObject.end);
  }
  copy(other) {
    this._refObject = new RouteEdgeRefStruct(other._refObject.toObject());
    this._edgeidIntialized = true;
    this._wrapperFields.start.copy(other._wrapperFields.start);
    this._wrapperFields.end.copy(other._wrapperFields.end);
  }
  static get classType() {
    return RouteEdgeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"edgeid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"end","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"RouteEdge","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RouteEdge"};
  }
  hasMember(name) {
    let memberNames = ["edgeid","start","end"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RouteEdgeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RouteEdgeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RouteEdgeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RouteEdgeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RouteEdgeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RouteEdgeRefArray(size);
    this._refObject = new RouteEdgeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RouteEdgeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RouteEdgeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RouteEdgeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RouteEdgeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RouteEdgeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RouteEdgeArrayWrapper;
  }
}
module.exports = RouteEdgeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "edgeid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "end",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "RouteEdge",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RouteEdge"
}
*/
