// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__TerrainReport.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const TerrainReportRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  latitude: primitiveTypes.float64,
  longitude: primitiveTypes.float64,
  spacing: primitiveTypes.uint16,
  terrain_height: primitiveTypes.float32,
  current_height: primitiveTypes.float32,
  pending: primitiveTypes.uint16,
  loaded: primitiveTypes.uint16,
});
const TerrainReportRefArray = ArrayType(TerrainReportRefStruct);
const TerrainReportRefStructArray = StructType({
  data: TerrainReportRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TerrainReportWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TerrainReportRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._latitudeIntialized = true;
      this._longitudeIntialized = true;
      this._spacingIntialized = true;
      this._terrain_heightIntialized = true;
      this._current_heightIntialized = true;
      this._pendingIntialized = true;
      this._loadedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TerrainReportRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._latitudeIntialized = false;
    this._longitudeIntialized = false;
    this._spacingIntialized = false;
    this._terrain_heightIntialized = false;
    this._current_heightIntialized = false;
    this._pendingIntialized = false;
    this._loadedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new TerrainReportWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TerrainReportArrayWrapper;
  }
  static get ArrayType() {
    return TerrainReportArrayWrapper;
  }
  static get refObjectArrayType() {
    return TerrainReportRefStructArray
  }
  static get refObjectType() {
    return TerrainReportRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._latitudeIntialized) {
        throw new TypeError('Invalid argument: latitude in TerrainReport');
      }
      if (!this._longitudeIntialized) {
        throw new TypeError('Invalid argument: longitude in TerrainReport');
      }
      if (!this._spacingIntialized) {
        throw new TypeError('Invalid argument: spacing in TerrainReport');
      }
      if (!this._terrain_heightIntialized) {
        throw new TypeError('Invalid argument: terrain_height in TerrainReport');
      }
      if (!this._current_heightIntialized) {
        throw new TypeError('Invalid argument: current_height in TerrainReport');
      }
      if (!this._pendingIntialized) {
        throw new TypeError('Invalid argument: pending in TerrainReport');
      }
      if (!this._loadedIntialized) {
        throw new TypeError('Invalid argument: loaded in TerrainReport');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._latitudeIntialized = true;
    this._refObject.latitude = refObject.latitude;
    this._longitudeIntialized = true;
    this._refObject.longitude = refObject.longitude;
    this._spacingIntialized = true;
    this._refObject.spacing = refObject.spacing;
    this._terrain_heightIntialized = true;
    this._refObject.terrain_height = refObject.terrain_height;
    this._current_heightIntialized = true;
    this._refObject.current_height = refObject.current_height;
    this._pendingIntialized = true;
    this._refObject.pending = refObject.pending;
    this._loadedIntialized = true;
    this._refObject.loaded = refObject.loaded;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    TerrainReportWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'TerrainReport'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get latitude() {
    if (!this._latitudeIntialized) {
      return undefined;
    }
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._latitudeIntialized = true;
    this._refObject.latitude = value;
  }
  get longitude() {
    if (!this._longitudeIntialized) {
      return undefined;
    }
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._longitudeIntialized = true;
    this._refObject.longitude = value;
  }
  get spacing() {
    if (!this._spacingIntialized) {
      return undefined;
    }
    return this._refObject.spacing;
  }
  set spacing(value) {
    this._spacingIntialized = true;
    this._refObject.spacing = value;
  }
  get terrain_height() {
    if (!this._terrain_heightIntialized) {
      return undefined;
    }
    return this._refObject.terrain_height;
  }
  set terrain_height(value) {
    this._terrain_heightIntialized = true;
    this._refObject.terrain_height = value;
  }
  get current_height() {
    if (!this._current_heightIntialized) {
      return undefined;
    }
    return this._refObject.current_height;
  }
  set current_height(value) {
    this._current_heightIntialized = true;
    this._refObject.current_height = value;
  }
  get pending() {
    if (!this._pendingIntialized) {
      return undefined;
    }
    return this._refObject.pending;
  }
  set pending(value) {
    this._pendingIntialized = true;
    this._refObject.pending = value;
  }
  get loaded() {
    if (!this._loadedIntialized) {
      return undefined;
    }
    return this._refObject.loaded;
  }
  set loaded(value) {
    this._loadedIntialized = true;
    this._refObject.loaded = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TerrainReportRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._spacingIntialized = true;
    this._terrain_heightIntialized = true;
    this._current_heightIntialized = true;
    this._pendingIntialized = true;
    this._loadedIntialized = true;
  }
  copy(other) {
    this._refObject = new TerrainReportRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._spacingIntialized = true;
    this._terrain_heightIntialized = true;
    this._current_heightIntialized = true;
    this._pendingIntialized = true;
    this._loadedIntialized = true;
  }
  static get classType() {
    return TerrainReportWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"spacing","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"terrain_height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pending","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"loaded","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"TerrainReport","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TerrainReport"};
  }
  hasMember(name) {
    let memberNames = ["header","latitude","longitude","spacing","terrain_height","current_height","pending","loaded"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TerrainReportArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TerrainReportWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TerrainReportWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TerrainReportArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TerrainReportArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TerrainReportRefArray(size);
    this._refObject = new TerrainReportRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TerrainReportWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TerrainReportArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TerrainReportArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TerrainReportWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TerrainReportWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TerrainReportArrayWrapper;
  }
}
module.exports = TerrainReportWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "spacing",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "terrain_height",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "current_height",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pending",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "loaded",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "TerrainReport",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TerrainReport"
}
*/
