// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__DebugValue.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const DebugValueRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  index: primitiveTypes.int32,
  array_id: primitiveTypes.int32,
  name: primitiveTypes.string,
  value_float: primitiveTypes.float32,
  value_int: primitiveTypes.int32,
  data: Float32Wrapper.refObjectArrayType,
  type: primitiveTypes.uint8,
});
const DebugValueRefArray = ArrayType(DebugValueRefStruct);
const DebugValueRefStructArray = StructType({
  data: DebugValueRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DebugValueWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DebugValueRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._indexIntialized = true;
      this._array_idIntialized = true;
      this._nameIntialized = true;
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._value_floatIntialized = true;
      this._value_intIntialized = true;
      this._wrapperFields.data = Float32Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
      this._typeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DebugValueRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._indexIntialized = false;
    this._array_idIntialized = false;
    this._nameIntialized = false;
    this._wrapperFields.name = new StringWrapper();
    this._value_floatIntialized = false;
    this._value_intIntialized = false;
    this._wrapperFields.data = Float32Wrapper.createArray();
    this._typeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new DebugValueWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DebugValueArrayWrapper;
  }
  static get ArrayType() {
    return DebugValueArrayWrapper;
  }
  static get refObjectArrayType() {
    return DebugValueRefStructArray
  }
  static get refObjectType() {
    return DebugValueRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._indexIntialized) {
        throw new TypeError('Invalid argument: index in DebugValue');
      }
      if (!this._array_idIntialized) {
        throw new TypeError('Invalid argument: array_id in DebugValue');
      }
      if (!this._nameIntialized) {
        throw new TypeError('Invalid argument: name in DebugValue');
      }
      if (!this._value_floatIntialized) {
        throw new TypeError('Invalid argument: value_float in DebugValue');
      }
      if (!this._value_intIntialized) {
        throw new TypeError('Invalid argument: value_int in DebugValue');
      }
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in DebugValue');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.name.freeze(own, checkConsistency);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    this._wrapperFields.data.freeze(own, checkConsistency);
    this._refObject.data = this._wrapperFields.data.refObject;
    if (own) {
      this._wrapperFields.data.fill(Float32Array.from([]));
      this._wrapperFields.data.freeze(own, checkConsistency);
      this._refObject.data = this._wrapperFields.data.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._indexIntialized = true;
    this._refObject.index = refObject.index;
    this._array_idIntialized = true;
    this._refObject.array_id = refObject.array_id;
    this._nameIntialized = true;
    this._wrapperFields.name.data = refObject.name.data;
    this._value_floatIntialized = true;
    this._refObject.value_float = refObject.value_float;
    this._value_intIntialized = true;
    this._refObject.value_int = refObject.value_int;
    this._wrapperFields.data.copyRefObject(refObject.data);
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.name);
    if (refObject.data.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.data);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.data, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    DebugValueWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'DebugValue'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get index() {
    if (!this._indexIntialized) {
      return undefined;
    }
    return this._refObject.index;
  }
  set index(value) {
    this._indexIntialized = true;
    this._refObject.index = value;
  }
  get array_id() {
    if (!this._array_idIntialized) {
      return undefined;
    }
    return this._refObject.array_id;
  }
  set array_id(value) {
    this._array_idIntialized = true;
    this._refObject.array_id = value;
  }
  get name() {
    if (!this._nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = value;
  }
  get value_float() {
    if (!this._value_floatIntialized) {
      return undefined;
    }
    return this._refObject.value_float;
  }
  set value_float(value) {
    this._value_floatIntialized = true;
    this._refObject.value_float = value;
  }
  get value_int() {
    if (!this._value_intIntialized) {
      return undefined;
    }
    return this._refObject.value_int;
  }
  set value_int(value) {
    this._value_intIntialized = true;
    this._refObject.value_int = value;
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    this._wrapperFields['data'].fill(value);
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DebugValueRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._indexIntialized = true;
    this._array_idIntialized = true;
    this._nameIntialized = true;
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._value_floatIntialized = true;
    this._value_intIntialized = true;
    this._wrapperFields.data.copyRefObject(this._refObject.data);
    this._typeIntialized = true;
  }
  copy(other) {
    this._refObject = new DebugValueRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._indexIntialized = true;
    this._array_idIntialized = true;
    this._nameIntialized = true;
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._value_floatIntialized = true;
    this._value_intIntialized = true;
    this._wrapperFields.data.copy(other._wrapperFields.data);
    this._typeIntialized = true;
  }
  static get classType() {
    return DebugValueWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"TYPE_DEBUG","value":0},{"type":"uint8","name":"TYPE_DEBUG_VECT","value":1},{"type":"uint8","name":"TYPE_DEBUG_FLOAT_ARRAY","value":2},{"type":"uint8","name":"TYPE_NAMED_VALUE_FLOAT","value":3},{"type":"uint8","name":"TYPE_NAMED_VALUE_INT","value":4}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"array_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"value_float","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"value_int","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"DebugValue","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DebugValue"};
  }
  hasMember(name) {
    let memberNames = ["header","index","array_id","name","value_float","value_int","data","type"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DebugValueArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DebugValueWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DebugValueWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DebugValueArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DebugValueArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DebugValueRefArray(size);
    this._refObject = new DebugValueRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DebugValueWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DebugValueArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DebugValueArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DebugValueWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DebugValueWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DebugValueArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(DebugValueWrapper, "TYPE_DEBUG", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(DebugValueWrapper, "TYPE_DEBUG_VECT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(DebugValueWrapper, "TYPE_DEBUG_FLOAT_ARRAY", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(DebugValueWrapper, "TYPE_NAMED_VALUE_FLOAT", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(DebugValueWrapper, "TYPE_NAMED_VALUE_INT", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = DebugValueWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "TYPE_DEBUG",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "TYPE_DEBUG_VECT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "TYPE_DEBUG_FLOAT_ARRAY",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "TYPE_NAMED_VALUE_FLOAT",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "TYPE_NAMED_VALUE_INT",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "array_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "value_float",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "value_int",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "DebugValue",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DebugValue"
}
*/
