// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__GimbalManagerSetRoi_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GimbalManagerSetRoi_RequestRefStruct = StructType({
  mode: primitiveTypes.uint8,
  gimbal_device_id: primitiveTypes.uint8,
  latitude: primitiveTypes.float32,
  longitude: primitiveTypes.float32,
  altitude: primitiveTypes.float32,
  pitch_offset: primitiveTypes.float32,
  roll_offset: primitiveTypes.float32,
  yaw_offset: primitiveTypes.float32,
  sysid: primitiveTypes.uint8,
});
const GimbalManagerSetRoi_RequestRefArray = ArrayType(GimbalManagerSetRoi_RequestRefStruct);
const GimbalManagerSetRoi_RequestRefStructArray = StructType({
  data: GimbalManagerSetRoi_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GimbalManagerSetRoi_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GimbalManagerSetRoi_RequestRefStruct(other._refObject.toObject());
      this._modeIntialized = true;
      this._gimbal_device_idIntialized = true;
      this._latitudeIntialized = true;
      this._longitudeIntialized = true;
      this._altitudeIntialized = true;
      this._pitch_offsetIntialized = true;
      this._roll_offsetIntialized = true;
      this._yaw_offsetIntialized = true;
      this._sysidIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GimbalManagerSetRoi_RequestRefStruct();
    this._modeIntialized = false;
    this._gimbal_device_idIntialized = false;
    this._latitudeIntialized = false;
    this._longitudeIntialized = false;
    this._altitudeIntialized = false;
    this._pitch_offsetIntialized = false;
    this._roll_offsetIntialized = false;
    this._yaw_offsetIntialized = false;
    this._sysidIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GimbalManagerSetRoi_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GimbalManagerSetRoi_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GimbalManagerSetRoi_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GimbalManagerSetRoi_RequestRefStructArray
  }
  static get refObjectType() {
    return GimbalManagerSetRoi_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in GimbalManagerSetRoi_Request');
      }
      if (!this._gimbal_device_idIntialized) {
        throw new TypeError('Invalid argument: gimbal_device_id in GimbalManagerSetRoi_Request');
      }
      if (!this._latitudeIntialized) {
        throw new TypeError('Invalid argument: latitude in GimbalManagerSetRoi_Request');
      }
      if (!this._longitudeIntialized) {
        throw new TypeError('Invalid argument: longitude in GimbalManagerSetRoi_Request');
      }
      if (!this._altitudeIntialized) {
        throw new TypeError('Invalid argument: altitude in GimbalManagerSetRoi_Request');
      }
      if (!this._pitch_offsetIntialized) {
        throw new TypeError('Invalid argument: pitch_offset in GimbalManagerSetRoi_Request');
      }
      if (!this._roll_offsetIntialized) {
        throw new TypeError('Invalid argument: roll_offset in GimbalManagerSetRoi_Request');
      }
      if (!this._yaw_offsetIntialized) {
        throw new TypeError('Invalid argument: yaw_offset in GimbalManagerSetRoi_Request');
      }
      if (!this._sysidIntialized) {
        throw new TypeError('Invalid argument: sysid in GimbalManagerSetRoi_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._modeIntialized = true;
    this._refObject.mode = refObject.mode;
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = refObject.gimbal_device_id;
    this._latitudeIntialized = true;
    this._refObject.latitude = refObject.latitude;
    this._longitudeIntialized = true;
    this._refObject.longitude = refObject.longitude;
    this._altitudeIntialized = true;
    this._refObject.altitude = refObject.altitude;
    this._pitch_offsetIntialized = true;
    this._refObject.pitch_offset = refObject.pitch_offset;
    this._roll_offsetIntialized = true;
    this._refObject.roll_offset = refObject.roll_offset;
    this._yaw_offsetIntialized = true;
    this._refObject.yaw_offset = refObject.yaw_offset;
    this._sysidIntialized = true;
    this._refObject.sysid = refObject.sysid;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GimbalManagerSetRoi_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'GimbalManagerSetRoi_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._refObject.mode;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._refObject.mode = value;
  }
  get gimbal_device_id() {
    if (!this._gimbal_device_idIntialized) {
      return undefined;
    }
    return this._refObject.gimbal_device_id;
  }
  set gimbal_device_id(value) {
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = value;
  }
  get latitude() {
    if (!this._latitudeIntialized) {
      return undefined;
    }
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._latitudeIntialized = true;
    this._refObject.latitude = value;
  }
  get longitude() {
    if (!this._longitudeIntialized) {
      return undefined;
    }
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._longitudeIntialized = true;
    this._refObject.longitude = value;
  }
  get altitude() {
    if (!this._altitudeIntialized) {
      return undefined;
    }
    return this._refObject.altitude;
  }
  set altitude(value) {
    this._altitudeIntialized = true;
    this._refObject.altitude = value;
  }
  get pitch_offset() {
    if (!this._pitch_offsetIntialized) {
      return undefined;
    }
    return this._refObject.pitch_offset;
  }
  set pitch_offset(value) {
    this._pitch_offsetIntialized = true;
    this._refObject.pitch_offset = value;
  }
  get roll_offset() {
    if (!this._roll_offsetIntialized) {
      return undefined;
    }
    return this._refObject.roll_offset;
  }
  set roll_offset(value) {
    this._roll_offsetIntialized = true;
    this._refObject.roll_offset = value;
  }
  get yaw_offset() {
    if (!this._yaw_offsetIntialized) {
      return undefined;
    }
    return this._refObject.yaw_offset;
  }
  set yaw_offset(value) {
    this._yaw_offsetIntialized = true;
    this._refObject.yaw_offset = value;
  }
  get sysid() {
    if (!this._sysidIntialized) {
      return undefined;
    }
    return this._refObject.sysid;
  }
  set sysid(value) {
    this._sysidIntialized = true;
    this._refObject.sysid = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GimbalManagerSetRoi_RequestRefStruct(refObject.toObject());
    this._modeIntialized = true;
    this._gimbal_device_idIntialized = true;
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._pitch_offsetIntialized = true;
    this._roll_offsetIntialized = true;
    this._yaw_offsetIntialized = true;
    this._sysidIntialized = true;
  }
  copy(other) {
    this._refObject = new GimbalManagerSetRoi_RequestRefStruct(other._refObject.toObject());
    this._modeIntialized = true;
    this._gimbal_device_idIntialized = true;
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._pitch_offsetIntialized = true;
    this._roll_offsetIntialized = true;
    this._yaw_offsetIntialized = true;
    this._sysidIntialized = true;
  }
  static get classType() {
    return GimbalManagerSetRoi_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ROI_MODE_LOCATION","value":0},{"type":"uint8","name":"ROI_MODE_WP_NEXT_OFFSET","value":1},{"type":"uint8","name":"ROI_MODE_SYSID","value":2},{"type":"uint8","name":"ROI_MODE_NONE","value":3}],"fields":[{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gimbal_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"roll_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sysid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"GimbalManagerSetRoi_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GimbalManagerSetRoi_Request"};
  }
  hasMember(name) {
    let memberNames = ["mode","gimbal_device_id","latitude","longitude","altitude","pitch_offset","roll_offset","yaw_offset","sysid"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GimbalManagerSetRoi_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GimbalManagerSetRoi_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GimbalManagerSetRoi_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerSetRoi_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerSetRoi_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GimbalManagerSetRoi_RequestRefArray(size);
    this._refObject = new GimbalManagerSetRoi_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GimbalManagerSetRoi_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GimbalManagerSetRoi_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GimbalManagerSetRoi_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GimbalManagerSetRoi_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GimbalManagerSetRoi_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GimbalManagerSetRoi_RequestArrayWrapper;
  }
}
// Define constants (4 in total)
Object.defineProperty(GimbalManagerSetRoi_RequestWrapper, "ROI_MODE_LOCATION", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerSetRoi_RequestWrapper, "ROI_MODE_WP_NEXT_OFFSET", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerSetRoi_RequestWrapper, "ROI_MODE_SYSID", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerSetRoi_RequestWrapper, "ROI_MODE_NONE", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = GimbalManagerSetRoi_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ROI_MODE_LOCATION",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ROI_MODE_WP_NEXT_OFFSET",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "ROI_MODE_SYSID",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "ROI_MODE_NONE",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gimbal_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "roll_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sysid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "GimbalManagerSetRoi_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GimbalManagerSetRoi_Request"
}
*/
