// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__State.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const StateRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  connected: primitiveTypes.bool,
  armed: primitiveTypes.bool,
  guided: primitiveTypes.bool,
  manual_input: primitiveTypes.bool,
  mode: primitiveTypes.string,
  system_status: primitiveTypes.uint8,
});
const StateRefArray = ArrayType(StateRefStruct);
const StateRefStructArray = StructType({
  data: StateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StateWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StateRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._connectedIntialized = true;
      this._armedIntialized = true;
      this._guidedIntialized = true;
      this._manual_inputIntialized = true;
      this._modeIntialized = true;
      this._wrapperFields.mode =  new StringWrapper(other._wrapperFields.mode);
      this._system_statusIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StateRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._connectedIntialized = false;
    this._armedIntialized = false;
    this._guidedIntialized = false;
    this._manual_inputIntialized = false;
    this._modeIntialized = false;
    this._wrapperFields.mode = new StringWrapper();
    this._system_statusIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new StateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StateArrayWrapper;
  }
  static get ArrayType() {
    return StateArrayWrapper;
  }
  static get refObjectArrayType() {
    return StateRefStructArray
  }
  static get refObjectType() {
    return StateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._connectedIntialized) {
        throw new TypeError('Invalid argument: connected in State');
      }
      if (!this._armedIntialized) {
        throw new TypeError('Invalid argument: armed in State');
      }
      if (!this._guidedIntialized) {
        throw new TypeError('Invalid argument: guided in State');
      }
      if (!this._manual_inputIntialized) {
        throw new TypeError('Invalid argument: manual_input in State');
      }
      if (!this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in State');
      }
      if (!this._system_statusIntialized) {
        throw new TypeError('Invalid argument: system_status in State');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.mode.freeze(own, checkConsistency);
    }
    this._refObject.mode = this._wrapperFields.mode.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._connectedIntialized = true;
    this._refObject.connected = refObject.connected;
    this._armedIntialized = true;
    this._refObject.armed = refObject.armed;
    this._guidedIntialized = true;
    this._refObject.guided = refObject.guided;
    this._manual_inputIntialized = true;
    this._refObject.manual_input = refObject.manual_input;
    this._modeIntialized = true;
    this._wrapperFields.mode.data = refObject.mode.data;
    this._system_statusIntialized = true;
    this._refObject.system_status = refObject.system_status;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.mode);
  }
  static destoryRawROS(msg) {
    StateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'State'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get connected() {
    if (!this._connectedIntialized) {
      return undefined;
    }
    return this._refObject.connected;
  }
  set connected(value) {
    this._connectedIntialized = true;
    this._refObject.connected = value;
  }
  get armed() {
    if (!this._armedIntialized) {
      return undefined;
    }
    return this._refObject.armed;
  }
  set armed(value) {
    this._armedIntialized = true;
    this._refObject.armed = value;
  }
  get guided() {
    if (!this._guidedIntialized) {
      return undefined;
    }
    return this._refObject.guided;
  }
  set guided(value) {
    this._guidedIntialized = true;
    this._refObject.guided = value;
  }
  get manual_input() {
    if (!this._manual_inputIntialized) {
      return undefined;
    }
    return this._refObject.manual_input;
  }
  set manual_input(value) {
    this._manual_inputIntialized = true;
    this._refObject.manual_input = value;
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._wrapperFields.mode.data;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._wrapperFields.mode.data = value;
  }
  get system_status() {
    if (!this._system_statusIntialized) {
      return undefined;
    }
    return this._refObject.system_status;
  }
  set system_status(value) {
    this._system_statusIntialized = true;
    this._refObject.system_status = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StateRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._connectedIntialized = true;
    this._armedIntialized = true;
    this._guidedIntialized = true;
    this._manual_inputIntialized = true;
    this._modeIntialized = true;
    this._wrapperFields.mode.copyRefObject(this._refObject.mode);
    this._system_statusIntialized = true;
  }
  copy(other) {
    this._refObject = new StateRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._connectedIntialized = true;
    this._armedIntialized = true;
    this._guidedIntialized = true;
    this._manual_inputIntialized = true;
    this._modeIntialized = true;
    this._wrapperFields.mode.copy(other._wrapperFields.mode);
    this._system_statusIntialized = true;
  }
  static get classType() {
    return StateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"string","name":"MODE_APM_PLANE_MANUAL","value":"MANUAL"},{"type":"string","name":"MODE_APM_PLANE_CIRCLE","value":"CIRCLE"},{"type":"string","name":"MODE_APM_PLANE_STABILIZE","value":"STABILIZE"},{"type":"string","name":"MODE_APM_PLANE_TRAINING","value":"TRAINING"},{"type":"string","name":"MODE_APM_PLANE_ACRO","value":"ACRO"},{"type":"string","name":"MODE_APM_PLANE_FBWA","value":"FBWA"},{"type":"string","name":"MODE_APM_PLANE_FBWB","value":"FBWB"},{"type":"string","name":"MODE_APM_PLANE_CRUISE","value":"CRUISE"},{"type":"string","name":"MODE_APM_PLANE_AUTOTUNE","value":"AUTOTUNE"},{"type":"string","name":"MODE_APM_PLANE_AUTO","value":"AUTO"},{"type":"string","name":"MODE_APM_PLANE_RTL","value":"RTL"},{"type":"string","name":"MODE_APM_PLANE_LOITER","value":"LOITER"},{"type":"string","name":"MODE_APM_PLANE_LAND","value":"LAND"},{"type":"string","name":"MODE_APM_PLANE_GUIDED","value":"GUIDED"},{"type":"string","name":"MODE_APM_PLANE_INITIALISING","value":"INITIALISING"},{"type":"string","name":"MODE_APM_PLANE_QSTABILIZE","value":"QSTABILIZE"},{"type":"string","name":"MODE_APM_PLANE_QHOVER","value":"QHOVER"},{"type":"string","name":"MODE_APM_PLANE_QLOITER","value":"QLOITER"},{"type":"string","name":"MODE_APM_PLANE_QLAND","value":"QLAND"},{"type":"string","name":"MODE_APM_PLANE_QRTL","value":"QRTL"},{"type":"string","name":"MODE_APM_COPTER_STABILIZE","value":"STABILIZE"},{"type":"string","name":"MODE_APM_COPTER_ACRO","value":"ACRO"},{"type":"string","name":"MODE_APM_COPTER_ALT_HOLD","value":"ALT_HOLD"},{"type":"string","name":"MODE_APM_COPTER_AUTO","value":"AUTO"},{"type":"string","name":"MODE_APM_COPTER_GUIDED","value":"GUIDED"},{"type":"string","name":"MODE_APM_COPTER_LOITER","value":"LOITER"},{"type":"string","name":"MODE_APM_COPTER_RTL","value":"RTL"},{"type":"string","name":"MODE_APM_COPTER_CIRCLE","value":"CIRCLE"},{"type":"string","name":"MODE_APM_COPTER_POSITION","value":"POSITION"},{"type":"string","name":"MODE_APM_COPTER_LAND","value":"LAND"},{"type":"string","name":"MODE_APM_COPTER_OF_LOITER","value":"OF_LOITER"},{"type":"string","name":"MODE_APM_COPTER_DRIFT","value":"DRIFT"},{"type":"string","name":"MODE_APM_COPTER_SPORT","value":"SPORT"},{"type":"string","name":"MODE_APM_COPTER_FLIP","value":"FLIP"},{"type":"string","name":"MODE_APM_COPTER_AUTOTUNE","value":"AUTOTUNE"},{"type":"string","name":"MODE_APM_COPTER_POSHOLD","value":"POSHOLD"},{"type":"string","name":"MODE_APM_COPTER_BRAKE","value":"BRAKE"},{"type":"string","name":"MODE_APM_COPTER_THROW","value":"THROW"},{"type":"string","name":"MODE_APM_COPTER_AVOID_ADSB","value":"AVOID_ADSB"},{"type":"string","name":"MODE_APM_COPTER_GUIDED_NOGPS","value":"GUIDED_NOGPS"},{"type":"string","name":"MODE_APM_ROVER_MANUAL","value":"MANUAL"},{"type":"string","name":"MODE_APM_ROVER_LEARNING","value":"LEARNING"},{"type":"string","name":"MODE_APM_ROVER_STEERING","value":"STEERING"},{"type":"string","name":"MODE_APM_ROVER_HOLD","value":"HOLD"},{"type":"string","name":"MODE_APM_ROVER_AUTO","value":"AUTO"},{"type":"string","name":"MODE_APM_ROVER_RTL","value":"RTL"},{"type":"string","name":"MODE_APM_ROVER_GUIDED","value":"GUIDED"},{"type":"string","name":"MODE_APM_ROVER_INITIALISING","value":"INITIALISING"},{"type":"string","name":"MODE_PX4_MANUAL","value":"MANUAL"},{"type":"string","name":"MODE_PX4_ACRO","value":"ACRO"},{"type":"string","name":"MODE_PX4_ALTITUDE","value":"ALTCTL"},{"type":"string","name":"MODE_PX4_POSITION","value":"POSCTL"},{"type":"string","name":"MODE_PX4_OFFBOARD","value":"OFFBOARD"},{"type":"string","name":"MODE_PX4_STABILIZED","value":"STABILIZED"},{"type":"string","name":"MODE_PX4_RATTITUDE","value":"RATTITUDE"},{"type":"string","name":"MODE_PX4_MISSION","value":"AUTO.MISSION"},{"type":"string","name":"MODE_PX4_LOITER","value":"AUTO.LOITER"},{"type":"string","name":"MODE_PX4_RTL","value":"AUTO.RTL"},{"type":"string","name":"MODE_PX4_LAND","value":"AUTO.LAND"},{"type":"string","name":"MODE_PX4_RTGS","value":"AUTO.RTGS"},{"type":"string","name":"MODE_PX4_READY","value":"AUTO.READY"},{"type":"string","name":"MODE_PX4_TAKEOFF","value":"AUTO.TAKEOFF"}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"connected","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"armed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"guided","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"manual_input","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"system_status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"State","stringUpperBound":null,"isPrimitiveType":false},"msgName":"State"};
  }
  hasMember(name) {
    let memberNames = ["header","connected","armed","guided","manual_input","mode","system_status"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StateRefArray(size);
    this._refObject = new StateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StateArrayWrapper;
  }
}
// Define constants (62 in total)
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_MANUAL", {value: "MANUAL", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_CIRCLE", {value: "CIRCLE", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_STABILIZE", {value: "STABILIZE", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_TRAINING", {value: "TRAINING", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_ACRO", {value: "ACRO", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_FBWA", {value: "FBWA", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_FBWB", {value: "FBWB", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_CRUISE", {value: "CRUISE", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_AUTOTUNE", {value: "AUTOTUNE", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_AUTO", {value: "AUTO", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_RTL", {value: "RTL", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_LOITER", {value: "LOITER", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_LAND", {value: "LAND", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_GUIDED", {value: "GUIDED", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_INITIALISING", {value: "INITIALISING", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_QSTABILIZE", {value: "QSTABILIZE", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_QHOVER", {value: "QHOVER", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_QLOITER", {value: "QLOITER", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_QLAND", {value: "QLAND", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_PLANE_QRTL", {value: "QRTL", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_STABILIZE", {value: "STABILIZE", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_ACRO", {value: "ACRO", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_ALT_HOLD", {value: "ALT_HOLD", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_AUTO", {value: "AUTO", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_GUIDED", {value: "GUIDED", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_LOITER", {value: "LOITER", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_RTL", {value: "RTL", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_CIRCLE", {value: "CIRCLE", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_POSITION", {value: "POSITION", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_LAND", {value: "LAND", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_OF_LOITER", {value: "OF_LOITER", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_DRIFT", {value: "DRIFT", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_SPORT", {value: "SPORT", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_FLIP", {value: "FLIP", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_AUTOTUNE", {value: "AUTOTUNE", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_POSHOLD", {value: "POSHOLD", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_BRAKE", {value: "BRAKE", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_THROW", {value: "THROW", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_AVOID_ADSB", {value: "AVOID_ADSB", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_COPTER_GUIDED_NOGPS", {value: "GUIDED_NOGPS", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_ROVER_MANUAL", {value: "MANUAL", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_ROVER_LEARNING", {value: "LEARNING", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_ROVER_STEERING", {value: "STEERING", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_ROVER_HOLD", {value: "HOLD", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_ROVER_AUTO", {value: "AUTO", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_ROVER_RTL", {value: "RTL", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_ROVER_GUIDED", {value: "GUIDED", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_APM_ROVER_INITIALISING", {value: "INITIALISING", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_MANUAL", {value: "MANUAL", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_ACRO", {value: "ACRO", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_ALTITUDE", {value: "ALTCTL", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_POSITION", {value: "POSCTL", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_OFFBOARD", {value: "OFFBOARD", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_STABILIZED", {value: "STABILIZED", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_RATTITUDE", {value: "RATTITUDE", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_MISSION", {value: "AUTO.MISSION", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_LOITER", {value: "AUTO.LOITER", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_RTL", {value: "AUTO.RTL", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_LAND", {value: "AUTO.LAND", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_RTGS", {value: "AUTO.RTGS", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_READY", {value: "AUTO.READY", writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "MODE_PX4_TAKEOFF", {value: "AUTO.TAKEOFF", writable: false, enumerable: true, configurable: true});
module.exports = StateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "string",
      "name": "MODE_APM_PLANE_MANUAL",
      "value": "MANUAL"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_CIRCLE",
      "value": "CIRCLE"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_STABILIZE",
      "value": "STABILIZE"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_TRAINING",
      "value": "TRAINING"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_ACRO",
      "value": "ACRO"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_FBWA",
      "value": "FBWA"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_FBWB",
      "value": "FBWB"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_CRUISE",
      "value": "CRUISE"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_AUTOTUNE",
      "value": "AUTOTUNE"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_AUTO",
      "value": "AUTO"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_RTL",
      "value": "RTL"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_LOITER",
      "value": "LOITER"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_LAND",
      "value": "LAND"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_GUIDED",
      "value": "GUIDED"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_INITIALISING",
      "value": "INITIALISING"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_QSTABILIZE",
      "value": "QSTABILIZE"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_QHOVER",
      "value": "QHOVER"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_QLOITER",
      "value": "QLOITER"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_QLAND",
      "value": "QLAND"
    },
    {
      "type": "string",
      "name": "MODE_APM_PLANE_QRTL",
      "value": "QRTL"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_STABILIZE",
      "value": "STABILIZE"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_ACRO",
      "value": "ACRO"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_ALT_HOLD",
      "value": "ALT_HOLD"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_AUTO",
      "value": "AUTO"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_GUIDED",
      "value": "GUIDED"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_LOITER",
      "value": "LOITER"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_RTL",
      "value": "RTL"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_CIRCLE",
      "value": "CIRCLE"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_POSITION",
      "value": "POSITION"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_LAND",
      "value": "LAND"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_OF_LOITER",
      "value": "OF_LOITER"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_DRIFT",
      "value": "DRIFT"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_SPORT",
      "value": "SPORT"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_FLIP",
      "value": "FLIP"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_AUTOTUNE",
      "value": "AUTOTUNE"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_POSHOLD",
      "value": "POSHOLD"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_BRAKE",
      "value": "BRAKE"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_THROW",
      "value": "THROW"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_AVOID_ADSB",
      "value": "AVOID_ADSB"
    },
    {
      "type": "string",
      "name": "MODE_APM_COPTER_GUIDED_NOGPS",
      "value": "GUIDED_NOGPS"
    },
    {
      "type": "string",
      "name": "MODE_APM_ROVER_MANUAL",
      "value": "MANUAL"
    },
    {
      "type": "string",
      "name": "MODE_APM_ROVER_LEARNING",
      "value": "LEARNING"
    },
    {
      "type": "string",
      "name": "MODE_APM_ROVER_STEERING",
      "value": "STEERING"
    },
    {
      "type": "string",
      "name": "MODE_APM_ROVER_HOLD",
      "value": "HOLD"
    },
    {
      "type": "string",
      "name": "MODE_APM_ROVER_AUTO",
      "value": "AUTO"
    },
    {
      "type": "string",
      "name": "MODE_APM_ROVER_RTL",
      "value": "RTL"
    },
    {
      "type": "string",
      "name": "MODE_APM_ROVER_GUIDED",
      "value": "GUIDED"
    },
    {
      "type": "string",
      "name": "MODE_APM_ROVER_INITIALISING",
      "value": "INITIALISING"
    },
    {
      "type": "string",
      "name": "MODE_PX4_MANUAL",
      "value": "MANUAL"
    },
    {
      "type": "string",
      "name": "MODE_PX4_ACRO",
      "value": "ACRO"
    },
    {
      "type": "string",
      "name": "MODE_PX4_ALTITUDE",
      "value": "ALTCTL"
    },
    {
      "type": "string",
      "name": "MODE_PX4_POSITION",
      "value": "POSCTL"
    },
    {
      "type": "string",
      "name": "MODE_PX4_OFFBOARD",
      "value": "OFFBOARD"
    },
    {
      "type": "string",
      "name": "MODE_PX4_STABILIZED",
      "value": "STABILIZED"
    },
    {
      "type": "string",
      "name": "MODE_PX4_RATTITUDE",
      "value": "RATTITUDE"
    },
    {
      "type": "string",
      "name": "MODE_PX4_MISSION",
      "value": "AUTO.MISSION"
    },
    {
      "type": "string",
      "name": "MODE_PX4_LOITER",
      "value": "AUTO.LOITER"
    },
    {
      "type": "string",
      "name": "MODE_PX4_RTL",
      "value": "AUTO.RTL"
    },
    {
      "type": "string",
      "name": "MODE_PX4_LAND",
      "value": "AUTO.LAND"
    },
    {
      "type": "string",
      "name": "MODE_PX4_RTGS",
      "value": "AUTO.RTGS"
    },
    {
      "type": "string",
      "name": "MODE_PX4_READY",
      "value": "AUTO.READY"
    },
    {
      "type": "string",
      "name": "MODE_PX4_TAKEOFF",
      "value": "AUTO.TAKEOFF"
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "connected",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "armed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "guided",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "manual_input",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "system_status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "State",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "State"
}
*/
