// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__CommandCode.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CommandCodeRefStruct = StructType({
  _dummy: primitiveTypes.bool,
});
const CommandCodeRefArray = ArrayType(CommandCodeRefStruct);
const CommandCodeRefStructArray = StructType({
  data: CommandCodeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CommandCodeWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CommandCodeRefStruct(other._refObject.toObject());
      this.__dummyIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CommandCodeRefStruct();
    this.__dummyIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new CommandCodeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CommandCodeArrayWrapper;
  }
  static get ArrayType() {
    return CommandCodeArrayWrapper;
  }
  static get refObjectArrayType() {
    return CommandCodeRefStructArray
  }
  static get refObjectType() {
    return CommandCodeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this.__dummyIntialized) {
        throw new TypeError('Invalid argument: _dummy in CommandCode');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this.__dummyIntialized = true;
    this._refObject._dummy = refObject._dummy;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    CommandCodeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'CommandCode'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get _dummy() {
    if (!this.__dummyIntialized) {
      return undefined;
    }
    return this._refObject._dummy;
  }
  set _dummy(value) {
    this.__dummyIntialized = true;
    this._refObject._dummy = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CommandCodeRefStruct(refObject.toObject());
    this.__dummyIntialized = true;
  }
  copy(other) {
    this._refObject = new CommandCodeRefStruct(other._refObject.toObject());
    this.__dummyIntialized = true;
  }
  static get classType() {
    return CommandCodeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint16","name":"AIRFRAME_CONFIGURATION","value":2520},{"type":"uint16","name":"ARM_AUTHORIZATION_REQUEST","value":3001},{"type":"uint16","name":"CAMERA_TRACK_POINT","value":2004},{"type":"uint16","name":"CAMERA_TRACK_RECTANGLE","value":2005},{"type":"uint16","name":"CAMERA_STOP_TRACKING","value":2010},{"type":"uint16","name":"CAN_FORWARD","value":32000},{"type":"uint16","name":"COMPONENT_ARM_DISARM","value":400},{"type":"uint16","name":"CONDITION_DELAY","value":112},{"type":"uint16","name":"CONDITION_CHANGE_ALT","value":113},{"type":"uint16","name":"CONDITION_DISTANCE","value":114},{"type":"uint16","name":"CONDITION_YAW","value":115},{"type":"uint16","name":"CONDITION_LAST","value":159},{"type":"uint16","name":"CONTROL_HIGH_LATENCY","value":2600},{"type":"uint16","name":"DO_FOLLOW","value":32},{"type":"uint16","name":"DO_FOLLOW_REPOSITION","value":33},{"type":"uint16","name":"DO_SET_MODE","value":176},{"type":"uint16","name":"DO_JUMP","value":177},{"type":"uint16","name":"DO_CHANGE_SPEED","value":178},{"type":"uint16","name":"DO_SET_HOME","value":179},{"type":"uint16","name":"DO_SET_PARAMETER","value":180},{"type":"uint16","name":"DO_SET_RELAY","value":181},{"type":"uint16","name":"DO_REPEAT_RELAY","value":182},{"type":"uint16","name":"DO_SET_SERVO","value":183},{"type":"uint16","name":"DO_REPEAT_SERVO","value":184},{"type":"uint16","name":"DO_FLIGHTTERMINATION","value":185},{"type":"uint16","name":"DO_CHANGE_ALTITUDE","value":186},{"type":"uint16","name":"DO_RETURN_PATH_START","value":188},{"type":"uint16","name":"DO_LAND_START","value":189},{"type":"uint16","name":"DO_RALLY_LAND","value":190},{"type":"uint16","name":"DO_GO_AROUND","value":191},{"type":"uint16","name":"DO_REPOSITION","value":192},{"type":"uint16","name":"DO_PAUSE_CONTINUE","value":193},{"type":"uint16","name":"DO_SET_REVERSE","value":194},{"type":"uint16","name":"DO_SET_ROI_LOCATION","value":195},{"type":"uint16","name":"DO_SET_ROI_WPNEXT_OFFSET","value":196},{"type":"uint16","name":"DO_SET_ROI_NONE","value":197},{"type":"uint16","name":"DO_SET_ROI_SYSID","value":198},{"type":"uint16","name":"DO_CONTROL_VIDEO","value":200},{"type":"uint16","name":"DO_SET_ROI","value":201},{"type":"uint16","name":"DO_DIGICAM_CONFIGURE","value":202},{"type":"uint16","name":"DO_DIGICAM_CONTROL","value":203},{"type":"uint16","name":"DO_MOUNT_CONFIGURE","value":204},{"type":"uint16","name":"DO_MOUNT_CONTROL","value":205},{"type":"uint16","name":"DO_SET_CAM_TRIGG_DIST","value":206},{"type":"uint16","name":"DO_FENCE_ENABLE","value":207},{"type":"uint16","name":"DO_PARACHUTE","value":208},{"type":"uint16","name":"DO_MOTOR_TEST","value":209},{"type":"uint16","name":"DO_INVERTED_FLIGHT","value":210},{"type":"uint16","name":"DO_GRIPPER","value":211},{"type":"uint16","name":"DO_AUTOTUNE_ENABLE","value":212},{"type":"uint16","name":"DO_SET_CAM_TRIGG_INTERVAL","value":214},{"type":"uint16","name":"DO_MOUNT_CONTROL_QUAT","value":220},{"type":"uint16","name":"DO_GUIDED_MASTER","value":221},{"type":"uint16","name":"DO_GUIDED_LIMITS","value":222},{"type":"uint16","name":"DO_ENGINE_CONTROL","value":223},{"type":"uint16","name":"DO_SET_MISSION_CURRENT","value":224},{"type":"uint16","name":"DO_LAST","value":240},{"type":"uint16","name":"DO_JUMP_TAG","value":601},{"type":"uint16","name":"DO_GIMBAL_MANAGER_PITCHYAW","value":1000},{"type":"uint16","name":"DO_GIMBAL_MANAGER_CONFIGURE","value":1001},{"type":"uint16","name":"DO_TRIGGER_CONTROL","value":2003},{"type":"uint16","name":"DO_VTOL_TRANSITION","value":3000},{"type":"uint16","name":"DO_SET_SAFETY_SWITCH_STATE","value":5300},{"type":"uint16","name":"DO_ADSB_OUT_IDENT","value":10001},{"type":"uint16","name":"DO_WINCH","value":42600},{"type":"uint16","name":"EXTERNAL_POSITION_ESTIMATE","value":43003},{"type":"uint16","name":"FIXED_MAG_CAL_YAW","value":42006},{"type":"uint16","name":"GET_HOME_POSITION","value":410},{"type":"uint16","name":"GET_MESSAGE_INTERVAL","value":510},{"type":"uint16","name":"IMAGE_START_CAPTURE","value":2000},{"type":"uint16","name":"IMAGE_STOP_CAPTURE","value":2001},{"type":"uint16","name":"JUMP_TAG","value":600},{"type":"uint16","name":"LOGGING_START","value":2510},{"type":"uint16","name":"LOGGING_STOP","value":2511},{"type":"uint16","name":"MISSION_START","value":300},{"type":"uint16","name":"NAV_WAYPOINT","value":16},{"type":"uint16","name":"NAV_LOITER_UNLIM","value":17},{"type":"uint16","name":"NAV_LOITER_TURNS","value":18},{"type":"uint16","name":"NAV_LOITER_TIME","value":19},{"type":"uint16","name":"NAV_RETURN_TO_LAUNCH","value":20},{"type":"uint16","name":"NAV_LAND","value":21},{"type":"uint16","name":"NAV_TAKEOFF","value":22},{"type":"uint16","name":"NAV_LAND_LOCAL","value":23},{"type":"uint16","name":"NAV_TAKEOFF_LOCAL","value":24},{"type":"uint16","name":"NAV_FOLLOW","value":25},{"type":"uint16","name":"NAV_CONTINUE_AND_CHANGE_ALT","value":30},{"type":"uint16","name":"NAV_LOITER_TO_ALT","value":31},{"type":"uint16","name":"NAV_ROI","value":80},{"type":"uint16","name":"NAV_PATHPLANNING","value":81},{"type":"uint16","name":"NAV_SPLINE_WAYPOINT","value":82},{"type":"uint16","name":"NAV_VTOL_TAKEOFF","value":84},{"type":"uint16","name":"NAV_VTOL_LAND","value":85},{"type":"uint16","name":"NAV_GUIDED_ENABLE","value":92},{"type":"uint16","name":"NAV_DELAY","value":93},{"type":"uint16","name":"NAV_PAYLOAD_PLACE","value":94},{"type":"uint16","name":"NAV_LAST","value":95},{"type":"uint16","name":"NAV_SET_YAW_SPEED","value":213},{"type":"uint16","name":"NAV_FENCE_RETURN_POINT","value":5000},{"type":"uint16","name":"NAV_FENCE_POLYGON_VERTEX_INCLUSION","value":5001},{"type":"uint16","name":"NAV_FENCE_POLYGON_VERTEX_EXCLUSION","value":5002},{"type":"uint16","name":"NAV_FENCE_CIRCLE_INCLUSION","value":5003},{"type":"uint16","name":"NAV_FENCE_CIRCLE_EXCLUSION","value":5004},{"type":"uint16","name":"NAV_RALLY_POINT","value":5100},{"type":"uint16","name":"OBLIQUE_SURVEY","value":260},{"type":"uint16","name":"OVERRIDE_GOTO","value":252},{"type":"uint16","name":"PANORAMA_CREATE","value":2800},{"type":"uint16","name":"PAYLOAD_PREPARE_DEPLOY","value":30001},{"type":"uint16","name":"PAYLOAD_CONTROL_DEPLOY","value":30002},{"type":"uint16","name":"PREFLIGHT_CALIBRATION","value":241},{"type":"uint16","name":"PREFLIGHT_SET_SENSOR_OFFSETS","value":242},{"type":"uint16","name":"PREFLIGHT_UAVCAN","value":243},{"type":"uint16","name":"PREFLIGHT_STORAGE","value":245},{"type":"uint16","name":"PREFLIGHT_REBOOT_SHUTDOWN","value":246},{"type":"uint16","name":"REQUEST_MESSAGE","value":512},{"type":"uint16","name":"REQUEST_PROTOCOL_VERSION","value":519},{"type":"uint16","name":"REQUEST_AUTOPILOT_CAPABILITIES","value":520},{"type":"uint16","name":"REQUEST_CAMERA_INFORMATION","value":521},{"type":"uint16","name":"REQUEST_CAMERA_SETTINGS","value":522},{"type":"uint16","name":"REQUEST_STORAGE_INFORMATION","value":525},{"type":"uint16","name":"REQUEST_CAMERA_CAPTURE_STATUS","value":527},{"type":"uint16","name":"REQUEST_FLIGHT_INFORMATION","value":528},{"type":"uint16","name":"REQUEST_VIDEO_STREAM_INFORMATION","value":2504},{"type":"uint16","name":"REQUEST_VIDEO_STREAM_STATUS","value":2505},{"type":"uint16","name":"RESET_CAMERA_SETTINGS","value":529},{"type":"uint16","name":"RUN_PREARM_CHECKS","value":401},{"type":"uint16","name":"SET_MESSAGE_INTERVAL","value":511},{"type":"uint16","name":"SET_CAMERA_MODE","value":530},{"type":"uint16","name":"SET_CAMERA_ZOOM","value":531},{"type":"uint16","name":"SET_CAMERA_FOCUS","value":532},{"type":"uint16","name":"SET_STORAGE_USAGE","value":533},{"type":"uint16","name":"SET_CAMERA_SOURCE","value":534},{"type":"uint16","name":"SET_GUIDED_SUBMODE_STANDARD","value":4000},{"type":"uint16","name":"SET_GUIDED_SUBMODE_CIRCLE","value":4001},{"type":"uint16","name":"START_RX_PAIR","value":500},{"type":"uint16","name":"STORAGE_FORMAT","value":526},{"type":"uint16","name":"UAVCAN_GET_NODE_INFO","value":5200},{"type":"uint16","name":"VIDEO_START_CAPTURE","value":2500},{"type":"uint16","name":"VIDEO_STOP_CAPTURE","value":2501},{"type":"uint16","name":"VIDEO_START_STREAMING","value":2502},{"type":"uint16","name":"VIDEO_STOP_STREAMING","value":2503}],"fields":[],"baseType":{"pkgName":"mavros_msgs","type":"CommandCode","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CommandCode"};
  }
  hasMember(name) {
    let memberNames = ["_dummy"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CommandCodeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CommandCodeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CommandCodeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandCodeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandCodeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CommandCodeRefArray(size);
    this._refObject = new CommandCodeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CommandCodeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CommandCodeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CommandCodeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CommandCodeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CommandCodeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CommandCodeArrayWrapper;
  }
}
// Define constants (140 in total)
Object.defineProperty(CommandCodeWrapper, "AIRFRAME_CONFIGURATION", {value: 2520, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "ARM_AUTHORIZATION_REQUEST", {value: 3001, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "CAMERA_TRACK_POINT", {value: 2004, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "CAMERA_TRACK_RECTANGLE", {value: 2005, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "CAMERA_STOP_TRACKING", {value: 2010, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "CAN_FORWARD", {value: 32000, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "COMPONENT_ARM_DISARM", {value: 400, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "CONDITION_DELAY", {value: 112, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "CONDITION_CHANGE_ALT", {value: 113, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "CONDITION_DISTANCE", {value: 114, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "CONDITION_YAW", {value: 115, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "CONDITION_LAST", {value: 159, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "CONTROL_HIGH_LATENCY", {value: 2600, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_FOLLOW", {value: 32, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_FOLLOW_REPOSITION", {value: 33, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_MODE", {value: 176, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_JUMP", {value: 177, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_CHANGE_SPEED", {value: 178, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_HOME", {value: 179, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_PARAMETER", {value: 180, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_RELAY", {value: 181, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_REPEAT_RELAY", {value: 182, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_SERVO", {value: 183, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_REPEAT_SERVO", {value: 184, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_FLIGHTTERMINATION", {value: 185, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_CHANGE_ALTITUDE", {value: 186, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_RETURN_PATH_START", {value: 188, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_LAND_START", {value: 189, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_RALLY_LAND", {value: 190, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_GO_AROUND", {value: 191, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_REPOSITION", {value: 192, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_PAUSE_CONTINUE", {value: 193, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_REVERSE", {value: 194, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_ROI_LOCATION", {value: 195, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_ROI_WPNEXT_OFFSET", {value: 196, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_ROI_NONE", {value: 197, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_ROI_SYSID", {value: 198, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_CONTROL_VIDEO", {value: 200, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_ROI", {value: 201, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_DIGICAM_CONFIGURE", {value: 202, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_DIGICAM_CONTROL", {value: 203, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_MOUNT_CONFIGURE", {value: 204, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_MOUNT_CONTROL", {value: 205, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_CAM_TRIGG_DIST", {value: 206, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_FENCE_ENABLE", {value: 207, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_PARACHUTE", {value: 208, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_MOTOR_TEST", {value: 209, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_INVERTED_FLIGHT", {value: 210, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_GRIPPER", {value: 211, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_AUTOTUNE_ENABLE", {value: 212, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_CAM_TRIGG_INTERVAL", {value: 214, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_MOUNT_CONTROL_QUAT", {value: 220, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_GUIDED_MASTER", {value: 221, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_GUIDED_LIMITS", {value: 222, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_ENGINE_CONTROL", {value: 223, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_MISSION_CURRENT", {value: 224, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_LAST", {value: 240, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_JUMP_TAG", {value: 601, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_GIMBAL_MANAGER_PITCHYAW", {value: 1000, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_GIMBAL_MANAGER_CONFIGURE", {value: 1001, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_TRIGGER_CONTROL", {value: 2003, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_VTOL_TRANSITION", {value: 3000, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_SET_SAFETY_SWITCH_STATE", {value: 5300, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_ADSB_OUT_IDENT", {value: 10001, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "DO_WINCH", {value: 42600, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "EXTERNAL_POSITION_ESTIMATE", {value: 43003, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "FIXED_MAG_CAL_YAW", {value: 42006, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "GET_HOME_POSITION", {value: 410, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "GET_MESSAGE_INTERVAL", {value: 510, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "IMAGE_START_CAPTURE", {value: 2000, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "IMAGE_STOP_CAPTURE", {value: 2001, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "JUMP_TAG", {value: 600, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "LOGGING_START", {value: 2510, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "LOGGING_STOP", {value: 2511, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "MISSION_START", {value: 300, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_WAYPOINT", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_LOITER_UNLIM", {value: 17, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_LOITER_TURNS", {value: 18, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_LOITER_TIME", {value: 19, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_RETURN_TO_LAUNCH", {value: 20, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_LAND", {value: 21, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_TAKEOFF", {value: 22, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_LAND_LOCAL", {value: 23, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_TAKEOFF_LOCAL", {value: 24, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_FOLLOW", {value: 25, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_CONTINUE_AND_CHANGE_ALT", {value: 30, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_LOITER_TO_ALT", {value: 31, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_ROI", {value: 80, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_PATHPLANNING", {value: 81, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_SPLINE_WAYPOINT", {value: 82, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_VTOL_TAKEOFF", {value: 84, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_VTOL_LAND", {value: 85, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_GUIDED_ENABLE", {value: 92, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_DELAY", {value: 93, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_PAYLOAD_PLACE", {value: 94, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_LAST", {value: 95, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_SET_YAW_SPEED", {value: 213, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_FENCE_RETURN_POINT", {value: 5000, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_FENCE_POLYGON_VERTEX_INCLUSION", {value: 5001, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_FENCE_POLYGON_VERTEX_EXCLUSION", {value: 5002, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_FENCE_CIRCLE_INCLUSION", {value: 5003, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_FENCE_CIRCLE_EXCLUSION", {value: 5004, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "NAV_RALLY_POINT", {value: 5100, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "OBLIQUE_SURVEY", {value: 260, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "OVERRIDE_GOTO", {value: 252, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "PANORAMA_CREATE", {value: 2800, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "PAYLOAD_PREPARE_DEPLOY", {value: 30001, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "PAYLOAD_CONTROL_DEPLOY", {value: 30002, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "PREFLIGHT_CALIBRATION", {value: 241, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "PREFLIGHT_SET_SENSOR_OFFSETS", {value: 242, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "PREFLIGHT_UAVCAN", {value: 243, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "PREFLIGHT_STORAGE", {value: 245, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "PREFLIGHT_REBOOT_SHUTDOWN", {value: 246, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "REQUEST_MESSAGE", {value: 512, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "REQUEST_PROTOCOL_VERSION", {value: 519, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "REQUEST_AUTOPILOT_CAPABILITIES", {value: 520, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "REQUEST_CAMERA_INFORMATION", {value: 521, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "REQUEST_CAMERA_SETTINGS", {value: 522, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "REQUEST_STORAGE_INFORMATION", {value: 525, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "REQUEST_CAMERA_CAPTURE_STATUS", {value: 527, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "REQUEST_FLIGHT_INFORMATION", {value: 528, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "REQUEST_VIDEO_STREAM_INFORMATION", {value: 2504, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "REQUEST_VIDEO_STREAM_STATUS", {value: 2505, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "RESET_CAMERA_SETTINGS", {value: 529, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "RUN_PREARM_CHECKS", {value: 401, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "SET_MESSAGE_INTERVAL", {value: 511, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "SET_CAMERA_MODE", {value: 530, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "SET_CAMERA_ZOOM", {value: 531, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "SET_CAMERA_FOCUS", {value: 532, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "SET_STORAGE_USAGE", {value: 533, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "SET_CAMERA_SOURCE", {value: 534, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "SET_GUIDED_SUBMODE_STANDARD", {value: 4000, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "SET_GUIDED_SUBMODE_CIRCLE", {value: 4001, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "START_RX_PAIR", {value: 500, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "STORAGE_FORMAT", {value: 526, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "UAVCAN_GET_NODE_INFO", {value: 5200, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "VIDEO_START_CAPTURE", {value: 2500, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "VIDEO_STOP_CAPTURE", {value: 2501, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "VIDEO_START_STREAMING", {value: 2502, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandCodeWrapper, "VIDEO_STOP_STREAMING", {value: 2503, writable: false, enumerable: true, configurable: true});
module.exports = CommandCodeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint16",
      "name": "AIRFRAME_CONFIGURATION",
      "value": 2520
    },
    {
      "type": "uint16",
      "name": "ARM_AUTHORIZATION_REQUEST",
      "value": 3001
    },
    {
      "type": "uint16",
      "name": "CAMERA_TRACK_POINT",
      "value": 2004
    },
    {
      "type": "uint16",
      "name": "CAMERA_TRACK_RECTANGLE",
      "value": 2005
    },
    {
      "type": "uint16",
      "name": "CAMERA_STOP_TRACKING",
      "value": 2010
    },
    {
      "type": "uint16",
      "name": "CAN_FORWARD",
      "value": 32000
    },
    {
      "type": "uint16",
      "name": "COMPONENT_ARM_DISARM",
      "value": 400
    },
    {
      "type": "uint16",
      "name": "CONDITION_DELAY",
      "value": 112
    },
    {
      "type": "uint16",
      "name": "CONDITION_CHANGE_ALT",
      "value": 113
    },
    {
      "type": "uint16",
      "name": "CONDITION_DISTANCE",
      "value": 114
    },
    {
      "type": "uint16",
      "name": "CONDITION_YAW",
      "value": 115
    },
    {
      "type": "uint16",
      "name": "CONDITION_LAST",
      "value": 159
    },
    {
      "type": "uint16",
      "name": "CONTROL_HIGH_LATENCY",
      "value": 2600
    },
    {
      "type": "uint16",
      "name": "DO_FOLLOW",
      "value": 32
    },
    {
      "type": "uint16",
      "name": "DO_FOLLOW_REPOSITION",
      "value": 33
    },
    {
      "type": "uint16",
      "name": "DO_SET_MODE",
      "value": 176
    },
    {
      "type": "uint16",
      "name": "DO_JUMP",
      "value": 177
    },
    {
      "type": "uint16",
      "name": "DO_CHANGE_SPEED",
      "value": 178
    },
    {
      "type": "uint16",
      "name": "DO_SET_HOME",
      "value": 179
    },
    {
      "type": "uint16",
      "name": "DO_SET_PARAMETER",
      "value": 180
    },
    {
      "type": "uint16",
      "name": "DO_SET_RELAY",
      "value": 181
    },
    {
      "type": "uint16",
      "name": "DO_REPEAT_RELAY",
      "value": 182
    },
    {
      "type": "uint16",
      "name": "DO_SET_SERVO",
      "value": 183
    },
    {
      "type": "uint16",
      "name": "DO_REPEAT_SERVO",
      "value": 184
    },
    {
      "type": "uint16",
      "name": "DO_FLIGHTTERMINATION",
      "value": 185
    },
    {
      "type": "uint16",
      "name": "DO_CHANGE_ALTITUDE",
      "value": 186
    },
    {
      "type": "uint16",
      "name": "DO_RETURN_PATH_START",
      "value": 188
    },
    {
      "type": "uint16",
      "name": "DO_LAND_START",
      "value": 189
    },
    {
      "type": "uint16",
      "name": "DO_RALLY_LAND",
      "value": 190
    },
    {
      "type": "uint16",
      "name": "DO_GO_AROUND",
      "value": 191
    },
    {
      "type": "uint16",
      "name": "DO_REPOSITION",
      "value": 192
    },
    {
      "type": "uint16",
      "name": "DO_PAUSE_CONTINUE",
      "value": 193
    },
    {
      "type": "uint16",
      "name": "DO_SET_REVERSE",
      "value": 194
    },
    {
      "type": "uint16",
      "name": "DO_SET_ROI_LOCATION",
      "value": 195
    },
    {
      "type": "uint16",
      "name": "DO_SET_ROI_WPNEXT_OFFSET",
      "value": 196
    },
    {
      "type": "uint16",
      "name": "DO_SET_ROI_NONE",
      "value": 197
    },
    {
      "type": "uint16",
      "name": "DO_SET_ROI_SYSID",
      "value": 198
    },
    {
      "type": "uint16",
      "name": "DO_CONTROL_VIDEO",
      "value": 200
    },
    {
      "type": "uint16",
      "name": "DO_SET_ROI",
      "value": 201
    },
    {
      "type": "uint16",
      "name": "DO_DIGICAM_CONFIGURE",
      "value": 202
    },
    {
      "type": "uint16",
      "name": "DO_DIGICAM_CONTROL",
      "value": 203
    },
    {
      "type": "uint16",
      "name": "DO_MOUNT_CONFIGURE",
      "value": 204
    },
    {
      "type": "uint16",
      "name": "DO_MOUNT_CONTROL",
      "value": 205
    },
    {
      "type": "uint16",
      "name": "DO_SET_CAM_TRIGG_DIST",
      "value": 206
    },
    {
      "type": "uint16",
      "name": "DO_FENCE_ENABLE",
      "value": 207
    },
    {
      "type": "uint16",
      "name": "DO_PARACHUTE",
      "value": 208
    },
    {
      "type": "uint16",
      "name": "DO_MOTOR_TEST",
      "value": 209
    },
    {
      "type": "uint16",
      "name": "DO_INVERTED_FLIGHT",
      "value": 210
    },
    {
      "type": "uint16",
      "name": "DO_GRIPPER",
      "value": 211
    },
    {
      "type": "uint16",
      "name": "DO_AUTOTUNE_ENABLE",
      "value": 212
    },
    {
      "type": "uint16",
      "name": "DO_SET_CAM_TRIGG_INTERVAL",
      "value": 214
    },
    {
      "type": "uint16",
      "name": "DO_MOUNT_CONTROL_QUAT",
      "value": 220
    },
    {
      "type": "uint16",
      "name": "DO_GUIDED_MASTER",
      "value": 221
    },
    {
      "type": "uint16",
      "name": "DO_GUIDED_LIMITS",
      "value": 222
    },
    {
      "type": "uint16",
      "name": "DO_ENGINE_CONTROL",
      "value": 223
    },
    {
      "type": "uint16",
      "name": "DO_SET_MISSION_CURRENT",
      "value": 224
    },
    {
      "type": "uint16",
      "name": "DO_LAST",
      "value": 240
    },
    {
      "type": "uint16",
      "name": "DO_JUMP_TAG",
      "value": 601
    },
    {
      "type": "uint16",
      "name": "DO_GIMBAL_MANAGER_PITCHYAW",
      "value": 1000
    },
    {
      "type": "uint16",
      "name": "DO_GIMBAL_MANAGER_CONFIGURE",
      "value": 1001
    },
    {
      "type": "uint16",
      "name": "DO_TRIGGER_CONTROL",
      "value": 2003
    },
    {
      "type": "uint16",
      "name": "DO_VTOL_TRANSITION",
      "value": 3000
    },
    {
      "type": "uint16",
      "name": "DO_SET_SAFETY_SWITCH_STATE",
      "value": 5300
    },
    {
      "type": "uint16",
      "name": "DO_ADSB_OUT_IDENT",
      "value": 10001
    },
    {
      "type": "uint16",
      "name": "DO_WINCH",
      "value": 42600
    },
    {
      "type": "uint16",
      "name": "EXTERNAL_POSITION_ESTIMATE",
      "value": 43003
    },
    {
      "type": "uint16",
      "name": "FIXED_MAG_CAL_YAW",
      "value": 42006
    },
    {
      "type": "uint16",
      "name": "GET_HOME_POSITION",
      "value": 410
    },
    {
      "type": "uint16",
      "name": "GET_MESSAGE_INTERVAL",
      "value": 510
    },
    {
      "type": "uint16",
      "name": "IMAGE_START_CAPTURE",
      "value": 2000
    },
    {
      "type": "uint16",
      "name": "IMAGE_STOP_CAPTURE",
      "value": 2001
    },
    {
      "type": "uint16",
      "name": "JUMP_TAG",
      "value": 600
    },
    {
      "type": "uint16",
      "name": "LOGGING_START",
      "value": 2510
    },
    {
      "type": "uint16",
      "name": "LOGGING_STOP",
      "value": 2511
    },
    {
      "type": "uint16",
      "name": "MISSION_START",
      "value": 300
    },
    {
      "type": "uint16",
      "name": "NAV_WAYPOINT",
      "value": 16
    },
    {
      "type": "uint16",
      "name": "NAV_LOITER_UNLIM",
      "value": 17
    },
    {
      "type": "uint16",
      "name": "NAV_LOITER_TURNS",
      "value": 18
    },
    {
      "type": "uint16",
      "name": "NAV_LOITER_TIME",
      "value": 19
    },
    {
      "type": "uint16",
      "name": "NAV_RETURN_TO_LAUNCH",
      "value": 20
    },
    {
      "type": "uint16",
      "name": "NAV_LAND",
      "value": 21
    },
    {
      "type": "uint16",
      "name": "NAV_TAKEOFF",
      "value": 22
    },
    {
      "type": "uint16",
      "name": "NAV_LAND_LOCAL",
      "value": 23
    },
    {
      "type": "uint16",
      "name": "NAV_TAKEOFF_LOCAL",
      "value": 24
    },
    {
      "type": "uint16",
      "name": "NAV_FOLLOW",
      "value": 25
    },
    {
      "type": "uint16",
      "name": "NAV_CONTINUE_AND_CHANGE_ALT",
      "value": 30
    },
    {
      "type": "uint16",
      "name": "NAV_LOITER_TO_ALT",
      "value": 31
    },
    {
      "type": "uint16",
      "name": "NAV_ROI",
      "value": 80
    },
    {
      "type": "uint16",
      "name": "NAV_PATHPLANNING",
      "value": 81
    },
    {
      "type": "uint16",
      "name": "NAV_SPLINE_WAYPOINT",
      "value": 82
    },
    {
      "type": "uint16",
      "name": "NAV_VTOL_TAKEOFF",
      "value": 84
    },
    {
      "type": "uint16",
      "name": "NAV_VTOL_LAND",
      "value": 85
    },
    {
      "type": "uint16",
      "name": "NAV_GUIDED_ENABLE",
      "value": 92
    },
    {
      "type": "uint16",
      "name": "NAV_DELAY",
      "value": 93
    },
    {
      "type": "uint16",
      "name": "NAV_PAYLOAD_PLACE",
      "value": 94
    },
    {
      "type": "uint16",
      "name": "NAV_LAST",
      "value": 95
    },
    {
      "type": "uint16",
      "name": "NAV_SET_YAW_SPEED",
      "value": 213
    },
    {
      "type": "uint16",
      "name": "NAV_FENCE_RETURN_POINT",
      "value": 5000
    },
    {
      "type": "uint16",
      "name": "NAV_FENCE_POLYGON_VERTEX_INCLUSION",
      "value": 5001
    },
    {
      "type": "uint16",
      "name": "NAV_FENCE_POLYGON_VERTEX_EXCLUSION",
      "value": 5002
    },
    {
      "type": "uint16",
      "name": "NAV_FENCE_CIRCLE_INCLUSION",
      "value": 5003
    },
    {
      "type": "uint16",
      "name": "NAV_FENCE_CIRCLE_EXCLUSION",
      "value": 5004
    },
    {
      "type": "uint16",
      "name": "NAV_RALLY_POINT",
      "value": 5100
    },
    {
      "type": "uint16",
      "name": "OBLIQUE_SURVEY",
      "value": 260
    },
    {
      "type": "uint16",
      "name": "OVERRIDE_GOTO",
      "value": 252
    },
    {
      "type": "uint16",
      "name": "PANORAMA_CREATE",
      "value": 2800
    },
    {
      "type": "uint16",
      "name": "PAYLOAD_PREPARE_DEPLOY",
      "value": 30001
    },
    {
      "type": "uint16",
      "name": "PAYLOAD_CONTROL_DEPLOY",
      "value": 30002
    },
    {
      "type": "uint16",
      "name": "PREFLIGHT_CALIBRATION",
      "value": 241
    },
    {
      "type": "uint16",
      "name": "PREFLIGHT_SET_SENSOR_OFFSETS",
      "value": 242
    },
    {
      "type": "uint16",
      "name": "PREFLIGHT_UAVCAN",
      "value": 243
    },
    {
      "type": "uint16",
      "name": "PREFLIGHT_STORAGE",
      "value": 245
    },
    {
      "type": "uint16",
      "name": "PREFLIGHT_REBOOT_SHUTDOWN",
      "value": 246
    },
    {
      "type": "uint16",
      "name": "REQUEST_MESSAGE",
      "value": 512
    },
    {
      "type": "uint16",
      "name": "REQUEST_PROTOCOL_VERSION",
      "value": 519
    },
    {
      "type": "uint16",
      "name": "REQUEST_AUTOPILOT_CAPABILITIES",
      "value": 520
    },
    {
      "type": "uint16",
      "name": "REQUEST_CAMERA_INFORMATION",
      "value": 521
    },
    {
      "type": "uint16",
      "name": "REQUEST_CAMERA_SETTINGS",
      "value": 522
    },
    {
      "type": "uint16",
      "name": "REQUEST_STORAGE_INFORMATION",
      "value": 525
    },
    {
      "type": "uint16",
      "name": "REQUEST_CAMERA_CAPTURE_STATUS",
      "value": 527
    },
    {
      "type": "uint16",
      "name": "REQUEST_FLIGHT_INFORMATION",
      "value": 528
    },
    {
      "type": "uint16",
      "name": "REQUEST_VIDEO_STREAM_INFORMATION",
      "value": 2504
    },
    {
      "type": "uint16",
      "name": "REQUEST_VIDEO_STREAM_STATUS",
      "value": 2505
    },
    {
      "type": "uint16",
      "name": "RESET_CAMERA_SETTINGS",
      "value": 529
    },
    {
      "type": "uint16",
      "name": "RUN_PREARM_CHECKS",
      "value": 401
    },
    {
      "type": "uint16",
      "name": "SET_MESSAGE_INTERVAL",
      "value": 511
    },
    {
      "type": "uint16",
      "name": "SET_CAMERA_MODE",
      "value": 530
    },
    {
      "type": "uint16",
      "name": "SET_CAMERA_ZOOM",
      "value": 531
    },
    {
      "type": "uint16",
      "name": "SET_CAMERA_FOCUS",
      "value": 532
    },
    {
      "type": "uint16",
      "name": "SET_STORAGE_USAGE",
      "value": 533
    },
    {
      "type": "uint16",
      "name": "SET_CAMERA_SOURCE",
      "value": 534
    },
    {
      "type": "uint16",
      "name": "SET_GUIDED_SUBMODE_STANDARD",
      "value": 4000
    },
    {
      "type": "uint16",
      "name": "SET_GUIDED_SUBMODE_CIRCLE",
      "value": 4001
    },
    {
      "type": "uint16",
      "name": "START_RX_PAIR",
      "value": 500
    },
    {
      "type": "uint16",
      "name": "STORAGE_FORMAT",
      "value": 526
    },
    {
      "type": "uint16",
      "name": "UAVCAN_GET_NODE_INFO",
      "value": 5200
    },
    {
      "type": "uint16",
      "name": "VIDEO_START_CAPTURE",
      "value": 2500
    },
    {
      "type": "uint16",
      "name": "VIDEO_STOP_CAPTURE",
      "value": 2501
    },
    {
      "type": "uint16",
      "name": "VIDEO_START_STREAMING",
      "value": 2502
    },
    {
      "type": "uint16",
      "name": "VIDEO_STOP_STREAMING",
      "value": 2503
    }
  ],
  "fields": [],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "CommandCode",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CommandCode"
}
*/
