// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__GimbalManagerPitchyaw_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GimbalManagerPitchyaw_RequestRefStruct = StructType({
  pitch: primitiveTypes.float32,
  yaw: primitiveTypes.float32,
  pitch_rate: primitiveTypes.float32,
  yaw_rate: primitiveTypes.float32,
  flags: primitiveTypes.uint32,
  gimbal_device_id: primitiveTypes.uint8,
});
const GimbalManagerPitchyaw_RequestRefArray = ArrayType(GimbalManagerPitchyaw_RequestRefStruct);
const GimbalManagerPitchyaw_RequestRefStructArray = StructType({
  data: GimbalManagerPitchyaw_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GimbalManagerPitchyaw_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GimbalManagerPitchyaw_RequestRefStruct(other._refObject.toObject());
      this._pitchIntialized = true;
      this._yawIntialized = true;
      this._pitch_rateIntialized = true;
      this._yaw_rateIntialized = true;
      this._flagsIntialized = true;
      this._gimbal_device_idIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GimbalManagerPitchyaw_RequestRefStruct();
    this._pitchIntialized = false;
    this._yawIntialized = false;
    this._pitch_rateIntialized = false;
    this._yaw_rateIntialized = false;
    this._flagsIntialized = false;
    this._gimbal_device_idIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GimbalManagerPitchyaw_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GimbalManagerPitchyaw_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GimbalManagerPitchyaw_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GimbalManagerPitchyaw_RequestRefStructArray
  }
  static get refObjectType() {
    return GimbalManagerPitchyaw_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._pitchIntialized) {
        throw new TypeError('Invalid argument: pitch in GimbalManagerPitchyaw_Request');
      }
      if (!this._yawIntialized) {
        throw new TypeError('Invalid argument: yaw in GimbalManagerPitchyaw_Request');
      }
      if (!this._pitch_rateIntialized) {
        throw new TypeError('Invalid argument: pitch_rate in GimbalManagerPitchyaw_Request');
      }
      if (!this._yaw_rateIntialized) {
        throw new TypeError('Invalid argument: yaw_rate in GimbalManagerPitchyaw_Request');
      }
      if (!this._flagsIntialized) {
        throw new TypeError('Invalid argument: flags in GimbalManagerPitchyaw_Request');
      }
      if (!this._gimbal_device_idIntialized) {
        throw new TypeError('Invalid argument: gimbal_device_id in GimbalManagerPitchyaw_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._pitchIntialized = true;
    this._refObject.pitch = refObject.pitch;
    this._yawIntialized = true;
    this._refObject.yaw = refObject.yaw;
    this._pitch_rateIntialized = true;
    this._refObject.pitch_rate = refObject.pitch_rate;
    this._yaw_rateIntialized = true;
    this._refObject.yaw_rate = refObject.yaw_rate;
    this._flagsIntialized = true;
    this._refObject.flags = refObject.flags;
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = refObject.gimbal_device_id;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GimbalManagerPitchyaw_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'GimbalManagerPitchyaw_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pitch() {
    if (!this._pitchIntialized) {
      return undefined;
    }
    return this._refObject.pitch;
  }
  set pitch(value) {
    this._pitchIntialized = true;
    this._refObject.pitch = value;
  }
  get yaw() {
    if (!this._yawIntialized) {
      return undefined;
    }
    return this._refObject.yaw;
  }
  set yaw(value) {
    this._yawIntialized = true;
    this._refObject.yaw = value;
  }
  get pitch_rate() {
    if (!this._pitch_rateIntialized) {
      return undefined;
    }
    return this._refObject.pitch_rate;
  }
  set pitch_rate(value) {
    this._pitch_rateIntialized = true;
    this._refObject.pitch_rate = value;
  }
  get yaw_rate() {
    if (!this._yaw_rateIntialized) {
      return undefined;
    }
    return this._refObject.yaw_rate;
  }
  set yaw_rate(value) {
    this._yaw_rateIntialized = true;
    this._refObject.yaw_rate = value;
  }
  get flags() {
    if (!this._flagsIntialized) {
      return undefined;
    }
    return this._refObject.flags;
  }
  set flags(value) {
    this._flagsIntialized = true;
    this._refObject.flags = value;
  }
  get gimbal_device_id() {
    if (!this._gimbal_device_idIntialized) {
      return undefined;
    }
    return this._refObject.gimbal_device_id;
  }
  set gimbal_device_id(value) {
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GimbalManagerPitchyaw_RequestRefStruct(refObject.toObject());
    this._pitchIntialized = true;
    this._yawIntialized = true;
    this._pitch_rateIntialized = true;
    this._yaw_rateIntialized = true;
    this._flagsIntialized = true;
    this._gimbal_device_idIntialized = true;
  }
  copy(other) {
    this._refObject = new GimbalManagerPitchyaw_RequestRefStruct(other._refObject.toObject());
    this._pitchIntialized = true;
    this._yawIntialized = true;
    this._pitch_rateIntialized = true;
    this._yaw_rateIntialized = true;
    this._flagsIntialized = true;
    this._gimbal_device_idIntialized = true;
  }
  static get classType() {
    return GimbalManagerPitchyaw_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_RETRACT","value":1},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_NEUTRAL","value":2},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_ROLL_LOCK","value":4},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_PITCH_LOCK","value":8},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_YAW_LOCK","value":16}],"fields":[{"name":"pitch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gimbal_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"GimbalManagerPitchyaw_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GimbalManagerPitchyaw_Request"};
  }
  hasMember(name) {
    let memberNames = ["pitch","yaw","pitch_rate","yaw_rate","flags","gimbal_device_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GimbalManagerPitchyaw_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GimbalManagerPitchyaw_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GimbalManagerPitchyaw_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerPitchyaw_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerPitchyaw_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GimbalManagerPitchyaw_RequestRefArray(size);
    this._refObject = new GimbalManagerPitchyaw_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GimbalManagerPitchyaw_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GimbalManagerPitchyaw_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GimbalManagerPitchyaw_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GimbalManagerPitchyaw_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GimbalManagerPitchyaw_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GimbalManagerPitchyaw_RequestArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(GimbalManagerPitchyaw_RequestWrapper, "GIMBAL_MANAGER_FLAGS_RETRACT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerPitchyaw_RequestWrapper, "GIMBAL_MANAGER_FLAGS_NEUTRAL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerPitchyaw_RequestWrapper, "GIMBAL_MANAGER_FLAGS_ROLL_LOCK", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerPitchyaw_RequestWrapper, "GIMBAL_MANAGER_FLAGS_PITCH_LOCK", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerPitchyaw_RequestWrapper, "GIMBAL_MANAGER_FLAGS_YAW_LOCK", {value: 16, writable: false, enumerable: true, configurable: true});
module.exports = GimbalManagerPitchyaw_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_RETRACT",
      "value": 1
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_NEUTRAL",
      "value": 2
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_ROLL_LOCK",
      "value": 4
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_PITCH_LOCK",
      "value": 8
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_YAW_LOCK",
      "value": 16
    }
  ],
  "fields": [
    {
      "name": "pitch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gimbal_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "GimbalManagerPitchyaw_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GimbalManagerPitchyaw_Request"
}
*/
