// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__ParamEvent.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ParameterValueWrapper = require('../../generated/rcl_interfaces/rcl_interfaces__msg__ParameterValue.js');
const ParamEventRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  param_id: primitiveTypes.string,
  value: ParameterValueWrapper.refObjectType,
  param_index: primitiveTypes.uint16,
  param_count: primitiveTypes.uint16,
});
const ParamEventRefArray = ArrayType(ParamEventRefStruct);
const ParamEventRefStructArray = StructType({
  data: ParamEventRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ParamEventWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ParamEventRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._param_idIntialized = true;
      this._wrapperFields.param_id =  new StringWrapper(other._wrapperFields.param_id);
      this._wrapperFields.value =  new ParameterValueWrapper(other._wrapperFields.value);
      this._param_indexIntialized = true;
      this._param_countIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ParamEventRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._param_idIntialized = false;
    this._wrapperFields.param_id = new StringWrapper();
    this._wrapperFields.value = new ParameterValueWrapper();
    this._param_indexIntialized = false;
    this._param_countIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ParamEventWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ParamEventArrayWrapper;
  }
  static get ArrayType() {
    return ParamEventArrayWrapper;
  }
  static get refObjectArrayType() {
    return ParamEventRefStructArray
  }
  static get refObjectType() {
    return ParamEventRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._param_idIntialized) {
        throw new TypeError('Invalid argument: param_id in ParamEvent');
      }
      if (!this._param_indexIntialized) {
        throw new TypeError('Invalid argument: param_index in ParamEvent');
      }
      if (!this._param_countIntialized) {
        throw new TypeError('Invalid argument: param_count in ParamEvent');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.param_id.freeze(own, checkConsistency);
    }
    this._refObject.param_id = this._wrapperFields.param_id.refObject;
    this._wrapperFields.value.freeze(own, checkConsistency);
    this._refObject.value = this._wrapperFields.value.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._param_idIntialized = true;
    this._wrapperFields.param_id.data = refObject.param_id.data;
    this._wrapperFields.value.copyRefObject(refObject.value);
    this._param_indexIntialized = true;
    this._refObject.param_index = refObject.param_index;
    this._param_countIntialized = true;
    this._refObject.param_count = refObject.param_count;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.param_id);
    ParameterValueWrapper.freeStruct(refObject.value);
  }
  static destoryRawROS(msg) {
    ParamEventWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'ParamEvent'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get param_id() {
    if (!this._param_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.param_id.data;
  }
  set param_id(value) {
    this._param_idIntialized = true;
    this._wrapperFields.param_id.data = value;
  }
  get value() {
    return this._wrapperFields.value;
  }
  set value(value) {
    if (value instanceof ParameterValueWrapper) {
      this._wrapperFields.value.copy(value);
    } else {
      this._wrapperFields.value.copy(new ParameterValueWrapper(value));
    }
  }
  get param_index() {
    if (!this._param_indexIntialized) {
      return undefined;
    }
    return this._refObject.param_index;
  }
  set param_index(value) {
    this._param_indexIntialized = true;
    this._refObject.param_index = value;
  }
  get param_count() {
    if (!this._param_countIntialized) {
      return undefined;
    }
    return this._refObject.param_count;
  }
  set param_count(value) {
    this._param_countIntialized = true;
    this._refObject.param_count = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ParamEventRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._param_idIntialized = true;
    this._wrapperFields.param_id.copyRefObject(this._refObject.param_id);
    this._wrapperFields.value.copyRefObject(this._refObject.value);
    this._param_indexIntialized = true;
    this._param_countIntialized = true;
  }
  copy(other) {
    this._refObject = new ParamEventRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._param_idIntialized = true;
    this._wrapperFields.param_id.copy(other._wrapperFields.param_id);
    this._wrapperFields.value.copy(other._wrapperFields.value);
    this._param_indexIntialized = true;
    this._param_countIntialized = true;
  }
  static get classType() {
    return ParamEventWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"param_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"rcl_interfaces","type":"ParameterValue","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"param_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"ParamEvent","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ParamEvent"};
  }
  hasMember(name) {
    let memberNames = ["header","param_id","value","param_index","param_count"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ParamEventArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ParamEventWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ParamEventWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParamEventArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParamEventArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ParamEventRefArray(size);
    this._refObject = new ParamEventRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ParamEventWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ParamEventArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ParamEventArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ParamEventWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ParamEventWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ParamEventArrayWrapper;
  }
}
module.exports = ParamEventWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "param_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "rcl_interfaces",
        "type": "ParameterValue",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "param_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "ParamEvent",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ParamEvent"
}
*/
