// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__Vibration.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const VibrationRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  vibration: Vector3Wrapper.refObjectType,
  clipping: ArrayType(primitiveTypes.float32, 3),
});
const VibrationRefArray = ArrayType(VibrationRefStruct);
const VibrationRefStructArray = StructType({
  data: VibrationRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VibrationWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VibrationRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.vibration =  new Vector3Wrapper(other._wrapperFields.vibration);
      this._wrapperFields.clipping = Float32Wrapper.createArray();
      this._wrapperFields.clipping.copy(other._wrapperFields.clipping);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VibrationRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.vibration = new Vector3Wrapper();
    this._wrapperFields.clipping = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new VibrationWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VibrationArrayWrapper;
  }
  static get ArrayType() {
    return VibrationArrayWrapper;
  }
  static get refObjectArrayType() {
    return VibrationRefStructArray
  }
  static get refObjectType() {
    return VibrationRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.vibration.freeze(own, checkConsistency);
    this._refObject.vibration = this._wrapperFields.vibration.refObject;
    for (let i = 0; i < this._wrapperFields.clipping.data.length; i++) {
      this._refObject.clipping[i] = this._wrapperFields.clipping.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.vibration.copyRefObject(refObject.vibration);
    this._refObject.clipping = refObject.clipping;
    this._wrapperFields.clipping.fill(refObject.clipping.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    Vector3Wrapper.freeStruct(refObject.vibration);
  }
  static destoryRawROS(msg) {
    VibrationWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'Vibration'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get vibration() {
    return this._wrapperFields.vibration;
  }
  set vibration(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.vibration.copy(value);
    } else {
      this._wrapperFields.vibration.copy(new Vector3Wrapper(value));
    }
  }
  get clipping() {
    return this._wrapperFields['clipping'].data;
  }
  set clipping(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['clipping'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new VibrationRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.vibration.copyRefObject(this._refObject.vibration);
    this._wrapperFields.clipping.fill(refObject.clipping.toArray());
  }
  copy(other) {
    this._refObject = new VibrationRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.vibration.copy(other._wrapperFields.vibration);
    this._wrapperFields.clipping.copy(other._wrapperFields.clipping);
  }
  static get classType() {
    return VibrationWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"vibration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"clipping","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"Vibration","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Vibration"};
  }
  hasMember(name) {
    let memberNames = ["header","vibration","clipping"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VibrationArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VibrationWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VibrationWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VibrationArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VibrationArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VibrationRefArray(size);
    this._refObject = new VibrationRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VibrationWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VibrationArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VibrationArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VibrationWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VibrationWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VibrationArrayWrapper;
  }
}
module.exports = VibrationWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "vibration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "clipping",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "Vibration",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Vibration"
}
*/
