// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__HilStateQuaternion.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const QuaternionWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Quaternion.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const GeoPointWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__GeoPoint.js');
const HilStateQuaternionRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  orientation: QuaternionWrapper.refObjectType,
  angular_velocity: Vector3Wrapper.refObjectType,
  linear_acceleration: Vector3Wrapper.refObjectType,
  linear_velocity: Vector3Wrapper.refObjectType,
  geo: GeoPointWrapper.refObjectType,
  ind_airspeed: primitiveTypes.float32,
  true_airspeed: primitiveTypes.float32,
});
const HilStateQuaternionRefArray = ArrayType(HilStateQuaternionRefStruct);
const HilStateQuaternionRefStructArray = StructType({
  data: HilStateQuaternionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HilStateQuaternionWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HilStateQuaternionRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.orientation =  new QuaternionWrapper(other._wrapperFields.orientation);
      this._wrapperFields.angular_velocity =  new Vector3Wrapper(other._wrapperFields.angular_velocity);
      this._wrapperFields.linear_acceleration =  new Vector3Wrapper(other._wrapperFields.linear_acceleration);
      this._wrapperFields.linear_velocity =  new Vector3Wrapper(other._wrapperFields.linear_velocity);
      this._wrapperFields.geo =  new GeoPointWrapper(other._wrapperFields.geo);
      this._ind_airspeedIntialized = true;
      this._true_airspeedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HilStateQuaternionRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.orientation = new QuaternionWrapper();
    this._wrapperFields.angular_velocity = new Vector3Wrapper();
    this._wrapperFields.linear_acceleration = new Vector3Wrapper();
    this._wrapperFields.linear_velocity = new Vector3Wrapper();
    this._wrapperFields.geo = new GeoPointWrapper();
    this._ind_airspeedIntialized = false;
    this._true_airspeedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new HilStateQuaternionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HilStateQuaternionArrayWrapper;
  }
  static get ArrayType() {
    return HilStateQuaternionArrayWrapper;
  }
  static get refObjectArrayType() {
    return HilStateQuaternionRefStructArray
  }
  static get refObjectType() {
    return HilStateQuaternionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._ind_airspeedIntialized) {
        throw new TypeError('Invalid argument: ind_airspeed in HilStateQuaternion');
      }
      if (!this._true_airspeedIntialized) {
        throw new TypeError('Invalid argument: true_airspeed in HilStateQuaternion');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.orientation.freeze(own, checkConsistency);
    this._refObject.orientation = this._wrapperFields.orientation.refObject;
    this._wrapperFields.angular_velocity.freeze(own, checkConsistency);
    this._refObject.angular_velocity = this._wrapperFields.angular_velocity.refObject;
    this._wrapperFields.linear_acceleration.freeze(own, checkConsistency);
    this._refObject.linear_acceleration = this._wrapperFields.linear_acceleration.refObject;
    this._wrapperFields.linear_velocity.freeze(own, checkConsistency);
    this._refObject.linear_velocity = this._wrapperFields.linear_velocity.refObject;
    this._wrapperFields.geo.freeze(own, checkConsistency);
    this._refObject.geo = this._wrapperFields.geo.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.orientation.copyRefObject(refObject.orientation);
    this._wrapperFields.angular_velocity.copyRefObject(refObject.angular_velocity);
    this._wrapperFields.linear_acceleration.copyRefObject(refObject.linear_acceleration);
    this._wrapperFields.linear_velocity.copyRefObject(refObject.linear_velocity);
    this._wrapperFields.geo.copyRefObject(refObject.geo);
    this._ind_airspeedIntialized = true;
    this._refObject.ind_airspeed = refObject.ind_airspeed;
    this._true_airspeedIntialized = true;
    this._refObject.true_airspeed = refObject.true_airspeed;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    QuaternionWrapper.freeStruct(refObject.orientation);
    Vector3Wrapper.freeStruct(refObject.angular_velocity);
    Vector3Wrapper.freeStruct(refObject.linear_acceleration);
    Vector3Wrapper.freeStruct(refObject.linear_velocity);
    GeoPointWrapper.freeStruct(refObject.geo);
  }
  static destoryRawROS(msg) {
    HilStateQuaternionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'HilStateQuaternion'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get orientation() {
    return this._wrapperFields.orientation;
  }
  set orientation(value) {
    if (value instanceof QuaternionWrapper) {
      this._wrapperFields.orientation.copy(value);
    } else {
      this._wrapperFields.orientation.copy(new QuaternionWrapper(value));
    }
  }
  get angular_velocity() {
    return this._wrapperFields.angular_velocity;
  }
  set angular_velocity(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.angular_velocity.copy(value);
    } else {
      this._wrapperFields.angular_velocity.copy(new Vector3Wrapper(value));
    }
  }
  get linear_acceleration() {
    return this._wrapperFields.linear_acceleration;
  }
  set linear_acceleration(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.linear_acceleration.copy(value);
    } else {
      this._wrapperFields.linear_acceleration.copy(new Vector3Wrapper(value));
    }
  }
  get linear_velocity() {
    return this._wrapperFields.linear_velocity;
  }
  set linear_velocity(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.linear_velocity.copy(value);
    } else {
      this._wrapperFields.linear_velocity.copy(new Vector3Wrapper(value));
    }
  }
  get geo() {
    return this._wrapperFields.geo;
  }
  set geo(value) {
    if (value instanceof GeoPointWrapper) {
      this._wrapperFields.geo.copy(value);
    } else {
      this._wrapperFields.geo.copy(new GeoPointWrapper(value));
    }
  }
  get ind_airspeed() {
    if (!this._ind_airspeedIntialized) {
      return undefined;
    }
    return this._refObject.ind_airspeed;
  }
  set ind_airspeed(value) {
    this._ind_airspeedIntialized = true;
    this._refObject.ind_airspeed = value;
  }
  get true_airspeed() {
    if (!this._true_airspeedIntialized) {
      return undefined;
    }
    return this._refObject.true_airspeed;
  }
  set true_airspeed(value) {
    this._true_airspeedIntialized = true;
    this._refObject.true_airspeed = value;
  }
  copyRefObject(refObject) {
    this._refObject = new HilStateQuaternionRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.orientation.copyRefObject(this._refObject.orientation);
    this._wrapperFields.angular_velocity.copyRefObject(this._refObject.angular_velocity);
    this._wrapperFields.linear_acceleration.copyRefObject(this._refObject.linear_acceleration);
    this._wrapperFields.linear_velocity.copyRefObject(this._refObject.linear_velocity);
    this._wrapperFields.geo.copyRefObject(this._refObject.geo);
    this._ind_airspeedIntialized = true;
    this._true_airspeedIntialized = true;
  }
  copy(other) {
    this._refObject = new HilStateQuaternionRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.orientation.copy(other._wrapperFields.orientation);
    this._wrapperFields.angular_velocity.copy(other._wrapperFields.angular_velocity);
    this._wrapperFields.linear_acceleration.copy(other._wrapperFields.linear_acceleration);
    this._wrapperFields.linear_velocity.copy(other._wrapperFields.linear_velocity);
    this._wrapperFields.geo.copy(other._wrapperFields.geo);
    this._ind_airspeedIntialized = true;
    this._true_airspeedIntialized = true;
  }
  static get classType() {
    return HilStateQuaternionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"orientation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Quaternion","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"angular_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"linear_acceleration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"linear_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"geo","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"GeoPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"ind_airspeed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"true_airspeed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"HilStateQuaternion","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HilStateQuaternion"};
  }
  hasMember(name) {
    let memberNames = ["header","orientation","angular_velocity","linear_acceleration","linear_velocity","geo","ind_airspeed","true_airspeed"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HilStateQuaternionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HilStateQuaternionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HilStateQuaternionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HilStateQuaternionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HilStateQuaternionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HilStateQuaternionRefArray(size);
    this._refObject = new HilStateQuaternionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HilStateQuaternionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HilStateQuaternionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HilStateQuaternionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HilStateQuaternionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HilStateQuaternionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HilStateQuaternionArrayWrapper;
  }
}
module.exports = HilStateQuaternionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "orientation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Quaternion",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "angular_velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "linear_acceleration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "linear_velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "geo",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "GeoPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "ind_airspeed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "true_airspeed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "HilStateQuaternion",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HilStateQuaternion"
}
*/
