// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__GimbalManagerCameraTrack_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GimbalManagerCameraTrack_RequestRefStruct = StructType({
  mode: primitiveTypes.uint8,
  x: primitiveTypes.float32,
  y: primitiveTypes.float32,
  radius: primitiveTypes.float32,
  top_left_x: primitiveTypes.float32,
  top_left_y: primitiveTypes.float32,
  bottom_right_x: primitiveTypes.float32,
  bottom_right_y: primitiveTypes.float32,
});
const GimbalManagerCameraTrack_RequestRefArray = ArrayType(GimbalManagerCameraTrack_RequestRefStruct);
const GimbalManagerCameraTrack_RequestRefStructArray = StructType({
  data: GimbalManagerCameraTrack_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GimbalManagerCameraTrack_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GimbalManagerCameraTrack_RequestRefStruct(other._refObject.toObject());
      this._modeIntialized = true;
      this._xIntialized = true;
      this._yIntialized = true;
      this._radiusIntialized = true;
      this._top_left_xIntialized = true;
      this._top_left_yIntialized = true;
      this._bottom_right_xIntialized = true;
      this._bottom_right_yIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GimbalManagerCameraTrack_RequestRefStruct();
    this._modeIntialized = false;
    this._xIntialized = false;
    this._yIntialized = false;
    this._radiusIntialized = false;
    this._top_left_xIntialized = false;
    this._top_left_yIntialized = false;
    this._bottom_right_xIntialized = false;
    this._bottom_right_yIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GimbalManagerCameraTrack_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GimbalManagerCameraTrack_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GimbalManagerCameraTrack_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GimbalManagerCameraTrack_RequestRefStructArray
  }
  static get refObjectType() {
    return GimbalManagerCameraTrack_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in GimbalManagerCameraTrack_Request');
      }
      if (!this._xIntialized) {
        throw new TypeError('Invalid argument: x in GimbalManagerCameraTrack_Request');
      }
      if (!this._yIntialized) {
        throw new TypeError('Invalid argument: y in GimbalManagerCameraTrack_Request');
      }
      if (!this._radiusIntialized) {
        throw new TypeError('Invalid argument: radius in GimbalManagerCameraTrack_Request');
      }
      if (!this._top_left_xIntialized) {
        throw new TypeError('Invalid argument: top_left_x in GimbalManagerCameraTrack_Request');
      }
      if (!this._top_left_yIntialized) {
        throw new TypeError('Invalid argument: top_left_y in GimbalManagerCameraTrack_Request');
      }
      if (!this._bottom_right_xIntialized) {
        throw new TypeError('Invalid argument: bottom_right_x in GimbalManagerCameraTrack_Request');
      }
      if (!this._bottom_right_yIntialized) {
        throw new TypeError('Invalid argument: bottom_right_y in GimbalManagerCameraTrack_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._modeIntialized = true;
    this._refObject.mode = refObject.mode;
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._radiusIntialized = true;
    this._refObject.radius = refObject.radius;
    this._top_left_xIntialized = true;
    this._refObject.top_left_x = refObject.top_left_x;
    this._top_left_yIntialized = true;
    this._refObject.top_left_y = refObject.top_left_y;
    this._bottom_right_xIntialized = true;
    this._refObject.bottom_right_x = refObject.bottom_right_x;
    this._bottom_right_yIntialized = true;
    this._refObject.bottom_right_y = refObject.bottom_right_y;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GimbalManagerCameraTrack_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'GimbalManagerCameraTrack_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._refObject.mode;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._refObject.mode = value;
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get radius() {
    if (!this._radiusIntialized) {
      return undefined;
    }
    return this._refObject.radius;
  }
  set radius(value) {
    this._radiusIntialized = true;
    this._refObject.radius = value;
  }
  get top_left_x() {
    if (!this._top_left_xIntialized) {
      return undefined;
    }
    return this._refObject.top_left_x;
  }
  set top_left_x(value) {
    this._top_left_xIntialized = true;
    this._refObject.top_left_x = value;
  }
  get top_left_y() {
    if (!this._top_left_yIntialized) {
      return undefined;
    }
    return this._refObject.top_left_y;
  }
  set top_left_y(value) {
    this._top_left_yIntialized = true;
    this._refObject.top_left_y = value;
  }
  get bottom_right_x() {
    if (!this._bottom_right_xIntialized) {
      return undefined;
    }
    return this._refObject.bottom_right_x;
  }
  set bottom_right_x(value) {
    this._bottom_right_xIntialized = true;
    this._refObject.bottom_right_x = value;
  }
  get bottom_right_y() {
    if (!this._bottom_right_yIntialized) {
      return undefined;
    }
    return this._refObject.bottom_right_y;
  }
  set bottom_right_y(value) {
    this._bottom_right_yIntialized = true;
    this._refObject.bottom_right_y = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GimbalManagerCameraTrack_RequestRefStruct(refObject.toObject());
    this._modeIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._radiusIntialized = true;
    this._top_left_xIntialized = true;
    this._top_left_yIntialized = true;
    this._bottom_right_xIntialized = true;
    this._bottom_right_yIntialized = true;
  }
  copy(other) {
    this._refObject = new GimbalManagerCameraTrack_RequestRefStruct(other._refObject.toObject());
    this._modeIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._radiusIntialized = true;
    this._top_left_xIntialized = true;
    this._top_left_yIntialized = true;
    this._bottom_right_xIntialized = true;
    this._bottom_right_yIntialized = true;
  }
  static get classType() {
    return GimbalManagerCameraTrack_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"CAMERA_TRACK_MODE_POINT","value":0},{"type":"uint8","name":"CAMERA_TRACK_MODE_RECTANGLE","value":1},{"type":"uint8","name":"CAMERA_TRACK_MODE_STOP_TRACKING","value":2}],"fields":[{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"radius","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"top_left_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"top_left_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"bottom_right_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"bottom_right_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"GimbalManagerCameraTrack_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GimbalManagerCameraTrack_Request"};
  }
  hasMember(name) {
    let memberNames = ["mode","x","y","radius","top_left_x","top_left_y","bottom_right_x","bottom_right_y"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GimbalManagerCameraTrack_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GimbalManagerCameraTrack_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GimbalManagerCameraTrack_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerCameraTrack_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerCameraTrack_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GimbalManagerCameraTrack_RequestRefArray(size);
    this._refObject = new GimbalManagerCameraTrack_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GimbalManagerCameraTrack_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GimbalManagerCameraTrack_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GimbalManagerCameraTrack_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GimbalManagerCameraTrack_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GimbalManagerCameraTrack_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GimbalManagerCameraTrack_RequestArrayWrapper;
  }
}
// Define constants (3 in total)
Object.defineProperty(GimbalManagerCameraTrack_RequestWrapper, "CAMERA_TRACK_MODE_POINT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerCameraTrack_RequestWrapper, "CAMERA_TRACK_MODE_RECTANGLE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerCameraTrack_RequestWrapper, "CAMERA_TRACK_MODE_STOP_TRACKING", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = GimbalManagerCameraTrack_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "CAMERA_TRACK_MODE_POINT",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "CAMERA_TRACK_MODE_RECTANGLE",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "CAMERA_TRACK_MODE_STOP_TRACKING",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "radius",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "top_left_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "top_left_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "bottom_right_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "bottom_right_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "GimbalManagerCameraTrack_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GimbalManagerCameraTrack_Request"
}
*/
