// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__HilSensor.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const HilSensorRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  acc: Vector3Wrapper.refObjectType,
  gyro: Vector3Wrapper.refObjectType,
  mag: Vector3Wrapper.refObjectType,
  abs_pressure: primitiveTypes.float32,
  diff_pressure: primitiveTypes.float32,
  pressure_alt: primitiveTypes.float32,
  temperature: primitiveTypes.float32,
  fields_updated: primitiveTypes.uint32,
});
const HilSensorRefArray = ArrayType(HilSensorRefStruct);
const HilSensorRefStructArray = StructType({
  data: HilSensorRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HilSensorWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HilSensorRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.acc =  new Vector3Wrapper(other._wrapperFields.acc);
      this._wrapperFields.gyro =  new Vector3Wrapper(other._wrapperFields.gyro);
      this._wrapperFields.mag =  new Vector3Wrapper(other._wrapperFields.mag);
      this._abs_pressureIntialized = true;
      this._diff_pressureIntialized = true;
      this._pressure_altIntialized = true;
      this._temperatureIntialized = true;
      this._fields_updatedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HilSensorRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.acc = new Vector3Wrapper();
    this._wrapperFields.gyro = new Vector3Wrapper();
    this._wrapperFields.mag = new Vector3Wrapper();
    this._abs_pressureIntialized = false;
    this._diff_pressureIntialized = false;
    this._pressure_altIntialized = false;
    this._temperatureIntialized = false;
    this._fields_updatedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new HilSensorWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HilSensorArrayWrapper;
  }
  static get ArrayType() {
    return HilSensorArrayWrapper;
  }
  static get refObjectArrayType() {
    return HilSensorRefStructArray
  }
  static get refObjectType() {
    return HilSensorRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._abs_pressureIntialized) {
        throw new TypeError('Invalid argument: abs_pressure in HilSensor');
      }
      if (!this._diff_pressureIntialized) {
        throw new TypeError('Invalid argument: diff_pressure in HilSensor');
      }
      if (!this._pressure_altIntialized) {
        throw new TypeError('Invalid argument: pressure_alt in HilSensor');
      }
      if (!this._temperatureIntialized) {
        throw new TypeError('Invalid argument: temperature in HilSensor');
      }
      if (!this._fields_updatedIntialized) {
        throw new TypeError('Invalid argument: fields_updated in HilSensor');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.acc.freeze(own, checkConsistency);
    this._refObject.acc = this._wrapperFields.acc.refObject;
    this._wrapperFields.gyro.freeze(own, checkConsistency);
    this._refObject.gyro = this._wrapperFields.gyro.refObject;
    this._wrapperFields.mag.freeze(own, checkConsistency);
    this._refObject.mag = this._wrapperFields.mag.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.acc.copyRefObject(refObject.acc);
    this._wrapperFields.gyro.copyRefObject(refObject.gyro);
    this._wrapperFields.mag.copyRefObject(refObject.mag);
    this._abs_pressureIntialized = true;
    this._refObject.abs_pressure = refObject.abs_pressure;
    this._diff_pressureIntialized = true;
    this._refObject.diff_pressure = refObject.diff_pressure;
    this._pressure_altIntialized = true;
    this._refObject.pressure_alt = refObject.pressure_alt;
    this._temperatureIntialized = true;
    this._refObject.temperature = refObject.temperature;
    this._fields_updatedIntialized = true;
    this._refObject.fields_updated = refObject.fields_updated;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    Vector3Wrapper.freeStruct(refObject.acc);
    Vector3Wrapper.freeStruct(refObject.gyro);
    Vector3Wrapper.freeStruct(refObject.mag);
  }
  static destoryRawROS(msg) {
    HilSensorWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'HilSensor'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get acc() {
    return this._wrapperFields.acc;
  }
  set acc(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.acc.copy(value);
    } else {
      this._wrapperFields.acc.copy(new Vector3Wrapper(value));
    }
  }
  get gyro() {
    return this._wrapperFields.gyro;
  }
  set gyro(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.gyro.copy(value);
    } else {
      this._wrapperFields.gyro.copy(new Vector3Wrapper(value));
    }
  }
  get mag() {
    return this._wrapperFields.mag;
  }
  set mag(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.mag.copy(value);
    } else {
      this._wrapperFields.mag.copy(new Vector3Wrapper(value));
    }
  }
  get abs_pressure() {
    if (!this._abs_pressureIntialized) {
      return undefined;
    }
    return this._refObject.abs_pressure;
  }
  set abs_pressure(value) {
    this._abs_pressureIntialized = true;
    this._refObject.abs_pressure = value;
  }
  get diff_pressure() {
    if (!this._diff_pressureIntialized) {
      return undefined;
    }
    return this._refObject.diff_pressure;
  }
  set diff_pressure(value) {
    this._diff_pressureIntialized = true;
    this._refObject.diff_pressure = value;
  }
  get pressure_alt() {
    if (!this._pressure_altIntialized) {
      return undefined;
    }
    return this._refObject.pressure_alt;
  }
  set pressure_alt(value) {
    this._pressure_altIntialized = true;
    this._refObject.pressure_alt = value;
  }
  get temperature() {
    if (!this._temperatureIntialized) {
      return undefined;
    }
    return this._refObject.temperature;
  }
  set temperature(value) {
    this._temperatureIntialized = true;
    this._refObject.temperature = value;
  }
  get fields_updated() {
    if (!this._fields_updatedIntialized) {
      return undefined;
    }
    return this._refObject.fields_updated;
  }
  set fields_updated(value) {
    this._fields_updatedIntialized = true;
    this._refObject.fields_updated = value;
  }
  copyRefObject(refObject) {
    this._refObject = new HilSensorRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.acc.copyRefObject(this._refObject.acc);
    this._wrapperFields.gyro.copyRefObject(this._refObject.gyro);
    this._wrapperFields.mag.copyRefObject(this._refObject.mag);
    this._abs_pressureIntialized = true;
    this._diff_pressureIntialized = true;
    this._pressure_altIntialized = true;
    this._temperatureIntialized = true;
    this._fields_updatedIntialized = true;
  }
  copy(other) {
    this._refObject = new HilSensorRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.acc.copy(other._wrapperFields.acc);
    this._wrapperFields.gyro.copy(other._wrapperFields.gyro);
    this._wrapperFields.mag.copy(other._wrapperFields.mag);
    this._abs_pressureIntialized = true;
    this._diff_pressureIntialized = true;
    this._pressure_altIntialized = true;
    this._temperatureIntialized = true;
    this._fields_updatedIntialized = true;
  }
  static get classType() {
    return HilSensorWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"acc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"gyro","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"mag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"abs_pressure","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"diff_pressure","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pressure_alt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"temperature","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fields_updated","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"HilSensor","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HilSensor"};
  }
  hasMember(name) {
    let memberNames = ["header","acc","gyro","mag","abs_pressure","diff_pressure","pressure_alt","temperature","fields_updated"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HilSensorArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HilSensorWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HilSensorWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HilSensorArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HilSensorArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HilSensorRefArray(size);
    this._refObject = new HilSensorRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HilSensorWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HilSensorArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HilSensorArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HilSensorWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HilSensorWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HilSensorArrayWrapper;
  }
}
module.exports = HilSensorWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "acc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "gyro",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "mag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "abs_pressure",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "diff_pressure",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pressure_alt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "temperature",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fields_updated",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "HilSensor",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HilSensor"
}
*/
