// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__WaypointPush_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const WaypointWrapper = require('../../generated/mavros_msgs/mavros_msgs__msg__Waypoint.js');
const WaypointPush_RequestRefStruct = StructType({
  start_index: primitiveTypes.uint16,
  waypoints: WaypointWrapper.refObjectArrayType,
});
const WaypointPush_RequestRefArray = ArrayType(WaypointPush_RequestRefStruct);
const WaypointPush_RequestRefStructArray = StructType({
  data: WaypointPush_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WaypointPush_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WaypointPush_RequestRefStruct(other._refObject.toObject());
      this._start_indexIntialized = true;
      this._wrapperFields.waypoints = WaypointWrapper.createArray();
      this._wrapperFields.waypoints.copy(other._wrapperFields.waypoints);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WaypointPush_RequestRefStruct();
    this._start_indexIntialized = false;
    this._wrapperFields.waypoints = WaypointWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new WaypointPush_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WaypointPush_RequestArrayWrapper;
  }
  static get ArrayType() {
    return WaypointPush_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return WaypointPush_RequestRefStructArray
  }
  static get refObjectType() {
    return WaypointPush_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._start_indexIntialized) {
        throw new TypeError('Invalid argument: start_index in WaypointPush_Request');
      }
    }
    this._wrapperFields.waypoints.freeze(own, checkConsistency);
    this._refObject.waypoints = this._wrapperFields.waypoints.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._start_indexIntialized = true;
    this._refObject.start_index = refObject.start_index;
    this._wrapperFields.waypoints.copyRefObject(refObject.waypoints);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.waypoints.size != 0) {
      WaypointWrapper.ArrayType.freeArray(refObject.waypoints);
      if (WaypointWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.waypoints, WaypointWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    WaypointPush_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'WaypointPush_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get start_index() {
    if (!this._start_indexIntialized) {
      return undefined;
    }
    return this._refObject.start_index;
  }
  set start_index(value) {
    this._start_indexIntialized = true;
    this._refObject.start_index = value;
  }
  get waypoints() {
    return this._wrapperFields.waypoints;
  }
  set waypoints(value) {
    this._wrapperFields.waypoints.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new WaypointPush_RequestRefStruct(refObject.toObject());
    this._start_indexIntialized = true;
    this._wrapperFields.waypoints.copyRefObject(this._refObject.waypoints);
  }
  copy(other) {
    this._refObject = new WaypointPush_RequestRefStruct(other._refObject.toObject());
    this._start_indexIntialized = true;
    this._wrapperFields.waypoints.copy(other._wrapperFields.waypoints);
  }
  static get classType() {
    return WaypointPush_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"start_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"waypoints","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"mavros_msgs","type":"Waypoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"WaypointPush_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"WaypointPush_Request"};
  }
  hasMember(name) {
    let memberNames = ["start_index","waypoints"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WaypointPush_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WaypointPush_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WaypointPush_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WaypointPush_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WaypointPush_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WaypointPush_RequestRefArray(size);
    this._refObject = new WaypointPush_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WaypointPush_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WaypointPush_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WaypointPush_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WaypointPush_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WaypointPush_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WaypointPush_RequestArrayWrapper;
  }
}
module.exports = WaypointPush_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "start_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "waypoints",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "mavros_msgs",
        "type": "Waypoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "WaypointPush_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "WaypointPush_Request"
}
*/
