// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__RadioStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const RadioStatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  rssi: primitiveTypes.uint8,
  remrssi: primitiveTypes.uint8,
  txbuf: primitiveTypes.uint8,
  noise: primitiveTypes.uint8,
  remnoise: primitiveTypes.uint8,
  rxerrors: primitiveTypes.uint16,
  fixed: primitiveTypes.uint16,
  rssi_dbm: primitiveTypes.float32,
  remrssi_dbm: primitiveTypes.float32,
});
const RadioStatusRefArray = ArrayType(RadioStatusRefStruct);
const RadioStatusRefStructArray = StructType({
  data: RadioStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RadioStatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RadioStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._rssiIntialized = true;
      this._remrssiIntialized = true;
      this._txbufIntialized = true;
      this._noiseIntialized = true;
      this._remnoiseIntialized = true;
      this._rxerrorsIntialized = true;
      this._fixedIntialized = true;
      this._rssi_dbmIntialized = true;
      this._remrssi_dbmIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RadioStatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._rssiIntialized = false;
    this._remrssiIntialized = false;
    this._txbufIntialized = false;
    this._noiseIntialized = false;
    this._remnoiseIntialized = false;
    this._rxerrorsIntialized = false;
    this._fixedIntialized = false;
    this._rssi_dbmIntialized = false;
    this._remrssi_dbmIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new RadioStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RadioStatusArrayWrapper;
  }
  static get ArrayType() {
    return RadioStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return RadioStatusRefStructArray
  }
  static get refObjectType() {
    return RadioStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._rssiIntialized) {
        throw new TypeError('Invalid argument: rssi in RadioStatus');
      }
      if (!this._remrssiIntialized) {
        throw new TypeError('Invalid argument: remrssi in RadioStatus');
      }
      if (!this._txbufIntialized) {
        throw new TypeError('Invalid argument: txbuf in RadioStatus');
      }
      if (!this._noiseIntialized) {
        throw new TypeError('Invalid argument: noise in RadioStatus');
      }
      if (!this._remnoiseIntialized) {
        throw new TypeError('Invalid argument: remnoise in RadioStatus');
      }
      if (!this._rxerrorsIntialized) {
        throw new TypeError('Invalid argument: rxerrors in RadioStatus');
      }
      if (!this._fixedIntialized) {
        throw new TypeError('Invalid argument: fixed in RadioStatus');
      }
      if (!this._rssi_dbmIntialized) {
        throw new TypeError('Invalid argument: rssi_dbm in RadioStatus');
      }
      if (!this._remrssi_dbmIntialized) {
        throw new TypeError('Invalid argument: remrssi_dbm in RadioStatus');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._rssiIntialized = true;
    this._refObject.rssi = refObject.rssi;
    this._remrssiIntialized = true;
    this._refObject.remrssi = refObject.remrssi;
    this._txbufIntialized = true;
    this._refObject.txbuf = refObject.txbuf;
    this._noiseIntialized = true;
    this._refObject.noise = refObject.noise;
    this._remnoiseIntialized = true;
    this._refObject.remnoise = refObject.remnoise;
    this._rxerrorsIntialized = true;
    this._refObject.rxerrors = refObject.rxerrors;
    this._fixedIntialized = true;
    this._refObject.fixed = refObject.fixed;
    this._rssi_dbmIntialized = true;
    this._refObject.rssi_dbm = refObject.rssi_dbm;
    this._remrssi_dbmIntialized = true;
    this._refObject.remrssi_dbm = refObject.remrssi_dbm;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    RadioStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'RadioStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get rssi() {
    if (!this._rssiIntialized) {
      return undefined;
    }
    return this._refObject.rssi;
  }
  set rssi(value) {
    this._rssiIntialized = true;
    this._refObject.rssi = value;
  }
  get remrssi() {
    if (!this._remrssiIntialized) {
      return undefined;
    }
    return this._refObject.remrssi;
  }
  set remrssi(value) {
    this._remrssiIntialized = true;
    this._refObject.remrssi = value;
  }
  get txbuf() {
    if (!this._txbufIntialized) {
      return undefined;
    }
    return this._refObject.txbuf;
  }
  set txbuf(value) {
    this._txbufIntialized = true;
    this._refObject.txbuf = value;
  }
  get noise() {
    if (!this._noiseIntialized) {
      return undefined;
    }
    return this._refObject.noise;
  }
  set noise(value) {
    this._noiseIntialized = true;
    this._refObject.noise = value;
  }
  get remnoise() {
    if (!this._remnoiseIntialized) {
      return undefined;
    }
    return this._refObject.remnoise;
  }
  set remnoise(value) {
    this._remnoiseIntialized = true;
    this._refObject.remnoise = value;
  }
  get rxerrors() {
    if (!this._rxerrorsIntialized) {
      return undefined;
    }
    return this._refObject.rxerrors;
  }
  set rxerrors(value) {
    this._rxerrorsIntialized = true;
    this._refObject.rxerrors = value;
  }
  get fixed() {
    if (!this._fixedIntialized) {
      return undefined;
    }
    return this._refObject.fixed;
  }
  set fixed(value) {
    this._fixedIntialized = true;
    this._refObject.fixed = value;
  }
  get rssi_dbm() {
    if (!this._rssi_dbmIntialized) {
      return undefined;
    }
    return this._refObject.rssi_dbm;
  }
  set rssi_dbm(value) {
    this._rssi_dbmIntialized = true;
    this._refObject.rssi_dbm = value;
  }
  get remrssi_dbm() {
    if (!this._remrssi_dbmIntialized) {
      return undefined;
    }
    return this._refObject.remrssi_dbm;
  }
  set remrssi_dbm(value) {
    this._remrssi_dbmIntialized = true;
    this._refObject.remrssi_dbm = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RadioStatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._rssiIntialized = true;
    this._remrssiIntialized = true;
    this._txbufIntialized = true;
    this._noiseIntialized = true;
    this._remnoiseIntialized = true;
    this._rxerrorsIntialized = true;
    this._fixedIntialized = true;
    this._rssi_dbmIntialized = true;
    this._remrssi_dbmIntialized = true;
  }
  copy(other) {
    this._refObject = new RadioStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._rssiIntialized = true;
    this._remrssiIntialized = true;
    this._txbufIntialized = true;
    this._noiseIntialized = true;
    this._remnoiseIntialized = true;
    this._rxerrorsIntialized = true;
    this._fixedIntialized = true;
    this._rssi_dbmIntialized = true;
    this._remrssi_dbmIntialized = true;
  }
  static get classType() {
    return RadioStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"rssi","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"remrssi","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"txbuf","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"noise","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"remnoise","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rxerrors","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fixed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rssi_dbm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"remrssi_dbm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"RadioStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RadioStatus"};
  }
  hasMember(name) {
    let memberNames = ["header","rssi","remrssi","txbuf","noise","remnoise","rxerrors","fixed","rssi_dbm","remrssi_dbm"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RadioStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RadioStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RadioStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RadioStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RadioStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RadioStatusRefArray(size);
    this._refObject = new RadioStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RadioStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RadioStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RadioStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RadioStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RadioStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RadioStatusArrayWrapper;
  }
}
module.exports = RadioStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "rssi",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "remrssi",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "txbuf",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "noise",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "remnoise",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rxerrors",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fixed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rssi_dbm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "remrssi_dbm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "RadioStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RadioStatus"
}
*/
