// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__Altitude.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const AltitudeRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  monotonic: primitiveTypes.float32,
  amsl: primitiveTypes.float32,
  local: primitiveTypes.float32,
  relative: primitiveTypes.float32,
  terrain: primitiveTypes.float32,
  bottom_clearance: primitiveTypes.float32,
});
const AltitudeRefArray = ArrayType(AltitudeRefStruct);
const AltitudeRefStructArray = StructType({
  data: AltitudeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AltitudeWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AltitudeRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._monotonicIntialized = true;
      this._amslIntialized = true;
      this._localIntialized = true;
      this._relativeIntialized = true;
      this._terrainIntialized = true;
      this._bottom_clearanceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AltitudeRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._monotonicIntialized = false;
    this._amslIntialized = false;
    this._localIntialized = false;
    this._relativeIntialized = false;
    this._terrainIntialized = false;
    this._bottom_clearanceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new AltitudeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AltitudeArrayWrapper;
  }
  static get ArrayType() {
    return AltitudeArrayWrapper;
  }
  static get refObjectArrayType() {
    return AltitudeRefStructArray
  }
  static get refObjectType() {
    return AltitudeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._monotonicIntialized) {
        throw new TypeError('Invalid argument: monotonic in Altitude');
      }
      if (!this._amslIntialized) {
        throw new TypeError('Invalid argument: amsl in Altitude');
      }
      if (!this._localIntialized) {
        throw new TypeError('Invalid argument: local in Altitude');
      }
      if (!this._relativeIntialized) {
        throw new TypeError('Invalid argument: relative in Altitude');
      }
      if (!this._terrainIntialized) {
        throw new TypeError('Invalid argument: terrain in Altitude');
      }
      if (!this._bottom_clearanceIntialized) {
        throw new TypeError('Invalid argument: bottom_clearance in Altitude');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._monotonicIntialized = true;
    this._refObject.monotonic = refObject.monotonic;
    this._amslIntialized = true;
    this._refObject.amsl = refObject.amsl;
    this._localIntialized = true;
    this._refObject.local = refObject.local;
    this._relativeIntialized = true;
    this._refObject.relative = refObject.relative;
    this._terrainIntialized = true;
    this._refObject.terrain = refObject.terrain;
    this._bottom_clearanceIntialized = true;
    this._refObject.bottom_clearance = refObject.bottom_clearance;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    AltitudeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'Altitude'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get monotonic() {
    if (!this._monotonicIntialized) {
      return undefined;
    }
    return this._refObject.monotonic;
  }
  set monotonic(value) {
    this._monotonicIntialized = true;
    this._refObject.monotonic = value;
  }
  get amsl() {
    if (!this._amslIntialized) {
      return undefined;
    }
    return this._refObject.amsl;
  }
  set amsl(value) {
    this._amslIntialized = true;
    this._refObject.amsl = value;
  }
  get local() {
    if (!this._localIntialized) {
      return undefined;
    }
    return this._refObject.local;
  }
  set local(value) {
    this._localIntialized = true;
    this._refObject.local = value;
  }
  get relative() {
    if (!this._relativeIntialized) {
      return undefined;
    }
    return this._refObject.relative;
  }
  set relative(value) {
    this._relativeIntialized = true;
    this._refObject.relative = value;
  }
  get terrain() {
    if (!this._terrainIntialized) {
      return undefined;
    }
    return this._refObject.terrain;
  }
  set terrain(value) {
    this._terrainIntialized = true;
    this._refObject.terrain = value;
  }
  get bottom_clearance() {
    if (!this._bottom_clearanceIntialized) {
      return undefined;
    }
    return this._refObject.bottom_clearance;
  }
  set bottom_clearance(value) {
    this._bottom_clearanceIntialized = true;
    this._refObject.bottom_clearance = value;
  }
  copyRefObject(refObject) {
    this._refObject = new AltitudeRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._monotonicIntialized = true;
    this._amslIntialized = true;
    this._localIntialized = true;
    this._relativeIntialized = true;
    this._terrainIntialized = true;
    this._bottom_clearanceIntialized = true;
  }
  copy(other) {
    this._refObject = new AltitudeRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._monotonicIntialized = true;
    this._amslIntialized = true;
    this._localIntialized = true;
    this._relativeIntialized = true;
    this._terrainIntialized = true;
    this._bottom_clearanceIntialized = true;
  }
  static get classType() {
    return AltitudeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"monotonic","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"amsl","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"local","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"relative","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"terrain","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"bottom_clearance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"Altitude","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Altitude"};
  }
  hasMember(name) {
    let memberNames = ["header","monotonic","amsl","local","relative","terrain","bottom_clearance"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AltitudeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AltitudeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AltitudeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AltitudeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AltitudeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AltitudeRefArray(size);
    this._refObject = new AltitudeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AltitudeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AltitudeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AltitudeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AltitudeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AltitudeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AltitudeArrayWrapper;
  }
}
module.exports = AltitudeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "monotonic",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "amsl",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "local",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "relative",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "terrain",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "bottom_clearance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "Altitude",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Altitude"
}
*/
