// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__FileList_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const FileList_RequestRefStruct = StructType({
  dir_path: primitiveTypes.string,
});
const FileList_RequestRefArray = ArrayType(FileList_RequestRefStruct);
const FileList_RequestRefStructArray = StructType({
  data: FileList_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FileList_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FileList_RequestRefStruct(other._refObject.toObject());
      this._dir_pathIntialized = true;
      this._wrapperFields.dir_path =  new StringWrapper(other._wrapperFields.dir_path);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FileList_RequestRefStruct();
    this._dir_pathIntialized = false;
    this._wrapperFields.dir_path = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FileList_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FileList_RequestArrayWrapper;
  }
  static get ArrayType() {
    return FileList_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return FileList_RequestRefStructArray
  }
  static get refObjectType() {
    return FileList_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._dir_pathIntialized) {
        throw new TypeError('Invalid argument: dir_path in FileList_Request');
      }
    }
    if (own) {
      this._wrapperFields.dir_path.freeze(own, checkConsistency);
    }
    this._refObject.dir_path = this._wrapperFields.dir_path.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._dir_pathIntialized = true;
    this._wrapperFields.dir_path.data = refObject.dir_path.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.dir_path);
  }
  static destoryRawROS(msg) {
    FileList_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'FileList_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get dir_path() {
    if (!this._dir_pathIntialized) {
      return undefined;
    }
    return this._wrapperFields.dir_path.data;
  }
  set dir_path(value) {
    this._dir_pathIntialized = true;
    this._wrapperFields.dir_path.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FileList_RequestRefStruct(refObject.toObject());
    this._dir_pathIntialized = true;
    this._wrapperFields.dir_path.copyRefObject(this._refObject.dir_path);
  }
  copy(other) {
    this._refObject = new FileList_RequestRefStruct(other._refObject.toObject());
    this._dir_pathIntialized = true;
    this._wrapperFields.dir_path.copy(other._wrapperFields.dir_path);
  }
  static get classType() {
    return FileList_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"dir_path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"FileList_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FileList_Request"};
  }
  hasMember(name) {
    let memberNames = ["dir_path"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FileList_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FileList_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FileList_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FileList_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FileList_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FileList_RequestRefArray(size);
    this._refObject = new FileList_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FileList_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FileList_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FileList_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FileList_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FileList_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FileList_RequestArrayWrapper;
  }
}
module.exports = FileList_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "dir_path",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "FileList_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FileList_Request"
}
*/
