// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__StreamRate_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StreamRate_RequestRefStruct = StructType({
  stream_id: primitiveTypes.uint8,
  message_rate: primitiveTypes.uint16,
  on_off: primitiveTypes.bool,
});
const StreamRate_RequestRefArray = ArrayType(StreamRate_RequestRefStruct);
const StreamRate_RequestRefStructArray = StructType({
  data: StreamRate_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StreamRate_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StreamRate_RequestRefStruct(other._refObject.toObject());
      this._stream_idIntialized = true;
      this._message_rateIntialized = true;
      this._on_offIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StreamRate_RequestRefStruct();
    this._stream_idIntialized = false;
    this._message_rateIntialized = false;
    this._on_offIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new StreamRate_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StreamRate_RequestArrayWrapper;
  }
  static get ArrayType() {
    return StreamRate_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return StreamRate_RequestRefStructArray
  }
  static get refObjectType() {
    return StreamRate_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._stream_idIntialized) {
        throw new TypeError('Invalid argument: stream_id in StreamRate_Request');
      }
      if (!this._message_rateIntialized) {
        throw new TypeError('Invalid argument: message_rate in StreamRate_Request');
      }
      if (!this._on_offIntialized) {
        throw new TypeError('Invalid argument: on_off in StreamRate_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._stream_idIntialized = true;
    this._refObject.stream_id = refObject.stream_id;
    this._message_rateIntialized = true;
    this._refObject.message_rate = refObject.message_rate;
    this._on_offIntialized = true;
    this._refObject.on_off = refObject.on_off;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    StreamRate_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'StreamRate_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get stream_id() {
    if (!this._stream_idIntialized) {
      return undefined;
    }
    return this._refObject.stream_id;
  }
  set stream_id(value) {
    this._stream_idIntialized = true;
    this._refObject.stream_id = value;
  }
  get message_rate() {
    if (!this._message_rateIntialized) {
      return undefined;
    }
    return this._refObject.message_rate;
  }
  set message_rate(value) {
    this._message_rateIntialized = true;
    this._refObject.message_rate = value;
  }
  get on_off() {
    if (!this._on_offIntialized) {
      return undefined;
    }
    return this._refObject.on_off;
  }
  set on_off(value) {
    this._on_offIntialized = true;
    this._refObject.on_off = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StreamRate_RequestRefStruct(refObject.toObject());
    this._stream_idIntialized = true;
    this._message_rateIntialized = true;
    this._on_offIntialized = true;
  }
  copy(other) {
    this._refObject = new StreamRate_RequestRefStruct(other._refObject.toObject());
    this._stream_idIntialized = true;
    this._message_rateIntialized = true;
    this._on_offIntialized = true;
  }
  static get classType() {
    return StreamRate_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"STREAM_ALL","value":0},{"type":"uint8","name":"STREAM_RAW_SENSORS","value":1},{"type":"uint8","name":"STREAM_EXTENDED_STATUS","value":2},{"type":"uint8","name":"STREAM_RC_CHANNELS","value":3},{"type":"uint8","name":"STREAM_RAW_CONTROLLER","value":4},{"type":"uint8","name":"STREAM_POSITION","value":6},{"type":"uint8","name":"STREAM_EXTRA1","value":10},{"type":"uint8","name":"STREAM_EXTRA2","value":11},{"type":"uint8","name":"STREAM_EXTRA3","value":12}],"fields":[{"name":"stream_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"message_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"on_off","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"StreamRate_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StreamRate_Request"};
  }
  hasMember(name) {
    let memberNames = ["stream_id","message_rate","on_off"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StreamRate_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StreamRate_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StreamRate_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StreamRate_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StreamRate_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StreamRate_RequestRefArray(size);
    this._refObject = new StreamRate_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StreamRate_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StreamRate_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StreamRate_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StreamRate_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StreamRate_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StreamRate_RequestArrayWrapper;
  }
}
// Define constants (9 in total)
Object.defineProperty(StreamRate_RequestWrapper, "STREAM_ALL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StreamRate_RequestWrapper, "STREAM_RAW_SENSORS", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StreamRate_RequestWrapper, "STREAM_EXTENDED_STATUS", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StreamRate_RequestWrapper, "STREAM_RC_CHANNELS", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StreamRate_RequestWrapper, "STREAM_RAW_CONTROLLER", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StreamRate_RequestWrapper, "STREAM_POSITION", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StreamRate_RequestWrapper, "STREAM_EXTRA1", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StreamRate_RequestWrapper, "STREAM_EXTRA2", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StreamRate_RequestWrapper, "STREAM_EXTRA3", {value: 12, writable: false, enumerable: true, configurable: true});
module.exports = StreamRate_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "STREAM_ALL",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "STREAM_RAW_SENSORS",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "STREAM_EXTENDED_STATUS",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "STREAM_RC_CHANNELS",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "STREAM_RAW_CONTROLLER",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "STREAM_POSITION",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "STREAM_EXTRA1",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "STREAM_EXTRA2",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "STREAM_EXTRA3",
      "value": 12
    }
  ],
  "fields": [
    {
      "name": "stream_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "message_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "on_off",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "StreamRate_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "StreamRate_Request"
}
*/
