// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__HomePosition.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const GeoPointWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__GeoPoint.js');
const PointWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point.js');
const QuaternionWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Quaternion.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const HomePositionRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  geo: GeoPointWrapper.refObjectType,
  position: PointWrapper.refObjectType,
  orientation: QuaternionWrapper.refObjectType,
  approach: Vector3Wrapper.refObjectType,
});
const HomePositionRefArray = ArrayType(HomePositionRefStruct);
const HomePositionRefStructArray = StructType({
  data: HomePositionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HomePositionWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HomePositionRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.geo =  new GeoPointWrapper(other._wrapperFields.geo);
      this._wrapperFields.position =  new PointWrapper(other._wrapperFields.position);
      this._wrapperFields.orientation =  new QuaternionWrapper(other._wrapperFields.orientation);
      this._wrapperFields.approach =  new Vector3Wrapper(other._wrapperFields.approach);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HomePositionRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.geo = new GeoPointWrapper();
    this._wrapperFields.position = new PointWrapper();
    this._wrapperFields.orientation = new QuaternionWrapper();
    this._wrapperFields.approach = new Vector3Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new HomePositionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HomePositionArrayWrapper;
  }
  static get ArrayType() {
    return HomePositionArrayWrapper;
  }
  static get refObjectArrayType() {
    return HomePositionRefStructArray
  }
  static get refObjectType() {
    return HomePositionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.geo.freeze(own, checkConsistency);
    this._refObject.geo = this._wrapperFields.geo.refObject;
    this._wrapperFields.position.freeze(own, checkConsistency);
    this._refObject.position = this._wrapperFields.position.refObject;
    this._wrapperFields.orientation.freeze(own, checkConsistency);
    this._refObject.orientation = this._wrapperFields.orientation.refObject;
    this._wrapperFields.approach.freeze(own, checkConsistency);
    this._refObject.approach = this._wrapperFields.approach.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.geo.copyRefObject(refObject.geo);
    this._wrapperFields.position.copyRefObject(refObject.position);
    this._wrapperFields.orientation.copyRefObject(refObject.orientation);
    this._wrapperFields.approach.copyRefObject(refObject.approach);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    GeoPointWrapper.freeStruct(refObject.geo);
    PointWrapper.freeStruct(refObject.position);
    QuaternionWrapper.freeStruct(refObject.orientation);
    Vector3Wrapper.freeStruct(refObject.approach);
  }
  static destoryRawROS(msg) {
    HomePositionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'HomePosition'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get geo() {
    return this._wrapperFields.geo;
  }
  set geo(value) {
    if (value instanceof GeoPointWrapper) {
      this._wrapperFields.geo.copy(value);
    } else {
      this._wrapperFields.geo.copy(new GeoPointWrapper(value));
    }
  }
  get position() {
    return this._wrapperFields.position;
  }
  set position(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.position.copy(value);
    } else {
      this._wrapperFields.position.copy(new PointWrapper(value));
    }
  }
  get orientation() {
    return this._wrapperFields.orientation;
  }
  set orientation(value) {
    if (value instanceof QuaternionWrapper) {
      this._wrapperFields.orientation.copy(value);
    } else {
      this._wrapperFields.orientation.copy(new QuaternionWrapper(value));
    }
  }
  get approach() {
    return this._wrapperFields.approach;
  }
  set approach(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.approach.copy(value);
    } else {
      this._wrapperFields.approach.copy(new Vector3Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new HomePositionRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.geo.copyRefObject(this._refObject.geo);
    this._wrapperFields.position.copyRefObject(this._refObject.position);
    this._wrapperFields.orientation.copyRefObject(this._refObject.orientation);
    this._wrapperFields.approach.copyRefObject(this._refObject.approach);
  }
  copy(other) {
    this._refObject = new HomePositionRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.geo.copy(other._wrapperFields.geo);
    this._wrapperFields.position.copy(other._wrapperFields.position);
    this._wrapperFields.orientation.copy(other._wrapperFields.orientation);
    this._wrapperFields.approach.copy(other._wrapperFields.approach);
  }
  static get classType() {
    return HomePositionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"geo","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"GeoPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"orientation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Quaternion","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"approach","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"HomePosition","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HomePosition"};
  }
  hasMember(name) {
    let memberNames = ["header","geo","position","orientation","approach"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HomePositionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HomePositionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HomePositionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HomePositionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HomePositionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HomePositionRefArray(size);
    this._refObject = new HomePositionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HomePositionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HomePositionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HomePositionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HomePositionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HomePositionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HomePositionArrayWrapper;
  }
}
module.exports = HomePositionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "geo",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "GeoPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "position",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "orientation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Quaternion",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "approach",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "HomePosition",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HomePosition"
}
*/
