// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__StatusText.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const StatusTextRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  severity: primitiveTypes.uint8,
  text: primitiveTypes.string,
});
const StatusTextRefArray = ArrayType(StatusTextRefStruct);
const StatusTextRefStructArray = StructType({
  data: StatusTextRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StatusTextWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StatusTextRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._severityIntialized = true;
      this._textIntialized = true;
      this._wrapperFields.text =  new StringWrapper(other._wrapperFields.text);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StatusTextRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._severityIntialized = false;
    this._textIntialized = false;
    this._wrapperFields.text = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new StatusTextWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StatusTextArrayWrapper;
  }
  static get ArrayType() {
    return StatusTextArrayWrapper;
  }
  static get refObjectArrayType() {
    return StatusTextRefStructArray
  }
  static get refObjectType() {
    return StatusTextRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._severityIntialized) {
        throw new TypeError('Invalid argument: severity in StatusText');
      }
      if (!this._textIntialized) {
        throw new TypeError('Invalid argument: text in StatusText');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.text.freeze(own, checkConsistency);
    }
    this._refObject.text = this._wrapperFields.text.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._severityIntialized = true;
    this._refObject.severity = refObject.severity;
    this._textIntialized = true;
    this._wrapperFields.text.data = refObject.text.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.text);
  }
  static destoryRawROS(msg) {
    StatusTextWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'StatusText'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get severity() {
    if (!this._severityIntialized) {
      return undefined;
    }
    return this._refObject.severity;
  }
  set severity(value) {
    this._severityIntialized = true;
    this._refObject.severity = value;
  }
  get text() {
    if (!this._textIntialized) {
      return undefined;
    }
    return this._wrapperFields.text.data;
  }
  set text(value) {
    this._textIntialized = true;
    this._wrapperFields.text.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StatusTextRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._severityIntialized = true;
    this._textIntialized = true;
    this._wrapperFields.text.copyRefObject(this._refObject.text);
  }
  copy(other) {
    this._refObject = new StatusTextRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._severityIntialized = true;
    this._textIntialized = true;
    this._wrapperFields.text.copy(other._wrapperFields.text);
  }
  static get classType() {
    return StatusTextWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"EMERGENCY","value":0},{"type":"uint8","name":"ALERT","value":1},{"type":"uint8","name":"CRITICAL","value":2},{"type":"uint8","name":"ERROR","value":3},{"type":"uint8","name":"WARNING","value":4},{"type":"uint8","name":"NOTICE","value":5},{"type":"uint8","name":"INFO","value":6},{"type":"uint8","name":"DEBUG","value":7}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"severity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"text","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"StatusText","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StatusText"};
  }
  hasMember(name) {
    let memberNames = ["header","severity","text"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StatusTextArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StatusTextWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StatusTextWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatusTextArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatusTextArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StatusTextRefArray(size);
    this._refObject = new StatusTextRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StatusTextWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StatusTextArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StatusTextArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StatusTextWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StatusTextWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StatusTextArrayWrapper;
  }
}
// Define constants (8 in total)
Object.defineProperty(StatusTextWrapper, "EMERGENCY", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusTextWrapper, "ALERT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusTextWrapper, "CRITICAL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusTextWrapper, "ERROR", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusTextWrapper, "WARNING", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusTextWrapper, "NOTICE", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusTextWrapper, "INFO", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusTextWrapper, "DEBUG", {value: 7, writable: false, enumerable: true, configurable: true});
module.exports = StatusTextWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "EMERGENCY",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ALERT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "CRITICAL",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "ERROR",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "WARNING",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "NOTICE",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "INFO",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "DEBUG",
      "value": 7
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "severity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "text",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "StatusText",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "StatusText"
}
*/
