// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__CompanionProcessStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const CompanionProcessStatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  state: primitiveTypes.uint8,
  component: primitiveTypes.uint8,
});
const CompanionProcessStatusRefArray = ArrayType(CompanionProcessStatusRefStruct);
const CompanionProcessStatusRefStructArray = StructType({
  data: CompanionProcessStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CompanionProcessStatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CompanionProcessStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._stateIntialized = true;
      this._componentIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CompanionProcessStatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._stateIntialized = false;
    this._componentIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CompanionProcessStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CompanionProcessStatusArrayWrapper;
  }
  static get ArrayType() {
    return CompanionProcessStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return CompanionProcessStatusRefStructArray
  }
  static get refObjectType() {
    return CompanionProcessStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._stateIntialized) {
        throw new TypeError('Invalid argument: state in CompanionProcessStatus');
      }
      if (!this._componentIntialized) {
        throw new TypeError('Invalid argument: component in CompanionProcessStatus');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._stateIntialized = true;
    this._refObject.state = refObject.state;
    this._componentIntialized = true;
    this._refObject.component = refObject.component;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    CompanionProcessStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'CompanionProcessStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get state() {
    if (!this._stateIntialized) {
      return undefined;
    }
    return this._refObject.state;
  }
  set state(value) {
    this._stateIntialized = true;
    this._refObject.state = value;
  }
  get component() {
    if (!this._componentIntialized) {
      return undefined;
    }
    return this._refObject.component;
  }
  set component(value) {
    this._componentIntialized = true;
    this._refObject.component = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CompanionProcessStatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._stateIntialized = true;
    this._componentIntialized = true;
  }
  copy(other) {
    this._refObject = new CompanionProcessStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._stateIntialized = true;
    this._componentIntialized = true;
  }
  static get classType() {
    return CompanionProcessStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"MAV_STATE_UNINIT","value":0},{"type":"uint8","name":"MAV_STATE_BOOT","value":1},{"type":"uint8","name":"MAV_STATE_CALIBRATING","value":2},{"type":"uint8","name":"MAV_STATE_STANDBY","value":3},{"type":"uint8","name":"MAV_STATE_ACTIVE","value":4},{"type":"uint8","name":"MAV_STATE_CRITICAL","value":5},{"type":"uint8","name":"MAV_STATE_EMERGENCY","value":6},{"type":"uint8","name":"MAV_STATE_POWEROFF","value":7},{"type":"uint8","name":"MAV_STATE_FLIGHT_TERMINATION","value":8},{"type":"uint8","name":"MAV_COMP_ID_OBSTACLE_AVOIDANCE","value":196},{"type":"uint8","name":"MAV_COMP_ID_VISUAL_INERTIAL_ODOMETRY","value":197}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"component","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"CompanionProcessStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CompanionProcessStatus"};
  }
  hasMember(name) {
    let memberNames = ["header","state","component"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CompanionProcessStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CompanionProcessStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CompanionProcessStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CompanionProcessStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CompanionProcessStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CompanionProcessStatusRefArray(size);
    this._refObject = new CompanionProcessStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CompanionProcessStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CompanionProcessStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CompanionProcessStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CompanionProcessStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CompanionProcessStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CompanionProcessStatusArrayWrapper;
  }
}
// Define constants (11 in total)
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_STATE_UNINIT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_STATE_BOOT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_STATE_CALIBRATING", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_STATE_STANDBY", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_STATE_ACTIVE", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_STATE_CRITICAL", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_STATE_EMERGENCY", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_STATE_POWEROFF", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_STATE_FLIGHT_TERMINATION", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_COMP_ID_OBSTACLE_AVOIDANCE", {value: 196, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CompanionProcessStatusWrapper, "MAV_COMP_ID_VISUAL_INERTIAL_ODOMETRY", {value: 197, writable: false, enumerable: true, configurable: true});
module.exports = CompanionProcessStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "MAV_STATE_UNINIT",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "MAV_STATE_BOOT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "MAV_STATE_CALIBRATING",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "MAV_STATE_STANDBY",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "MAV_STATE_ACTIVE",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "MAV_STATE_CRITICAL",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "MAV_STATE_EMERGENCY",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "MAV_STATE_POWEROFF",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "MAV_STATE_FLIGHT_TERMINATION",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "MAV_COMP_ID_OBSTACLE_AVOIDANCE",
      "value": 196
    },
    {
      "type": "uint8",
      "name": "MAV_COMP_ID_VISUAL_INERTIAL_ODOMETRY",
      "value": 197
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "component",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "CompanionProcessStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CompanionProcessStatus"
}
*/
