// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__RTKBaseline.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const RTKBaselineRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  time_last_baseline_ms: primitiveTypes.uint32,
  rtk_receiver_id: primitiveTypes.uint8,
  wn: primitiveTypes.uint16,
  tow: primitiveTypes.uint32,
  rtk_health: primitiveTypes.uint8,
  rtk_rate: primitiveTypes.uint8,
  nsats: primitiveTypes.uint8,
  baseline_coords_type: primitiveTypes.uint8,
  baseline_a_mm: primitiveTypes.int32,
  baseline_b_mm: primitiveTypes.int32,
  baseline_c_mm: primitiveTypes.int32,
  accuracy: primitiveTypes.uint32,
  iar_num_hypotheses: primitiveTypes.int32,
});
const RTKBaselineRefArray = ArrayType(RTKBaselineRefStruct);
const RTKBaselineRefStructArray = StructType({
  data: RTKBaselineRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RTKBaselineWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RTKBaselineRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._time_last_baseline_msIntialized = true;
      this._rtk_receiver_idIntialized = true;
      this._wnIntialized = true;
      this._towIntialized = true;
      this._rtk_healthIntialized = true;
      this._rtk_rateIntialized = true;
      this._nsatsIntialized = true;
      this._baseline_coords_typeIntialized = true;
      this._baseline_a_mmIntialized = true;
      this._baseline_b_mmIntialized = true;
      this._baseline_c_mmIntialized = true;
      this._accuracyIntialized = true;
      this._iar_num_hypothesesIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RTKBaselineRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._time_last_baseline_msIntialized = false;
    this._rtk_receiver_idIntialized = false;
    this._wnIntialized = false;
    this._towIntialized = false;
    this._rtk_healthIntialized = false;
    this._rtk_rateIntialized = false;
    this._nsatsIntialized = false;
    this._baseline_coords_typeIntialized = false;
    this._baseline_a_mmIntialized = false;
    this._baseline_b_mmIntialized = false;
    this._baseline_c_mmIntialized = false;
    this._accuracyIntialized = false;
    this._iar_num_hypothesesIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new RTKBaselineWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RTKBaselineArrayWrapper;
  }
  static get ArrayType() {
    return RTKBaselineArrayWrapper;
  }
  static get refObjectArrayType() {
    return RTKBaselineRefStructArray
  }
  static get refObjectType() {
    return RTKBaselineRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._time_last_baseline_msIntialized) {
        throw new TypeError('Invalid argument: time_last_baseline_ms in RTKBaseline');
      }
      if (!this._rtk_receiver_idIntialized) {
        throw new TypeError('Invalid argument: rtk_receiver_id in RTKBaseline');
      }
      if (!this._wnIntialized) {
        throw new TypeError('Invalid argument: wn in RTKBaseline');
      }
      if (!this._towIntialized) {
        throw new TypeError('Invalid argument: tow in RTKBaseline');
      }
      if (!this._rtk_healthIntialized) {
        throw new TypeError('Invalid argument: rtk_health in RTKBaseline');
      }
      if (!this._rtk_rateIntialized) {
        throw new TypeError('Invalid argument: rtk_rate in RTKBaseline');
      }
      if (!this._nsatsIntialized) {
        throw new TypeError('Invalid argument: nsats in RTKBaseline');
      }
      if (!this._baseline_coords_typeIntialized) {
        throw new TypeError('Invalid argument: baseline_coords_type in RTKBaseline');
      }
      if (!this._baseline_a_mmIntialized) {
        throw new TypeError('Invalid argument: baseline_a_mm in RTKBaseline');
      }
      if (!this._baseline_b_mmIntialized) {
        throw new TypeError('Invalid argument: baseline_b_mm in RTKBaseline');
      }
      if (!this._baseline_c_mmIntialized) {
        throw new TypeError('Invalid argument: baseline_c_mm in RTKBaseline');
      }
      if (!this._accuracyIntialized) {
        throw new TypeError('Invalid argument: accuracy in RTKBaseline');
      }
      if (!this._iar_num_hypothesesIntialized) {
        throw new TypeError('Invalid argument: iar_num_hypotheses in RTKBaseline');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._time_last_baseline_msIntialized = true;
    this._refObject.time_last_baseline_ms = refObject.time_last_baseline_ms;
    this._rtk_receiver_idIntialized = true;
    this._refObject.rtk_receiver_id = refObject.rtk_receiver_id;
    this._wnIntialized = true;
    this._refObject.wn = refObject.wn;
    this._towIntialized = true;
    this._refObject.tow = refObject.tow;
    this._rtk_healthIntialized = true;
    this._refObject.rtk_health = refObject.rtk_health;
    this._rtk_rateIntialized = true;
    this._refObject.rtk_rate = refObject.rtk_rate;
    this._nsatsIntialized = true;
    this._refObject.nsats = refObject.nsats;
    this._baseline_coords_typeIntialized = true;
    this._refObject.baseline_coords_type = refObject.baseline_coords_type;
    this._baseline_a_mmIntialized = true;
    this._refObject.baseline_a_mm = refObject.baseline_a_mm;
    this._baseline_b_mmIntialized = true;
    this._refObject.baseline_b_mm = refObject.baseline_b_mm;
    this._baseline_c_mmIntialized = true;
    this._refObject.baseline_c_mm = refObject.baseline_c_mm;
    this._accuracyIntialized = true;
    this._refObject.accuracy = refObject.accuracy;
    this._iar_num_hypothesesIntialized = true;
    this._refObject.iar_num_hypotheses = refObject.iar_num_hypotheses;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    RTKBaselineWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'RTKBaseline'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get time_last_baseline_ms() {
    if (!this._time_last_baseline_msIntialized) {
      return undefined;
    }
    return this._refObject.time_last_baseline_ms;
  }
  set time_last_baseline_ms(value) {
    this._time_last_baseline_msIntialized = true;
    this._refObject.time_last_baseline_ms = value;
  }
  get rtk_receiver_id() {
    if (!this._rtk_receiver_idIntialized) {
      return undefined;
    }
    return this._refObject.rtk_receiver_id;
  }
  set rtk_receiver_id(value) {
    this._rtk_receiver_idIntialized = true;
    this._refObject.rtk_receiver_id = value;
  }
  get wn() {
    if (!this._wnIntialized) {
      return undefined;
    }
    return this._refObject.wn;
  }
  set wn(value) {
    this._wnIntialized = true;
    this._refObject.wn = value;
  }
  get tow() {
    if (!this._towIntialized) {
      return undefined;
    }
    return this._refObject.tow;
  }
  set tow(value) {
    this._towIntialized = true;
    this._refObject.tow = value;
  }
  get rtk_health() {
    if (!this._rtk_healthIntialized) {
      return undefined;
    }
    return this._refObject.rtk_health;
  }
  set rtk_health(value) {
    this._rtk_healthIntialized = true;
    this._refObject.rtk_health = value;
  }
  get rtk_rate() {
    if (!this._rtk_rateIntialized) {
      return undefined;
    }
    return this._refObject.rtk_rate;
  }
  set rtk_rate(value) {
    this._rtk_rateIntialized = true;
    this._refObject.rtk_rate = value;
  }
  get nsats() {
    if (!this._nsatsIntialized) {
      return undefined;
    }
    return this._refObject.nsats;
  }
  set nsats(value) {
    this._nsatsIntialized = true;
    this._refObject.nsats = value;
  }
  get baseline_coords_type() {
    if (!this._baseline_coords_typeIntialized) {
      return undefined;
    }
    return this._refObject.baseline_coords_type;
  }
  set baseline_coords_type(value) {
    this._baseline_coords_typeIntialized = true;
    this._refObject.baseline_coords_type = value;
  }
  get baseline_a_mm() {
    if (!this._baseline_a_mmIntialized) {
      return undefined;
    }
    return this._refObject.baseline_a_mm;
  }
  set baseline_a_mm(value) {
    this._baseline_a_mmIntialized = true;
    this._refObject.baseline_a_mm = value;
  }
  get baseline_b_mm() {
    if (!this._baseline_b_mmIntialized) {
      return undefined;
    }
    return this._refObject.baseline_b_mm;
  }
  set baseline_b_mm(value) {
    this._baseline_b_mmIntialized = true;
    this._refObject.baseline_b_mm = value;
  }
  get baseline_c_mm() {
    if (!this._baseline_c_mmIntialized) {
      return undefined;
    }
    return this._refObject.baseline_c_mm;
  }
  set baseline_c_mm(value) {
    this._baseline_c_mmIntialized = true;
    this._refObject.baseline_c_mm = value;
  }
  get accuracy() {
    if (!this._accuracyIntialized) {
      return undefined;
    }
    return this._refObject.accuracy;
  }
  set accuracy(value) {
    this._accuracyIntialized = true;
    this._refObject.accuracy = value;
  }
  get iar_num_hypotheses() {
    if (!this._iar_num_hypothesesIntialized) {
      return undefined;
    }
    return this._refObject.iar_num_hypotheses;
  }
  set iar_num_hypotheses(value) {
    this._iar_num_hypothesesIntialized = true;
    this._refObject.iar_num_hypotheses = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RTKBaselineRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._time_last_baseline_msIntialized = true;
    this._rtk_receiver_idIntialized = true;
    this._wnIntialized = true;
    this._towIntialized = true;
    this._rtk_healthIntialized = true;
    this._rtk_rateIntialized = true;
    this._nsatsIntialized = true;
    this._baseline_coords_typeIntialized = true;
    this._baseline_a_mmIntialized = true;
    this._baseline_b_mmIntialized = true;
    this._baseline_c_mmIntialized = true;
    this._accuracyIntialized = true;
    this._iar_num_hypothesesIntialized = true;
  }
  copy(other) {
    this._refObject = new RTKBaselineRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._time_last_baseline_msIntialized = true;
    this._rtk_receiver_idIntialized = true;
    this._wnIntialized = true;
    this._towIntialized = true;
    this._rtk_healthIntialized = true;
    this._rtk_rateIntialized = true;
    this._nsatsIntialized = true;
    this._baseline_coords_typeIntialized = true;
    this._baseline_a_mmIntialized = true;
    this._baseline_b_mmIntialized = true;
    this._baseline_c_mmIntialized = true;
    this._accuracyIntialized = true;
    this._iar_num_hypothesesIntialized = true;
  }
  static get classType() {
    return RTKBaselineWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"RTK_BASELINE_COORDINATE_SYSTEM_ECEF","value":0},{"type":"uint8","name":"RTK_BASELINE_COORDINATE_SYSTEM_NED","value":1}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"time_last_baseline_ms","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rtk_receiver_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"wn","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tow","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rtk_health","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rtk_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nsats","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baseline_coords_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baseline_a_mm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baseline_b_mm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baseline_c_mm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accuracy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"iar_num_hypotheses","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"RTKBaseline","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RTKBaseline"};
  }
  hasMember(name) {
    let memberNames = ["header","time_last_baseline_ms","rtk_receiver_id","wn","tow","rtk_health","rtk_rate","nsats","baseline_coords_type","baseline_a_mm","baseline_b_mm","baseline_c_mm","accuracy","iar_num_hypotheses"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RTKBaselineArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RTKBaselineWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RTKBaselineWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RTKBaselineArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RTKBaselineArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RTKBaselineRefArray(size);
    this._refObject = new RTKBaselineRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RTKBaselineWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RTKBaselineArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RTKBaselineArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RTKBaselineWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RTKBaselineWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RTKBaselineArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(RTKBaselineWrapper, "RTK_BASELINE_COORDINATE_SYSTEM_ECEF", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RTKBaselineWrapper, "RTK_BASELINE_COORDINATE_SYSTEM_NED", {value: 1, writable: false, enumerable: true, configurable: true});
module.exports = RTKBaselineWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "RTK_BASELINE_COORDINATE_SYSTEM_ECEF",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "RTK_BASELINE_COORDINATE_SYSTEM_NED",
      "value": 1
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "time_last_baseline_ms",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rtk_receiver_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "wn",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tow",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rtk_health",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rtk_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "nsats",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baseline_coords_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baseline_a_mm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baseline_b_mm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baseline_c_mm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accuracy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "iar_num_hypotheses",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "RTKBaseline",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RTKBaseline"
}
*/
