// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__SysStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const SysStatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  sensors_present: primitiveTypes.uint32,
  sensors_enabled: primitiveTypes.uint32,
  sensors_health: primitiveTypes.uint32,
  load: primitiveTypes.uint16,
  voltage_battery: primitiveTypes.uint16,
  current_battery: primitiveTypes.int16,
  battery_remaining: primitiveTypes.int8,
  drop_rate_comm: primitiveTypes.uint16,
  errors_comm: primitiveTypes.uint16,
  errors_count1: primitiveTypes.uint16,
  errors_count2: primitiveTypes.uint16,
  errors_count3: primitiveTypes.uint16,
  errors_count4: primitiveTypes.uint16,
});
const SysStatusRefArray = ArrayType(SysStatusRefStruct);
const SysStatusRefStructArray = StructType({
  data: SysStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SysStatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SysStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._sensors_presentIntialized = true;
      this._sensors_enabledIntialized = true;
      this._sensors_healthIntialized = true;
      this._loadIntialized = true;
      this._voltage_batteryIntialized = true;
      this._current_batteryIntialized = true;
      this._battery_remainingIntialized = true;
      this._drop_rate_commIntialized = true;
      this._errors_commIntialized = true;
      this._errors_count1Intialized = true;
      this._errors_count2Intialized = true;
      this._errors_count3Intialized = true;
      this._errors_count4Intialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SysStatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._sensors_presentIntialized = false;
    this._sensors_enabledIntialized = false;
    this._sensors_healthIntialized = false;
    this._loadIntialized = false;
    this._voltage_batteryIntialized = false;
    this._current_batteryIntialized = false;
    this._battery_remainingIntialized = false;
    this._drop_rate_commIntialized = false;
    this._errors_commIntialized = false;
    this._errors_count1Intialized = false;
    this._errors_count2Intialized = false;
    this._errors_count3Intialized = false;
    this._errors_count4Intialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SysStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SysStatusArrayWrapper;
  }
  static get ArrayType() {
    return SysStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return SysStatusRefStructArray
  }
  static get refObjectType() {
    return SysStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._sensors_presentIntialized) {
        throw new TypeError('Invalid argument: sensors_present in SysStatus');
      }
      if (!this._sensors_enabledIntialized) {
        throw new TypeError('Invalid argument: sensors_enabled in SysStatus');
      }
      if (!this._sensors_healthIntialized) {
        throw new TypeError('Invalid argument: sensors_health in SysStatus');
      }
      if (!this._loadIntialized) {
        throw new TypeError('Invalid argument: load in SysStatus');
      }
      if (!this._voltage_batteryIntialized) {
        throw new TypeError('Invalid argument: voltage_battery in SysStatus');
      }
      if (!this._current_batteryIntialized) {
        throw new TypeError('Invalid argument: current_battery in SysStatus');
      }
      if (!this._battery_remainingIntialized) {
        throw new TypeError('Invalid argument: battery_remaining in SysStatus');
      }
      if (!this._drop_rate_commIntialized) {
        throw new TypeError('Invalid argument: drop_rate_comm in SysStatus');
      }
      if (!this._errors_commIntialized) {
        throw new TypeError('Invalid argument: errors_comm in SysStatus');
      }
      if (!this._errors_count1Intialized) {
        throw new TypeError('Invalid argument: errors_count1 in SysStatus');
      }
      if (!this._errors_count2Intialized) {
        throw new TypeError('Invalid argument: errors_count2 in SysStatus');
      }
      if (!this._errors_count3Intialized) {
        throw new TypeError('Invalid argument: errors_count3 in SysStatus');
      }
      if (!this._errors_count4Intialized) {
        throw new TypeError('Invalid argument: errors_count4 in SysStatus');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._sensors_presentIntialized = true;
    this._refObject.sensors_present = refObject.sensors_present;
    this._sensors_enabledIntialized = true;
    this._refObject.sensors_enabled = refObject.sensors_enabled;
    this._sensors_healthIntialized = true;
    this._refObject.sensors_health = refObject.sensors_health;
    this._loadIntialized = true;
    this._refObject.load = refObject.load;
    this._voltage_batteryIntialized = true;
    this._refObject.voltage_battery = refObject.voltage_battery;
    this._current_batteryIntialized = true;
    this._refObject.current_battery = refObject.current_battery;
    this._battery_remainingIntialized = true;
    this._refObject.battery_remaining = refObject.battery_remaining;
    this._drop_rate_commIntialized = true;
    this._refObject.drop_rate_comm = refObject.drop_rate_comm;
    this._errors_commIntialized = true;
    this._refObject.errors_comm = refObject.errors_comm;
    this._errors_count1Intialized = true;
    this._refObject.errors_count1 = refObject.errors_count1;
    this._errors_count2Intialized = true;
    this._refObject.errors_count2 = refObject.errors_count2;
    this._errors_count3Intialized = true;
    this._refObject.errors_count3 = refObject.errors_count3;
    this._errors_count4Intialized = true;
    this._refObject.errors_count4 = refObject.errors_count4;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    SysStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'SysStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get sensors_present() {
    if (!this._sensors_presentIntialized) {
      return undefined;
    }
    return this._refObject.sensors_present;
  }
  set sensors_present(value) {
    this._sensors_presentIntialized = true;
    this._refObject.sensors_present = value;
  }
  get sensors_enabled() {
    if (!this._sensors_enabledIntialized) {
      return undefined;
    }
    return this._refObject.sensors_enabled;
  }
  set sensors_enabled(value) {
    this._sensors_enabledIntialized = true;
    this._refObject.sensors_enabled = value;
  }
  get sensors_health() {
    if (!this._sensors_healthIntialized) {
      return undefined;
    }
    return this._refObject.sensors_health;
  }
  set sensors_health(value) {
    this._sensors_healthIntialized = true;
    this._refObject.sensors_health = value;
  }
  get load() {
    if (!this._loadIntialized) {
      return undefined;
    }
    return this._refObject.load;
  }
  set load(value) {
    this._loadIntialized = true;
    this._refObject.load = value;
  }
  get voltage_battery() {
    if (!this._voltage_batteryIntialized) {
      return undefined;
    }
    return this._refObject.voltage_battery;
  }
  set voltage_battery(value) {
    this._voltage_batteryIntialized = true;
    this._refObject.voltage_battery = value;
  }
  get current_battery() {
    if (!this._current_batteryIntialized) {
      return undefined;
    }
    return this._refObject.current_battery;
  }
  set current_battery(value) {
    this._current_batteryIntialized = true;
    this._refObject.current_battery = value;
  }
  get battery_remaining() {
    if (!this._battery_remainingIntialized) {
      return undefined;
    }
    return this._refObject.battery_remaining;
  }
  set battery_remaining(value) {
    this._battery_remainingIntialized = true;
    this._refObject.battery_remaining = value;
  }
  get drop_rate_comm() {
    if (!this._drop_rate_commIntialized) {
      return undefined;
    }
    return this._refObject.drop_rate_comm;
  }
  set drop_rate_comm(value) {
    this._drop_rate_commIntialized = true;
    this._refObject.drop_rate_comm = value;
  }
  get errors_comm() {
    if (!this._errors_commIntialized) {
      return undefined;
    }
    return this._refObject.errors_comm;
  }
  set errors_comm(value) {
    this._errors_commIntialized = true;
    this._refObject.errors_comm = value;
  }
  get errors_count1() {
    if (!this._errors_count1Intialized) {
      return undefined;
    }
    return this._refObject.errors_count1;
  }
  set errors_count1(value) {
    this._errors_count1Intialized = true;
    this._refObject.errors_count1 = value;
  }
  get errors_count2() {
    if (!this._errors_count2Intialized) {
      return undefined;
    }
    return this._refObject.errors_count2;
  }
  set errors_count2(value) {
    this._errors_count2Intialized = true;
    this._refObject.errors_count2 = value;
  }
  get errors_count3() {
    if (!this._errors_count3Intialized) {
      return undefined;
    }
    return this._refObject.errors_count3;
  }
  set errors_count3(value) {
    this._errors_count3Intialized = true;
    this._refObject.errors_count3 = value;
  }
  get errors_count4() {
    if (!this._errors_count4Intialized) {
      return undefined;
    }
    return this._refObject.errors_count4;
  }
  set errors_count4(value) {
    this._errors_count4Intialized = true;
    this._refObject.errors_count4 = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SysStatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._sensors_presentIntialized = true;
    this._sensors_enabledIntialized = true;
    this._sensors_healthIntialized = true;
    this._loadIntialized = true;
    this._voltage_batteryIntialized = true;
    this._current_batteryIntialized = true;
    this._battery_remainingIntialized = true;
    this._drop_rate_commIntialized = true;
    this._errors_commIntialized = true;
    this._errors_count1Intialized = true;
    this._errors_count2Intialized = true;
    this._errors_count3Intialized = true;
    this._errors_count4Intialized = true;
  }
  copy(other) {
    this._refObject = new SysStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._sensors_presentIntialized = true;
    this._sensors_enabledIntialized = true;
    this._sensors_healthIntialized = true;
    this._loadIntialized = true;
    this._voltage_batteryIntialized = true;
    this._current_batteryIntialized = true;
    this._battery_remainingIntialized = true;
    this._drop_rate_commIntialized = true;
    this._errors_commIntialized = true;
    this._errors_count1Intialized = true;
    this._errors_count2Intialized = true;
    this._errors_count3Intialized = true;
    this._errors_count4Intialized = true;
  }
  static get classType() {
    return SysStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"sensors_present","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sensors_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sensors_health","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"load","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"voltage_battery","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_battery","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"battery_remaining","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"drop_rate_comm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"errors_comm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"errors_count1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"errors_count2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"errors_count3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"errors_count4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"SysStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SysStatus"};
  }
  hasMember(name) {
    let memberNames = ["header","sensors_present","sensors_enabled","sensors_health","load","voltage_battery","current_battery","battery_remaining","drop_rate_comm","errors_comm","errors_count1","errors_count2","errors_count3","errors_count4"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SysStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SysStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SysStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SysStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SysStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SysStatusRefArray(size);
    this._refObject = new SysStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SysStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SysStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SysStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SysStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SysStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SysStatusArrayWrapper;
  }
}
module.exports = SysStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "sensors_present",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sensors_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sensors_health",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "load",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "voltage_battery",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "current_battery",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "battery_remaining",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "drop_rate_comm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "errors_comm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "errors_count1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "errors_count2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "errors_count3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "errors_count4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "SysStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SysStatus"
}
*/
