// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__CommandTOL_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CommandTOL_RequestRefStruct = StructType({
  min_pitch: primitiveTypes.float32,
  yaw: primitiveTypes.float32,
  latitude: primitiveTypes.float32,
  longitude: primitiveTypes.float32,
  altitude: primitiveTypes.float32,
});
const CommandTOL_RequestRefArray = ArrayType(CommandTOL_RequestRefStruct);
const CommandTOL_RequestRefStructArray = StructType({
  data: CommandTOL_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CommandTOL_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CommandTOL_RequestRefStruct(other._refObject.toObject());
      this._min_pitchIntialized = true;
      this._yawIntialized = true;
      this._latitudeIntialized = true;
      this._longitudeIntialized = true;
      this._altitudeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CommandTOL_RequestRefStruct();
    this._min_pitchIntialized = false;
    this._yawIntialized = false;
    this._latitudeIntialized = false;
    this._longitudeIntialized = false;
    this._altitudeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CommandTOL_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CommandTOL_RequestArrayWrapper;
  }
  static get ArrayType() {
    return CommandTOL_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return CommandTOL_RequestRefStructArray
  }
  static get refObjectType() {
    return CommandTOL_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._min_pitchIntialized) {
        throw new TypeError('Invalid argument: min_pitch in CommandTOL_Request');
      }
      if (!this._yawIntialized) {
        throw new TypeError('Invalid argument: yaw in CommandTOL_Request');
      }
      if (!this._latitudeIntialized) {
        throw new TypeError('Invalid argument: latitude in CommandTOL_Request');
      }
      if (!this._longitudeIntialized) {
        throw new TypeError('Invalid argument: longitude in CommandTOL_Request');
      }
      if (!this._altitudeIntialized) {
        throw new TypeError('Invalid argument: altitude in CommandTOL_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._min_pitchIntialized = true;
    this._refObject.min_pitch = refObject.min_pitch;
    this._yawIntialized = true;
    this._refObject.yaw = refObject.yaw;
    this._latitudeIntialized = true;
    this._refObject.latitude = refObject.latitude;
    this._longitudeIntialized = true;
    this._refObject.longitude = refObject.longitude;
    this._altitudeIntialized = true;
    this._refObject.altitude = refObject.altitude;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    CommandTOL_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'CommandTOL_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get min_pitch() {
    if (!this._min_pitchIntialized) {
      return undefined;
    }
    return this._refObject.min_pitch;
  }
  set min_pitch(value) {
    this._min_pitchIntialized = true;
    this._refObject.min_pitch = value;
  }
  get yaw() {
    if (!this._yawIntialized) {
      return undefined;
    }
    return this._refObject.yaw;
  }
  set yaw(value) {
    this._yawIntialized = true;
    this._refObject.yaw = value;
  }
  get latitude() {
    if (!this._latitudeIntialized) {
      return undefined;
    }
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._latitudeIntialized = true;
    this._refObject.latitude = value;
  }
  get longitude() {
    if (!this._longitudeIntialized) {
      return undefined;
    }
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._longitudeIntialized = true;
    this._refObject.longitude = value;
  }
  get altitude() {
    if (!this._altitudeIntialized) {
      return undefined;
    }
    return this._refObject.altitude;
  }
  set altitude(value) {
    this._altitudeIntialized = true;
    this._refObject.altitude = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CommandTOL_RequestRefStruct(refObject.toObject());
    this._min_pitchIntialized = true;
    this._yawIntialized = true;
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
  }
  copy(other) {
    this._refObject = new CommandTOL_RequestRefStruct(other._refObject.toObject());
    this._min_pitchIntialized = true;
    this._yawIntialized = true;
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
  }
  static get classType() {
    return CommandTOL_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"min_pitch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"CommandTOL_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CommandTOL_Request"};
  }
  hasMember(name) {
    let memberNames = ["min_pitch","yaw","latitude","longitude","altitude"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CommandTOL_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CommandTOL_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CommandTOL_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandTOL_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandTOL_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CommandTOL_RequestRefArray(size);
    this._refObject = new CommandTOL_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CommandTOL_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CommandTOL_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CommandTOL_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CommandTOL_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CommandTOL_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CommandTOL_RequestArrayWrapper;
  }
}
module.exports = CommandTOL_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "min_pitch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "CommandTOL_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CommandTOL_Request"
}
*/
