// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__VehicleInfoGet_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const VehicleInfoGet_RequestRefStruct = StructType({
  sysid: primitiveTypes.uint8,
  compid: primitiveTypes.uint8,
  get_all: primitiveTypes.bool,
});
const VehicleInfoGet_RequestRefArray = ArrayType(VehicleInfoGet_RequestRefStruct);
const VehicleInfoGet_RequestRefStructArray = StructType({
  data: VehicleInfoGet_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleInfoGet_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleInfoGet_RequestRefStruct(other._refObject.toObject());
      this._sysidIntialized = true;
      this._compidIntialized = true;
      this._get_allIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleInfoGet_RequestRefStruct();
    this._sysidIntialized = false;
    this._compidIntialized = false;
    this._get_allIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleInfoGet_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleInfoGet_RequestArrayWrapper;
  }
  static get ArrayType() {
    return VehicleInfoGet_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleInfoGet_RequestRefStructArray
  }
  static get refObjectType() {
    return VehicleInfoGet_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._sysidIntialized) {
        throw new TypeError('Invalid argument: sysid in VehicleInfoGet_Request');
      }
      if (!this._compidIntialized) {
        throw new TypeError('Invalid argument: compid in VehicleInfoGet_Request');
      }
      if (!this._get_allIntialized) {
        throw new TypeError('Invalid argument: get_all in VehicleInfoGet_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._sysidIntialized = true;
    this._refObject.sysid = refObject.sysid;
    this._compidIntialized = true;
    this._refObject.compid = refObject.compid;
    this._get_allIntialized = true;
    this._refObject.get_all = refObject.get_all;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleInfoGet_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'VehicleInfoGet_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get sysid() {
    if (!this._sysidIntialized) {
      return undefined;
    }
    return this._refObject.sysid;
  }
  set sysid(value) {
    this._sysidIntialized = true;
    this._refObject.sysid = value;
  }
  get compid() {
    if (!this._compidIntialized) {
      return undefined;
    }
    return this._refObject.compid;
  }
  set compid(value) {
    this._compidIntialized = true;
    this._refObject.compid = value;
  }
  get get_all() {
    if (!this._get_allIntialized) {
      return undefined;
    }
    return this._refObject.get_all;
  }
  set get_all(value) {
    this._get_allIntialized = true;
    this._refObject.get_all = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleInfoGet_RequestRefStruct(refObject.toObject());
    this._sysidIntialized = true;
    this._compidIntialized = true;
    this._get_allIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleInfoGet_RequestRefStruct(other._refObject.toObject());
    this._sysidIntialized = true;
    this._compidIntialized = true;
    this._get_allIntialized = true;
  }
  static get classType() {
    return VehicleInfoGet_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"GET_MY_SYSID","value":0},{"type":"uint8","name":"GET_MY_COMPID","value":0}],"fields":[{"name":"sysid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"compid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"get_all","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"VehicleInfoGet_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleInfoGet_Request"};
  }
  hasMember(name) {
    let memberNames = ["sysid","compid","get_all"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleInfoGet_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleInfoGet_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleInfoGet_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleInfoGet_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleInfoGet_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleInfoGet_RequestRefArray(size);
    this._refObject = new VehicleInfoGet_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleInfoGet_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleInfoGet_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleInfoGet_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleInfoGet_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleInfoGet_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleInfoGet_RequestArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(VehicleInfoGet_RequestWrapper, "GET_MY_SYSID", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleInfoGet_RequestWrapper, "GET_MY_COMPID", {value: 0, writable: false, enumerable: true, configurable: true});
module.exports = VehicleInfoGet_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "GET_MY_SYSID",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "GET_MY_COMPID",
      "value": 0
    }
  ],
  "fields": [
    {
      "name": "sysid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "compid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "get_all",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "VehicleInfoGet_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleInfoGet_Request"
}
*/
