// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__AttitudeTarget.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const QuaternionWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Quaternion.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const AttitudeTargetRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  type_mask: primitiveTypes.uint8,
  orientation: QuaternionWrapper.refObjectType,
  body_rate: Vector3Wrapper.refObjectType,
  thrust: primitiveTypes.float32,
});
const AttitudeTargetRefArray = ArrayType(AttitudeTargetRefStruct);
const AttitudeTargetRefStructArray = StructType({
  data: AttitudeTargetRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AttitudeTargetWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AttitudeTargetRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._type_maskIntialized = true;
      this._wrapperFields.orientation =  new QuaternionWrapper(other._wrapperFields.orientation);
      this._wrapperFields.body_rate =  new Vector3Wrapper(other._wrapperFields.body_rate);
      this._thrustIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AttitudeTargetRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._type_maskIntialized = false;
    this._wrapperFields.orientation = new QuaternionWrapper();
    this._wrapperFields.body_rate = new Vector3Wrapper();
    this._thrustIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new AttitudeTargetWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AttitudeTargetArrayWrapper;
  }
  static get ArrayType() {
    return AttitudeTargetArrayWrapper;
  }
  static get refObjectArrayType() {
    return AttitudeTargetRefStructArray
  }
  static get refObjectType() {
    return AttitudeTargetRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._type_maskIntialized) {
        throw new TypeError('Invalid argument: type_mask in AttitudeTarget');
      }
      if (!this._thrustIntialized) {
        throw new TypeError('Invalid argument: thrust in AttitudeTarget');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.orientation.freeze(own, checkConsistency);
    this._refObject.orientation = this._wrapperFields.orientation.refObject;
    this._wrapperFields.body_rate.freeze(own, checkConsistency);
    this._refObject.body_rate = this._wrapperFields.body_rate.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._type_maskIntialized = true;
    this._refObject.type_mask = refObject.type_mask;
    this._wrapperFields.orientation.copyRefObject(refObject.orientation);
    this._wrapperFields.body_rate.copyRefObject(refObject.body_rate);
    this._thrustIntialized = true;
    this._refObject.thrust = refObject.thrust;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    QuaternionWrapper.freeStruct(refObject.orientation);
    Vector3Wrapper.freeStruct(refObject.body_rate);
  }
  static destoryRawROS(msg) {
    AttitudeTargetWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'AttitudeTarget'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get type_mask() {
    if (!this._type_maskIntialized) {
      return undefined;
    }
    return this._refObject.type_mask;
  }
  set type_mask(value) {
    this._type_maskIntialized = true;
    this._refObject.type_mask = value;
  }
  get orientation() {
    return this._wrapperFields.orientation;
  }
  set orientation(value) {
    if (value instanceof QuaternionWrapper) {
      this._wrapperFields.orientation.copy(value);
    } else {
      this._wrapperFields.orientation.copy(new QuaternionWrapper(value));
    }
  }
  get body_rate() {
    return this._wrapperFields.body_rate;
  }
  set body_rate(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.body_rate.copy(value);
    } else {
      this._wrapperFields.body_rate.copy(new Vector3Wrapper(value));
    }
  }
  get thrust() {
    if (!this._thrustIntialized) {
      return undefined;
    }
    return this._refObject.thrust;
  }
  set thrust(value) {
    this._thrustIntialized = true;
    this._refObject.thrust = value;
  }
  copyRefObject(refObject) {
    this._refObject = new AttitudeTargetRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._type_maskIntialized = true;
    this._wrapperFields.orientation.copyRefObject(this._refObject.orientation);
    this._wrapperFields.body_rate.copyRefObject(this._refObject.body_rate);
    this._thrustIntialized = true;
  }
  copy(other) {
    this._refObject = new AttitudeTargetRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._type_maskIntialized = true;
    this._wrapperFields.orientation.copy(other._wrapperFields.orientation);
    this._wrapperFields.body_rate.copy(other._wrapperFields.body_rate);
    this._thrustIntialized = true;
  }
  static get classType() {
    return AttitudeTargetWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"IGNORE_ROLL_RATE","value":1},{"type":"uint8","name":"IGNORE_PITCH_RATE","value":2},{"type":"uint8","name":"IGNORE_YAW_RATE","value":4},{"type":"uint8","name":"IGNORE_THRUST","value":64},{"type":"uint8","name":"IGNORE_ATTITUDE","value":128}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"type_mask","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"orientation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Quaternion","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"body_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"thrust","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"AttitudeTarget","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AttitudeTarget"};
  }
  hasMember(name) {
    let memberNames = ["header","type_mask","orientation","body_rate","thrust"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AttitudeTargetArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AttitudeTargetWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AttitudeTargetWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AttitudeTargetArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AttitudeTargetArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AttitudeTargetRefArray(size);
    this._refObject = new AttitudeTargetRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AttitudeTargetWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AttitudeTargetArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AttitudeTargetArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AttitudeTargetWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AttitudeTargetWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AttitudeTargetArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(AttitudeTargetWrapper, "IGNORE_ROLL_RATE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AttitudeTargetWrapper, "IGNORE_PITCH_RATE", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AttitudeTargetWrapper, "IGNORE_YAW_RATE", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AttitudeTargetWrapper, "IGNORE_THRUST", {value: 64, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AttitudeTargetWrapper, "IGNORE_ATTITUDE", {value: 128, writable: false, enumerable: true, configurable: true});
module.exports = AttitudeTargetWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "IGNORE_ROLL_RATE",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "IGNORE_PITCH_RATE",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "IGNORE_YAW_RATE",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "IGNORE_THRUST",
      "value": 64
    },
    {
      "type": "uint8",
      "name": "IGNORE_ATTITUDE",
      "value": 128
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "type_mask",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "orientation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Quaternion",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "body_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "thrust",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "AttitudeTarget",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "AttitudeTarget"
}
*/
