// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__OpticalFlowRad.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const OpticalFlowRadRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  integration_time_us: primitiveTypes.uint32,
  integrated_x: primitiveTypes.float32,
  integrated_y: primitiveTypes.float32,
  integrated_xgyro: primitiveTypes.float32,
  integrated_ygyro: primitiveTypes.float32,
  integrated_zgyro: primitiveTypes.float32,
  temperature: primitiveTypes.int16,
  quality: primitiveTypes.uint8,
  time_delta_distance_us: primitiveTypes.uint32,
  distance: primitiveTypes.float32,
});
const OpticalFlowRadRefArray = ArrayType(OpticalFlowRadRefStruct);
const OpticalFlowRadRefStructArray = StructType({
  data: OpticalFlowRadRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OpticalFlowRadWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OpticalFlowRadRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._integration_time_usIntialized = true;
      this._integrated_xIntialized = true;
      this._integrated_yIntialized = true;
      this._integrated_xgyroIntialized = true;
      this._integrated_ygyroIntialized = true;
      this._integrated_zgyroIntialized = true;
      this._temperatureIntialized = true;
      this._qualityIntialized = true;
      this._time_delta_distance_usIntialized = true;
      this._distanceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OpticalFlowRadRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._integration_time_usIntialized = false;
    this._integrated_xIntialized = false;
    this._integrated_yIntialized = false;
    this._integrated_xgyroIntialized = false;
    this._integrated_ygyroIntialized = false;
    this._integrated_zgyroIntialized = false;
    this._temperatureIntialized = false;
    this._qualityIntialized = false;
    this._time_delta_distance_usIntialized = false;
    this._distanceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new OpticalFlowRadWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OpticalFlowRadArrayWrapper;
  }
  static get ArrayType() {
    return OpticalFlowRadArrayWrapper;
  }
  static get refObjectArrayType() {
    return OpticalFlowRadRefStructArray
  }
  static get refObjectType() {
    return OpticalFlowRadRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._integration_time_usIntialized) {
        throw new TypeError('Invalid argument: integration_time_us in OpticalFlowRad');
      }
      if (!this._integrated_xIntialized) {
        throw new TypeError('Invalid argument: integrated_x in OpticalFlowRad');
      }
      if (!this._integrated_yIntialized) {
        throw new TypeError('Invalid argument: integrated_y in OpticalFlowRad');
      }
      if (!this._integrated_xgyroIntialized) {
        throw new TypeError('Invalid argument: integrated_xgyro in OpticalFlowRad');
      }
      if (!this._integrated_ygyroIntialized) {
        throw new TypeError('Invalid argument: integrated_ygyro in OpticalFlowRad');
      }
      if (!this._integrated_zgyroIntialized) {
        throw new TypeError('Invalid argument: integrated_zgyro in OpticalFlowRad');
      }
      if (!this._temperatureIntialized) {
        throw new TypeError('Invalid argument: temperature in OpticalFlowRad');
      }
      if (!this._qualityIntialized) {
        throw new TypeError('Invalid argument: quality in OpticalFlowRad');
      }
      if (!this._time_delta_distance_usIntialized) {
        throw new TypeError('Invalid argument: time_delta_distance_us in OpticalFlowRad');
      }
      if (!this._distanceIntialized) {
        throw new TypeError('Invalid argument: distance in OpticalFlowRad');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._integration_time_usIntialized = true;
    this._refObject.integration_time_us = refObject.integration_time_us;
    this._integrated_xIntialized = true;
    this._refObject.integrated_x = refObject.integrated_x;
    this._integrated_yIntialized = true;
    this._refObject.integrated_y = refObject.integrated_y;
    this._integrated_xgyroIntialized = true;
    this._refObject.integrated_xgyro = refObject.integrated_xgyro;
    this._integrated_ygyroIntialized = true;
    this._refObject.integrated_ygyro = refObject.integrated_ygyro;
    this._integrated_zgyroIntialized = true;
    this._refObject.integrated_zgyro = refObject.integrated_zgyro;
    this._temperatureIntialized = true;
    this._refObject.temperature = refObject.temperature;
    this._qualityIntialized = true;
    this._refObject.quality = refObject.quality;
    this._time_delta_distance_usIntialized = true;
    this._refObject.time_delta_distance_us = refObject.time_delta_distance_us;
    this._distanceIntialized = true;
    this._refObject.distance = refObject.distance;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    OpticalFlowRadWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'OpticalFlowRad'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get integration_time_us() {
    if (!this._integration_time_usIntialized) {
      return undefined;
    }
    return this._refObject.integration_time_us;
  }
  set integration_time_us(value) {
    this._integration_time_usIntialized = true;
    this._refObject.integration_time_us = value;
  }
  get integrated_x() {
    if (!this._integrated_xIntialized) {
      return undefined;
    }
    return this._refObject.integrated_x;
  }
  set integrated_x(value) {
    this._integrated_xIntialized = true;
    this._refObject.integrated_x = value;
  }
  get integrated_y() {
    if (!this._integrated_yIntialized) {
      return undefined;
    }
    return this._refObject.integrated_y;
  }
  set integrated_y(value) {
    this._integrated_yIntialized = true;
    this._refObject.integrated_y = value;
  }
  get integrated_xgyro() {
    if (!this._integrated_xgyroIntialized) {
      return undefined;
    }
    return this._refObject.integrated_xgyro;
  }
  set integrated_xgyro(value) {
    this._integrated_xgyroIntialized = true;
    this._refObject.integrated_xgyro = value;
  }
  get integrated_ygyro() {
    if (!this._integrated_ygyroIntialized) {
      return undefined;
    }
    return this._refObject.integrated_ygyro;
  }
  set integrated_ygyro(value) {
    this._integrated_ygyroIntialized = true;
    this._refObject.integrated_ygyro = value;
  }
  get integrated_zgyro() {
    if (!this._integrated_zgyroIntialized) {
      return undefined;
    }
    return this._refObject.integrated_zgyro;
  }
  set integrated_zgyro(value) {
    this._integrated_zgyroIntialized = true;
    this._refObject.integrated_zgyro = value;
  }
  get temperature() {
    if (!this._temperatureIntialized) {
      return undefined;
    }
    return this._refObject.temperature;
  }
  set temperature(value) {
    this._temperatureIntialized = true;
    this._refObject.temperature = value;
  }
  get quality() {
    if (!this._qualityIntialized) {
      return undefined;
    }
    return this._refObject.quality;
  }
  set quality(value) {
    this._qualityIntialized = true;
    this._refObject.quality = value;
  }
  get time_delta_distance_us() {
    if (!this._time_delta_distance_usIntialized) {
      return undefined;
    }
    return this._refObject.time_delta_distance_us;
  }
  set time_delta_distance_us(value) {
    this._time_delta_distance_usIntialized = true;
    this._refObject.time_delta_distance_us = value;
  }
  get distance() {
    if (!this._distanceIntialized) {
      return undefined;
    }
    return this._refObject.distance;
  }
  set distance(value) {
    this._distanceIntialized = true;
    this._refObject.distance = value;
  }
  copyRefObject(refObject) {
    this._refObject = new OpticalFlowRadRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._integration_time_usIntialized = true;
    this._integrated_xIntialized = true;
    this._integrated_yIntialized = true;
    this._integrated_xgyroIntialized = true;
    this._integrated_ygyroIntialized = true;
    this._integrated_zgyroIntialized = true;
    this._temperatureIntialized = true;
    this._qualityIntialized = true;
    this._time_delta_distance_usIntialized = true;
    this._distanceIntialized = true;
  }
  copy(other) {
    this._refObject = new OpticalFlowRadRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._integration_time_usIntialized = true;
    this._integrated_xIntialized = true;
    this._integrated_yIntialized = true;
    this._integrated_xgyroIntialized = true;
    this._integrated_ygyroIntialized = true;
    this._integrated_zgyroIntialized = true;
    this._temperatureIntialized = true;
    this._qualityIntialized = true;
    this._time_delta_distance_usIntialized = true;
    this._distanceIntialized = true;
  }
  static get classType() {
    return OpticalFlowRadWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"integration_time_us","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"integrated_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"integrated_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"integrated_xgyro","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"integrated_ygyro","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"integrated_zgyro","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"temperature","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"quality","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_delta_distance_us","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"distance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"OpticalFlowRad","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OpticalFlowRad"};
  }
  hasMember(name) {
    let memberNames = ["header","integration_time_us","integrated_x","integrated_y","integrated_xgyro","integrated_ygyro","integrated_zgyro","temperature","quality","time_delta_distance_us","distance"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OpticalFlowRadArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OpticalFlowRadWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OpticalFlowRadWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpticalFlowRadArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpticalFlowRadArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OpticalFlowRadRefArray(size);
    this._refObject = new OpticalFlowRadRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OpticalFlowRadWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OpticalFlowRadArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OpticalFlowRadArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OpticalFlowRadWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OpticalFlowRadWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OpticalFlowRadArrayWrapper;
  }
}
module.exports = OpticalFlowRadWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "integration_time_us",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "integrated_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "integrated_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "integrated_xgyro",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "integrated_ygyro",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "integrated_zgyro",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "temperature",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "quality",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_delta_distance_us",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "distance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "OpticalFlowRad",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OpticalFlowRad"
}
*/
