// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__StatusEvent.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const StatusEventRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  severity: primitiveTypes.uint8,
  px4_id: primitiveTypes.uint32,
  arguments: ArrayType(primitiveTypes.uint8, 40),
  sequence: primitiveTypes.uint16,
});
const StatusEventRefArray = ArrayType(StatusEventRefStruct);
const StatusEventRefStructArray = StructType({
  data: StatusEventRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StatusEventWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StatusEventRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._severityIntialized = true;
      this._px4_idIntialized = true;
      this._wrapperFields.arguments = UInt8Wrapper.createArray();
      this._wrapperFields.arguments.copy(other._wrapperFields.arguments);
      this._sequenceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StatusEventRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._severityIntialized = false;
    this._px4_idIntialized = false;
    this._wrapperFields.arguments = UInt8Wrapper.createArray();
    this._sequenceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new StatusEventWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StatusEventArrayWrapper;
  }
  static get ArrayType() {
    return StatusEventArrayWrapper;
  }
  static get refObjectArrayType() {
    return StatusEventRefStructArray
  }
  static get refObjectType() {
    return StatusEventRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._severityIntialized) {
        throw new TypeError('Invalid argument: severity in StatusEvent');
      }
      if (!this._px4_idIntialized) {
        throw new TypeError('Invalid argument: px4_id in StatusEvent');
      }
      if (!this._sequenceIntialized) {
        throw new TypeError('Invalid argument: sequence in StatusEvent');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    for (let i = 0; i < this._wrapperFields.arguments.data.length; i++) {
      this._refObject.arguments[i] = this._wrapperFields.arguments.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._severityIntialized = true;
    this._refObject.severity = refObject.severity;
    this._px4_idIntialized = true;
    this._refObject.px4_id = refObject.px4_id;
    this._refObject.arguments = refObject.arguments;
    this._wrapperFields.arguments.fill(refObject.arguments.toArray());
    this._sequenceIntialized = true;
    this._refObject.sequence = refObject.sequence;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    StatusEventWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'StatusEvent'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get severity() {
    if (!this._severityIntialized) {
      return undefined;
    }
    return this._refObject.severity;
  }
  set severity(value) {
    this._severityIntialized = true;
    this._refObject.severity = value;
  }
  get px4_id() {
    if (!this._px4_idIntialized) {
      return undefined;
    }
    return this._refObject.px4_id;
  }
  set px4_id(value) {
    this._px4_idIntialized = true;
    this._refObject.px4_id = value;
  }
  get arguments() {
    return this._wrapperFields['arguments'].data;
  }
  set arguments(value) {
    if (value.length !== 40) {
      throw new RangeError('The length of the array must be 40.');
    }
    this._wrapperFields['arguments'].fill(value);
  }
  get sequence() {
    if (!this._sequenceIntialized) {
      return undefined;
    }
    return this._refObject.sequence;
  }
  set sequence(value) {
    this._sequenceIntialized = true;
    this._refObject.sequence = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StatusEventRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._severityIntialized = true;
    this._px4_idIntialized = true;
    this._wrapperFields.arguments.fill(refObject.arguments.toArray());
    this._sequenceIntialized = true;
  }
  copy(other) {
    this._refObject = new StatusEventRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._severityIntialized = true;
    this._px4_idIntialized = true;
    this._wrapperFields.arguments.copy(other._wrapperFields.arguments);
    this._sequenceIntialized = true;
  }
  static get classType() {
    return StatusEventWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"EMERGENCY","value":0},{"type":"uint8","name":"ALERT","value":1},{"type":"uint8","name":"CRITICAL","value":2},{"type":"uint8","name":"ERROR","value":3},{"type":"uint8","name":"WARNING","value":4},{"type":"uint8","name":"NOTICE","value":5},{"type":"uint8","name":"INFO","value":6},{"type":"uint8","name":"DEBUG","value":7}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"severity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"px4_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arguments","type":{"isArray":true,"arraySize":40,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sequence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"StatusEvent","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StatusEvent"};
  }
  hasMember(name) {
    let memberNames = ["header","severity","px4_id","arguments","sequence"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StatusEventArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StatusEventWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StatusEventWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatusEventArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatusEventArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StatusEventRefArray(size);
    this._refObject = new StatusEventRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StatusEventWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StatusEventArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StatusEventArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StatusEventWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StatusEventWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StatusEventArrayWrapper;
  }
}
// Define constants (8 in total)
Object.defineProperty(StatusEventWrapper, "EMERGENCY", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusEventWrapper, "ALERT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusEventWrapper, "CRITICAL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusEventWrapper, "ERROR", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusEventWrapper, "WARNING", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusEventWrapper, "NOTICE", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusEventWrapper, "INFO", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusEventWrapper, "DEBUG", {value: 7, writable: false, enumerable: true, configurable: true});
module.exports = StatusEventWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "EMERGENCY",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ALERT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "CRITICAL",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "ERROR",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "WARNING",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "NOTICE",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "INFO",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "DEBUG",
      "value": 7
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "severity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "px4_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arguments",
      "type": {
        "isArray": true,
        "arraySize": 40,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sequence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "StatusEvent",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "StatusEvent"
}
*/
