// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__GPSRTK.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const GPSRTKRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  rtk_receiver_id: primitiveTypes.uint8,
  wn: primitiveTypes.int16,
  tow: primitiveTypes.uint32,
  rtk_health: primitiveTypes.uint8,
  rtk_rate: primitiveTypes.uint8,
  nsats: primitiveTypes.uint8,
  baseline_a: primitiveTypes.int32,
  baseline_b: primitiveTypes.int32,
  baseline_c: primitiveTypes.int32,
  accuracy: primitiveTypes.uint32,
  iar_num_hypotheses: primitiveTypes.int32,
});
const GPSRTKRefArray = ArrayType(GPSRTKRefStruct);
const GPSRTKRefStructArray = StructType({
  data: GPSRTKRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GPSRTKWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GPSRTKRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._rtk_receiver_idIntialized = true;
      this._wnIntialized = true;
      this._towIntialized = true;
      this._rtk_healthIntialized = true;
      this._rtk_rateIntialized = true;
      this._nsatsIntialized = true;
      this._baseline_aIntialized = true;
      this._baseline_bIntialized = true;
      this._baseline_cIntialized = true;
      this._accuracyIntialized = true;
      this._iar_num_hypothesesIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GPSRTKRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._rtk_receiver_idIntialized = false;
    this._wnIntialized = false;
    this._towIntialized = false;
    this._rtk_healthIntialized = false;
    this._rtk_rateIntialized = false;
    this._nsatsIntialized = false;
    this._baseline_aIntialized = false;
    this._baseline_bIntialized = false;
    this._baseline_cIntialized = false;
    this._accuracyIntialized = false;
    this._iar_num_hypothesesIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GPSRTKWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GPSRTKArrayWrapper;
  }
  static get ArrayType() {
    return GPSRTKArrayWrapper;
  }
  static get refObjectArrayType() {
    return GPSRTKRefStructArray
  }
  static get refObjectType() {
    return GPSRTKRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._rtk_receiver_idIntialized) {
        throw new TypeError('Invalid argument: rtk_receiver_id in GPSRTK');
      }
      if (!this._wnIntialized) {
        throw new TypeError('Invalid argument: wn in GPSRTK');
      }
      if (!this._towIntialized) {
        throw new TypeError('Invalid argument: tow in GPSRTK');
      }
      if (!this._rtk_healthIntialized) {
        throw new TypeError('Invalid argument: rtk_health in GPSRTK');
      }
      if (!this._rtk_rateIntialized) {
        throw new TypeError('Invalid argument: rtk_rate in GPSRTK');
      }
      if (!this._nsatsIntialized) {
        throw new TypeError('Invalid argument: nsats in GPSRTK');
      }
      if (!this._baseline_aIntialized) {
        throw new TypeError('Invalid argument: baseline_a in GPSRTK');
      }
      if (!this._baseline_bIntialized) {
        throw new TypeError('Invalid argument: baseline_b in GPSRTK');
      }
      if (!this._baseline_cIntialized) {
        throw new TypeError('Invalid argument: baseline_c in GPSRTK');
      }
      if (!this._accuracyIntialized) {
        throw new TypeError('Invalid argument: accuracy in GPSRTK');
      }
      if (!this._iar_num_hypothesesIntialized) {
        throw new TypeError('Invalid argument: iar_num_hypotheses in GPSRTK');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._rtk_receiver_idIntialized = true;
    this._refObject.rtk_receiver_id = refObject.rtk_receiver_id;
    this._wnIntialized = true;
    this._refObject.wn = refObject.wn;
    this._towIntialized = true;
    this._refObject.tow = refObject.tow;
    this._rtk_healthIntialized = true;
    this._refObject.rtk_health = refObject.rtk_health;
    this._rtk_rateIntialized = true;
    this._refObject.rtk_rate = refObject.rtk_rate;
    this._nsatsIntialized = true;
    this._refObject.nsats = refObject.nsats;
    this._baseline_aIntialized = true;
    this._refObject.baseline_a = refObject.baseline_a;
    this._baseline_bIntialized = true;
    this._refObject.baseline_b = refObject.baseline_b;
    this._baseline_cIntialized = true;
    this._refObject.baseline_c = refObject.baseline_c;
    this._accuracyIntialized = true;
    this._refObject.accuracy = refObject.accuracy;
    this._iar_num_hypothesesIntialized = true;
    this._refObject.iar_num_hypotheses = refObject.iar_num_hypotheses;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    GPSRTKWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'GPSRTK'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get rtk_receiver_id() {
    if (!this._rtk_receiver_idIntialized) {
      return undefined;
    }
    return this._refObject.rtk_receiver_id;
  }
  set rtk_receiver_id(value) {
    this._rtk_receiver_idIntialized = true;
    this._refObject.rtk_receiver_id = value;
  }
  get wn() {
    if (!this._wnIntialized) {
      return undefined;
    }
    return this._refObject.wn;
  }
  set wn(value) {
    this._wnIntialized = true;
    this._refObject.wn = value;
  }
  get tow() {
    if (!this._towIntialized) {
      return undefined;
    }
    return this._refObject.tow;
  }
  set tow(value) {
    this._towIntialized = true;
    this._refObject.tow = value;
  }
  get rtk_health() {
    if (!this._rtk_healthIntialized) {
      return undefined;
    }
    return this._refObject.rtk_health;
  }
  set rtk_health(value) {
    this._rtk_healthIntialized = true;
    this._refObject.rtk_health = value;
  }
  get rtk_rate() {
    if (!this._rtk_rateIntialized) {
      return undefined;
    }
    return this._refObject.rtk_rate;
  }
  set rtk_rate(value) {
    this._rtk_rateIntialized = true;
    this._refObject.rtk_rate = value;
  }
  get nsats() {
    if (!this._nsatsIntialized) {
      return undefined;
    }
    return this._refObject.nsats;
  }
  set nsats(value) {
    this._nsatsIntialized = true;
    this._refObject.nsats = value;
  }
  get baseline_a() {
    if (!this._baseline_aIntialized) {
      return undefined;
    }
    return this._refObject.baseline_a;
  }
  set baseline_a(value) {
    this._baseline_aIntialized = true;
    this._refObject.baseline_a = value;
  }
  get baseline_b() {
    if (!this._baseline_bIntialized) {
      return undefined;
    }
    return this._refObject.baseline_b;
  }
  set baseline_b(value) {
    this._baseline_bIntialized = true;
    this._refObject.baseline_b = value;
  }
  get baseline_c() {
    if (!this._baseline_cIntialized) {
      return undefined;
    }
    return this._refObject.baseline_c;
  }
  set baseline_c(value) {
    this._baseline_cIntialized = true;
    this._refObject.baseline_c = value;
  }
  get accuracy() {
    if (!this._accuracyIntialized) {
      return undefined;
    }
    return this._refObject.accuracy;
  }
  set accuracy(value) {
    this._accuracyIntialized = true;
    this._refObject.accuracy = value;
  }
  get iar_num_hypotheses() {
    if (!this._iar_num_hypothesesIntialized) {
      return undefined;
    }
    return this._refObject.iar_num_hypotheses;
  }
  set iar_num_hypotheses(value) {
    this._iar_num_hypothesesIntialized = true;
    this._refObject.iar_num_hypotheses = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GPSRTKRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._rtk_receiver_idIntialized = true;
    this._wnIntialized = true;
    this._towIntialized = true;
    this._rtk_healthIntialized = true;
    this._rtk_rateIntialized = true;
    this._nsatsIntialized = true;
    this._baseline_aIntialized = true;
    this._baseline_bIntialized = true;
    this._baseline_cIntialized = true;
    this._accuracyIntialized = true;
    this._iar_num_hypothesesIntialized = true;
  }
  copy(other) {
    this._refObject = new GPSRTKRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._rtk_receiver_idIntialized = true;
    this._wnIntialized = true;
    this._towIntialized = true;
    this._rtk_healthIntialized = true;
    this._rtk_rateIntialized = true;
    this._nsatsIntialized = true;
    this._baseline_aIntialized = true;
    this._baseline_bIntialized = true;
    this._baseline_cIntialized = true;
    this._accuracyIntialized = true;
    this._iar_num_hypothesesIntialized = true;
  }
  static get classType() {
    return GPSRTKWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"rtk_receiver_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"wn","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tow","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rtk_health","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rtk_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nsats","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baseline_a","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baseline_b","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baseline_c","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accuracy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"iar_num_hypotheses","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"GPSRTK","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GPSRTK"};
  }
  hasMember(name) {
    let memberNames = ["header","rtk_receiver_id","wn","tow","rtk_health","rtk_rate","nsats","baseline_a","baseline_b","baseline_c","accuracy","iar_num_hypotheses"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GPSRTKArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GPSRTKWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GPSRTKWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPSRTKArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPSRTKArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GPSRTKRefArray(size);
    this._refObject = new GPSRTKRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GPSRTKWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GPSRTKArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GPSRTKArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GPSRTKWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GPSRTKWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GPSRTKArrayWrapper;
  }
}
module.exports = GPSRTKWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "rtk_receiver_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "wn",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tow",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rtk_health",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rtk_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "nsats",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baseline_a",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baseline_b",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baseline_c",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accuracy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "iar_num_hypotheses",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "GPSRTK",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GPSRTK"
}
*/
