// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__GimbalManagerStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const GimbalManagerStatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  flags: primitiveTypes.uint32,
  gimbal_device_id: primitiveTypes.uint8,
  sysid_primary: primitiveTypes.uint8,
  compid_primary: primitiveTypes.uint8,
  sysid_secondary: primitiveTypes.uint8,
  compid_secondary: primitiveTypes.uint8,
});
const GimbalManagerStatusRefArray = ArrayType(GimbalManagerStatusRefStruct);
const GimbalManagerStatusRefStructArray = StructType({
  data: GimbalManagerStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GimbalManagerStatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GimbalManagerStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._flagsIntialized = true;
      this._gimbal_device_idIntialized = true;
      this._sysid_primaryIntialized = true;
      this._compid_primaryIntialized = true;
      this._sysid_secondaryIntialized = true;
      this._compid_secondaryIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GimbalManagerStatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._flagsIntialized = false;
    this._gimbal_device_idIntialized = false;
    this._sysid_primaryIntialized = false;
    this._compid_primaryIntialized = false;
    this._sysid_secondaryIntialized = false;
    this._compid_secondaryIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GimbalManagerStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GimbalManagerStatusArrayWrapper;
  }
  static get ArrayType() {
    return GimbalManagerStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return GimbalManagerStatusRefStructArray
  }
  static get refObjectType() {
    return GimbalManagerStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._flagsIntialized) {
        throw new TypeError('Invalid argument: flags in GimbalManagerStatus');
      }
      if (!this._gimbal_device_idIntialized) {
        throw new TypeError('Invalid argument: gimbal_device_id in GimbalManagerStatus');
      }
      if (!this._sysid_primaryIntialized) {
        throw new TypeError('Invalid argument: sysid_primary in GimbalManagerStatus');
      }
      if (!this._compid_primaryIntialized) {
        throw new TypeError('Invalid argument: compid_primary in GimbalManagerStatus');
      }
      if (!this._sysid_secondaryIntialized) {
        throw new TypeError('Invalid argument: sysid_secondary in GimbalManagerStatus');
      }
      if (!this._compid_secondaryIntialized) {
        throw new TypeError('Invalid argument: compid_secondary in GimbalManagerStatus');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._flagsIntialized = true;
    this._refObject.flags = refObject.flags;
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = refObject.gimbal_device_id;
    this._sysid_primaryIntialized = true;
    this._refObject.sysid_primary = refObject.sysid_primary;
    this._compid_primaryIntialized = true;
    this._refObject.compid_primary = refObject.compid_primary;
    this._sysid_secondaryIntialized = true;
    this._refObject.sysid_secondary = refObject.sysid_secondary;
    this._compid_secondaryIntialized = true;
    this._refObject.compid_secondary = refObject.compid_secondary;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    GimbalManagerStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'GimbalManagerStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get flags() {
    if (!this._flagsIntialized) {
      return undefined;
    }
    return this._refObject.flags;
  }
  set flags(value) {
    this._flagsIntialized = true;
    this._refObject.flags = value;
  }
  get gimbal_device_id() {
    if (!this._gimbal_device_idIntialized) {
      return undefined;
    }
    return this._refObject.gimbal_device_id;
  }
  set gimbal_device_id(value) {
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = value;
  }
  get sysid_primary() {
    if (!this._sysid_primaryIntialized) {
      return undefined;
    }
    return this._refObject.sysid_primary;
  }
  set sysid_primary(value) {
    this._sysid_primaryIntialized = true;
    this._refObject.sysid_primary = value;
  }
  get compid_primary() {
    if (!this._compid_primaryIntialized) {
      return undefined;
    }
    return this._refObject.compid_primary;
  }
  set compid_primary(value) {
    this._compid_primaryIntialized = true;
    this._refObject.compid_primary = value;
  }
  get sysid_secondary() {
    if (!this._sysid_secondaryIntialized) {
      return undefined;
    }
    return this._refObject.sysid_secondary;
  }
  set sysid_secondary(value) {
    this._sysid_secondaryIntialized = true;
    this._refObject.sysid_secondary = value;
  }
  get compid_secondary() {
    if (!this._compid_secondaryIntialized) {
      return undefined;
    }
    return this._refObject.compid_secondary;
  }
  set compid_secondary(value) {
    this._compid_secondaryIntialized = true;
    this._refObject.compid_secondary = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GimbalManagerStatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._flagsIntialized = true;
    this._gimbal_device_idIntialized = true;
    this._sysid_primaryIntialized = true;
    this._compid_primaryIntialized = true;
    this._sysid_secondaryIntialized = true;
    this._compid_secondaryIntialized = true;
  }
  copy(other) {
    this._refObject = new GimbalManagerStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._flagsIntialized = true;
    this._gimbal_device_idIntialized = true;
    this._sysid_primaryIntialized = true;
    this._compid_primaryIntialized = true;
    this._sysid_secondaryIntialized = true;
    this._compid_secondaryIntialized = true;
  }
  static get classType() {
    return GimbalManagerStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_RETRACT","value":1},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_NEUTRAL","value":2},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_ROLL_LOCK","value":4},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_PITCH_LOCK","value":8},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_YAW_LOCK","value":16}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gimbal_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sysid_primary","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"compid_primary","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sysid_secondary","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"compid_secondary","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"GimbalManagerStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GimbalManagerStatus"};
  }
  hasMember(name) {
    let memberNames = ["header","flags","gimbal_device_id","sysid_primary","compid_primary","sysid_secondary","compid_secondary"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GimbalManagerStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GimbalManagerStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GimbalManagerStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GimbalManagerStatusRefArray(size);
    this._refObject = new GimbalManagerStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GimbalManagerStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GimbalManagerStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GimbalManagerStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GimbalManagerStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GimbalManagerStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GimbalManagerStatusArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(GimbalManagerStatusWrapper, "GIMBAL_MANAGER_FLAGS_RETRACT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerStatusWrapper, "GIMBAL_MANAGER_FLAGS_NEUTRAL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerStatusWrapper, "GIMBAL_MANAGER_FLAGS_ROLL_LOCK", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerStatusWrapper, "GIMBAL_MANAGER_FLAGS_PITCH_LOCK", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerStatusWrapper, "GIMBAL_MANAGER_FLAGS_YAW_LOCK", {value: 16, writable: false, enumerable: true, configurable: true});
module.exports = GimbalManagerStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_RETRACT",
      "value": 1
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_NEUTRAL",
      "value": 2
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_ROLL_LOCK",
      "value": 4
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_PITCH_LOCK",
      "value": 8
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_YAW_LOCK",
      "value": 16
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gimbal_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sysid_primary",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "compid_primary",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sysid_secondary",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "compid_secondary",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "GimbalManagerStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GimbalManagerStatus"
}
*/
