// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__ActuatorControl.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const ActuatorControlRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  group_mix: primitiveTypes.uint8,
  controls: ArrayType(primitiveTypes.float32, 8),
});
const ActuatorControlRefArray = ArrayType(ActuatorControlRefStruct);
const ActuatorControlRefStructArray = StructType({
  data: ActuatorControlRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ActuatorControlWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ActuatorControlRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._group_mixIntialized = true;
      this._wrapperFields.controls = Float32Wrapper.createArray();
      this._wrapperFields.controls.copy(other._wrapperFields.controls);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ActuatorControlRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._group_mixIntialized = false;
    this._wrapperFields.controls = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ActuatorControlWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ActuatorControlArrayWrapper;
  }
  static get ArrayType() {
    return ActuatorControlArrayWrapper;
  }
  static get refObjectArrayType() {
    return ActuatorControlRefStructArray
  }
  static get refObjectType() {
    return ActuatorControlRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._group_mixIntialized) {
        throw new TypeError('Invalid argument: group_mix in ActuatorControl');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    for (let i = 0; i < this._wrapperFields.controls.data.length; i++) {
      this._refObject.controls[i] = this._wrapperFields.controls.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._group_mixIntialized = true;
    this._refObject.group_mix = refObject.group_mix;
    this._refObject.controls = refObject.controls;
    this._wrapperFields.controls.fill(refObject.controls.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    ActuatorControlWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'ActuatorControl'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get group_mix() {
    if (!this._group_mixIntialized) {
      return undefined;
    }
    return this._refObject.group_mix;
  }
  set group_mix(value) {
    this._group_mixIntialized = true;
    this._refObject.group_mix = value;
  }
  get controls() {
    return this._wrapperFields['controls'].data;
  }
  set controls(value) {
    if (value.length !== 8) {
      throw new RangeError('The length of the array must be 8.');
    }
    this._wrapperFields['controls'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ActuatorControlRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._group_mixIntialized = true;
    this._wrapperFields.controls.fill(refObject.controls.toArray());
  }
  copy(other) {
    this._refObject = new ActuatorControlRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._group_mixIntialized = true;
    this._wrapperFields.controls.copy(other._wrapperFields.controls);
  }
  static get classType() {
    return ActuatorControlWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"PX4_MIX_FLIGHT_CONTROL","value":0},{"type":"uint8","name":"PX4_MIX_FLIGHT_CONTROL_VTOL_ALT","value":1},{"type":"uint8","name":"PX4_MIX_PAYLOAD","value":2},{"type":"uint8","name":"PX4_MIX_MANUAL_PASSTHROUGH","value":3}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"group_mix","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"controls","type":{"isArray":true,"arraySize":8,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"ActuatorControl","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ActuatorControl"};
  }
  hasMember(name) {
    let memberNames = ["header","group_mix","controls"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ActuatorControlArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ActuatorControlWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ActuatorControlWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorControlArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorControlArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ActuatorControlRefArray(size);
    this._refObject = new ActuatorControlRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ActuatorControlWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ActuatorControlArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ActuatorControlArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ActuatorControlWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ActuatorControlWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ActuatorControlArrayWrapper;
  }
}
// Define constants (4 in total)
Object.defineProperty(ActuatorControlWrapper, "PX4_MIX_FLIGHT_CONTROL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ActuatorControlWrapper, "PX4_MIX_FLIGHT_CONTROL_VTOL_ALT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ActuatorControlWrapper, "PX4_MIX_PAYLOAD", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ActuatorControlWrapper, "PX4_MIX_MANUAL_PASSTHROUGH", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = ActuatorControlWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "PX4_MIX_FLIGHT_CONTROL",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "PX4_MIX_FLIGHT_CONTROL_VTOL_ALT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "PX4_MIX_PAYLOAD",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "PX4_MIX_MANUAL_PASSTHROUGH",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "group_mix",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "controls",
      "type": {
        "isArray": true,
        "arraySize": 8,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "ActuatorControl",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ActuatorControl"
}
*/
