// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__HilGPS.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const GeoPointWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__GeoPoint.js');
const HilGPSRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  fix_type: primitiveTypes.uint8,
  geo: GeoPointWrapper.refObjectType,
  eph: primitiveTypes.uint16,
  epv: primitiveTypes.uint16,
  vel: primitiveTypes.uint16,
  vn: primitiveTypes.int16,
  ve: primitiveTypes.int16,
  vd: primitiveTypes.int16,
  cog: primitiveTypes.uint16,
  satellites_visible: primitiveTypes.uint8,
});
const HilGPSRefArray = ArrayType(HilGPSRefStruct);
const HilGPSRefStructArray = StructType({
  data: HilGPSRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HilGPSWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HilGPSRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._fix_typeIntialized = true;
      this._wrapperFields.geo =  new GeoPointWrapper(other._wrapperFields.geo);
      this._ephIntialized = true;
      this._epvIntialized = true;
      this._velIntialized = true;
      this._vnIntialized = true;
      this._veIntialized = true;
      this._vdIntialized = true;
      this._cogIntialized = true;
      this._satellites_visibleIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HilGPSRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._fix_typeIntialized = false;
    this._wrapperFields.geo = new GeoPointWrapper();
    this._ephIntialized = false;
    this._epvIntialized = false;
    this._velIntialized = false;
    this._vnIntialized = false;
    this._veIntialized = false;
    this._vdIntialized = false;
    this._cogIntialized = false;
    this._satellites_visibleIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new HilGPSWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HilGPSArrayWrapper;
  }
  static get ArrayType() {
    return HilGPSArrayWrapper;
  }
  static get refObjectArrayType() {
    return HilGPSRefStructArray
  }
  static get refObjectType() {
    return HilGPSRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._fix_typeIntialized) {
        throw new TypeError('Invalid argument: fix_type in HilGPS');
      }
      if (!this._ephIntialized) {
        throw new TypeError('Invalid argument: eph in HilGPS');
      }
      if (!this._epvIntialized) {
        throw new TypeError('Invalid argument: epv in HilGPS');
      }
      if (!this._velIntialized) {
        throw new TypeError('Invalid argument: vel in HilGPS');
      }
      if (!this._vnIntialized) {
        throw new TypeError('Invalid argument: vn in HilGPS');
      }
      if (!this._veIntialized) {
        throw new TypeError('Invalid argument: ve in HilGPS');
      }
      if (!this._vdIntialized) {
        throw new TypeError('Invalid argument: vd in HilGPS');
      }
      if (!this._cogIntialized) {
        throw new TypeError('Invalid argument: cog in HilGPS');
      }
      if (!this._satellites_visibleIntialized) {
        throw new TypeError('Invalid argument: satellites_visible in HilGPS');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.geo.freeze(own, checkConsistency);
    this._refObject.geo = this._wrapperFields.geo.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._fix_typeIntialized = true;
    this._refObject.fix_type = refObject.fix_type;
    this._wrapperFields.geo.copyRefObject(refObject.geo);
    this._ephIntialized = true;
    this._refObject.eph = refObject.eph;
    this._epvIntialized = true;
    this._refObject.epv = refObject.epv;
    this._velIntialized = true;
    this._refObject.vel = refObject.vel;
    this._vnIntialized = true;
    this._refObject.vn = refObject.vn;
    this._veIntialized = true;
    this._refObject.ve = refObject.ve;
    this._vdIntialized = true;
    this._refObject.vd = refObject.vd;
    this._cogIntialized = true;
    this._refObject.cog = refObject.cog;
    this._satellites_visibleIntialized = true;
    this._refObject.satellites_visible = refObject.satellites_visible;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    GeoPointWrapper.freeStruct(refObject.geo);
  }
  static destoryRawROS(msg) {
    HilGPSWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'HilGPS'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get fix_type() {
    if (!this._fix_typeIntialized) {
      return undefined;
    }
    return this._refObject.fix_type;
  }
  set fix_type(value) {
    this._fix_typeIntialized = true;
    this._refObject.fix_type = value;
  }
  get geo() {
    return this._wrapperFields.geo;
  }
  set geo(value) {
    if (value instanceof GeoPointWrapper) {
      this._wrapperFields.geo.copy(value);
    } else {
      this._wrapperFields.geo.copy(new GeoPointWrapper(value));
    }
  }
  get eph() {
    if (!this._ephIntialized) {
      return undefined;
    }
    return this._refObject.eph;
  }
  set eph(value) {
    this._ephIntialized = true;
    this._refObject.eph = value;
  }
  get epv() {
    if (!this._epvIntialized) {
      return undefined;
    }
    return this._refObject.epv;
  }
  set epv(value) {
    this._epvIntialized = true;
    this._refObject.epv = value;
  }
  get vel() {
    if (!this._velIntialized) {
      return undefined;
    }
    return this._refObject.vel;
  }
  set vel(value) {
    this._velIntialized = true;
    this._refObject.vel = value;
  }
  get vn() {
    if (!this._vnIntialized) {
      return undefined;
    }
    return this._refObject.vn;
  }
  set vn(value) {
    this._vnIntialized = true;
    this._refObject.vn = value;
  }
  get ve() {
    if (!this._veIntialized) {
      return undefined;
    }
    return this._refObject.ve;
  }
  set ve(value) {
    this._veIntialized = true;
    this._refObject.ve = value;
  }
  get vd() {
    if (!this._vdIntialized) {
      return undefined;
    }
    return this._refObject.vd;
  }
  set vd(value) {
    this._vdIntialized = true;
    this._refObject.vd = value;
  }
  get cog() {
    if (!this._cogIntialized) {
      return undefined;
    }
    return this._refObject.cog;
  }
  set cog(value) {
    this._cogIntialized = true;
    this._refObject.cog = value;
  }
  get satellites_visible() {
    if (!this._satellites_visibleIntialized) {
      return undefined;
    }
    return this._refObject.satellites_visible;
  }
  set satellites_visible(value) {
    this._satellites_visibleIntialized = true;
    this._refObject.satellites_visible = value;
  }
  copyRefObject(refObject) {
    this._refObject = new HilGPSRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._fix_typeIntialized = true;
    this._wrapperFields.geo.copyRefObject(this._refObject.geo);
    this._ephIntialized = true;
    this._epvIntialized = true;
    this._velIntialized = true;
    this._vnIntialized = true;
    this._veIntialized = true;
    this._vdIntialized = true;
    this._cogIntialized = true;
    this._satellites_visibleIntialized = true;
  }
  copy(other) {
    this._refObject = new HilGPSRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._fix_typeIntialized = true;
    this._wrapperFields.geo.copy(other._wrapperFields.geo);
    this._ephIntialized = true;
    this._epvIntialized = true;
    this._velIntialized = true;
    this._vnIntialized = true;
    this._veIntialized = true;
    this._vdIntialized = true;
    this._cogIntialized = true;
    this._satellites_visibleIntialized = true;
  }
  static get classType() {
    return HilGPSWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"fix_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"geo","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"GeoPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"eph","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"epv","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vn","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ve","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vd","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cog","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"satellites_visible","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"HilGPS","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HilGPS"};
  }
  hasMember(name) {
    let memberNames = ["header","fix_type","geo","eph","epv","vel","vn","ve","vd","cog","satellites_visible"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HilGPSArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HilGPSWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HilGPSWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HilGPSArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HilGPSArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HilGPSRefArray(size);
    this._refObject = new HilGPSRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HilGPSWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HilGPSArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HilGPSArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HilGPSWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HilGPSWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HilGPSArrayWrapper;
  }
}
module.exports = HilGPSWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "fix_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "geo",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "GeoPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "eph",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "epv",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vn",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ve",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vd",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cog",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "satellites_visible",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "HilGPS",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HilGPS"
}
*/
