// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__CommandInt_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CommandInt_RequestRefStruct = StructType({
  broadcast: primitiveTypes.bool,
  frame: primitiveTypes.uint8,
  command: primitiveTypes.uint16,
  current: primitiveTypes.uint8,
  autocontinue: primitiveTypes.uint8,
  param1: primitiveTypes.float32,
  param2: primitiveTypes.float32,
  param3: primitiveTypes.float32,
  param4: primitiveTypes.float32,
  x: primitiveTypes.int32,
  y: primitiveTypes.int32,
  z: primitiveTypes.float32,
});
const CommandInt_RequestRefArray = ArrayType(CommandInt_RequestRefStruct);
const CommandInt_RequestRefStructArray = StructType({
  data: CommandInt_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CommandInt_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CommandInt_RequestRefStruct(other._refObject.toObject());
      this._broadcastIntialized = true;
      this._frameIntialized = true;
      this._commandIntialized = true;
      this._currentIntialized = true;
      this._autocontinueIntialized = true;
      this._param1Intialized = true;
      this._param2Intialized = true;
      this._param3Intialized = true;
      this._param4Intialized = true;
      this._xIntialized = true;
      this._yIntialized = true;
      this._zIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CommandInt_RequestRefStruct();
    this._broadcastIntialized = false;
    this._frameIntialized = false;
    this._commandIntialized = false;
    this._currentIntialized = false;
    this._autocontinueIntialized = false;
    this._param1Intialized = false;
    this._param2Intialized = false;
    this._param3Intialized = false;
    this._param4Intialized = false;
    this._xIntialized = false;
    this._yIntialized = false;
    this._zIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CommandInt_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CommandInt_RequestArrayWrapper;
  }
  static get ArrayType() {
    return CommandInt_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return CommandInt_RequestRefStructArray
  }
  static get refObjectType() {
    return CommandInt_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._broadcastIntialized) {
        throw new TypeError('Invalid argument: broadcast in CommandInt_Request');
      }
      if (!this._frameIntialized) {
        throw new TypeError('Invalid argument: frame in CommandInt_Request');
      }
      if (!this._commandIntialized) {
        throw new TypeError('Invalid argument: command in CommandInt_Request');
      }
      if (!this._currentIntialized) {
        throw new TypeError('Invalid argument: current in CommandInt_Request');
      }
      if (!this._autocontinueIntialized) {
        throw new TypeError('Invalid argument: autocontinue in CommandInt_Request');
      }
      if (!this._param1Intialized) {
        throw new TypeError('Invalid argument: param1 in CommandInt_Request');
      }
      if (!this._param2Intialized) {
        throw new TypeError('Invalid argument: param2 in CommandInt_Request');
      }
      if (!this._param3Intialized) {
        throw new TypeError('Invalid argument: param3 in CommandInt_Request');
      }
      if (!this._param4Intialized) {
        throw new TypeError('Invalid argument: param4 in CommandInt_Request');
      }
      if (!this._xIntialized) {
        throw new TypeError('Invalid argument: x in CommandInt_Request');
      }
      if (!this._yIntialized) {
        throw new TypeError('Invalid argument: y in CommandInt_Request');
      }
      if (!this._zIntialized) {
        throw new TypeError('Invalid argument: z in CommandInt_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._broadcastIntialized = true;
    this._refObject.broadcast = refObject.broadcast;
    this._frameIntialized = true;
    this._refObject.frame = refObject.frame;
    this._commandIntialized = true;
    this._refObject.command = refObject.command;
    this._currentIntialized = true;
    this._refObject.current = refObject.current;
    this._autocontinueIntialized = true;
    this._refObject.autocontinue = refObject.autocontinue;
    this._param1Intialized = true;
    this._refObject.param1 = refObject.param1;
    this._param2Intialized = true;
    this._refObject.param2 = refObject.param2;
    this._param3Intialized = true;
    this._refObject.param3 = refObject.param3;
    this._param4Intialized = true;
    this._refObject.param4 = refObject.param4;
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._zIntialized = true;
    this._refObject.z = refObject.z;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    CommandInt_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'CommandInt_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get broadcast() {
    if (!this._broadcastIntialized) {
      return undefined;
    }
    return this._refObject.broadcast;
  }
  set broadcast(value) {
    this._broadcastIntialized = true;
    this._refObject.broadcast = value;
  }
  get frame() {
    if (!this._frameIntialized) {
      return undefined;
    }
    return this._refObject.frame;
  }
  set frame(value) {
    this._frameIntialized = true;
    this._refObject.frame = value;
  }
  get command() {
    if (!this._commandIntialized) {
      return undefined;
    }
    return this._refObject.command;
  }
  set command(value) {
    this._commandIntialized = true;
    this._refObject.command = value;
  }
  get current() {
    if (!this._currentIntialized) {
      return undefined;
    }
    return this._refObject.current;
  }
  set current(value) {
    this._currentIntialized = true;
    this._refObject.current = value;
  }
  get autocontinue() {
    if (!this._autocontinueIntialized) {
      return undefined;
    }
    return this._refObject.autocontinue;
  }
  set autocontinue(value) {
    this._autocontinueIntialized = true;
    this._refObject.autocontinue = value;
  }
  get param1() {
    if (!this._param1Intialized) {
      return undefined;
    }
    return this._refObject.param1;
  }
  set param1(value) {
    this._param1Intialized = true;
    this._refObject.param1 = value;
  }
  get param2() {
    if (!this._param2Intialized) {
      return undefined;
    }
    return this._refObject.param2;
  }
  set param2(value) {
    this._param2Intialized = true;
    this._refObject.param2 = value;
  }
  get param3() {
    if (!this._param3Intialized) {
      return undefined;
    }
    return this._refObject.param3;
  }
  set param3(value) {
    this._param3Intialized = true;
    this._refObject.param3 = value;
  }
  get param4() {
    if (!this._param4Intialized) {
      return undefined;
    }
    return this._refObject.param4;
  }
  set param4(value) {
    this._param4Intialized = true;
    this._refObject.param4 = value;
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get z() {
    if (!this._zIntialized) {
      return undefined;
    }
    return this._refObject.z;
  }
  set z(value) {
    this._zIntialized = true;
    this._refObject.z = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CommandInt_RequestRefStruct(refObject.toObject());
    this._broadcastIntialized = true;
    this._frameIntialized = true;
    this._commandIntialized = true;
    this._currentIntialized = true;
    this._autocontinueIntialized = true;
    this._param1Intialized = true;
    this._param2Intialized = true;
    this._param3Intialized = true;
    this._param4Intialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
  }
  copy(other) {
    this._refObject = new CommandInt_RequestRefStruct(other._refObject.toObject());
    this._broadcastIntialized = true;
    this._frameIntialized = true;
    this._commandIntialized = true;
    this._currentIntialized = true;
    this._autocontinueIntialized = true;
    this._param1Intialized = true;
    this._param2Intialized = true;
    this._param3Intialized = true;
    this._param4Intialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
  }
  static get classType() {
    return CommandInt_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"broadcast","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"command","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"autocontinue","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"CommandInt_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CommandInt_Request"};
  }
  hasMember(name) {
    let memberNames = ["broadcast","frame","command","current","autocontinue","param1","param2","param3","param4","x","y","z"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CommandInt_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CommandInt_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CommandInt_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandInt_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandInt_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CommandInt_RequestRefArray(size);
    this._refObject = new CommandInt_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CommandInt_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CommandInt_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CommandInt_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CommandInt_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CommandInt_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CommandInt_RequestArrayWrapper;
  }
}
module.exports = CommandInt_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "broadcast",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "command",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "current",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "autocontinue",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "CommandInt_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CommandInt_Request"
}
*/
