// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__CommandAck_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CommandAck_RequestRefStruct = StructType({
  command: primitiveTypes.uint16,
  result: primitiveTypes.uint8,
  progress: primitiveTypes.uint8,
  result_param2: primitiveTypes.uint32,
});
const CommandAck_RequestRefArray = ArrayType(CommandAck_RequestRefStruct);
const CommandAck_RequestRefStructArray = StructType({
  data: CommandAck_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CommandAck_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CommandAck_RequestRefStruct(other._refObject.toObject());
      this._commandIntialized = true;
      this._resultIntialized = true;
      this._progressIntialized = true;
      this._result_param2Intialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CommandAck_RequestRefStruct();
    this._commandIntialized = false;
    this._resultIntialized = false;
    this._progressIntialized = false;
    this._result_param2Intialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CommandAck_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CommandAck_RequestArrayWrapper;
  }
  static get ArrayType() {
    return CommandAck_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return CommandAck_RequestRefStructArray
  }
  static get refObjectType() {
    return CommandAck_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._commandIntialized) {
        throw new TypeError('Invalid argument: command in CommandAck_Request');
      }
      if (!this._resultIntialized) {
        throw new TypeError('Invalid argument: result in CommandAck_Request');
      }
      if (!this._progressIntialized) {
        throw new TypeError('Invalid argument: progress in CommandAck_Request');
      }
      if (!this._result_param2Intialized) {
        throw new TypeError('Invalid argument: result_param2 in CommandAck_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._commandIntialized = true;
    this._refObject.command = refObject.command;
    this._resultIntialized = true;
    this._refObject.result = refObject.result;
    this._progressIntialized = true;
    this._refObject.progress = refObject.progress;
    this._result_param2Intialized = true;
    this._refObject.result_param2 = refObject.result_param2;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    CommandAck_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'CommandAck_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get command() {
    if (!this._commandIntialized) {
      return undefined;
    }
    return this._refObject.command;
  }
  set command(value) {
    this._commandIntialized = true;
    this._refObject.command = value;
  }
  get result() {
    if (!this._resultIntialized) {
      return undefined;
    }
    return this._refObject.result;
  }
  set result(value) {
    this._resultIntialized = true;
    this._refObject.result = value;
  }
  get progress() {
    if (!this._progressIntialized) {
      return undefined;
    }
    return this._refObject.progress;
  }
  set progress(value) {
    this._progressIntialized = true;
    this._refObject.progress = value;
  }
  get result_param2() {
    if (!this._result_param2Intialized) {
      return undefined;
    }
    return this._refObject.result_param2;
  }
  set result_param2(value) {
    this._result_param2Intialized = true;
    this._refObject.result_param2 = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CommandAck_RequestRefStruct(refObject.toObject());
    this._commandIntialized = true;
    this._resultIntialized = true;
    this._progressIntialized = true;
    this._result_param2Intialized = true;
  }
  copy(other) {
    this._refObject = new CommandAck_RequestRefStruct(other._refObject.toObject());
    this._commandIntialized = true;
    this._resultIntialized = true;
    this._progressIntialized = true;
    this._result_param2Intialized = true;
  }
  static get classType() {
    return CommandAck_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"command","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"result","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"progress","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"result_param2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"CommandAck_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CommandAck_Request"};
  }
  hasMember(name) {
    let memberNames = ["command","result","progress","result_param2"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CommandAck_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CommandAck_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CommandAck_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandAck_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandAck_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CommandAck_RequestRefArray(size);
    this._refObject = new CommandAck_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CommandAck_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CommandAck_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CommandAck_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CommandAck_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CommandAck_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CommandAck_RequestArrayWrapper;
  }
}
module.exports = CommandAck_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "command",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "result",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "progress",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "result_param2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "CommandAck_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CommandAck_Request"
}
*/
