// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__EstimatorStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const EstimatorStatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  attitude_status_flag: primitiveTypes.bool,
  velocity_horiz_status_flag: primitiveTypes.bool,
  velocity_vert_status_flag: primitiveTypes.bool,
  pos_horiz_rel_status_flag: primitiveTypes.bool,
  pos_horiz_abs_status_flag: primitiveTypes.bool,
  pos_vert_abs_status_flag: primitiveTypes.bool,
  pos_vert_agl_status_flag: primitiveTypes.bool,
  const_pos_mode_status_flag: primitiveTypes.bool,
  pred_pos_horiz_rel_status_flag: primitiveTypes.bool,
  pred_pos_horiz_abs_status_flag: primitiveTypes.bool,
  gps_glitch_status_flag: primitiveTypes.bool,
  accel_error_status_flag: primitiveTypes.bool,
});
const EstimatorStatusRefArray = ArrayType(EstimatorStatusRefStruct);
const EstimatorStatusRefStructArray = StructType({
  data: EstimatorStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EstimatorStatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EstimatorStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._attitude_status_flagIntialized = true;
      this._velocity_horiz_status_flagIntialized = true;
      this._velocity_vert_status_flagIntialized = true;
      this._pos_horiz_rel_status_flagIntialized = true;
      this._pos_horiz_abs_status_flagIntialized = true;
      this._pos_vert_abs_status_flagIntialized = true;
      this._pos_vert_agl_status_flagIntialized = true;
      this._const_pos_mode_status_flagIntialized = true;
      this._pred_pos_horiz_rel_status_flagIntialized = true;
      this._pred_pos_horiz_abs_status_flagIntialized = true;
      this._gps_glitch_status_flagIntialized = true;
      this._accel_error_status_flagIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EstimatorStatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._attitude_status_flagIntialized = false;
    this._velocity_horiz_status_flagIntialized = false;
    this._velocity_vert_status_flagIntialized = false;
    this._pos_horiz_rel_status_flagIntialized = false;
    this._pos_horiz_abs_status_flagIntialized = false;
    this._pos_vert_abs_status_flagIntialized = false;
    this._pos_vert_agl_status_flagIntialized = false;
    this._const_pos_mode_status_flagIntialized = false;
    this._pred_pos_horiz_rel_status_flagIntialized = false;
    this._pred_pos_horiz_abs_status_flagIntialized = false;
    this._gps_glitch_status_flagIntialized = false;
    this._accel_error_status_flagIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EstimatorStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EstimatorStatusArrayWrapper;
  }
  static get ArrayType() {
    return EstimatorStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return EstimatorStatusRefStructArray
  }
  static get refObjectType() {
    return EstimatorStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._attitude_status_flagIntialized) {
        throw new TypeError('Invalid argument: attitude_status_flag in EstimatorStatus');
      }
      if (!this._velocity_horiz_status_flagIntialized) {
        throw new TypeError('Invalid argument: velocity_horiz_status_flag in EstimatorStatus');
      }
      if (!this._velocity_vert_status_flagIntialized) {
        throw new TypeError('Invalid argument: velocity_vert_status_flag in EstimatorStatus');
      }
      if (!this._pos_horiz_rel_status_flagIntialized) {
        throw new TypeError('Invalid argument: pos_horiz_rel_status_flag in EstimatorStatus');
      }
      if (!this._pos_horiz_abs_status_flagIntialized) {
        throw new TypeError('Invalid argument: pos_horiz_abs_status_flag in EstimatorStatus');
      }
      if (!this._pos_vert_abs_status_flagIntialized) {
        throw new TypeError('Invalid argument: pos_vert_abs_status_flag in EstimatorStatus');
      }
      if (!this._pos_vert_agl_status_flagIntialized) {
        throw new TypeError('Invalid argument: pos_vert_agl_status_flag in EstimatorStatus');
      }
      if (!this._const_pos_mode_status_flagIntialized) {
        throw new TypeError('Invalid argument: const_pos_mode_status_flag in EstimatorStatus');
      }
      if (!this._pred_pos_horiz_rel_status_flagIntialized) {
        throw new TypeError('Invalid argument: pred_pos_horiz_rel_status_flag in EstimatorStatus');
      }
      if (!this._pred_pos_horiz_abs_status_flagIntialized) {
        throw new TypeError('Invalid argument: pred_pos_horiz_abs_status_flag in EstimatorStatus');
      }
      if (!this._gps_glitch_status_flagIntialized) {
        throw new TypeError('Invalid argument: gps_glitch_status_flag in EstimatorStatus');
      }
      if (!this._accel_error_status_flagIntialized) {
        throw new TypeError('Invalid argument: accel_error_status_flag in EstimatorStatus');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._attitude_status_flagIntialized = true;
    this._refObject.attitude_status_flag = refObject.attitude_status_flag;
    this._velocity_horiz_status_flagIntialized = true;
    this._refObject.velocity_horiz_status_flag = refObject.velocity_horiz_status_flag;
    this._velocity_vert_status_flagIntialized = true;
    this._refObject.velocity_vert_status_flag = refObject.velocity_vert_status_flag;
    this._pos_horiz_rel_status_flagIntialized = true;
    this._refObject.pos_horiz_rel_status_flag = refObject.pos_horiz_rel_status_flag;
    this._pos_horiz_abs_status_flagIntialized = true;
    this._refObject.pos_horiz_abs_status_flag = refObject.pos_horiz_abs_status_flag;
    this._pos_vert_abs_status_flagIntialized = true;
    this._refObject.pos_vert_abs_status_flag = refObject.pos_vert_abs_status_flag;
    this._pos_vert_agl_status_flagIntialized = true;
    this._refObject.pos_vert_agl_status_flag = refObject.pos_vert_agl_status_flag;
    this._const_pos_mode_status_flagIntialized = true;
    this._refObject.const_pos_mode_status_flag = refObject.const_pos_mode_status_flag;
    this._pred_pos_horiz_rel_status_flagIntialized = true;
    this._refObject.pred_pos_horiz_rel_status_flag = refObject.pred_pos_horiz_rel_status_flag;
    this._pred_pos_horiz_abs_status_flagIntialized = true;
    this._refObject.pred_pos_horiz_abs_status_flag = refObject.pred_pos_horiz_abs_status_flag;
    this._gps_glitch_status_flagIntialized = true;
    this._refObject.gps_glitch_status_flag = refObject.gps_glitch_status_flag;
    this._accel_error_status_flagIntialized = true;
    this._refObject.accel_error_status_flag = refObject.accel_error_status_flag;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    EstimatorStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'EstimatorStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get attitude_status_flag() {
    if (!this._attitude_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.attitude_status_flag;
  }
  set attitude_status_flag(value) {
    this._attitude_status_flagIntialized = true;
    this._refObject.attitude_status_flag = value;
  }
  get velocity_horiz_status_flag() {
    if (!this._velocity_horiz_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.velocity_horiz_status_flag;
  }
  set velocity_horiz_status_flag(value) {
    this._velocity_horiz_status_flagIntialized = true;
    this._refObject.velocity_horiz_status_flag = value;
  }
  get velocity_vert_status_flag() {
    if (!this._velocity_vert_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.velocity_vert_status_flag;
  }
  set velocity_vert_status_flag(value) {
    this._velocity_vert_status_flagIntialized = true;
    this._refObject.velocity_vert_status_flag = value;
  }
  get pos_horiz_rel_status_flag() {
    if (!this._pos_horiz_rel_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.pos_horiz_rel_status_flag;
  }
  set pos_horiz_rel_status_flag(value) {
    this._pos_horiz_rel_status_flagIntialized = true;
    this._refObject.pos_horiz_rel_status_flag = value;
  }
  get pos_horiz_abs_status_flag() {
    if (!this._pos_horiz_abs_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.pos_horiz_abs_status_flag;
  }
  set pos_horiz_abs_status_flag(value) {
    this._pos_horiz_abs_status_flagIntialized = true;
    this._refObject.pos_horiz_abs_status_flag = value;
  }
  get pos_vert_abs_status_flag() {
    if (!this._pos_vert_abs_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.pos_vert_abs_status_flag;
  }
  set pos_vert_abs_status_flag(value) {
    this._pos_vert_abs_status_flagIntialized = true;
    this._refObject.pos_vert_abs_status_flag = value;
  }
  get pos_vert_agl_status_flag() {
    if (!this._pos_vert_agl_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.pos_vert_agl_status_flag;
  }
  set pos_vert_agl_status_flag(value) {
    this._pos_vert_agl_status_flagIntialized = true;
    this._refObject.pos_vert_agl_status_flag = value;
  }
  get const_pos_mode_status_flag() {
    if (!this._const_pos_mode_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.const_pos_mode_status_flag;
  }
  set const_pos_mode_status_flag(value) {
    this._const_pos_mode_status_flagIntialized = true;
    this._refObject.const_pos_mode_status_flag = value;
  }
  get pred_pos_horiz_rel_status_flag() {
    if (!this._pred_pos_horiz_rel_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.pred_pos_horiz_rel_status_flag;
  }
  set pred_pos_horiz_rel_status_flag(value) {
    this._pred_pos_horiz_rel_status_flagIntialized = true;
    this._refObject.pred_pos_horiz_rel_status_flag = value;
  }
  get pred_pos_horiz_abs_status_flag() {
    if (!this._pred_pos_horiz_abs_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.pred_pos_horiz_abs_status_flag;
  }
  set pred_pos_horiz_abs_status_flag(value) {
    this._pred_pos_horiz_abs_status_flagIntialized = true;
    this._refObject.pred_pos_horiz_abs_status_flag = value;
  }
  get gps_glitch_status_flag() {
    if (!this._gps_glitch_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.gps_glitch_status_flag;
  }
  set gps_glitch_status_flag(value) {
    this._gps_glitch_status_flagIntialized = true;
    this._refObject.gps_glitch_status_flag = value;
  }
  get accel_error_status_flag() {
    if (!this._accel_error_status_flagIntialized) {
      return undefined;
    }
    return this._refObject.accel_error_status_flag;
  }
  set accel_error_status_flag(value) {
    this._accel_error_status_flagIntialized = true;
    this._refObject.accel_error_status_flag = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EstimatorStatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._attitude_status_flagIntialized = true;
    this._velocity_horiz_status_flagIntialized = true;
    this._velocity_vert_status_flagIntialized = true;
    this._pos_horiz_rel_status_flagIntialized = true;
    this._pos_horiz_abs_status_flagIntialized = true;
    this._pos_vert_abs_status_flagIntialized = true;
    this._pos_vert_agl_status_flagIntialized = true;
    this._const_pos_mode_status_flagIntialized = true;
    this._pred_pos_horiz_rel_status_flagIntialized = true;
    this._pred_pos_horiz_abs_status_flagIntialized = true;
    this._gps_glitch_status_flagIntialized = true;
    this._accel_error_status_flagIntialized = true;
  }
  copy(other) {
    this._refObject = new EstimatorStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._attitude_status_flagIntialized = true;
    this._velocity_horiz_status_flagIntialized = true;
    this._velocity_vert_status_flagIntialized = true;
    this._pos_horiz_rel_status_flagIntialized = true;
    this._pos_horiz_abs_status_flagIntialized = true;
    this._pos_vert_abs_status_flagIntialized = true;
    this._pos_vert_agl_status_flagIntialized = true;
    this._const_pos_mode_status_flagIntialized = true;
    this._pred_pos_horiz_rel_status_flagIntialized = true;
    this._pred_pos_horiz_abs_status_flagIntialized = true;
    this._gps_glitch_status_flagIntialized = true;
    this._accel_error_status_flagIntialized = true;
  }
  static get classType() {
    return EstimatorStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"attitude_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity_horiz_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity_vert_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pos_horiz_rel_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pos_horiz_abs_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pos_vert_abs_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pos_vert_agl_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"const_pos_mode_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pred_pos_horiz_rel_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pred_pos_horiz_abs_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_glitch_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accel_error_status_flag","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"EstimatorStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EstimatorStatus"};
  }
  hasMember(name) {
    let memberNames = ["header","attitude_status_flag","velocity_horiz_status_flag","velocity_vert_status_flag","pos_horiz_rel_status_flag","pos_horiz_abs_status_flag","pos_vert_abs_status_flag","pos_vert_agl_status_flag","const_pos_mode_status_flag","pred_pos_horiz_rel_status_flag","pred_pos_horiz_abs_status_flag","gps_glitch_status_flag","accel_error_status_flag"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EstimatorStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EstimatorStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EstimatorStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EstimatorStatusRefArray(size);
    this._refObject = new EstimatorStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EstimatorStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EstimatorStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EstimatorStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EstimatorStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EstimatorStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EstimatorStatusArrayWrapper;
  }
}
module.exports = EstimatorStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "attitude_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velocity_horiz_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velocity_vert_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pos_horiz_rel_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pos_horiz_abs_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pos_vert_abs_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pos_vert_agl_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "const_pos_mode_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pred_pos_horiz_rel_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pred_pos_horiz_abs_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_glitch_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accel_error_status_flag",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "EstimatorStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EstimatorStatus"
}
*/
