// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__PlayTuneV2.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PlayTuneV2RefStruct = StructType({
  format: primitiveTypes.uint8,
  tune: primitiveTypes.string,
});
const PlayTuneV2RefArray = ArrayType(PlayTuneV2RefStruct);
const PlayTuneV2RefStructArray = StructType({
  data: PlayTuneV2RefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PlayTuneV2Wrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PlayTuneV2RefStruct(other._refObject.toObject());
      this._formatIntialized = true;
      this._tuneIntialized = true;
      this._wrapperFields.tune =  new StringWrapper(other._wrapperFields.tune);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PlayTuneV2RefStruct();
    this._formatIntialized = false;
    this._tuneIntialized = false;
    this._wrapperFields.tune = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new PlayTuneV2Wrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PlayTuneV2ArrayWrapper;
  }
  static get ArrayType() {
    return PlayTuneV2ArrayWrapper;
  }
  static get refObjectArrayType() {
    return PlayTuneV2RefStructArray
  }
  static get refObjectType() {
    return PlayTuneV2RefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._formatIntialized) {
        throw new TypeError('Invalid argument: format in PlayTuneV2');
      }
      if (!this._tuneIntialized) {
        throw new TypeError('Invalid argument: tune in PlayTuneV2');
      }
    }
    if (own) {
      this._wrapperFields.tune.freeze(own, checkConsistency);
    }
    this._refObject.tune = this._wrapperFields.tune.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._formatIntialized = true;
    this._refObject.format = refObject.format;
    this._tuneIntialized = true;
    this._wrapperFields.tune.data = refObject.tune.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.tune);
  }
  static destoryRawROS(msg) {
    PlayTuneV2Wrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'PlayTuneV2'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get format() {
    if (!this._formatIntialized) {
      return undefined;
    }
    return this._refObject.format;
  }
  set format(value) {
    this._formatIntialized = true;
    this._refObject.format = value;
  }
  get tune() {
    if (!this._tuneIntialized) {
      return undefined;
    }
    return this._wrapperFields.tune.data;
  }
  set tune(value) {
    this._tuneIntialized = true;
    this._wrapperFields.tune.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PlayTuneV2RefStruct(refObject.toObject());
    this._formatIntialized = true;
    this._tuneIntialized = true;
    this._wrapperFields.tune.copyRefObject(this._refObject.tune);
  }
  copy(other) {
    this._refObject = new PlayTuneV2RefStruct(other._refObject.toObject());
    this._formatIntialized = true;
    this._tuneIntialized = true;
    this._wrapperFields.tune.copy(other._wrapperFields.tune);
  }
  static get classType() {
    return PlayTuneV2Wrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"QBASIC1_1","value":1},{"type":"uint8","name":"MML_MODERN","value":2}],"fields":[{"name":"format","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tune","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"PlayTuneV2","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PlayTuneV2"};
  }
  hasMember(name) {
    let memberNames = ["format","tune"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PlayTuneV2ArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PlayTuneV2Wrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PlayTuneV2Wrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PlayTuneV2ArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PlayTuneV2ArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PlayTuneV2RefArray(size);
    this._refObject = new PlayTuneV2RefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PlayTuneV2Wrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PlayTuneV2ArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PlayTuneV2ArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PlayTuneV2Wrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PlayTuneV2Wrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PlayTuneV2ArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(PlayTuneV2Wrapper, "QBASIC1_1", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PlayTuneV2Wrapper, "MML_MODERN", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = PlayTuneV2Wrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "QBASIC1_1",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "MML_MODERN",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "format",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tune",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "PlayTuneV2",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PlayTuneV2"
}
*/
