// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__NavControllerOutput.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const NavControllerOutputRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  nav_roll: primitiveTypes.float32,
  nav_pitch: primitiveTypes.float32,
  nav_bearing: primitiveTypes.int16,
  target_bearing: primitiveTypes.int16,
  wp_dist: primitiveTypes.uint16,
  alt_error: primitiveTypes.float32,
  aspd_error: primitiveTypes.float32,
  xtrack_error: primitiveTypes.float32,
});
const NavControllerOutputRefArray = ArrayType(NavControllerOutputRefStruct);
const NavControllerOutputRefStructArray = StructType({
  data: NavControllerOutputRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NavControllerOutputWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NavControllerOutputRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._nav_rollIntialized = true;
      this._nav_pitchIntialized = true;
      this._nav_bearingIntialized = true;
      this._target_bearingIntialized = true;
      this._wp_distIntialized = true;
      this._alt_errorIntialized = true;
      this._aspd_errorIntialized = true;
      this._xtrack_errorIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NavControllerOutputRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._nav_rollIntialized = false;
    this._nav_pitchIntialized = false;
    this._nav_bearingIntialized = false;
    this._target_bearingIntialized = false;
    this._wp_distIntialized = false;
    this._alt_errorIntialized = false;
    this._aspd_errorIntialized = false;
    this._xtrack_errorIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new NavControllerOutputWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NavControllerOutputArrayWrapper;
  }
  static get ArrayType() {
    return NavControllerOutputArrayWrapper;
  }
  static get refObjectArrayType() {
    return NavControllerOutputRefStructArray
  }
  static get refObjectType() {
    return NavControllerOutputRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._nav_rollIntialized) {
        throw new TypeError('Invalid argument: nav_roll in NavControllerOutput');
      }
      if (!this._nav_pitchIntialized) {
        throw new TypeError('Invalid argument: nav_pitch in NavControllerOutput');
      }
      if (!this._nav_bearingIntialized) {
        throw new TypeError('Invalid argument: nav_bearing in NavControllerOutput');
      }
      if (!this._target_bearingIntialized) {
        throw new TypeError('Invalid argument: target_bearing in NavControllerOutput');
      }
      if (!this._wp_distIntialized) {
        throw new TypeError('Invalid argument: wp_dist in NavControllerOutput');
      }
      if (!this._alt_errorIntialized) {
        throw new TypeError('Invalid argument: alt_error in NavControllerOutput');
      }
      if (!this._aspd_errorIntialized) {
        throw new TypeError('Invalid argument: aspd_error in NavControllerOutput');
      }
      if (!this._xtrack_errorIntialized) {
        throw new TypeError('Invalid argument: xtrack_error in NavControllerOutput');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._nav_rollIntialized = true;
    this._refObject.nav_roll = refObject.nav_roll;
    this._nav_pitchIntialized = true;
    this._refObject.nav_pitch = refObject.nav_pitch;
    this._nav_bearingIntialized = true;
    this._refObject.nav_bearing = refObject.nav_bearing;
    this._target_bearingIntialized = true;
    this._refObject.target_bearing = refObject.target_bearing;
    this._wp_distIntialized = true;
    this._refObject.wp_dist = refObject.wp_dist;
    this._alt_errorIntialized = true;
    this._refObject.alt_error = refObject.alt_error;
    this._aspd_errorIntialized = true;
    this._refObject.aspd_error = refObject.aspd_error;
    this._xtrack_errorIntialized = true;
    this._refObject.xtrack_error = refObject.xtrack_error;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    NavControllerOutputWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'NavControllerOutput'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get nav_roll() {
    if (!this._nav_rollIntialized) {
      return undefined;
    }
    return this._refObject.nav_roll;
  }
  set nav_roll(value) {
    this._nav_rollIntialized = true;
    this._refObject.nav_roll = value;
  }
  get nav_pitch() {
    if (!this._nav_pitchIntialized) {
      return undefined;
    }
    return this._refObject.nav_pitch;
  }
  set nav_pitch(value) {
    this._nav_pitchIntialized = true;
    this._refObject.nav_pitch = value;
  }
  get nav_bearing() {
    if (!this._nav_bearingIntialized) {
      return undefined;
    }
    return this._refObject.nav_bearing;
  }
  set nav_bearing(value) {
    this._nav_bearingIntialized = true;
    this._refObject.nav_bearing = value;
  }
  get target_bearing() {
    if (!this._target_bearingIntialized) {
      return undefined;
    }
    return this._refObject.target_bearing;
  }
  set target_bearing(value) {
    this._target_bearingIntialized = true;
    this._refObject.target_bearing = value;
  }
  get wp_dist() {
    if (!this._wp_distIntialized) {
      return undefined;
    }
    return this._refObject.wp_dist;
  }
  set wp_dist(value) {
    this._wp_distIntialized = true;
    this._refObject.wp_dist = value;
  }
  get alt_error() {
    if (!this._alt_errorIntialized) {
      return undefined;
    }
    return this._refObject.alt_error;
  }
  set alt_error(value) {
    this._alt_errorIntialized = true;
    this._refObject.alt_error = value;
  }
  get aspd_error() {
    if (!this._aspd_errorIntialized) {
      return undefined;
    }
    return this._refObject.aspd_error;
  }
  set aspd_error(value) {
    this._aspd_errorIntialized = true;
    this._refObject.aspd_error = value;
  }
  get xtrack_error() {
    if (!this._xtrack_errorIntialized) {
      return undefined;
    }
    return this._refObject.xtrack_error;
  }
  set xtrack_error(value) {
    this._xtrack_errorIntialized = true;
    this._refObject.xtrack_error = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NavControllerOutputRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._nav_rollIntialized = true;
    this._nav_pitchIntialized = true;
    this._nav_bearingIntialized = true;
    this._target_bearingIntialized = true;
    this._wp_distIntialized = true;
    this._alt_errorIntialized = true;
    this._aspd_errorIntialized = true;
    this._xtrack_errorIntialized = true;
  }
  copy(other) {
    this._refObject = new NavControllerOutputRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._nav_rollIntialized = true;
    this._nav_pitchIntialized = true;
    this._nav_bearingIntialized = true;
    this._target_bearingIntialized = true;
    this._wp_distIntialized = true;
    this._alt_errorIntialized = true;
    this._aspd_errorIntialized = true;
    this._xtrack_errorIntialized = true;
  }
  static get classType() {
    return NavControllerOutputWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"nav_roll","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nav_pitch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nav_bearing","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"target_bearing","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"wp_dist","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"alt_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aspd_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"xtrack_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"NavControllerOutput","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NavControllerOutput"};
  }
  hasMember(name) {
    let memberNames = ["header","nav_roll","nav_pitch","nav_bearing","target_bearing","wp_dist","alt_error","aspd_error","xtrack_error"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NavControllerOutputArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NavControllerOutputWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NavControllerOutputWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavControllerOutputArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavControllerOutputArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NavControllerOutputRefArray(size);
    this._refObject = new NavControllerOutputRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NavControllerOutputWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NavControllerOutputArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NavControllerOutputArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NavControllerOutputWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NavControllerOutputWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NavControllerOutputArrayWrapper;
  }
}
module.exports = NavControllerOutputWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "nav_roll",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "nav_pitch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "nav_bearing",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "target_bearing",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "wp_dist",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "alt_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aspd_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "xtrack_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "NavControllerOutput",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "NavControllerOutput"
}
*/
