// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__LogRequestList_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LogRequestList_RequestRefStruct = StructType({
  start: primitiveTypes.uint16,
  end: primitiveTypes.uint16,
});
const LogRequestList_RequestRefArray = ArrayType(LogRequestList_RequestRefStruct);
const LogRequestList_RequestRefStructArray = StructType({
  data: LogRequestList_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LogRequestList_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LogRequestList_RequestRefStruct(other._refObject.toObject());
      this._startIntialized = true;
      this._endIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LogRequestList_RequestRefStruct();
    this._startIntialized = false;
    this._endIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LogRequestList_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LogRequestList_RequestArrayWrapper;
  }
  static get ArrayType() {
    return LogRequestList_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return LogRequestList_RequestRefStructArray
  }
  static get refObjectType() {
    return LogRequestList_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._startIntialized) {
        throw new TypeError('Invalid argument: start in LogRequestList_Request');
      }
      if (!this._endIntialized) {
        throw new TypeError('Invalid argument: end in LogRequestList_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._startIntialized = true;
    this._refObject.start = refObject.start;
    this._endIntialized = true;
    this._refObject.end = refObject.end;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LogRequestList_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'LogRequestList_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get start() {
    if (!this._startIntialized) {
      return undefined;
    }
    return this._refObject.start;
  }
  set start(value) {
    this._startIntialized = true;
    this._refObject.start = value;
  }
  get end() {
    if (!this._endIntialized) {
      return undefined;
    }
    return this._refObject.end;
  }
  set end(value) {
    this._endIntialized = true;
    this._refObject.end = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LogRequestList_RequestRefStruct(refObject.toObject());
    this._startIntialized = true;
    this._endIntialized = true;
  }
  copy(other) {
    this._refObject = new LogRequestList_RequestRefStruct(other._refObject.toObject());
    this._startIntialized = true;
    this._endIntialized = true;
  }
  static get classType() {
    return LogRequestList_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"end","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"LogRequestList_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LogRequestList_Request"};
  }
  hasMember(name) {
    let memberNames = ["start","end"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LogRequestList_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LogRequestList_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LogRequestList_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LogRequestList_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LogRequestList_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LogRequestList_RequestRefArray(size);
    this._refObject = new LogRequestList_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LogRequestList_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LogRequestList_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LogRequestList_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LogRequestList_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LogRequestList_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LogRequestList_RequestArrayWrapper;
  }
}
module.exports = LogRequestList_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "end",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "LogRequestList_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LogRequestList_Request"
}
*/
