// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__VehicleInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const VehicleInfoRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  available_info: primitiveTypes.uint8,
  sysid: primitiveTypes.uint8,
  compid: primitiveTypes.uint8,
  autopilot: primitiveTypes.uint8,
  type: primitiveTypes.uint8,
  system_status: primitiveTypes.uint8,
  base_mode: primitiveTypes.uint8,
  custom_mode: primitiveTypes.uint32,
  mode: primitiveTypes.string,
  mode_id: primitiveTypes.uint32,
  capabilities: primitiveTypes.uint64,
  flight_sw_version: primitiveTypes.uint32,
  middleware_sw_version: primitiveTypes.uint32,
  os_sw_version: primitiveTypes.uint32,
  board_version: primitiveTypes.uint32,
  flight_custom_version: primitiveTypes.string,
  vendor_id: primitiveTypes.uint16,
  product_id: primitiveTypes.uint16,
  uid: primitiveTypes.uint64,
});
const VehicleInfoRefArray = ArrayType(VehicleInfoRefStruct);
const VehicleInfoRefStructArray = StructType({
  data: VehicleInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleInfoWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._available_infoIntialized = true;
      this._sysidIntialized = true;
      this._compidIntialized = true;
      this._autopilotIntialized = true;
      this._typeIntialized = true;
      this._system_statusIntialized = true;
      this._base_modeIntialized = true;
      this._custom_modeIntialized = true;
      this._modeIntialized = true;
      this._wrapperFields.mode =  new StringWrapper(other._wrapperFields.mode);
      this._mode_idIntialized = true;
      this._capabilitiesIntialized = true;
      this._flight_sw_versionIntialized = true;
      this._middleware_sw_versionIntialized = true;
      this._os_sw_versionIntialized = true;
      this._board_versionIntialized = true;
      this._flight_custom_versionIntialized = true;
      this._wrapperFields.flight_custom_version =  new StringWrapper(other._wrapperFields.flight_custom_version);
      this._vendor_idIntialized = true;
      this._product_idIntialized = true;
      this._uidIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleInfoRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._available_infoIntialized = false;
    this._sysidIntialized = false;
    this._compidIntialized = false;
    this._autopilotIntialized = false;
    this._typeIntialized = false;
    this._system_statusIntialized = false;
    this._base_modeIntialized = false;
    this._custom_modeIntialized = false;
    this._modeIntialized = false;
    this._wrapperFields.mode = new StringWrapper();
    this._mode_idIntialized = false;
    this._capabilitiesIntialized = false;
    this._flight_sw_versionIntialized = false;
    this._middleware_sw_versionIntialized = false;
    this._os_sw_versionIntialized = false;
    this._board_versionIntialized = false;
    this._flight_custom_versionIntialized = false;
    this._wrapperFields.flight_custom_version = new StringWrapper();
    this._vendor_idIntialized = false;
    this._product_idIntialized = false;
    this._uidIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleInfoArrayWrapper;
  }
  static get ArrayType() {
    return VehicleInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleInfoRefStructArray
  }
  static get refObjectType() {
    return VehicleInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._available_infoIntialized) {
        throw new TypeError('Invalid argument: available_info in VehicleInfo');
      }
      if (!this._sysidIntialized) {
        throw new TypeError('Invalid argument: sysid in VehicleInfo');
      }
      if (!this._compidIntialized) {
        throw new TypeError('Invalid argument: compid in VehicleInfo');
      }
      if (!this._autopilotIntialized) {
        throw new TypeError('Invalid argument: autopilot in VehicleInfo');
      }
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in VehicleInfo');
      }
      if (!this._system_statusIntialized) {
        throw new TypeError('Invalid argument: system_status in VehicleInfo');
      }
      if (!this._base_modeIntialized) {
        throw new TypeError('Invalid argument: base_mode in VehicleInfo');
      }
      if (!this._custom_modeIntialized) {
        throw new TypeError('Invalid argument: custom_mode in VehicleInfo');
      }
      if (!this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in VehicleInfo');
      }
      if (!this._mode_idIntialized) {
        throw new TypeError('Invalid argument: mode_id in VehicleInfo');
      }
      if (!this._capabilitiesIntialized) {
        throw new TypeError('Invalid argument: capabilities in VehicleInfo');
      }
      if (!this._flight_sw_versionIntialized) {
        throw new TypeError('Invalid argument: flight_sw_version in VehicleInfo');
      }
      if (!this._middleware_sw_versionIntialized) {
        throw new TypeError('Invalid argument: middleware_sw_version in VehicleInfo');
      }
      if (!this._os_sw_versionIntialized) {
        throw new TypeError('Invalid argument: os_sw_version in VehicleInfo');
      }
      if (!this._board_versionIntialized) {
        throw new TypeError('Invalid argument: board_version in VehicleInfo');
      }
      if (!this._flight_custom_versionIntialized) {
        throw new TypeError('Invalid argument: flight_custom_version in VehicleInfo');
      }
      if (!this._vendor_idIntialized) {
        throw new TypeError('Invalid argument: vendor_id in VehicleInfo');
      }
      if (!this._product_idIntialized) {
        throw new TypeError('Invalid argument: product_id in VehicleInfo');
      }
      if (!this._uidIntialized) {
        throw new TypeError('Invalid argument: uid in VehicleInfo');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.mode.freeze(own, checkConsistency);
    }
    this._refObject.mode = this._wrapperFields.mode.refObject;
    if (own) {
      this._wrapperFields.flight_custom_version.freeze(own, checkConsistency);
    }
    this._refObject.flight_custom_version = this._wrapperFields.flight_custom_version.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._available_infoIntialized = true;
    this._refObject.available_info = refObject.available_info;
    this._sysidIntialized = true;
    this._refObject.sysid = refObject.sysid;
    this._compidIntialized = true;
    this._refObject.compid = refObject.compid;
    this._autopilotIntialized = true;
    this._refObject.autopilot = refObject.autopilot;
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
    this._system_statusIntialized = true;
    this._refObject.system_status = refObject.system_status;
    this._base_modeIntialized = true;
    this._refObject.base_mode = refObject.base_mode;
    this._custom_modeIntialized = true;
    this._refObject.custom_mode = refObject.custom_mode;
    this._modeIntialized = true;
    this._wrapperFields.mode.data = refObject.mode.data;
    this._mode_idIntialized = true;
    this._refObject.mode_id = refObject.mode_id;
    this._capabilitiesIntialized = true;
    this._refObject.capabilities = refObject.capabilities;
    this._flight_sw_versionIntialized = true;
    this._refObject.flight_sw_version = refObject.flight_sw_version;
    this._middleware_sw_versionIntialized = true;
    this._refObject.middleware_sw_version = refObject.middleware_sw_version;
    this._os_sw_versionIntialized = true;
    this._refObject.os_sw_version = refObject.os_sw_version;
    this._board_versionIntialized = true;
    this._refObject.board_version = refObject.board_version;
    this._flight_custom_versionIntialized = true;
    this._wrapperFields.flight_custom_version.data = refObject.flight_custom_version.data;
    this._vendor_idIntialized = true;
    this._refObject.vendor_id = refObject.vendor_id;
    this._product_idIntialized = true;
    this._refObject.product_id = refObject.product_id;
    this._uidIntialized = true;
    this._refObject.uid = refObject.uid;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.mode);
    StringWrapper.freeStruct(refObject.flight_custom_version);
  }
  static destoryRawROS(msg) {
    VehicleInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'VehicleInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get available_info() {
    if (!this._available_infoIntialized) {
      return undefined;
    }
    return this._refObject.available_info;
  }
  set available_info(value) {
    this._available_infoIntialized = true;
    this._refObject.available_info = value;
  }
  get sysid() {
    if (!this._sysidIntialized) {
      return undefined;
    }
    return this._refObject.sysid;
  }
  set sysid(value) {
    this._sysidIntialized = true;
    this._refObject.sysid = value;
  }
  get compid() {
    if (!this._compidIntialized) {
      return undefined;
    }
    return this._refObject.compid;
  }
  set compid(value) {
    this._compidIntialized = true;
    this._refObject.compid = value;
  }
  get autopilot() {
    if (!this._autopilotIntialized) {
      return undefined;
    }
    return this._refObject.autopilot;
  }
  set autopilot(value) {
    this._autopilotIntialized = true;
    this._refObject.autopilot = value;
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  get system_status() {
    if (!this._system_statusIntialized) {
      return undefined;
    }
    return this._refObject.system_status;
  }
  set system_status(value) {
    this._system_statusIntialized = true;
    this._refObject.system_status = value;
  }
  get base_mode() {
    if (!this._base_modeIntialized) {
      return undefined;
    }
    return this._refObject.base_mode;
  }
  set base_mode(value) {
    this._base_modeIntialized = true;
    this._refObject.base_mode = value;
  }
  get custom_mode() {
    if (!this._custom_modeIntialized) {
      return undefined;
    }
    return this._refObject.custom_mode;
  }
  set custom_mode(value) {
    this._custom_modeIntialized = true;
    this._refObject.custom_mode = value;
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._wrapperFields.mode.data;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._wrapperFields.mode.data = value;
  }
  get mode_id() {
    if (!this._mode_idIntialized) {
      return undefined;
    }
    return this._refObject.mode_id;
  }
  set mode_id(value) {
    this._mode_idIntialized = true;
    this._refObject.mode_id = value;
  }
  get capabilities() {
    if (!this._capabilitiesIntialized) {
      return undefined;
    }
    return this._refObject.capabilities;
  }
  set capabilities(value) {
    this._capabilitiesIntialized = true;
    this._refObject.capabilities = value;
  }
  get flight_sw_version() {
    if (!this._flight_sw_versionIntialized) {
      return undefined;
    }
    return this._refObject.flight_sw_version;
  }
  set flight_sw_version(value) {
    this._flight_sw_versionIntialized = true;
    this._refObject.flight_sw_version = value;
  }
  get middleware_sw_version() {
    if (!this._middleware_sw_versionIntialized) {
      return undefined;
    }
    return this._refObject.middleware_sw_version;
  }
  set middleware_sw_version(value) {
    this._middleware_sw_versionIntialized = true;
    this._refObject.middleware_sw_version = value;
  }
  get os_sw_version() {
    if (!this._os_sw_versionIntialized) {
      return undefined;
    }
    return this._refObject.os_sw_version;
  }
  set os_sw_version(value) {
    this._os_sw_versionIntialized = true;
    this._refObject.os_sw_version = value;
  }
  get board_version() {
    if (!this._board_versionIntialized) {
      return undefined;
    }
    return this._refObject.board_version;
  }
  set board_version(value) {
    this._board_versionIntialized = true;
    this._refObject.board_version = value;
  }
  get flight_custom_version() {
    if (!this._flight_custom_versionIntialized) {
      return undefined;
    }
    return this._wrapperFields.flight_custom_version.data;
  }
  set flight_custom_version(value) {
    this._flight_custom_versionIntialized = true;
    this._wrapperFields.flight_custom_version.data = value;
  }
  get vendor_id() {
    if (!this._vendor_idIntialized) {
      return undefined;
    }
    return this._refObject.vendor_id;
  }
  set vendor_id(value) {
    this._vendor_idIntialized = true;
    this._refObject.vendor_id = value;
  }
  get product_id() {
    if (!this._product_idIntialized) {
      return undefined;
    }
    return this._refObject.product_id;
  }
  set product_id(value) {
    this._product_idIntialized = true;
    this._refObject.product_id = value;
  }
  get uid() {
    if (!this._uidIntialized) {
      return undefined;
    }
    return this._refObject.uid;
  }
  set uid(value) {
    this._uidIntialized = true;
    this._refObject.uid = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleInfoRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._available_infoIntialized = true;
    this._sysidIntialized = true;
    this._compidIntialized = true;
    this._autopilotIntialized = true;
    this._typeIntialized = true;
    this._system_statusIntialized = true;
    this._base_modeIntialized = true;
    this._custom_modeIntialized = true;
    this._modeIntialized = true;
    this._wrapperFields.mode.copyRefObject(this._refObject.mode);
    this._mode_idIntialized = true;
    this._capabilitiesIntialized = true;
    this._flight_sw_versionIntialized = true;
    this._middleware_sw_versionIntialized = true;
    this._os_sw_versionIntialized = true;
    this._board_versionIntialized = true;
    this._flight_custom_versionIntialized = true;
    this._wrapperFields.flight_custom_version.copyRefObject(this._refObject.flight_custom_version);
    this._vendor_idIntialized = true;
    this._product_idIntialized = true;
    this._uidIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._available_infoIntialized = true;
    this._sysidIntialized = true;
    this._compidIntialized = true;
    this._autopilotIntialized = true;
    this._typeIntialized = true;
    this._system_statusIntialized = true;
    this._base_modeIntialized = true;
    this._custom_modeIntialized = true;
    this._modeIntialized = true;
    this._wrapperFields.mode.copy(other._wrapperFields.mode);
    this._mode_idIntialized = true;
    this._capabilitiesIntialized = true;
    this._flight_sw_versionIntialized = true;
    this._middleware_sw_versionIntialized = true;
    this._os_sw_versionIntialized = true;
    this._board_versionIntialized = true;
    this._flight_custom_versionIntialized = true;
    this._wrapperFields.flight_custom_version.copy(other._wrapperFields.flight_custom_version);
    this._vendor_idIntialized = true;
    this._product_idIntialized = true;
    this._uidIntialized = true;
  }
  static get classType() {
    return VehicleInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"HAVE_INFO_HEARTBEAT","value":1},{"type":"uint8","name":"HAVE_INFO_AUTOPILOT_VERSION","value":2}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"available_info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sysid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"compid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"autopilot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"system_status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"base_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"custom_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"capabilities","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flight_sw_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"middleware_sw_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"os_sw_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"board_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flight_custom_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vendor_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"product_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"uid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"VehicleInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleInfo"};
  }
  hasMember(name) {
    let memberNames = ["header","available_info","sysid","compid","autopilot","type","system_status","base_mode","custom_mode","mode","mode_id","capabilities","flight_sw_version","middleware_sw_version","os_sw_version","board_version","flight_custom_version","vendor_id","product_id","uid"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleInfoRefArray(size);
    this._refObject = new VehicleInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleInfoArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(VehicleInfoWrapper, "HAVE_INFO_HEARTBEAT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleInfoWrapper, "HAVE_INFO_AUTOPILOT_VERSION", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = VehicleInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "HAVE_INFO_HEARTBEAT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "HAVE_INFO_AUTOPILOT_VERSION",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "available_info",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sysid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "compid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "autopilot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "system_status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "base_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "custom_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "capabilities",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flight_sw_version",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "middleware_sw_version",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "os_sw_version",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "board_version",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flight_custom_version",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vendor_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "product_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "uid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "VehicleInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleInfo"
}
*/
