// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__FileList_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const FileEntryWrapper = require('../../generated/mavros_msgs/mavros_msgs__msg__FileEntry.js');
const FileList_ResponseRefStruct = StructType({
  list: FileEntryWrapper.refObjectArrayType,
  success: primitiveTypes.bool,
  r_errno: primitiveTypes.int32,
});
const FileList_ResponseRefArray = ArrayType(FileList_ResponseRefStruct);
const FileList_ResponseRefStructArray = StructType({
  data: FileList_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FileList_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FileList_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.list = FileEntryWrapper.createArray();
      this._wrapperFields.list.copy(other._wrapperFields.list);
      this._successIntialized = true;
      this._r_errnoIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FileList_ResponseRefStruct();
    this._wrapperFields.list = FileEntryWrapper.createArray();
    this._successIntialized = false;
    this._r_errnoIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new FileList_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FileList_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return FileList_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return FileList_ResponseRefStructArray
  }
  static get refObjectType() {
    return FileList_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in FileList_Response');
      }
      if (!this._r_errnoIntialized) {
        throw new TypeError('Invalid argument: r_errno in FileList_Response');
      }
    }
    this._wrapperFields.list.freeze(own, checkConsistency);
    this._refObject.list = this._wrapperFields.list.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.list.copyRefObject(refObject.list);
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._r_errnoIntialized = true;
    this._refObject.r_errno = refObject.r_errno;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.list.size != 0) {
      FileEntryWrapper.ArrayType.freeArray(refObject.list);
      if (FileEntryWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.list, FileEntryWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    FileList_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'FileList_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get list() {
    return this._wrapperFields.list;
  }
  set list(value) {
    this._wrapperFields.list.fill(value);
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get r_errno() {
    if (!this._r_errnoIntialized) {
      return undefined;
    }
    return this._refObject.r_errno;
  }
  set r_errno(value) {
    this._r_errnoIntialized = true;
    this._refObject.r_errno = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FileList_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.list.copyRefObject(this._refObject.list);
    this._successIntialized = true;
    this._r_errnoIntialized = true;
  }
  copy(other) {
    this._refObject = new FileList_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.list.copy(other._wrapperFields.list);
    this._successIntialized = true;
    this._r_errnoIntialized = true;
  }
  static get classType() {
    return FileList_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"list","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"mavros_msgs","type":"FileEntry","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"r_errno","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"FileList_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FileList_Response"};
  }
  hasMember(name) {
    let memberNames = ["list","success","r_errno"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FileList_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FileList_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FileList_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FileList_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FileList_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FileList_ResponseRefArray(size);
    this._refObject = new FileList_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FileList_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FileList_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FileList_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FileList_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FileList_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FileList_ResponseArrayWrapper;
  }
}
module.exports = FileList_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "list",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "mavros_msgs",
        "type": "FileEntry",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "r_errno",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "FileList_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FileList_Response"
}
*/
