// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__GPSRAW.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const GPSRAWRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  fix_type: primitiveTypes.uint8,
  lat: primitiveTypes.int32,
  lon: primitiveTypes.int32,
  alt: primitiveTypes.int32,
  eph: primitiveTypes.uint16,
  epv: primitiveTypes.uint16,
  vel: primitiveTypes.uint16,
  cog: primitiveTypes.uint16,
  satellites_visible: primitiveTypes.uint8,
  alt_ellipsoid: primitiveTypes.int32,
  h_acc: primitiveTypes.uint32,
  v_acc: primitiveTypes.uint32,
  vel_acc: primitiveTypes.uint32,
  hdg_acc: primitiveTypes.int32,
  yaw: primitiveTypes.uint16,
  dgps_numch: primitiveTypes.uint8,
  dgps_age: primitiveTypes.uint32,
});
const GPSRAWRefArray = ArrayType(GPSRAWRefStruct);
const GPSRAWRefStructArray = StructType({
  data: GPSRAWRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GPSRAWWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GPSRAWRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._fix_typeIntialized = true;
      this._latIntialized = true;
      this._lonIntialized = true;
      this._altIntialized = true;
      this._ephIntialized = true;
      this._epvIntialized = true;
      this._velIntialized = true;
      this._cogIntialized = true;
      this._satellites_visibleIntialized = true;
      this._alt_ellipsoidIntialized = true;
      this._h_accIntialized = true;
      this._v_accIntialized = true;
      this._vel_accIntialized = true;
      this._hdg_accIntialized = true;
      this._yawIntialized = true;
      this._dgps_numchIntialized = true;
      this._dgps_ageIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GPSRAWRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._fix_typeIntialized = false;
    this._latIntialized = false;
    this._lonIntialized = false;
    this._altIntialized = false;
    this._ephIntialized = false;
    this._epvIntialized = false;
    this._velIntialized = false;
    this._cogIntialized = false;
    this._satellites_visibleIntialized = false;
    this._alt_ellipsoidIntialized = false;
    this._h_accIntialized = false;
    this._v_accIntialized = false;
    this._vel_accIntialized = false;
    this._hdg_accIntialized = false;
    this._yawIntialized = false;
    this._dgps_numchIntialized = false;
    this._dgps_ageIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GPSRAWWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GPSRAWArrayWrapper;
  }
  static get ArrayType() {
    return GPSRAWArrayWrapper;
  }
  static get refObjectArrayType() {
    return GPSRAWRefStructArray
  }
  static get refObjectType() {
    return GPSRAWRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._fix_typeIntialized) {
        throw new TypeError('Invalid argument: fix_type in GPSRAW');
      }
      if (!this._latIntialized) {
        throw new TypeError('Invalid argument: lat in GPSRAW');
      }
      if (!this._lonIntialized) {
        throw new TypeError('Invalid argument: lon in GPSRAW');
      }
      if (!this._altIntialized) {
        throw new TypeError('Invalid argument: alt in GPSRAW');
      }
      if (!this._ephIntialized) {
        throw new TypeError('Invalid argument: eph in GPSRAW');
      }
      if (!this._epvIntialized) {
        throw new TypeError('Invalid argument: epv in GPSRAW');
      }
      if (!this._velIntialized) {
        throw new TypeError('Invalid argument: vel in GPSRAW');
      }
      if (!this._cogIntialized) {
        throw new TypeError('Invalid argument: cog in GPSRAW');
      }
      if (!this._satellites_visibleIntialized) {
        throw new TypeError('Invalid argument: satellites_visible in GPSRAW');
      }
      if (!this._alt_ellipsoidIntialized) {
        throw new TypeError('Invalid argument: alt_ellipsoid in GPSRAW');
      }
      if (!this._h_accIntialized) {
        throw new TypeError('Invalid argument: h_acc in GPSRAW');
      }
      if (!this._v_accIntialized) {
        throw new TypeError('Invalid argument: v_acc in GPSRAW');
      }
      if (!this._vel_accIntialized) {
        throw new TypeError('Invalid argument: vel_acc in GPSRAW');
      }
      if (!this._hdg_accIntialized) {
        throw new TypeError('Invalid argument: hdg_acc in GPSRAW');
      }
      if (!this._yawIntialized) {
        throw new TypeError('Invalid argument: yaw in GPSRAW');
      }
      if (!this._dgps_numchIntialized) {
        throw new TypeError('Invalid argument: dgps_numch in GPSRAW');
      }
      if (!this._dgps_ageIntialized) {
        throw new TypeError('Invalid argument: dgps_age in GPSRAW');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._fix_typeIntialized = true;
    this._refObject.fix_type = refObject.fix_type;
    this._latIntialized = true;
    this._refObject.lat = refObject.lat;
    this._lonIntialized = true;
    this._refObject.lon = refObject.lon;
    this._altIntialized = true;
    this._refObject.alt = refObject.alt;
    this._ephIntialized = true;
    this._refObject.eph = refObject.eph;
    this._epvIntialized = true;
    this._refObject.epv = refObject.epv;
    this._velIntialized = true;
    this._refObject.vel = refObject.vel;
    this._cogIntialized = true;
    this._refObject.cog = refObject.cog;
    this._satellites_visibleIntialized = true;
    this._refObject.satellites_visible = refObject.satellites_visible;
    this._alt_ellipsoidIntialized = true;
    this._refObject.alt_ellipsoid = refObject.alt_ellipsoid;
    this._h_accIntialized = true;
    this._refObject.h_acc = refObject.h_acc;
    this._v_accIntialized = true;
    this._refObject.v_acc = refObject.v_acc;
    this._vel_accIntialized = true;
    this._refObject.vel_acc = refObject.vel_acc;
    this._hdg_accIntialized = true;
    this._refObject.hdg_acc = refObject.hdg_acc;
    this._yawIntialized = true;
    this._refObject.yaw = refObject.yaw;
    this._dgps_numchIntialized = true;
    this._refObject.dgps_numch = refObject.dgps_numch;
    this._dgps_ageIntialized = true;
    this._refObject.dgps_age = refObject.dgps_age;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    GPSRAWWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'GPSRAW'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get fix_type() {
    if (!this._fix_typeIntialized) {
      return undefined;
    }
    return this._refObject.fix_type;
  }
  set fix_type(value) {
    this._fix_typeIntialized = true;
    this._refObject.fix_type = value;
  }
  get lat() {
    if (!this._latIntialized) {
      return undefined;
    }
    return this._refObject.lat;
  }
  set lat(value) {
    this._latIntialized = true;
    this._refObject.lat = value;
  }
  get lon() {
    if (!this._lonIntialized) {
      return undefined;
    }
    return this._refObject.lon;
  }
  set lon(value) {
    this._lonIntialized = true;
    this._refObject.lon = value;
  }
  get alt() {
    if (!this._altIntialized) {
      return undefined;
    }
    return this._refObject.alt;
  }
  set alt(value) {
    this._altIntialized = true;
    this._refObject.alt = value;
  }
  get eph() {
    if (!this._ephIntialized) {
      return undefined;
    }
    return this._refObject.eph;
  }
  set eph(value) {
    this._ephIntialized = true;
    this._refObject.eph = value;
  }
  get epv() {
    if (!this._epvIntialized) {
      return undefined;
    }
    return this._refObject.epv;
  }
  set epv(value) {
    this._epvIntialized = true;
    this._refObject.epv = value;
  }
  get vel() {
    if (!this._velIntialized) {
      return undefined;
    }
    return this._refObject.vel;
  }
  set vel(value) {
    this._velIntialized = true;
    this._refObject.vel = value;
  }
  get cog() {
    if (!this._cogIntialized) {
      return undefined;
    }
    return this._refObject.cog;
  }
  set cog(value) {
    this._cogIntialized = true;
    this._refObject.cog = value;
  }
  get satellites_visible() {
    if (!this._satellites_visibleIntialized) {
      return undefined;
    }
    return this._refObject.satellites_visible;
  }
  set satellites_visible(value) {
    this._satellites_visibleIntialized = true;
    this._refObject.satellites_visible = value;
  }
  get alt_ellipsoid() {
    if (!this._alt_ellipsoidIntialized) {
      return undefined;
    }
    return this._refObject.alt_ellipsoid;
  }
  set alt_ellipsoid(value) {
    this._alt_ellipsoidIntialized = true;
    this._refObject.alt_ellipsoid = value;
  }
  get h_acc() {
    if (!this._h_accIntialized) {
      return undefined;
    }
    return this._refObject.h_acc;
  }
  set h_acc(value) {
    this._h_accIntialized = true;
    this._refObject.h_acc = value;
  }
  get v_acc() {
    if (!this._v_accIntialized) {
      return undefined;
    }
    return this._refObject.v_acc;
  }
  set v_acc(value) {
    this._v_accIntialized = true;
    this._refObject.v_acc = value;
  }
  get vel_acc() {
    if (!this._vel_accIntialized) {
      return undefined;
    }
    return this._refObject.vel_acc;
  }
  set vel_acc(value) {
    this._vel_accIntialized = true;
    this._refObject.vel_acc = value;
  }
  get hdg_acc() {
    if (!this._hdg_accIntialized) {
      return undefined;
    }
    return this._refObject.hdg_acc;
  }
  set hdg_acc(value) {
    this._hdg_accIntialized = true;
    this._refObject.hdg_acc = value;
  }
  get yaw() {
    if (!this._yawIntialized) {
      return undefined;
    }
    return this._refObject.yaw;
  }
  set yaw(value) {
    this._yawIntialized = true;
    this._refObject.yaw = value;
  }
  get dgps_numch() {
    if (!this._dgps_numchIntialized) {
      return undefined;
    }
    return this._refObject.dgps_numch;
  }
  set dgps_numch(value) {
    this._dgps_numchIntialized = true;
    this._refObject.dgps_numch = value;
  }
  get dgps_age() {
    if (!this._dgps_ageIntialized) {
      return undefined;
    }
    return this._refObject.dgps_age;
  }
  set dgps_age(value) {
    this._dgps_ageIntialized = true;
    this._refObject.dgps_age = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GPSRAWRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._fix_typeIntialized = true;
    this._latIntialized = true;
    this._lonIntialized = true;
    this._altIntialized = true;
    this._ephIntialized = true;
    this._epvIntialized = true;
    this._velIntialized = true;
    this._cogIntialized = true;
    this._satellites_visibleIntialized = true;
    this._alt_ellipsoidIntialized = true;
    this._h_accIntialized = true;
    this._v_accIntialized = true;
    this._vel_accIntialized = true;
    this._hdg_accIntialized = true;
    this._yawIntialized = true;
    this._dgps_numchIntialized = true;
    this._dgps_ageIntialized = true;
  }
  copy(other) {
    this._refObject = new GPSRAWRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._fix_typeIntialized = true;
    this._latIntialized = true;
    this._lonIntialized = true;
    this._altIntialized = true;
    this._ephIntialized = true;
    this._epvIntialized = true;
    this._velIntialized = true;
    this._cogIntialized = true;
    this._satellites_visibleIntialized = true;
    this._alt_ellipsoidIntialized = true;
    this._h_accIntialized = true;
    this._v_accIntialized = true;
    this._vel_accIntialized = true;
    this._hdg_accIntialized = true;
    this._yawIntialized = true;
    this._dgps_numchIntialized = true;
    this._dgps_ageIntialized = true;
  }
  static get classType() {
    return GPSRAWWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"GPS_FIX_TYPE_NO_GPS","value":0},{"type":"uint8","name":"GPS_FIX_TYPE_NO_FIX","value":1},{"type":"uint8","name":"GPS_FIX_TYPE_2D_FIX","value":2},{"type":"uint8","name":"GPS_FIX_TYPE_3D_FIX","value":3},{"type":"uint8","name":"GPS_FIX_TYPE_DGPS","value":4},{"type":"uint8","name":"GPS_FIX_TYPE_RTK_FLOAT","value":5},{"type":"uint8","name":"GPS_FIX_TYPE_RTK_FIXED","value":6},{"type":"uint8","name":"GPS_FIX_TYPE_STATIC","value":7},{"type":"uint8","name":"GPS_FIX_TYPE_PPP","value":8}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"fix_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lat","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lon","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"alt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"eph","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"epv","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cog","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"satellites_visible","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"alt_ellipsoid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"h_acc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"v_acc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vel_acc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hdg_acc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dgps_numch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dgps_age","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"GPSRAW","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GPSRAW"};
  }
  hasMember(name) {
    let memberNames = ["header","fix_type","lat","lon","alt","eph","epv","vel","cog","satellites_visible","alt_ellipsoid","h_acc","v_acc","vel_acc","hdg_acc","yaw","dgps_numch","dgps_age"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GPSRAWArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GPSRAWWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GPSRAWWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPSRAWArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPSRAWArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GPSRAWRefArray(size);
    this._refObject = new GPSRAWRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GPSRAWWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GPSRAWArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GPSRAWArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GPSRAWWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GPSRAWWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GPSRAWArrayWrapper;
  }
}
// Define constants (9 in total)
Object.defineProperty(GPSRAWWrapper, "GPS_FIX_TYPE_NO_GPS", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSRAWWrapper, "GPS_FIX_TYPE_NO_FIX", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSRAWWrapper, "GPS_FIX_TYPE_2D_FIX", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSRAWWrapper, "GPS_FIX_TYPE_3D_FIX", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSRAWWrapper, "GPS_FIX_TYPE_DGPS", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSRAWWrapper, "GPS_FIX_TYPE_RTK_FLOAT", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSRAWWrapper, "GPS_FIX_TYPE_RTK_FIXED", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSRAWWrapper, "GPS_FIX_TYPE_STATIC", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSRAWWrapper, "GPS_FIX_TYPE_PPP", {value: 8, writable: false, enumerable: true, configurable: true});
module.exports = GPSRAWWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_NO_GPS",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_NO_FIX",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_2D_FIX",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_3D_FIX",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_DGPS",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_RTK_FLOAT",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_RTK_FIXED",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_STATIC",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_PPP",
      "value": 8
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "fix_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lat",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lon",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "alt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "eph",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "epv",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cog",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "satellites_visible",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "alt_ellipsoid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "h_acc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "v_acc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vel_acc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hdg_acc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dgps_numch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dgps_age",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "GPSRAW",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GPSRAW"
}
*/
