// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__OpenDroneIDOperatorID.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const OpenDroneIDOperatorIDRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  id_or_mac: primitiveTypes.string,
  operator_id_type: primitiveTypes.uint8,
  operator_id: primitiveTypes.string,
});
const OpenDroneIDOperatorIDRefArray = ArrayType(OpenDroneIDOperatorIDRefStruct);
const OpenDroneIDOperatorIDRefStructArray = StructType({
  data: OpenDroneIDOperatorIDRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OpenDroneIDOperatorIDWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OpenDroneIDOperatorIDRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._id_or_macIntialized = true;
      this._wrapperFields.id_or_mac =  new StringWrapper(other._wrapperFields.id_or_mac);
      this._operator_id_typeIntialized = true;
      this._operator_idIntialized = true;
      this._wrapperFields.operator_id =  new StringWrapper(other._wrapperFields.operator_id);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OpenDroneIDOperatorIDRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._id_or_macIntialized = false;
    this._wrapperFields.id_or_mac = new StringWrapper();
    this._operator_id_typeIntialized = false;
    this._operator_idIntialized = false;
    this._wrapperFields.operator_id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new OpenDroneIDOperatorIDWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OpenDroneIDOperatorIDArrayWrapper;
  }
  static get ArrayType() {
    return OpenDroneIDOperatorIDArrayWrapper;
  }
  static get refObjectArrayType() {
    return OpenDroneIDOperatorIDRefStructArray
  }
  static get refObjectType() {
    return OpenDroneIDOperatorIDRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._id_or_macIntialized) {
        throw new TypeError('Invalid argument: id_or_mac in OpenDroneIDOperatorID');
      }
      if (!this._operator_id_typeIntialized) {
        throw new TypeError('Invalid argument: operator_id_type in OpenDroneIDOperatorID');
      }
      if (!this._operator_idIntialized) {
        throw new TypeError('Invalid argument: operator_id in OpenDroneIDOperatorID');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.id_or_mac.freeze(own, checkConsistency);
    }
    this._refObject.id_or_mac = this._wrapperFields.id_or_mac.refObject;
    if (own) {
      this._wrapperFields.operator_id.freeze(own, checkConsistency);
    }
    this._refObject.operator_id = this._wrapperFields.operator_id.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.data = refObject.id_or_mac.data;
    this._operator_id_typeIntialized = true;
    this._refObject.operator_id_type = refObject.operator_id_type;
    this._operator_idIntialized = true;
    this._wrapperFields.operator_id.data = refObject.operator_id.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.id_or_mac);
    StringWrapper.freeStruct(refObject.operator_id);
  }
  static destoryRawROS(msg) {
    OpenDroneIDOperatorIDWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'OpenDroneIDOperatorID'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get id_or_mac() {
    if (!this._id_or_macIntialized) {
      return undefined;
    }
    return this._wrapperFields.id_or_mac.data;
  }
  set id_or_mac(value) {
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.data = value;
  }
  get operator_id_type() {
    if (!this._operator_id_typeIntialized) {
      return undefined;
    }
    return this._refObject.operator_id_type;
  }
  set operator_id_type(value) {
    this._operator_id_typeIntialized = true;
    this._refObject.operator_id_type = value;
  }
  get operator_id() {
    if (!this._operator_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.operator_id.data;
  }
  set operator_id(value) {
    this._operator_idIntialized = true;
    this._wrapperFields.operator_id.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new OpenDroneIDOperatorIDRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.copyRefObject(this._refObject.id_or_mac);
    this._operator_id_typeIntialized = true;
    this._operator_idIntialized = true;
    this._wrapperFields.operator_id.copyRefObject(this._refObject.operator_id);
  }
  copy(other) {
    this._refObject = new OpenDroneIDOperatorIDRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.copy(other._wrapperFields.id_or_mac);
    this._operator_id_typeIntialized = true;
    this._operator_idIntialized = true;
    this._wrapperFields.operator_id.copy(other._wrapperFields.operator_id);
  }
  static get classType() {
    return OpenDroneIDOperatorIDWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ID_TYPE_CAA","value":0}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id_or_mac","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"operator_id_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"operator_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"OpenDroneIDOperatorID","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OpenDroneIDOperatorID"};
  }
  hasMember(name) {
    let memberNames = ["header","id_or_mac","operator_id_type","operator_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OpenDroneIDOperatorIDArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OpenDroneIDOperatorIDWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OpenDroneIDOperatorIDWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpenDroneIDOperatorIDArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpenDroneIDOperatorIDArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OpenDroneIDOperatorIDRefArray(size);
    this._refObject = new OpenDroneIDOperatorIDRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OpenDroneIDOperatorIDWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OpenDroneIDOperatorIDArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OpenDroneIDOperatorIDArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OpenDroneIDOperatorIDWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OpenDroneIDOperatorIDWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OpenDroneIDOperatorIDArrayWrapper;
  }
}
// Define constants (1 in total)
Object.defineProperty(OpenDroneIDOperatorIDWrapper, "ID_TYPE_CAA", {value: 0, writable: false, enumerable: true, configurable: true});
module.exports = OpenDroneIDOperatorIDWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ID_TYPE_CAA",
      "value": 0
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "id_or_mac",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "operator_id_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "operator_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "OpenDroneIDOperatorID",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OpenDroneIDOperatorID"
}
*/
