// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__ExtendedState.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const ExtendedStateRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  vtol_state: primitiveTypes.uint8,
  landed_state: primitiveTypes.uint8,
});
const ExtendedStateRefArray = ArrayType(ExtendedStateRefStruct);
const ExtendedStateRefStructArray = StructType({
  data: ExtendedStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ExtendedStateWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ExtendedStateRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._vtol_stateIntialized = true;
      this._landed_stateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ExtendedStateRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._vtol_stateIntialized = false;
    this._landed_stateIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ExtendedStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ExtendedStateArrayWrapper;
  }
  static get ArrayType() {
    return ExtendedStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return ExtendedStateRefStructArray
  }
  static get refObjectType() {
    return ExtendedStateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._vtol_stateIntialized) {
        throw new TypeError('Invalid argument: vtol_state in ExtendedState');
      }
      if (!this._landed_stateIntialized) {
        throw new TypeError('Invalid argument: landed_state in ExtendedState');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._vtol_stateIntialized = true;
    this._refObject.vtol_state = refObject.vtol_state;
    this._landed_stateIntialized = true;
    this._refObject.landed_state = refObject.landed_state;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    ExtendedStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'ExtendedState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get vtol_state() {
    if (!this._vtol_stateIntialized) {
      return undefined;
    }
    return this._refObject.vtol_state;
  }
  set vtol_state(value) {
    this._vtol_stateIntialized = true;
    this._refObject.vtol_state = value;
  }
  get landed_state() {
    if (!this._landed_stateIntialized) {
      return undefined;
    }
    return this._refObject.landed_state;
  }
  set landed_state(value) {
    this._landed_stateIntialized = true;
    this._refObject.landed_state = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ExtendedStateRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._vtol_stateIntialized = true;
    this._landed_stateIntialized = true;
  }
  copy(other) {
    this._refObject = new ExtendedStateRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._vtol_stateIntialized = true;
    this._landed_stateIntialized = true;
  }
  static get classType() {
    return ExtendedStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"VTOL_STATE_UNDEFINED","value":0},{"type":"uint8","name":"VTOL_STATE_TRANSITION_TO_FW","value":1},{"type":"uint8","name":"VTOL_STATE_TRANSITION_TO_MC","value":2},{"type":"uint8","name":"VTOL_STATE_MC","value":3},{"type":"uint8","name":"VTOL_STATE_FW","value":4},{"type":"uint8","name":"LANDED_STATE_UNDEFINED","value":0},{"type":"uint8","name":"LANDED_STATE_ON_GROUND","value":1},{"type":"uint8","name":"LANDED_STATE_IN_AIR","value":2},{"type":"uint8","name":"LANDED_STATE_TAKEOFF","value":3},{"type":"uint8","name":"LANDED_STATE_LANDING","value":4}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"vtol_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"landed_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"ExtendedState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ExtendedState"};
  }
  hasMember(name) {
    let memberNames = ["header","vtol_state","landed_state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ExtendedStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ExtendedStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ExtendedStateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ExtendedStateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ExtendedStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ExtendedStateRefArray(size);
    this._refObject = new ExtendedStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ExtendedStateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ExtendedStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ExtendedStateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ExtendedStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ExtendedStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ExtendedStateArrayWrapper;
  }
}
// Define constants (10 in total)
Object.defineProperty(ExtendedStateWrapper, "VTOL_STATE_UNDEFINED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ExtendedStateWrapper, "VTOL_STATE_TRANSITION_TO_FW", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ExtendedStateWrapper, "VTOL_STATE_TRANSITION_TO_MC", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ExtendedStateWrapper, "VTOL_STATE_MC", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ExtendedStateWrapper, "VTOL_STATE_FW", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ExtendedStateWrapper, "LANDED_STATE_UNDEFINED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ExtendedStateWrapper, "LANDED_STATE_ON_GROUND", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ExtendedStateWrapper, "LANDED_STATE_IN_AIR", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ExtendedStateWrapper, "LANDED_STATE_TAKEOFF", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ExtendedStateWrapper, "LANDED_STATE_LANDING", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = ExtendedStateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "VTOL_STATE_UNDEFINED",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "VTOL_STATE_TRANSITION_TO_FW",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "VTOL_STATE_TRANSITION_TO_MC",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "VTOL_STATE_MC",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "VTOL_STATE_FW",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "LANDED_STATE_UNDEFINED",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "LANDED_STATE_ON_GROUND",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "LANDED_STATE_IN_AIR",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "LANDED_STATE_TAKEOFF",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "LANDED_STATE_LANDING",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "vtol_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "landed_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "ExtendedState",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ExtendedState"
}
*/
