// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__Mavlink.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const UInt64Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt64.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const MavlinkRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  framing_status: primitiveTypes.uint8,
  magic: primitiveTypes.uint8,
  len: primitiveTypes.uint8,
  incompat_flags: primitiveTypes.uint8,
  compat_flags: primitiveTypes.uint8,
  seq: primitiveTypes.uint8,
  sysid: primitiveTypes.uint8,
  compid: primitiveTypes.uint8,
  msgid: primitiveTypes.uint32,
  checksum: primitiveTypes.uint16,
  payload64: UInt64Wrapper.refObjectArrayType,
  signature: UInt8Wrapper.refObjectArrayType,
});
const MavlinkRefArray = ArrayType(MavlinkRefStruct);
const MavlinkRefStructArray = StructType({
  data: MavlinkRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MavlinkWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._payload64Array = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MavlinkRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._framing_statusIntialized = true;
      this._magicIntialized = true;
      this._lenIntialized = true;
      this._incompat_flagsIntialized = true;
      this._compat_flagsIntialized = true;
      this._seqIntialized = true;
      this._sysidIntialized = true;
      this._compidIntialized = true;
      this._msgidIntialized = true;
      this._checksumIntialized = true;
      this._wrapperFields.payload64 = UInt64Wrapper.createArray();
      this._wrapperFields.payload64.copy(other._wrapperFields.payload64);
      this.payload64 = other.payload64;
      this._wrapperFields.signature = UInt8Wrapper.createArray();
      this._wrapperFields.signature.copy(other._wrapperFields.signature);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MavlinkRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._framing_statusIntialized = false;
    this._magicIntialized = false;
    this._lenIntialized = false;
    this._incompat_flagsIntialized = false;
    this._compat_flagsIntialized = false;
    this._seqIntialized = false;
    this._sysidIntialized = false;
    this._compidIntialized = false;
    this._msgidIntialized = false;
    this._checksumIntialized = false;
    this._wrapperFields.payload64 = UInt64Wrapper.createArray();
    this._wrapperFields.signature = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MavlinkWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MavlinkArrayWrapper;
  }
  static get ArrayType() {
    return MavlinkArrayWrapper;
  }
  static get refObjectArrayType() {
    return MavlinkRefStructArray
  }
  static get refObjectType() {
    return MavlinkRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._framing_statusIntialized) {
        throw new TypeError('Invalid argument: framing_status in Mavlink');
      }
      if (!this._magicIntialized) {
        throw new TypeError('Invalid argument: magic in Mavlink');
      }
      if (!this._lenIntialized) {
        throw new TypeError('Invalid argument: len in Mavlink');
      }
      if (!this._incompat_flagsIntialized) {
        throw new TypeError('Invalid argument: incompat_flags in Mavlink');
      }
      if (!this._compat_flagsIntialized) {
        throw new TypeError('Invalid argument: compat_flags in Mavlink');
      }
      if (!this._seqIntialized) {
        throw new TypeError('Invalid argument: seq in Mavlink');
      }
      if (!this._sysidIntialized) {
        throw new TypeError('Invalid argument: sysid in Mavlink');
      }
      if (!this._compidIntialized) {
        throw new TypeError('Invalid argument: compid in Mavlink');
      }
      if (!this._msgidIntialized) {
        throw new TypeError('Invalid argument: msgid in Mavlink');
      }
      if (!this._checksumIntialized) {
        throw new TypeError('Invalid argument: checksum in Mavlink');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (!own) {
      this._wrapperFields.payload64.fill(this._payload64Array);
      this._wrapperFields.payload64.freeze(own, checkConsistency);
      this._refObject.payload64 = this._wrapperFields.payload64.refObject;
    } else {
      this._wrapperFields.payload64.fill([]);
      this._wrapperFields.payload64.freeze(own, checkConsistency);
      this._refObject.payload64 = this._wrapperFields.payload64.refObject;
    }
    this._wrapperFields.signature.freeze(own, checkConsistency);
    this._refObject.signature = this._wrapperFields.signature.refObject;
    if (own) {
      this._wrapperFields.signature.fill(Uint8Array.from([]));
      this._wrapperFields.signature.freeze(own, checkConsistency);
      this._refObject.signature = this._wrapperFields.signature.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._framing_statusIntialized = true;
    this._refObject.framing_status = refObject.framing_status;
    this._magicIntialized = true;
    this._refObject.magic = refObject.magic;
    this._lenIntialized = true;
    this._refObject.len = refObject.len;
    this._incompat_flagsIntialized = true;
    this._refObject.incompat_flags = refObject.incompat_flags;
    this._compat_flagsIntialized = true;
    this._refObject.compat_flags = refObject.compat_flags;
    this._seqIntialized = true;
    this._refObject.seq = refObject.seq;
    this._sysidIntialized = true;
    this._refObject.sysid = refObject.sysid;
    this._compidIntialized = true;
    this._refObject.compid = refObject.compid;
    this._msgidIntialized = true;
    this._refObject.msgid = refObject.msgid;
    this._checksumIntialized = true;
    this._refObject.checksum = refObject.checksum;
    refObject.payload64.data.length = refObject.payload64.size;
    for (let index = 0; index < refObject.payload64.size; index++) {
      this._payload64Array[index] = refObject.payload64.data[index].data;
    }
    this._wrapperFields.signature.copyRefObject(refObject.signature);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.payload64.size != 0) {
      UInt64Wrapper.ArrayType.freeArray(refObject.payload64);
      if (UInt64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.payload64, UInt64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.signature.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.signature);
      if (UInt8Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.signature, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MavlinkWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'Mavlink'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get framing_status() {
    if (!this._framing_statusIntialized) {
      return undefined;
    }
    return this._refObject.framing_status;
  }
  set framing_status(value) {
    this._framing_statusIntialized = true;
    this._refObject.framing_status = value;
  }
  get magic() {
    if (!this._magicIntialized) {
      return undefined;
    }
    return this._refObject.magic;
  }
  set magic(value) {
    this._magicIntialized = true;
    this._refObject.magic = value;
  }
  get len() {
    if (!this._lenIntialized) {
      return undefined;
    }
    return this._refObject.len;
  }
  set len(value) {
    this._lenIntialized = true;
    this._refObject.len = value;
  }
  get incompat_flags() {
    if (!this._incompat_flagsIntialized) {
      return undefined;
    }
    return this._refObject.incompat_flags;
  }
  set incompat_flags(value) {
    this._incompat_flagsIntialized = true;
    this._refObject.incompat_flags = value;
  }
  get compat_flags() {
    if (!this._compat_flagsIntialized) {
      return undefined;
    }
    return this._refObject.compat_flags;
  }
  set compat_flags(value) {
    this._compat_flagsIntialized = true;
    this._refObject.compat_flags = value;
  }
  get seq() {
    if (!this._seqIntialized) {
      return undefined;
    }
    return this._refObject.seq;
  }
  set seq(value) {
    this._seqIntialized = true;
    this._refObject.seq = value;
  }
  get sysid() {
    if (!this._sysidIntialized) {
      return undefined;
    }
    return this._refObject.sysid;
  }
  set sysid(value) {
    this._sysidIntialized = true;
    this._refObject.sysid = value;
  }
  get compid() {
    if (!this._compidIntialized) {
      return undefined;
    }
    return this._refObject.compid;
  }
  set compid(value) {
    this._compidIntialized = true;
    this._refObject.compid = value;
  }
  get msgid() {
    if (!this._msgidIntialized) {
      return undefined;
    }
    return this._refObject.msgid;
  }
  set msgid(value) {
    this._msgidIntialized = true;
    this._refObject.msgid = value;
  }
  get checksum() {
    if (!this._checksumIntialized) {
      return undefined;
    }
    return this._refObject.checksum;
  }
  set checksum(value) {
    this._checksumIntialized = true;
    this._refObject.checksum = value;
  }
  get payload64() {
    return this._payload64Array;
  }
  set payload64(value) {
    if (value.length > 33) {
      throw new RangeError('The length of array payload64 must be <= 33.');
    }
    this._payload64Array = value;
  }
  get signature() {
    return this._wrapperFields['signature'].data;
  }
  set signature(value) {
    if (value.length > 13) {
      throw new RangeError('The length of array signature must be <= 13.');
    }
    this._wrapperFields['signature'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MavlinkRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._framing_statusIntialized = true;
    this._magicIntialized = true;
    this._lenIntialized = true;
    this._incompat_flagsIntialized = true;
    this._compat_flagsIntialized = true;
    this._seqIntialized = true;
    this._sysidIntialized = true;
    this._compidIntialized = true;
    this._msgidIntialized = true;
    this._checksumIntialized = true;
    refObject.payload64.data.length = refObject.payload64.size;
    for (let index = 0; index < refObject.payload64.size; index++) {
      this._payload64Array[index] = refObject.payload64.data[index].data;
    }
    this._wrapperFields.signature.copyRefObject(this._refObject.signature);
  }
  copy(other) {
    this._refObject = new MavlinkRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._framing_statusIntialized = true;
    this._magicIntialized = true;
    this._lenIntialized = true;
    this._incompat_flagsIntialized = true;
    this._compat_flagsIntialized = true;
    this._seqIntialized = true;
    this._sysidIntialized = true;
    this._compidIntialized = true;
    this._msgidIntialized = true;
    this._checksumIntialized = true;
    this._payload64Array = other._payload64Array.slice();
    this._wrapperFields.signature.copy(other._wrapperFields.signature);
  }
  static get classType() {
    return MavlinkWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"FRAMING_OK","value":1},{"type":"uint8","name":"FRAMING_BAD_CRC","value":2},{"type":"uint8","name":"FRAMING_BAD_SIGNATURE","value":3},{"type":"uint8","name":"MAVLINK_V10","value":254},{"type":"uint8","name":"MAVLINK_V20","value":253}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"framing_status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"magic","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"len","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"incompat_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"compat_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"seq","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sysid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"compid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"msgid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"checksum","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"payload64","type":{"isArray":true,"arraySize":33,"isUpperBound":true,"isDynamicArray":true,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"signature","type":{"isArray":true,"arraySize":13,"isUpperBound":true,"isDynamicArray":true,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"Mavlink","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Mavlink"};
  }
  hasMember(name) {
    let memberNames = ["header","framing_status","magic","len","incompat_flags","compat_flags","seq","sysid","compid","msgid","checksum","payload64","signature"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MavlinkArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MavlinkWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MavlinkWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MavlinkArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MavlinkArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MavlinkRefArray(size);
    this._refObject = new MavlinkRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MavlinkWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MavlinkArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MavlinkArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MavlinkWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MavlinkWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MavlinkArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(MavlinkWrapper, "FRAMING_OK", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MavlinkWrapper, "FRAMING_BAD_CRC", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MavlinkWrapper, "FRAMING_BAD_SIGNATURE", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MavlinkWrapper, "MAVLINK_V10", {value: 254, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MavlinkWrapper, "MAVLINK_V20", {value: 253, writable: false, enumerable: true, configurable: true});
module.exports = MavlinkWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "FRAMING_OK",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "FRAMING_BAD_CRC",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "FRAMING_BAD_SIGNATURE",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "MAVLINK_V10",
      "value": 254
    },
    {
      "type": "uint8",
      "name": "MAVLINK_V20",
      "value": 253
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "framing_status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "magic",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "len",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "incompat_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "compat_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "seq",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sysid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "compid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "msgid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "checksum",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "payload64",
      "type": {
        "isArray": true,
        "arraySize": 33,
        "isUpperBound": true,
        "isDynamicArray": true,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "signature",
      "type": {
        "isArray": true,
        "arraySize": 13,
        "isUpperBound": true,
        "isDynamicArray": true,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "Mavlink",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Mavlink"
}
*/
