// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__MountConfigure_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const MountConfigure_RequestRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  mode: primitiveTypes.uint8,
  stabilize_roll: primitiveTypes.bool,
  stabilize_pitch: primitiveTypes.bool,
  stabilize_yaw: primitiveTypes.bool,
  roll_input: primitiveTypes.uint8,
  pitch_input: primitiveTypes.uint8,
  yaw_input: primitiveTypes.uint8,
});
const MountConfigure_RequestRefArray = ArrayType(MountConfigure_RequestRefStruct);
const MountConfigure_RequestRefStructArray = StructType({
  data: MountConfigure_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MountConfigure_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MountConfigure_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._modeIntialized = true;
      this._stabilize_rollIntialized = true;
      this._stabilize_pitchIntialized = true;
      this._stabilize_yawIntialized = true;
      this._roll_inputIntialized = true;
      this._pitch_inputIntialized = true;
      this._yaw_inputIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MountConfigure_RequestRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._modeIntialized = false;
    this._stabilize_rollIntialized = false;
    this._stabilize_pitchIntialized = false;
    this._stabilize_yawIntialized = false;
    this._roll_inputIntialized = false;
    this._pitch_inputIntialized = false;
    this._yaw_inputIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MountConfigure_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MountConfigure_RequestArrayWrapper;
  }
  static get ArrayType() {
    return MountConfigure_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return MountConfigure_RequestRefStructArray
  }
  static get refObjectType() {
    return MountConfigure_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in MountConfigure_Request');
      }
      if (!this._stabilize_rollIntialized) {
        throw new TypeError('Invalid argument: stabilize_roll in MountConfigure_Request');
      }
      if (!this._stabilize_pitchIntialized) {
        throw new TypeError('Invalid argument: stabilize_pitch in MountConfigure_Request');
      }
      if (!this._stabilize_yawIntialized) {
        throw new TypeError('Invalid argument: stabilize_yaw in MountConfigure_Request');
      }
      if (!this._roll_inputIntialized) {
        throw new TypeError('Invalid argument: roll_input in MountConfigure_Request');
      }
      if (!this._pitch_inputIntialized) {
        throw new TypeError('Invalid argument: pitch_input in MountConfigure_Request');
      }
      if (!this._yaw_inputIntialized) {
        throw new TypeError('Invalid argument: yaw_input in MountConfigure_Request');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._modeIntialized = true;
    this._refObject.mode = refObject.mode;
    this._stabilize_rollIntialized = true;
    this._refObject.stabilize_roll = refObject.stabilize_roll;
    this._stabilize_pitchIntialized = true;
    this._refObject.stabilize_pitch = refObject.stabilize_pitch;
    this._stabilize_yawIntialized = true;
    this._refObject.stabilize_yaw = refObject.stabilize_yaw;
    this._roll_inputIntialized = true;
    this._refObject.roll_input = refObject.roll_input;
    this._pitch_inputIntialized = true;
    this._refObject.pitch_input = refObject.pitch_input;
    this._yaw_inputIntialized = true;
    this._refObject.yaw_input = refObject.yaw_input;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    MountConfigure_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'MountConfigure_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._refObject.mode;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._refObject.mode = value;
  }
  get stabilize_roll() {
    if (!this._stabilize_rollIntialized) {
      return undefined;
    }
    return this._refObject.stabilize_roll;
  }
  set stabilize_roll(value) {
    this._stabilize_rollIntialized = true;
    this._refObject.stabilize_roll = value;
  }
  get stabilize_pitch() {
    if (!this._stabilize_pitchIntialized) {
      return undefined;
    }
    return this._refObject.stabilize_pitch;
  }
  set stabilize_pitch(value) {
    this._stabilize_pitchIntialized = true;
    this._refObject.stabilize_pitch = value;
  }
  get stabilize_yaw() {
    if (!this._stabilize_yawIntialized) {
      return undefined;
    }
    return this._refObject.stabilize_yaw;
  }
  set stabilize_yaw(value) {
    this._stabilize_yawIntialized = true;
    this._refObject.stabilize_yaw = value;
  }
  get roll_input() {
    if (!this._roll_inputIntialized) {
      return undefined;
    }
    return this._refObject.roll_input;
  }
  set roll_input(value) {
    this._roll_inputIntialized = true;
    this._refObject.roll_input = value;
  }
  get pitch_input() {
    if (!this._pitch_inputIntialized) {
      return undefined;
    }
    return this._refObject.pitch_input;
  }
  set pitch_input(value) {
    this._pitch_inputIntialized = true;
    this._refObject.pitch_input = value;
  }
  get yaw_input() {
    if (!this._yaw_inputIntialized) {
      return undefined;
    }
    return this._refObject.yaw_input;
  }
  set yaw_input(value) {
    this._yaw_inputIntialized = true;
    this._refObject.yaw_input = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MountConfigure_RequestRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._modeIntialized = true;
    this._stabilize_rollIntialized = true;
    this._stabilize_pitchIntialized = true;
    this._stabilize_yawIntialized = true;
    this._roll_inputIntialized = true;
    this._pitch_inputIntialized = true;
    this._yaw_inputIntialized = true;
  }
  copy(other) {
    this._refObject = new MountConfigure_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._modeIntialized = true;
    this._stabilize_rollIntialized = true;
    this._stabilize_pitchIntialized = true;
    this._stabilize_yawIntialized = true;
    this._roll_inputIntialized = true;
    this._pitch_inputIntialized = true;
    this._yaw_inputIntialized = true;
  }
  static get classType() {
    return MountConfigure_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"MODE_RETRACT","value":0},{"type":"uint8","name":"MODE_NEUTRAL","value":1},{"type":"uint8","name":"MODE_MAVLINK_TARGETING","value":2},{"type":"uint8","name":"MODE_RC_TARGETING","value":3},{"type":"uint8","name":"MODE_GPS_POINT","value":4},{"type":"uint8","name":"INPUT_ANGLE_BODY_FRAME","value":0},{"type":"uint8","name":"INPUT_ANGULAR_RATE","value":1},{"type":"uint8","name":"INPUT_ANGLE_ABSOLUTE_FRAME","value":2}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stabilize_roll","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stabilize_pitch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stabilize_yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"roll_input","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch_input","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_input","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"MountConfigure_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MountConfigure_Request"};
  }
  hasMember(name) {
    let memberNames = ["header","mode","stabilize_roll","stabilize_pitch","stabilize_yaw","roll_input","pitch_input","yaw_input"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MountConfigure_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MountConfigure_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MountConfigure_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MountConfigure_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MountConfigure_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MountConfigure_RequestRefArray(size);
    this._refObject = new MountConfigure_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MountConfigure_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MountConfigure_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MountConfigure_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MountConfigure_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MountConfigure_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MountConfigure_RequestArrayWrapper;
  }
}
// Define constants (8 in total)
Object.defineProperty(MountConfigure_RequestWrapper, "MODE_RETRACT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountConfigure_RequestWrapper, "MODE_NEUTRAL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountConfigure_RequestWrapper, "MODE_MAVLINK_TARGETING", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountConfigure_RequestWrapper, "MODE_RC_TARGETING", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountConfigure_RequestWrapper, "MODE_GPS_POINT", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountConfigure_RequestWrapper, "INPUT_ANGLE_BODY_FRAME", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountConfigure_RequestWrapper, "INPUT_ANGULAR_RATE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountConfigure_RequestWrapper, "INPUT_ANGLE_ABSOLUTE_FRAME", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = MountConfigure_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "MODE_RETRACT",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "MODE_NEUTRAL",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "MODE_MAVLINK_TARGETING",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "MODE_RC_TARGETING",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "MODE_GPS_POINT",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "INPUT_ANGLE_BODY_FRAME",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "INPUT_ANGULAR_RATE",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "INPUT_ANGLE_ABSOLUTE_FRAME",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stabilize_roll",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stabilize_pitch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stabilize_yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "roll_input",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch_input",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_input",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "MountConfigure_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MountConfigure_Request"
}
*/
