// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__HilActuatorControls.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const HilActuatorControlsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  controls: ArrayType(primitiveTypes.float32, 16),
  mode: primitiveTypes.uint8,
  flags: primitiveTypes.uint64,
});
const HilActuatorControlsRefArray = ArrayType(HilActuatorControlsRefStruct);
const HilActuatorControlsRefStructArray = StructType({
  data: HilActuatorControlsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HilActuatorControlsWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HilActuatorControlsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.controls = Float32Wrapper.createArray();
      this._wrapperFields.controls.copy(other._wrapperFields.controls);
      this._modeIntialized = true;
      this._flagsIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HilActuatorControlsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.controls = Float32Wrapper.createArray();
    this._modeIntialized = false;
    this._flagsIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new HilActuatorControlsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HilActuatorControlsArrayWrapper;
  }
  static get ArrayType() {
    return HilActuatorControlsArrayWrapper;
  }
  static get refObjectArrayType() {
    return HilActuatorControlsRefStructArray
  }
  static get refObjectType() {
    return HilActuatorControlsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in HilActuatorControls');
      }
      if (!this._flagsIntialized) {
        throw new TypeError('Invalid argument: flags in HilActuatorControls');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    for (let i = 0; i < this._wrapperFields.controls.data.length; i++) {
      this._refObject.controls[i] = this._wrapperFields.controls.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.controls = refObject.controls;
    this._wrapperFields.controls.fill(refObject.controls.toArray());
    this._modeIntialized = true;
    this._refObject.mode = refObject.mode;
    this._flagsIntialized = true;
    this._refObject.flags = refObject.flags;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    HilActuatorControlsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'HilActuatorControls'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get controls() {
    return this._wrapperFields['controls'].data;
  }
  set controls(value) {
    if (value.length !== 16) {
      throw new RangeError('The length of the array must be 16.');
    }
    this._wrapperFields['controls'].fill(value);
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._refObject.mode;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._refObject.mode = value;
  }
  get flags() {
    if (!this._flagsIntialized) {
      return undefined;
    }
    return this._refObject.flags;
  }
  set flags(value) {
    this._flagsIntialized = true;
    this._refObject.flags = value;
  }
  copyRefObject(refObject) {
    this._refObject = new HilActuatorControlsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.controls.fill(refObject.controls.toArray());
    this._modeIntialized = true;
    this._flagsIntialized = true;
  }
  copy(other) {
    this._refObject = new HilActuatorControlsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.controls.copy(other._wrapperFields.controls);
    this._modeIntialized = true;
    this._flagsIntialized = true;
  }
  static get classType() {
    return HilActuatorControlsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"controls","type":{"isArray":true,"arraySize":16,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"HilActuatorControls","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HilActuatorControls"};
  }
  hasMember(name) {
    let memberNames = ["header","controls","mode","flags"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HilActuatorControlsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HilActuatorControlsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HilActuatorControlsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HilActuatorControlsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HilActuatorControlsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HilActuatorControlsRefArray(size);
    this._refObject = new HilActuatorControlsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HilActuatorControlsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HilActuatorControlsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HilActuatorControlsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HilActuatorControlsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HilActuatorControlsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HilActuatorControlsArrayWrapper;
  }
}
module.exports = HilActuatorControlsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "controls",
      "type": {
        "isArray": true,
        "arraySize": 16,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "HilActuatorControls",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HilActuatorControls"
}
*/
