// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__MessageInterval_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MessageInterval_RequestRefStruct = StructType({
  message_id: primitiveTypes.uint32,
  message_rate: primitiveTypes.float32,
});
const MessageInterval_RequestRefArray = ArrayType(MessageInterval_RequestRefStruct);
const MessageInterval_RequestRefStructArray = StructType({
  data: MessageInterval_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MessageInterval_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MessageInterval_RequestRefStruct(other._refObject.toObject());
      this._message_idIntialized = true;
      this._message_rateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MessageInterval_RequestRefStruct();
    this._message_idIntialized = false;
    this._message_rateIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MessageInterval_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MessageInterval_RequestArrayWrapper;
  }
  static get ArrayType() {
    return MessageInterval_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return MessageInterval_RequestRefStructArray
  }
  static get refObjectType() {
    return MessageInterval_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._message_idIntialized) {
        throw new TypeError('Invalid argument: message_id in MessageInterval_Request');
      }
      if (!this._message_rateIntialized) {
        throw new TypeError('Invalid argument: message_rate in MessageInterval_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._message_idIntialized = true;
    this._refObject.message_id = refObject.message_id;
    this._message_rateIntialized = true;
    this._refObject.message_rate = refObject.message_rate;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    MessageInterval_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'MessageInterval_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get message_id() {
    if (!this._message_idIntialized) {
      return undefined;
    }
    return this._refObject.message_id;
  }
  set message_id(value) {
    this._message_idIntialized = true;
    this._refObject.message_id = value;
  }
  get message_rate() {
    if (!this._message_rateIntialized) {
      return undefined;
    }
    return this._refObject.message_rate;
  }
  set message_rate(value) {
    this._message_rateIntialized = true;
    this._refObject.message_rate = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MessageInterval_RequestRefStruct(refObject.toObject());
    this._message_idIntialized = true;
    this._message_rateIntialized = true;
  }
  copy(other) {
    this._refObject = new MessageInterval_RequestRefStruct(other._refObject.toObject());
    this._message_idIntialized = true;
    this._message_rateIntialized = true;
  }
  static get classType() {
    return MessageInterval_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"message_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"message_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"MessageInterval_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MessageInterval_Request"};
  }
  hasMember(name) {
    let memberNames = ["message_id","message_rate"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MessageInterval_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MessageInterval_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MessageInterval_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MessageInterval_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MessageInterval_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MessageInterval_RequestRefArray(size);
    this._refObject = new MessageInterval_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MessageInterval_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MessageInterval_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MessageInterval_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MessageInterval_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MessageInterval_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MessageInterval_RequestArrayWrapper;
  }
}
module.exports = MessageInterval_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "message_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "message_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "MessageInterval_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MessageInterval_Request"
}
*/
