// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__ParamSetV2_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ParameterValueWrapper = require('../../generated/rcl_interfaces/rcl_interfaces__msg__ParameterValue.js');
const ParamSetV2_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  value: ParameterValueWrapper.refObjectType,
});
const ParamSetV2_ResponseRefArray = ArrayType(ParamSetV2_ResponseRefStruct);
const ParamSetV2_ResponseRefStructArray = StructType({
  data: ParamSetV2_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ParamSetV2_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ParamSetV2_ResponseRefStruct(other._refObject.toObject());
      this._successIntialized = true;
      this._wrapperFields.value =  new ParameterValueWrapper(other._wrapperFields.value);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ParamSetV2_ResponseRefStruct();
    this._successIntialized = false;
    this._wrapperFields.value = new ParameterValueWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ParamSetV2_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ParamSetV2_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return ParamSetV2_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ParamSetV2_ResponseRefStructArray
  }
  static get refObjectType() {
    return ParamSetV2_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in ParamSetV2_Response');
      }
    }
    this._wrapperFields.value.freeze(own, checkConsistency);
    this._refObject.value = this._wrapperFields.value.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._wrapperFields.value.copyRefObject(refObject.value);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    ParameterValueWrapper.freeStruct(refObject.value);
  }
  static destoryRawROS(msg) {
    ParamSetV2_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'ParamSetV2_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get value() {
    return this._wrapperFields.value;
  }
  set value(value) {
    if (value instanceof ParameterValueWrapper) {
      this._wrapperFields.value.copy(value);
    } else {
      this._wrapperFields.value.copy(new ParameterValueWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ParamSetV2_ResponseRefStruct(refObject.toObject());
    this._successIntialized = true;
    this._wrapperFields.value.copyRefObject(this._refObject.value);
  }
  copy(other) {
    this._refObject = new ParamSetV2_ResponseRefStruct(other._refObject.toObject());
    this._successIntialized = true;
    this._wrapperFields.value.copy(other._wrapperFields.value);
  }
  static get classType() {
    return ParamSetV2_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"rcl_interfaces","type":"ParameterValue","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"ParamSetV2_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ParamSetV2_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","value"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ParamSetV2_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ParamSetV2_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ParamSetV2_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParamSetV2_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParamSetV2_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ParamSetV2_ResponseRefArray(size);
    this._refObject = new ParamSetV2_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ParamSetV2_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ParamSetV2_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ParamSetV2_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ParamSetV2_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ParamSetV2_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ParamSetV2_ResponseArrayWrapper;
  }
}
module.exports = ParamSetV2_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "rcl_interfaces",
        "type": "ParameterValue",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "ParamSetV2_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ParamSetV2_Response"
}
*/
