// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__GlobalPositionTarget.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const GlobalPositionTargetRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  coordinate_frame: primitiveTypes.uint8,
  type_mask: primitiveTypes.uint16,
  latitude: primitiveTypes.float64,
  longitude: primitiveTypes.float64,
  altitude: primitiveTypes.float32,
  velocity: Vector3Wrapper.refObjectType,
  acceleration_or_force: Vector3Wrapper.refObjectType,
  yaw: primitiveTypes.float32,
  yaw_rate: primitiveTypes.float32,
});
const GlobalPositionTargetRefArray = ArrayType(GlobalPositionTargetRefStruct);
const GlobalPositionTargetRefStructArray = StructType({
  data: GlobalPositionTargetRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GlobalPositionTargetWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GlobalPositionTargetRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._coordinate_frameIntialized = true;
      this._type_maskIntialized = true;
      this._latitudeIntialized = true;
      this._longitudeIntialized = true;
      this._altitudeIntialized = true;
      this._wrapperFields.velocity =  new Vector3Wrapper(other._wrapperFields.velocity);
      this._wrapperFields.acceleration_or_force =  new Vector3Wrapper(other._wrapperFields.acceleration_or_force);
      this._yawIntialized = true;
      this._yaw_rateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GlobalPositionTargetRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._coordinate_frameIntialized = false;
    this._type_maskIntialized = false;
    this._latitudeIntialized = false;
    this._longitudeIntialized = false;
    this._altitudeIntialized = false;
    this._wrapperFields.velocity = new Vector3Wrapper();
    this._wrapperFields.acceleration_or_force = new Vector3Wrapper();
    this._yawIntialized = false;
    this._yaw_rateIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GlobalPositionTargetWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GlobalPositionTargetArrayWrapper;
  }
  static get ArrayType() {
    return GlobalPositionTargetArrayWrapper;
  }
  static get refObjectArrayType() {
    return GlobalPositionTargetRefStructArray
  }
  static get refObjectType() {
    return GlobalPositionTargetRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._coordinate_frameIntialized) {
        throw new TypeError('Invalid argument: coordinate_frame in GlobalPositionTarget');
      }
      if (!this._type_maskIntialized) {
        throw new TypeError('Invalid argument: type_mask in GlobalPositionTarget');
      }
      if (!this._latitudeIntialized) {
        throw new TypeError('Invalid argument: latitude in GlobalPositionTarget');
      }
      if (!this._longitudeIntialized) {
        throw new TypeError('Invalid argument: longitude in GlobalPositionTarget');
      }
      if (!this._altitudeIntialized) {
        throw new TypeError('Invalid argument: altitude in GlobalPositionTarget');
      }
      if (!this._yawIntialized) {
        throw new TypeError('Invalid argument: yaw in GlobalPositionTarget');
      }
      if (!this._yaw_rateIntialized) {
        throw new TypeError('Invalid argument: yaw_rate in GlobalPositionTarget');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.velocity.freeze(own, checkConsistency);
    this._refObject.velocity = this._wrapperFields.velocity.refObject;
    this._wrapperFields.acceleration_or_force.freeze(own, checkConsistency);
    this._refObject.acceleration_or_force = this._wrapperFields.acceleration_or_force.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._coordinate_frameIntialized = true;
    this._refObject.coordinate_frame = refObject.coordinate_frame;
    this._type_maskIntialized = true;
    this._refObject.type_mask = refObject.type_mask;
    this._latitudeIntialized = true;
    this._refObject.latitude = refObject.latitude;
    this._longitudeIntialized = true;
    this._refObject.longitude = refObject.longitude;
    this._altitudeIntialized = true;
    this._refObject.altitude = refObject.altitude;
    this._wrapperFields.velocity.copyRefObject(refObject.velocity);
    this._wrapperFields.acceleration_or_force.copyRefObject(refObject.acceleration_or_force);
    this._yawIntialized = true;
    this._refObject.yaw = refObject.yaw;
    this._yaw_rateIntialized = true;
    this._refObject.yaw_rate = refObject.yaw_rate;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    Vector3Wrapper.freeStruct(refObject.velocity);
    Vector3Wrapper.freeStruct(refObject.acceleration_or_force);
  }
  static destoryRawROS(msg) {
    GlobalPositionTargetWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'GlobalPositionTarget'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get coordinate_frame() {
    if (!this._coordinate_frameIntialized) {
      return undefined;
    }
    return this._refObject.coordinate_frame;
  }
  set coordinate_frame(value) {
    this._coordinate_frameIntialized = true;
    this._refObject.coordinate_frame = value;
  }
  get type_mask() {
    if (!this._type_maskIntialized) {
      return undefined;
    }
    return this._refObject.type_mask;
  }
  set type_mask(value) {
    this._type_maskIntialized = true;
    this._refObject.type_mask = value;
  }
  get latitude() {
    if (!this._latitudeIntialized) {
      return undefined;
    }
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._latitudeIntialized = true;
    this._refObject.latitude = value;
  }
  get longitude() {
    if (!this._longitudeIntialized) {
      return undefined;
    }
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._longitudeIntialized = true;
    this._refObject.longitude = value;
  }
  get altitude() {
    if (!this._altitudeIntialized) {
      return undefined;
    }
    return this._refObject.altitude;
  }
  set altitude(value) {
    this._altitudeIntialized = true;
    this._refObject.altitude = value;
  }
  get velocity() {
    return this._wrapperFields.velocity;
  }
  set velocity(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.velocity.copy(value);
    } else {
      this._wrapperFields.velocity.copy(new Vector3Wrapper(value));
    }
  }
  get acceleration_or_force() {
    return this._wrapperFields.acceleration_or_force;
  }
  set acceleration_or_force(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.acceleration_or_force.copy(value);
    } else {
      this._wrapperFields.acceleration_or_force.copy(new Vector3Wrapper(value));
    }
  }
  get yaw() {
    if (!this._yawIntialized) {
      return undefined;
    }
    return this._refObject.yaw;
  }
  set yaw(value) {
    this._yawIntialized = true;
    this._refObject.yaw = value;
  }
  get yaw_rate() {
    if (!this._yaw_rateIntialized) {
      return undefined;
    }
    return this._refObject.yaw_rate;
  }
  set yaw_rate(value) {
    this._yaw_rateIntialized = true;
    this._refObject.yaw_rate = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GlobalPositionTargetRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._coordinate_frameIntialized = true;
    this._type_maskIntialized = true;
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._wrapperFields.velocity.copyRefObject(this._refObject.velocity);
    this._wrapperFields.acceleration_or_force.copyRefObject(this._refObject.acceleration_or_force);
    this._yawIntialized = true;
    this._yaw_rateIntialized = true;
  }
  copy(other) {
    this._refObject = new GlobalPositionTargetRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._coordinate_frameIntialized = true;
    this._type_maskIntialized = true;
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
    this._wrapperFields.acceleration_or_force.copy(other._wrapperFields.acceleration_or_force);
    this._yawIntialized = true;
    this._yaw_rateIntialized = true;
  }
  static get classType() {
    return GlobalPositionTargetWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"FRAME_GLOBAL_INT","value":5},{"type":"uint8","name":"FRAME_GLOBAL_REL_ALT","value":6},{"type":"uint8","name":"FRAME_GLOBAL_TERRAIN_ALT","value":11},{"type":"uint16","name":"IGNORE_LATITUDE","value":1},{"type":"uint16","name":"IGNORE_LONGITUDE","value":2},{"type":"uint16","name":"IGNORE_ALTITUDE","value":4},{"type":"uint16","name":"IGNORE_VX","value":8},{"type":"uint16","name":"IGNORE_VY","value":16},{"type":"uint16","name":"IGNORE_VZ","value":32},{"type":"uint16","name":"IGNORE_AFX","value":64},{"type":"uint16","name":"IGNORE_AFY","value":128},{"type":"uint16","name":"IGNORE_AFZ","value":256},{"type":"uint16","name":"FORCE","value":512},{"type":"uint16","name":"IGNORE_YAW","value":1024},{"type":"uint16","name":"IGNORE_YAW_RATE","value":2048}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"coordinate_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type_mask","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"acceleration_or_force","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"GlobalPositionTarget","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GlobalPositionTarget"};
  }
  hasMember(name) {
    let memberNames = ["header","coordinate_frame","type_mask","latitude","longitude","altitude","velocity","acceleration_or_force","yaw","yaw_rate"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GlobalPositionTargetArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GlobalPositionTargetWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GlobalPositionTargetWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GlobalPositionTargetArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GlobalPositionTargetArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GlobalPositionTargetRefArray(size);
    this._refObject = new GlobalPositionTargetRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GlobalPositionTargetWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GlobalPositionTargetArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GlobalPositionTargetArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GlobalPositionTargetWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GlobalPositionTargetWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GlobalPositionTargetArrayWrapper;
  }
}
// Define constants (15 in total)
Object.defineProperty(GlobalPositionTargetWrapper, "FRAME_GLOBAL_INT", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "FRAME_GLOBAL_REL_ALT", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "FRAME_GLOBAL_TERRAIN_ALT", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_LATITUDE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_LONGITUDE", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_ALTITUDE", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_VX", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_VY", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_VZ", {value: 32, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_AFX", {value: 64, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_AFY", {value: 128, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_AFZ", {value: 256, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "FORCE", {value: 512, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_YAW", {value: 1024, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GlobalPositionTargetWrapper, "IGNORE_YAW_RATE", {value: 2048, writable: false, enumerable: true, configurable: true});
module.exports = GlobalPositionTargetWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_INT",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_REL_ALT",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_TERRAIN_ALT",
      "value": 11
    },
    {
      "type": "uint16",
      "name": "IGNORE_LATITUDE",
      "value": 1
    },
    {
      "type": "uint16",
      "name": "IGNORE_LONGITUDE",
      "value": 2
    },
    {
      "type": "uint16",
      "name": "IGNORE_ALTITUDE",
      "value": 4
    },
    {
      "type": "uint16",
      "name": "IGNORE_VX",
      "value": 8
    },
    {
      "type": "uint16",
      "name": "IGNORE_VY",
      "value": 16
    },
    {
      "type": "uint16",
      "name": "IGNORE_VZ",
      "value": 32
    },
    {
      "type": "uint16",
      "name": "IGNORE_AFX",
      "value": 64
    },
    {
      "type": "uint16",
      "name": "IGNORE_AFY",
      "value": 128
    },
    {
      "type": "uint16",
      "name": "IGNORE_AFZ",
      "value": 256
    },
    {
      "type": "uint16",
      "name": "FORCE",
      "value": 512
    },
    {
      "type": "uint16",
      "name": "IGNORE_YAW",
      "value": 1024
    },
    {
      "type": "uint16",
      "name": "IGNORE_YAW_RATE",
      "value": 2048
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "coordinate_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "type_mask",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "acceleration_or_force",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "GlobalPositionTarget",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GlobalPositionTarget"
}
*/
