// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__ESCInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const ESCInfoItemWrapper = require('../../generated/mavros_msgs/mavros_msgs__msg__ESCInfoItem.js');
const ESCInfoRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  counter: primitiveTypes.uint16,
  count: primitiveTypes.uint8,
  connection_type: primitiveTypes.uint8,
  info: primitiveTypes.uint8,
  esc_info: ESCInfoItemWrapper.refObjectArrayType,
});
const ESCInfoRefArray = ArrayType(ESCInfoRefStruct);
const ESCInfoRefStructArray = StructType({
  data: ESCInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ESCInfoWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ESCInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._counterIntialized = true;
      this._countIntialized = true;
      this._connection_typeIntialized = true;
      this._infoIntialized = true;
      this._wrapperFields.esc_info = ESCInfoItemWrapper.createArray();
      this._wrapperFields.esc_info.copy(other._wrapperFields.esc_info);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ESCInfoRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._counterIntialized = false;
    this._countIntialized = false;
    this._connection_typeIntialized = false;
    this._infoIntialized = false;
    this._wrapperFields.esc_info = ESCInfoItemWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ESCInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ESCInfoArrayWrapper;
  }
  static get ArrayType() {
    return ESCInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return ESCInfoRefStructArray
  }
  static get refObjectType() {
    return ESCInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._counterIntialized) {
        throw new TypeError('Invalid argument: counter in ESCInfo');
      }
      if (!this._countIntialized) {
        throw new TypeError('Invalid argument: count in ESCInfo');
      }
      if (!this._connection_typeIntialized) {
        throw new TypeError('Invalid argument: connection_type in ESCInfo');
      }
      if (!this._infoIntialized) {
        throw new TypeError('Invalid argument: info in ESCInfo');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.esc_info.freeze(own, checkConsistency);
    this._refObject.esc_info = this._wrapperFields.esc_info.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._counterIntialized = true;
    this._refObject.counter = refObject.counter;
    this._countIntialized = true;
    this._refObject.count = refObject.count;
    this._connection_typeIntialized = true;
    this._refObject.connection_type = refObject.connection_type;
    this._infoIntialized = true;
    this._refObject.info = refObject.info;
    this._wrapperFields.esc_info.copyRefObject(refObject.esc_info);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.esc_info.size != 0) {
      ESCInfoItemWrapper.ArrayType.freeArray(refObject.esc_info);
      if (ESCInfoItemWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.esc_info, ESCInfoItemWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ESCInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'ESCInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get counter() {
    if (!this._counterIntialized) {
      return undefined;
    }
    return this._refObject.counter;
  }
  set counter(value) {
    this._counterIntialized = true;
    this._refObject.counter = value;
  }
  get count() {
    if (!this._countIntialized) {
      return undefined;
    }
    return this._refObject.count;
  }
  set count(value) {
    this._countIntialized = true;
    this._refObject.count = value;
  }
  get connection_type() {
    if (!this._connection_typeIntialized) {
      return undefined;
    }
    return this._refObject.connection_type;
  }
  set connection_type(value) {
    this._connection_typeIntialized = true;
    this._refObject.connection_type = value;
  }
  get info() {
    if (!this._infoIntialized) {
      return undefined;
    }
    return this._refObject.info;
  }
  set info(value) {
    this._infoIntialized = true;
    this._refObject.info = value;
  }
  get esc_info() {
    return this._wrapperFields.esc_info;
  }
  set esc_info(value) {
    this._wrapperFields.esc_info.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ESCInfoRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._counterIntialized = true;
    this._countIntialized = true;
    this._connection_typeIntialized = true;
    this._infoIntialized = true;
    this._wrapperFields.esc_info.copyRefObject(this._refObject.esc_info);
  }
  copy(other) {
    this._refObject = new ESCInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._counterIntialized = true;
    this._countIntialized = true;
    this._connection_typeIntialized = true;
    this._infoIntialized = true;
    this._wrapperFields.esc_info.copy(other._wrapperFields.esc_info);
  }
  static get classType() {
    return ESCInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"connection_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"esc_info","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"mavros_msgs","type":"ESCInfoItem","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"ESCInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ESCInfo"};
  }
  hasMember(name) {
    let memberNames = ["header","counter","count","connection_type","info","esc_info"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ESCInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ESCInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ESCInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ESCInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ESCInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ESCInfoRefArray(size);
    this._refObject = new ESCInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ESCInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ESCInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ESCInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ESCInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ESCInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ESCInfoArrayWrapper;
  }
}
module.exports = ESCInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "connection_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "info",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "esc_info",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "mavros_msgs",
        "type": "ESCInfoItem",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "ESCInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ESCInfo"
}
*/
