// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__OpenDroneIDSystem.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const OpenDroneIDSystemRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  id_or_mac: primitiveTypes.string,
  operator_location_type: primitiveTypes.uint8,
  classification_type: primitiveTypes.uint8,
  operator_latitude: primitiveTypes.int32,
  operator_longitude: primitiveTypes.int32,
  area_count: primitiveTypes.uint16,
  area_radius: primitiveTypes.uint16,
  area_ceiling: primitiveTypes.float32,
  area_floor: primitiveTypes.float32,
  category_eu: primitiveTypes.uint8,
  class_eu: primitiveTypes.uint8,
  operator_altitude_geo: primitiveTypes.float32,
});
const OpenDroneIDSystemRefArray = ArrayType(OpenDroneIDSystemRefStruct);
const OpenDroneIDSystemRefStructArray = StructType({
  data: OpenDroneIDSystemRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OpenDroneIDSystemWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OpenDroneIDSystemRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._id_or_macIntialized = true;
      this._wrapperFields.id_or_mac =  new StringWrapper(other._wrapperFields.id_or_mac);
      this._operator_location_typeIntialized = true;
      this._classification_typeIntialized = true;
      this._operator_latitudeIntialized = true;
      this._operator_longitudeIntialized = true;
      this._area_countIntialized = true;
      this._area_radiusIntialized = true;
      this._area_ceilingIntialized = true;
      this._area_floorIntialized = true;
      this._category_euIntialized = true;
      this._class_euIntialized = true;
      this._operator_altitude_geoIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OpenDroneIDSystemRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._id_or_macIntialized = false;
    this._wrapperFields.id_or_mac = new StringWrapper();
    this._operator_location_typeIntialized = false;
    this._classification_typeIntialized = false;
    this._operator_latitudeIntialized = false;
    this._operator_longitudeIntialized = false;
    this._area_countIntialized = false;
    this._area_radiusIntialized = false;
    this._area_ceilingIntialized = false;
    this._area_floorIntialized = false;
    this._category_euIntialized = false;
    this._class_euIntialized = false;
    this._operator_altitude_geoIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new OpenDroneIDSystemWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OpenDroneIDSystemArrayWrapper;
  }
  static get ArrayType() {
    return OpenDroneIDSystemArrayWrapper;
  }
  static get refObjectArrayType() {
    return OpenDroneIDSystemRefStructArray
  }
  static get refObjectType() {
    return OpenDroneIDSystemRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._id_or_macIntialized) {
        throw new TypeError('Invalid argument: id_or_mac in OpenDroneIDSystem');
      }
      if (!this._operator_location_typeIntialized) {
        throw new TypeError('Invalid argument: operator_location_type in OpenDroneIDSystem');
      }
      if (!this._classification_typeIntialized) {
        throw new TypeError('Invalid argument: classification_type in OpenDroneIDSystem');
      }
      if (!this._operator_latitudeIntialized) {
        throw new TypeError('Invalid argument: operator_latitude in OpenDroneIDSystem');
      }
      if (!this._operator_longitudeIntialized) {
        throw new TypeError('Invalid argument: operator_longitude in OpenDroneIDSystem');
      }
      if (!this._area_countIntialized) {
        throw new TypeError('Invalid argument: area_count in OpenDroneIDSystem');
      }
      if (!this._area_radiusIntialized) {
        throw new TypeError('Invalid argument: area_radius in OpenDroneIDSystem');
      }
      if (!this._area_ceilingIntialized) {
        throw new TypeError('Invalid argument: area_ceiling in OpenDroneIDSystem');
      }
      if (!this._area_floorIntialized) {
        throw new TypeError('Invalid argument: area_floor in OpenDroneIDSystem');
      }
      if (!this._category_euIntialized) {
        throw new TypeError('Invalid argument: category_eu in OpenDroneIDSystem');
      }
      if (!this._class_euIntialized) {
        throw new TypeError('Invalid argument: class_eu in OpenDroneIDSystem');
      }
      if (!this._operator_altitude_geoIntialized) {
        throw new TypeError('Invalid argument: operator_altitude_geo in OpenDroneIDSystem');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.id_or_mac.freeze(own, checkConsistency);
    }
    this._refObject.id_or_mac = this._wrapperFields.id_or_mac.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.data = refObject.id_or_mac.data;
    this._operator_location_typeIntialized = true;
    this._refObject.operator_location_type = refObject.operator_location_type;
    this._classification_typeIntialized = true;
    this._refObject.classification_type = refObject.classification_type;
    this._operator_latitudeIntialized = true;
    this._refObject.operator_latitude = refObject.operator_latitude;
    this._operator_longitudeIntialized = true;
    this._refObject.operator_longitude = refObject.operator_longitude;
    this._area_countIntialized = true;
    this._refObject.area_count = refObject.area_count;
    this._area_radiusIntialized = true;
    this._refObject.area_radius = refObject.area_radius;
    this._area_ceilingIntialized = true;
    this._refObject.area_ceiling = refObject.area_ceiling;
    this._area_floorIntialized = true;
    this._refObject.area_floor = refObject.area_floor;
    this._category_euIntialized = true;
    this._refObject.category_eu = refObject.category_eu;
    this._class_euIntialized = true;
    this._refObject.class_eu = refObject.class_eu;
    this._operator_altitude_geoIntialized = true;
    this._refObject.operator_altitude_geo = refObject.operator_altitude_geo;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.id_or_mac);
  }
  static destoryRawROS(msg) {
    OpenDroneIDSystemWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'OpenDroneIDSystem'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get id_or_mac() {
    if (!this._id_or_macIntialized) {
      return undefined;
    }
    return this._wrapperFields.id_or_mac.data;
  }
  set id_or_mac(value) {
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.data = value;
  }
  get operator_location_type() {
    if (!this._operator_location_typeIntialized) {
      return undefined;
    }
    return this._refObject.operator_location_type;
  }
  set operator_location_type(value) {
    this._operator_location_typeIntialized = true;
    this._refObject.operator_location_type = value;
  }
  get classification_type() {
    if (!this._classification_typeIntialized) {
      return undefined;
    }
    return this._refObject.classification_type;
  }
  set classification_type(value) {
    this._classification_typeIntialized = true;
    this._refObject.classification_type = value;
  }
  get operator_latitude() {
    if (!this._operator_latitudeIntialized) {
      return undefined;
    }
    return this._refObject.operator_latitude;
  }
  set operator_latitude(value) {
    this._operator_latitudeIntialized = true;
    this._refObject.operator_latitude = value;
  }
  get operator_longitude() {
    if (!this._operator_longitudeIntialized) {
      return undefined;
    }
    return this._refObject.operator_longitude;
  }
  set operator_longitude(value) {
    this._operator_longitudeIntialized = true;
    this._refObject.operator_longitude = value;
  }
  get area_count() {
    if (!this._area_countIntialized) {
      return undefined;
    }
    return this._refObject.area_count;
  }
  set area_count(value) {
    this._area_countIntialized = true;
    this._refObject.area_count = value;
  }
  get area_radius() {
    if (!this._area_radiusIntialized) {
      return undefined;
    }
    return this._refObject.area_radius;
  }
  set area_radius(value) {
    this._area_radiusIntialized = true;
    this._refObject.area_radius = value;
  }
  get area_ceiling() {
    if (!this._area_ceilingIntialized) {
      return undefined;
    }
    return this._refObject.area_ceiling;
  }
  set area_ceiling(value) {
    this._area_ceilingIntialized = true;
    this._refObject.area_ceiling = value;
  }
  get area_floor() {
    if (!this._area_floorIntialized) {
      return undefined;
    }
    return this._refObject.area_floor;
  }
  set area_floor(value) {
    this._area_floorIntialized = true;
    this._refObject.area_floor = value;
  }
  get category_eu() {
    if (!this._category_euIntialized) {
      return undefined;
    }
    return this._refObject.category_eu;
  }
  set category_eu(value) {
    this._category_euIntialized = true;
    this._refObject.category_eu = value;
  }
  get class_eu() {
    if (!this._class_euIntialized) {
      return undefined;
    }
    return this._refObject.class_eu;
  }
  set class_eu(value) {
    this._class_euIntialized = true;
    this._refObject.class_eu = value;
  }
  get operator_altitude_geo() {
    if (!this._operator_altitude_geoIntialized) {
      return undefined;
    }
    return this._refObject.operator_altitude_geo;
  }
  set operator_altitude_geo(value) {
    this._operator_altitude_geoIntialized = true;
    this._refObject.operator_altitude_geo = value;
  }
  copyRefObject(refObject) {
    this._refObject = new OpenDroneIDSystemRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.copyRefObject(this._refObject.id_or_mac);
    this._operator_location_typeIntialized = true;
    this._classification_typeIntialized = true;
    this._operator_latitudeIntialized = true;
    this._operator_longitudeIntialized = true;
    this._area_countIntialized = true;
    this._area_radiusIntialized = true;
    this._area_ceilingIntialized = true;
    this._area_floorIntialized = true;
    this._category_euIntialized = true;
    this._class_euIntialized = true;
    this._operator_altitude_geoIntialized = true;
  }
  copy(other) {
    this._refObject = new OpenDroneIDSystemRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.copy(other._wrapperFields.id_or_mac);
    this._operator_location_typeIntialized = true;
    this._classification_typeIntialized = true;
    this._operator_latitudeIntialized = true;
    this._operator_longitudeIntialized = true;
    this._area_countIntialized = true;
    this._area_radiusIntialized = true;
    this._area_ceilingIntialized = true;
    this._area_floorIntialized = true;
    this._category_euIntialized = true;
    this._class_euIntialized = true;
    this._operator_altitude_geoIntialized = true;
  }
  static get classType() {
    return OpenDroneIDSystemWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"LOCATION_TYPE_TAKEOFF","value":0},{"type":"uint8","name":"LOCATION_TYPE_LIVE_GNSS","value":1},{"type":"uint8","name":"LOCATION_TYPE_FIXED","value":2},{"type":"uint8","name":"CLASSIFICATION_TYPE_UNDECLARED","value":0},{"type":"uint8","name":"CLASSIFICATION_TYPE_EU","value":1}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id_or_mac","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"operator_location_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"classification_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"operator_latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"operator_longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"area_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"area_radius","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"area_ceiling","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"area_floor","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"category_eu","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"class_eu","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"operator_altitude_geo","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"OpenDroneIDSystem","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OpenDroneIDSystem"};
  }
  hasMember(name) {
    let memberNames = ["header","id_or_mac","operator_location_type","classification_type","operator_latitude","operator_longitude","area_count","area_radius","area_ceiling","area_floor","category_eu","class_eu","operator_altitude_geo"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OpenDroneIDSystemArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OpenDroneIDSystemWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OpenDroneIDSystemWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpenDroneIDSystemArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpenDroneIDSystemArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OpenDroneIDSystemRefArray(size);
    this._refObject = new OpenDroneIDSystemRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OpenDroneIDSystemWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OpenDroneIDSystemArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OpenDroneIDSystemArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OpenDroneIDSystemWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OpenDroneIDSystemWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OpenDroneIDSystemArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(OpenDroneIDSystemWrapper, "LOCATION_TYPE_TAKEOFF", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDSystemWrapper, "LOCATION_TYPE_LIVE_GNSS", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDSystemWrapper, "LOCATION_TYPE_FIXED", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDSystemWrapper, "CLASSIFICATION_TYPE_UNDECLARED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDSystemWrapper, "CLASSIFICATION_TYPE_EU", {value: 1, writable: false, enumerable: true, configurable: true});
module.exports = OpenDroneIDSystemWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "LOCATION_TYPE_TAKEOFF",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "LOCATION_TYPE_LIVE_GNSS",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "LOCATION_TYPE_FIXED",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "CLASSIFICATION_TYPE_UNDECLARED",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "CLASSIFICATION_TYPE_EU",
      "value": 1
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "id_or_mac",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "operator_location_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "classification_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "operator_latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "operator_longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "area_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "area_radius",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "area_ceiling",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "area_floor",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "category_eu",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "class_eu",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "operator_altitude_geo",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "OpenDroneIDSystem",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OpenDroneIDSystem"
}
*/
