// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__LogEntry.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const LogEntryRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  id: primitiveTypes.uint16,
  num_logs: primitiveTypes.uint16,
  last_log_num: primitiveTypes.uint16,
  time_utc: TimeWrapper.refObjectType,
  size: primitiveTypes.uint32,
});
const LogEntryRefArray = ArrayType(LogEntryRefStruct);
const LogEntryRefStructArray = StructType({
  data: LogEntryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LogEntryWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LogEntryRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._idIntialized = true;
      this._num_logsIntialized = true;
      this._last_log_numIntialized = true;
      this._wrapperFields.time_utc =  new TimeWrapper(other._wrapperFields.time_utc);
      this._sizeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LogEntryRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._idIntialized = false;
    this._num_logsIntialized = false;
    this._last_log_numIntialized = false;
    this._wrapperFields.time_utc = new TimeWrapper();
    this._sizeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LogEntryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LogEntryArrayWrapper;
  }
  static get ArrayType() {
    return LogEntryArrayWrapper;
  }
  static get refObjectArrayType() {
    return LogEntryRefStructArray
  }
  static get refObjectType() {
    return LogEntryRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in LogEntry');
      }
      if (!this._num_logsIntialized) {
        throw new TypeError('Invalid argument: num_logs in LogEntry');
      }
      if (!this._last_log_numIntialized) {
        throw new TypeError('Invalid argument: last_log_num in LogEntry');
      }
      if (!this._sizeIntialized) {
        throw new TypeError('Invalid argument: size in LogEntry');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.time_utc.freeze(own, checkConsistency);
    this._refObject.time_utc = this._wrapperFields.time_utc.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._num_logsIntialized = true;
    this._refObject.num_logs = refObject.num_logs;
    this._last_log_numIntialized = true;
    this._refObject.last_log_num = refObject.last_log_num;
    this._wrapperFields.time_utc.copyRefObject(refObject.time_utc);
    this._sizeIntialized = true;
    this._refObject.size = refObject.size;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    TimeWrapper.freeStruct(refObject.time_utc);
  }
  static destoryRawROS(msg) {
    LogEntryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'LogEntry'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get num_logs() {
    if (!this._num_logsIntialized) {
      return undefined;
    }
    return this._refObject.num_logs;
  }
  set num_logs(value) {
    this._num_logsIntialized = true;
    this._refObject.num_logs = value;
  }
  get last_log_num() {
    if (!this._last_log_numIntialized) {
      return undefined;
    }
    return this._refObject.last_log_num;
  }
  set last_log_num(value) {
    this._last_log_numIntialized = true;
    this._refObject.last_log_num = value;
  }
  get time_utc() {
    return this._wrapperFields.time_utc;
  }
  set time_utc(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.time_utc.copy(value);
    } else {
      this._wrapperFields.time_utc.copy(new TimeWrapper(value));
    }
  }
  get size() {
    if (!this._sizeIntialized) {
      return undefined;
    }
    return this._refObject.size;
  }
  set size(value) {
    this._sizeIntialized = true;
    this._refObject.size = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LogEntryRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._idIntialized = true;
    this._num_logsIntialized = true;
    this._last_log_numIntialized = true;
    this._wrapperFields.time_utc.copyRefObject(this._refObject.time_utc);
    this._sizeIntialized = true;
  }
  copy(other) {
    this._refObject = new LogEntryRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._idIntialized = true;
    this._num_logsIntialized = true;
    this._last_log_numIntialized = true;
    this._wrapperFields.time_utc.copy(other._wrapperFields.time_utc);
    this._sizeIntialized = true;
  }
  static get classType() {
    return LogEntryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"num_logs","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"last_log_num","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_utc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"size","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"LogEntry","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LogEntry"};
  }
  hasMember(name) {
    let memberNames = ["header","id","num_logs","last_log_num","time_utc","size"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LogEntryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LogEntryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LogEntryWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LogEntryArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LogEntryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LogEntryRefArray(size);
    this._refObject = new LogEntryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LogEntryWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LogEntryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LogEntryArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LogEntryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LogEntryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LogEntryArrayWrapper;
  }
}
module.exports = LogEntryWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "num_logs",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "last_log_num",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_utc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "size",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "LogEntry",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LogEntry"
}
*/
