// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__MagnetometerReporter.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const MagnetometerReporterRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  report: primitiveTypes.uint8,
  confidence: primitiveTypes.float32,
  compass_id: primitiveTypes.uint8,
  cal_mask: primitiveTypes.uint8,
  cal_status: primitiveTypes.uint8,
  autosaved: primitiveTypes.uint8,
  fitness: primitiveTypes.float32,
  ofs_x: primitiveTypes.float32,
  ofs_y: primitiveTypes.float32,
  ofs_z: primitiveTypes.float32,
  diag_x: primitiveTypes.float32,
  diag_y: primitiveTypes.float32,
  diag_z: primitiveTypes.float32,
  offdiag_x: primitiveTypes.float32,
  offdiag_y: primitiveTypes.float32,
  offdiag_z: primitiveTypes.float32,
  orientation_confidence: primitiveTypes.float32,
  old_orientation: primitiveTypes.uint8,
  new_orientation: primitiveTypes.uint8,
  scale_factor: primitiveTypes.float32,
});
const MagnetometerReporterRefArray = ArrayType(MagnetometerReporterRefStruct);
const MagnetometerReporterRefStructArray = StructType({
  data: MagnetometerReporterRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MagnetometerReporterWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MagnetometerReporterRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._reportIntialized = true;
      this._confidenceIntialized = true;
      this._compass_idIntialized = true;
      this._cal_maskIntialized = true;
      this._cal_statusIntialized = true;
      this._autosavedIntialized = true;
      this._fitnessIntialized = true;
      this._ofs_xIntialized = true;
      this._ofs_yIntialized = true;
      this._ofs_zIntialized = true;
      this._diag_xIntialized = true;
      this._diag_yIntialized = true;
      this._diag_zIntialized = true;
      this._offdiag_xIntialized = true;
      this._offdiag_yIntialized = true;
      this._offdiag_zIntialized = true;
      this._orientation_confidenceIntialized = true;
      this._old_orientationIntialized = true;
      this._new_orientationIntialized = true;
      this._scale_factorIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MagnetometerReporterRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._reportIntialized = false;
    this._confidenceIntialized = false;
    this._compass_idIntialized = false;
    this._cal_maskIntialized = false;
    this._cal_statusIntialized = false;
    this._autosavedIntialized = false;
    this._fitnessIntialized = false;
    this._ofs_xIntialized = false;
    this._ofs_yIntialized = false;
    this._ofs_zIntialized = false;
    this._diag_xIntialized = false;
    this._diag_yIntialized = false;
    this._diag_zIntialized = false;
    this._offdiag_xIntialized = false;
    this._offdiag_yIntialized = false;
    this._offdiag_zIntialized = false;
    this._orientation_confidenceIntialized = false;
    this._old_orientationIntialized = false;
    this._new_orientationIntialized = false;
    this._scale_factorIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MagnetometerReporterWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MagnetometerReporterArrayWrapper;
  }
  static get ArrayType() {
    return MagnetometerReporterArrayWrapper;
  }
  static get refObjectArrayType() {
    return MagnetometerReporterRefStructArray
  }
  static get refObjectType() {
    return MagnetometerReporterRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._reportIntialized) {
        throw new TypeError('Invalid argument: report in MagnetometerReporter');
      }
      if (!this._confidenceIntialized) {
        throw new TypeError('Invalid argument: confidence in MagnetometerReporter');
      }
      if (!this._compass_idIntialized) {
        throw new TypeError('Invalid argument: compass_id in MagnetometerReporter');
      }
      if (!this._cal_maskIntialized) {
        throw new TypeError('Invalid argument: cal_mask in MagnetometerReporter');
      }
      if (!this._cal_statusIntialized) {
        throw new TypeError('Invalid argument: cal_status in MagnetometerReporter');
      }
      if (!this._autosavedIntialized) {
        throw new TypeError('Invalid argument: autosaved in MagnetometerReporter');
      }
      if (!this._fitnessIntialized) {
        throw new TypeError('Invalid argument: fitness in MagnetometerReporter');
      }
      if (!this._ofs_xIntialized) {
        throw new TypeError('Invalid argument: ofs_x in MagnetometerReporter');
      }
      if (!this._ofs_yIntialized) {
        throw new TypeError('Invalid argument: ofs_y in MagnetometerReporter');
      }
      if (!this._ofs_zIntialized) {
        throw new TypeError('Invalid argument: ofs_z in MagnetometerReporter');
      }
      if (!this._diag_xIntialized) {
        throw new TypeError('Invalid argument: diag_x in MagnetometerReporter');
      }
      if (!this._diag_yIntialized) {
        throw new TypeError('Invalid argument: diag_y in MagnetometerReporter');
      }
      if (!this._diag_zIntialized) {
        throw new TypeError('Invalid argument: diag_z in MagnetometerReporter');
      }
      if (!this._offdiag_xIntialized) {
        throw new TypeError('Invalid argument: offdiag_x in MagnetometerReporter');
      }
      if (!this._offdiag_yIntialized) {
        throw new TypeError('Invalid argument: offdiag_y in MagnetometerReporter');
      }
      if (!this._offdiag_zIntialized) {
        throw new TypeError('Invalid argument: offdiag_z in MagnetometerReporter');
      }
      if (!this._orientation_confidenceIntialized) {
        throw new TypeError('Invalid argument: orientation_confidence in MagnetometerReporter');
      }
      if (!this._old_orientationIntialized) {
        throw new TypeError('Invalid argument: old_orientation in MagnetometerReporter');
      }
      if (!this._new_orientationIntialized) {
        throw new TypeError('Invalid argument: new_orientation in MagnetometerReporter');
      }
      if (!this._scale_factorIntialized) {
        throw new TypeError('Invalid argument: scale_factor in MagnetometerReporter');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._reportIntialized = true;
    this._refObject.report = refObject.report;
    this._confidenceIntialized = true;
    this._refObject.confidence = refObject.confidence;
    this._compass_idIntialized = true;
    this._refObject.compass_id = refObject.compass_id;
    this._cal_maskIntialized = true;
    this._refObject.cal_mask = refObject.cal_mask;
    this._cal_statusIntialized = true;
    this._refObject.cal_status = refObject.cal_status;
    this._autosavedIntialized = true;
    this._refObject.autosaved = refObject.autosaved;
    this._fitnessIntialized = true;
    this._refObject.fitness = refObject.fitness;
    this._ofs_xIntialized = true;
    this._refObject.ofs_x = refObject.ofs_x;
    this._ofs_yIntialized = true;
    this._refObject.ofs_y = refObject.ofs_y;
    this._ofs_zIntialized = true;
    this._refObject.ofs_z = refObject.ofs_z;
    this._diag_xIntialized = true;
    this._refObject.diag_x = refObject.diag_x;
    this._diag_yIntialized = true;
    this._refObject.diag_y = refObject.diag_y;
    this._diag_zIntialized = true;
    this._refObject.diag_z = refObject.diag_z;
    this._offdiag_xIntialized = true;
    this._refObject.offdiag_x = refObject.offdiag_x;
    this._offdiag_yIntialized = true;
    this._refObject.offdiag_y = refObject.offdiag_y;
    this._offdiag_zIntialized = true;
    this._refObject.offdiag_z = refObject.offdiag_z;
    this._orientation_confidenceIntialized = true;
    this._refObject.orientation_confidence = refObject.orientation_confidence;
    this._old_orientationIntialized = true;
    this._refObject.old_orientation = refObject.old_orientation;
    this._new_orientationIntialized = true;
    this._refObject.new_orientation = refObject.new_orientation;
    this._scale_factorIntialized = true;
    this._refObject.scale_factor = refObject.scale_factor;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    MagnetometerReporterWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'MagnetometerReporter'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get report() {
    if (!this._reportIntialized) {
      return undefined;
    }
    return this._refObject.report;
  }
  set report(value) {
    this._reportIntialized = true;
    this._refObject.report = value;
  }
  get confidence() {
    if (!this._confidenceIntialized) {
      return undefined;
    }
    return this._refObject.confidence;
  }
  set confidence(value) {
    this._confidenceIntialized = true;
    this._refObject.confidence = value;
  }
  get compass_id() {
    if (!this._compass_idIntialized) {
      return undefined;
    }
    return this._refObject.compass_id;
  }
  set compass_id(value) {
    this._compass_idIntialized = true;
    this._refObject.compass_id = value;
  }
  get cal_mask() {
    if (!this._cal_maskIntialized) {
      return undefined;
    }
    return this._refObject.cal_mask;
  }
  set cal_mask(value) {
    this._cal_maskIntialized = true;
    this._refObject.cal_mask = value;
  }
  get cal_status() {
    if (!this._cal_statusIntialized) {
      return undefined;
    }
    return this._refObject.cal_status;
  }
  set cal_status(value) {
    this._cal_statusIntialized = true;
    this._refObject.cal_status = value;
  }
  get autosaved() {
    if (!this._autosavedIntialized) {
      return undefined;
    }
    return this._refObject.autosaved;
  }
  set autosaved(value) {
    this._autosavedIntialized = true;
    this._refObject.autosaved = value;
  }
  get fitness() {
    if (!this._fitnessIntialized) {
      return undefined;
    }
    return this._refObject.fitness;
  }
  set fitness(value) {
    this._fitnessIntialized = true;
    this._refObject.fitness = value;
  }
  get ofs_x() {
    if (!this._ofs_xIntialized) {
      return undefined;
    }
    return this._refObject.ofs_x;
  }
  set ofs_x(value) {
    this._ofs_xIntialized = true;
    this._refObject.ofs_x = value;
  }
  get ofs_y() {
    if (!this._ofs_yIntialized) {
      return undefined;
    }
    return this._refObject.ofs_y;
  }
  set ofs_y(value) {
    this._ofs_yIntialized = true;
    this._refObject.ofs_y = value;
  }
  get ofs_z() {
    if (!this._ofs_zIntialized) {
      return undefined;
    }
    return this._refObject.ofs_z;
  }
  set ofs_z(value) {
    this._ofs_zIntialized = true;
    this._refObject.ofs_z = value;
  }
  get diag_x() {
    if (!this._diag_xIntialized) {
      return undefined;
    }
    return this._refObject.diag_x;
  }
  set diag_x(value) {
    this._diag_xIntialized = true;
    this._refObject.diag_x = value;
  }
  get diag_y() {
    if (!this._diag_yIntialized) {
      return undefined;
    }
    return this._refObject.diag_y;
  }
  set diag_y(value) {
    this._diag_yIntialized = true;
    this._refObject.diag_y = value;
  }
  get diag_z() {
    if (!this._diag_zIntialized) {
      return undefined;
    }
    return this._refObject.diag_z;
  }
  set diag_z(value) {
    this._diag_zIntialized = true;
    this._refObject.diag_z = value;
  }
  get offdiag_x() {
    if (!this._offdiag_xIntialized) {
      return undefined;
    }
    return this._refObject.offdiag_x;
  }
  set offdiag_x(value) {
    this._offdiag_xIntialized = true;
    this._refObject.offdiag_x = value;
  }
  get offdiag_y() {
    if (!this._offdiag_yIntialized) {
      return undefined;
    }
    return this._refObject.offdiag_y;
  }
  set offdiag_y(value) {
    this._offdiag_yIntialized = true;
    this._refObject.offdiag_y = value;
  }
  get offdiag_z() {
    if (!this._offdiag_zIntialized) {
      return undefined;
    }
    return this._refObject.offdiag_z;
  }
  set offdiag_z(value) {
    this._offdiag_zIntialized = true;
    this._refObject.offdiag_z = value;
  }
  get orientation_confidence() {
    if (!this._orientation_confidenceIntialized) {
      return undefined;
    }
    return this._refObject.orientation_confidence;
  }
  set orientation_confidence(value) {
    this._orientation_confidenceIntialized = true;
    this._refObject.orientation_confidence = value;
  }
  get old_orientation() {
    if (!this._old_orientationIntialized) {
      return undefined;
    }
    return this._refObject.old_orientation;
  }
  set old_orientation(value) {
    this._old_orientationIntialized = true;
    this._refObject.old_orientation = value;
  }
  get new_orientation() {
    if (!this._new_orientationIntialized) {
      return undefined;
    }
    return this._refObject.new_orientation;
  }
  set new_orientation(value) {
    this._new_orientationIntialized = true;
    this._refObject.new_orientation = value;
  }
  get scale_factor() {
    if (!this._scale_factorIntialized) {
      return undefined;
    }
    return this._refObject.scale_factor;
  }
  set scale_factor(value) {
    this._scale_factorIntialized = true;
    this._refObject.scale_factor = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MagnetometerReporterRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._reportIntialized = true;
    this._confidenceIntialized = true;
    this._compass_idIntialized = true;
    this._cal_maskIntialized = true;
    this._cal_statusIntialized = true;
    this._autosavedIntialized = true;
    this._fitnessIntialized = true;
    this._ofs_xIntialized = true;
    this._ofs_yIntialized = true;
    this._ofs_zIntialized = true;
    this._diag_xIntialized = true;
    this._diag_yIntialized = true;
    this._diag_zIntialized = true;
    this._offdiag_xIntialized = true;
    this._offdiag_yIntialized = true;
    this._offdiag_zIntialized = true;
    this._orientation_confidenceIntialized = true;
    this._old_orientationIntialized = true;
    this._new_orientationIntialized = true;
    this._scale_factorIntialized = true;
  }
  copy(other) {
    this._refObject = new MagnetometerReporterRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._reportIntialized = true;
    this._confidenceIntialized = true;
    this._compass_idIntialized = true;
    this._cal_maskIntialized = true;
    this._cal_statusIntialized = true;
    this._autosavedIntialized = true;
    this._fitnessIntialized = true;
    this._ofs_xIntialized = true;
    this._ofs_yIntialized = true;
    this._ofs_zIntialized = true;
    this._diag_xIntialized = true;
    this._diag_yIntialized = true;
    this._diag_zIntialized = true;
    this._offdiag_xIntialized = true;
    this._offdiag_yIntialized = true;
    this._offdiag_zIntialized = true;
    this._orientation_confidenceIntialized = true;
    this._old_orientationIntialized = true;
    this._new_orientationIntialized = true;
    this._scale_factorIntialized = true;
  }
  static get classType() {
    return MagnetometerReporterWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"report","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"confidence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"compass_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cal_mask","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cal_status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"autosaved","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fitness","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ofs_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ofs_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ofs_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"diag_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"diag_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"diag_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"offdiag_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"offdiag_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"offdiag_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"orientation_confidence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"old_orientation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"new_orientation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"scale_factor","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"MagnetometerReporter","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MagnetometerReporter"};
  }
  hasMember(name) {
    let memberNames = ["header","report","confidence","compass_id","cal_mask","cal_status","autosaved","fitness","ofs_x","ofs_y","ofs_z","diag_x","diag_y","diag_z","offdiag_x","offdiag_y","offdiag_z","orientation_confidence","old_orientation","new_orientation","scale_factor"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MagnetometerReporterArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MagnetometerReporterWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MagnetometerReporterWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MagnetometerReporterArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MagnetometerReporterArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MagnetometerReporterRefArray(size);
    this._refObject = new MagnetometerReporterRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MagnetometerReporterWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MagnetometerReporterArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MagnetometerReporterArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MagnetometerReporterWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MagnetometerReporterWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MagnetometerReporterArrayWrapper;
  }
}
module.exports = MagnetometerReporterWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "report",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "confidence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "compass_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cal_mask",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cal_status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "autosaved",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fitness",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ofs_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ofs_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ofs_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "diag_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "diag_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "diag_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "offdiag_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "offdiag_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "offdiag_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "orientation_confidence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "old_orientation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "new_orientation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "scale_factor",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "MagnetometerReporter",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MagnetometerReporter"
}
*/
