// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__GPSINPUT.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const GPSINPUTRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  fix_type: primitiveTypes.uint8,
  gps_id: primitiveTypes.uint8,
  ignore_flags: primitiveTypes.uint16,
  time_week_ms: primitiveTypes.uint32,
  time_week: primitiveTypes.uint16,
  lat: primitiveTypes.int32,
  lon: primitiveTypes.int32,
  alt: primitiveTypes.float32,
  hdop: primitiveTypes.float32,
  vdop: primitiveTypes.float32,
  vn: primitiveTypes.float32,
  ve: primitiveTypes.float32,
  vd: primitiveTypes.float32,
  speed_accuracy: primitiveTypes.float32,
  horiz_accuracy: primitiveTypes.float32,
  vert_accuracy: primitiveTypes.float32,
  satellites_visible: primitiveTypes.uint8,
  yaw: primitiveTypes.uint16,
});
const GPSINPUTRefArray = ArrayType(GPSINPUTRefStruct);
const GPSINPUTRefStructArray = StructType({
  data: GPSINPUTRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GPSINPUTWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GPSINPUTRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._fix_typeIntialized = true;
      this._gps_idIntialized = true;
      this._ignore_flagsIntialized = true;
      this._time_week_msIntialized = true;
      this._time_weekIntialized = true;
      this._latIntialized = true;
      this._lonIntialized = true;
      this._altIntialized = true;
      this._hdopIntialized = true;
      this._vdopIntialized = true;
      this._vnIntialized = true;
      this._veIntialized = true;
      this._vdIntialized = true;
      this._speed_accuracyIntialized = true;
      this._horiz_accuracyIntialized = true;
      this._vert_accuracyIntialized = true;
      this._satellites_visibleIntialized = true;
      this._yawIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GPSINPUTRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._fix_typeIntialized = false;
    this._gps_idIntialized = false;
    this._ignore_flagsIntialized = false;
    this._time_week_msIntialized = false;
    this._time_weekIntialized = false;
    this._latIntialized = false;
    this._lonIntialized = false;
    this._altIntialized = false;
    this._hdopIntialized = false;
    this._vdopIntialized = false;
    this._vnIntialized = false;
    this._veIntialized = false;
    this._vdIntialized = false;
    this._speed_accuracyIntialized = false;
    this._horiz_accuracyIntialized = false;
    this._vert_accuracyIntialized = false;
    this._satellites_visibleIntialized = false;
    this._yawIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GPSINPUTWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GPSINPUTArrayWrapper;
  }
  static get ArrayType() {
    return GPSINPUTArrayWrapper;
  }
  static get refObjectArrayType() {
    return GPSINPUTRefStructArray
  }
  static get refObjectType() {
    return GPSINPUTRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._fix_typeIntialized) {
        throw new TypeError('Invalid argument: fix_type in GPSINPUT');
      }
      if (!this._gps_idIntialized) {
        throw new TypeError('Invalid argument: gps_id in GPSINPUT');
      }
      if (!this._ignore_flagsIntialized) {
        throw new TypeError('Invalid argument: ignore_flags in GPSINPUT');
      }
      if (!this._time_week_msIntialized) {
        throw new TypeError('Invalid argument: time_week_ms in GPSINPUT');
      }
      if (!this._time_weekIntialized) {
        throw new TypeError('Invalid argument: time_week in GPSINPUT');
      }
      if (!this._latIntialized) {
        throw new TypeError('Invalid argument: lat in GPSINPUT');
      }
      if (!this._lonIntialized) {
        throw new TypeError('Invalid argument: lon in GPSINPUT');
      }
      if (!this._altIntialized) {
        throw new TypeError('Invalid argument: alt in GPSINPUT');
      }
      if (!this._hdopIntialized) {
        throw new TypeError('Invalid argument: hdop in GPSINPUT');
      }
      if (!this._vdopIntialized) {
        throw new TypeError('Invalid argument: vdop in GPSINPUT');
      }
      if (!this._vnIntialized) {
        throw new TypeError('Invalid argument: vn in GPSINPUT');
      }
      if (!this._veIntialized) {
        throw new TypeError('Invalid argument: ve in GPSINPUT');
      }
      if (!this._vdIntialized) {
        throw new TypeError('Invalid argument: vd in GPSINPUT');
      }
      if (!this._speed_accuracyIntialized) {
        throw new TypeError('Invalid argument: speed_accuracy in GPSINPUT');
      }
      if (!this._horiz_accuracyIntialized) {
        throw new TypeError('Invalid argument: horiz_accuracy in GPSINPUT');
      }
      if (!this._vert_accuracyIntialized) {
        throw new TypeError('Invalid argument: vert_accuracy in GPSINPUT');
      }
      if (!this._satellites_visibleIntialized) {
        throw new TypeError('Invalid argument: satellites_visible in GPSINPUT');
      }
      if (!this._yawIntialized) {
        throw new TypeError('Invalid argument: yaw in GPSINPUT');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._fix_typeIntialized = true;
    this._refObject.fix_type = refObject.fix_type;
    this._gps_idIntialized = true;
    this._refObject.gps_id = refObject.gps_id;
    this._ignore_flagsIntialized = true;
    this._refObject.ignore_flags = refObject.ignore_flags;
    this._time_week_msIntialized = true;
    this._refObject.time_week_ms = refObject.time_week_ms;
    this._time_weekIntialized = true;
    this._refObject.time_week = refObject.time_week;
    this._latIntialized = true;
    this._refObject.lat = refObject.lat;
    this._lonIntialized = true;
    this._refObject.lon = refObject.lon;
    this._altIntialized = true;
    this._refObject.alt = refObject.alt;
    this._hdopIntialized = true;
    this._refObject.hdop = refObject.hdop;
    this._vdopIntialized = true;
    this._refObject.vdop = refObject.vdop;
    this._vnIntialized = true;
    this._refObject.vn = refObject.vn;
    this._veIntialized = true;
    this._refObject.ve = refObject.ve;
    this._vdIntialized = true;
    this._refObject.vd = refObject.vd;
    this._speed_accuracyIntialized = true;
    this._refObject.speed_accuracy = refObject.speed_accuracy;
    this._horiz_accuracyIntialized = true;
    this._refObject.horiz_accuracy = refObject.horiz_accuracy;
    this._vert_accuracyIntialized = true;
    this._refObject.vert_accuracy = refObject.vert_accuracy;
    this._satellites_visibleIntialized = true;
    this._refObject.satellites_visible = refObject.satellites_visible;
    this._yawIntialized = true;
    this._refObject.yaw = refObject.yaw;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    GPSINPUTWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'GPSINPUT'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get fix_type() {
    if (!this._fix_typeIntialized) {
      return undefined;
    }
    return this._refObject.fix_type;
  }
  set fix_type(value) {
    this._fix_typeIntialized = true;
    this._refObject.fix_type = value;
  }
  get gps_id() {
    if (!this._gps_idIntialized) {
      return undefined;
    }
    return this._refObject.gps_id;
  }
  set gps_id(value) {
    this._gps_idIntialized = true;
    this._refObject.gps_id = value;
  }
  get ignore_flags() {
    if (!this._ignore_flagsIntialized) {
      return undefined;
    }
    return this._refObject.ignore_flags;
  }
  set ignore_flags(value) {
    this._ignore_flagsIntialized = true;
    this._refObject.ignore_flags = value;
  }
  get time_week_ms() {
    if (!this._time_week_msIntialized) {
      return undefined;
    }
    return this._refObject.time_week_ms;
  }
  set time_week_ms(value) {
    this._time_week_msIntialized = true;
    this._refObject.time_week_ms = value;
  }
  get time_week() {
    if (!this._time_weekIntialized) {
      return undefined;
    }
    return this._refObject.time_week;
  }
  set time_week(value) {
    this._time_weekIntialized = true;
    this._refObject.time_week = value;
  }
  get lat() {
    if (!this._latIntialized) {
      return undefined;
    }
    return this._refObject.lat;
  }
  set lat(value) {
    this._latIntialized = true;
    this._refObject.lat = value;
  }
  get lon() {
    if (!this._lonIntialized) {
      return undefined;
    }
    return this._refObject.lon;
  }
  set lon(value) {
    this._lonIntialized = true;
    this._refObject.lon = value;
  }
  get alt() {
    if (!this._altIntialized) {
      return undefined;
    }
    return this._refObject.alt;
  }
  set alt(value) {
    this._altIntialized = true;
    this._refObject.alt = value;
  }
  get hdop() {
    if (!this._hdopIntialized) {
      return undefined;
    }
    return this._refObject.hdop;
  }
  set hdop(value) {
    this._hdopIntialized = true;
    this._refObject.hdop = value;
  }
  get vdop() {
    if (!this._vdopIntialized) {
      return undefined;
    }
    return this._refObject.vdop;
  }
  set vdop(value) {
    this._vdopIntialized = true;
    this._refObject.vdop = value;
  }
  get vn() {
    if (!this._vnIntialized) {
      return undefined;
    }
    return this._refObject.vn;
  }
  set vn(value) {
    this._vnIntialized = true;
    this._refObject.vn = value;
  }
  get ve() {
    if (!this._veIntialized) {
      return undefined;
    }
    return this._refObject.ve;
  }
  set ve(value) {
    this._veIntialized = true;
    this._refObject.ve = value;
  }
  get vd() {
    if (!this._vdIntialized) {
      return undefined;
    }
    return this._refObject.vd;
  }
  set vd(value) {
    this._vdIntialized = true;
    this._refObject.vd = value;
  }
  get speed_accuracy() {
    if (!this._speed_accuracyIntialized) {
      return undefined;
    }
    return this._refObject.speed_accuracy;
  }
  set speed_accuracy(value) {
    this._speed_accuracyIntialized = true;
    this._refObject.speed_accuracy = value;
  }
  get horiz_accuracy() {
    if (!this._horiz_accuracyIntialized) {
      return undefined;
    }
    return this._refObject.horiz_accuracy;
  }
  set horiz_accuracy(value) {
    this._horiz_accuracyIntialized = true;
    this._refObject.horiz_accuracy = value;
  }
  get vert_accuracy() {
    if (!this._vert_accuracyIntialized) {
      return undefined;
    }
    return this._refObject.vert_accuracy;
  }
  set vert_accuracy(value) {
    this._vert_accuracyIntialized = true;
    this._refObject.vert_accuracy = value;
  }
  get satellites_visible() {
    if (!this._satellites_visibleIntialized) {
      return undefined;
    }
    return this._refObject.satellites_visible;
  }
  set satellites_visible(value) {
    this._satellites_visibleIntialized = true;
    this._refObject.satellites_visible = value;
  }
  get yaw() {
    if (!this._yawIntialized) {
      return undefined;
    }
    return this._refObject.yaw;
  }
  set yaw(value) {
    this._yawIntialized = true;
    this._refObject.yaw = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GPSINPUTRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._fix_typeIntialized = true;
    this._gps_idIntialized = true;
    this._ignore_flagsIntialized = true;
    this._time_week_msIntialized = true;
    this._time_weekIntialized = true;
    this._latIntialized = true;
    this._lonIntialized = true;
    this._altIntialized = true;
    this._hdopIntialized = true;
    this._vdopIntialized = true;
    this._vnIntialized = true;
    this._veIntialized = true;
    this._vdIntialized = true;
    this._speed_accuracyIntialized = true;
    this._horiz_accuracyIntialized = true;
    this._vert_accuracyIntialized = true;
    this._satellites_visibleIntialized = true;
    this._yawIntialized = true;
  }
  copy(other) {
    this._refObject = new GPSINPUTRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._fix_typeIntialized = true;
    this._gps_idIntialized = true;
    this._ignore_flagsIntialized = true;
    this._time_week_msIntialized = true;
    this._time_weekIntialized = true;
    this._latIntialized = true;
    this._lonIntialized = true;
    this._altIntialized = true;
    this._hdopIntialized = true;
    this._vdopIntialized = true;
    this._vnIntialized = true;
    this._veIntialized = true;
    this._vdIntialized = true;
    this._speed_accuracyIntialized = true;
    this._horiz_accuracyIntialized = true;
    this._vert_accuracyIntialized = true;
    this._satellites_visibleIntialized = true;
    this._yawIntialized = true;
  }
  static get classType() {
    return GPSINPUTWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"GPS_FIX_TYPE_NO_GPS","value":0},{"type":"uint8","name":"GPS_FIX_TYPE_NO_FIX","value":1},{"type":"uint8","name":"GPS_FIX_TYPE_2D_FIX","value":2},{"type":"uint8","name":"GPS_FIX_TYPE_3D_FIX","value":3},{"type":"uint8","name":"GPS_FIX_TYPE_DGPS","value":4},{"type":"uint8","name":"GPS_FIX_TYPE_RTK_FLOATR","value":5},{"type":"uint8","name":"GPS_FIX_TYPE_RTK_FIXEDR","value":6},{"type":"uint8","name":"GPS_FIX_TYPE_STATIC","value":7},{"type":"uint8","name":"GPS_FIX_TYPE_PPP","value":8}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"fix_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ignore_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_week_ms","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_week","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lat","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lon","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"alt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hdop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vdop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vn","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ve","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vd","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"speed_accuracy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"horiz_accuracy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vert_accuracy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"satellites_visible","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"GPSINPUT","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GPSINPUT"};
  }
  hasMember(name) {
    let memberNames = ["header","fix_type","gps_id","ignore_flags","time_week_ms","time_week","lat","lon","alt","hdop","vdop","vn","ve","vd","speed_accuracy","horiz_accuracy","vert_accuracy","satellites_visible","yaw"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GPSINPUTArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GPSINPUTWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GPSINPUTWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPSINPUTArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPSINPUTArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GPSINPUTRefArray(size);
    this._refObject = new GPSINPUTRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GPSINPUTWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GPSINPUTArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GPSINPUTArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GPSINPUTWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GPSINPUTWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GPSINPUTArrayWrapper;
  }
}
// Define constants (9 in total)
Object.defineProperty(GPSINPUTWrapper, "GPS_FIX_TYPE_NO_GPS", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSINPUTWrapper, "GPS_FIX_TYPE_NO_FIX", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSINPUTWrapper, "GPS_FIX_TYPE_2D_FIX", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSINPUTWrapper, "GPS_FIX_TYPE_3D_FIX", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSINPUTWrapper, "GPS_FIX_TYPE_DGPS", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSINPUTWrapper, "GPS_FIX_TYPE_RTK_FLOATR", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSINPUTWrapper, "GPS_FIX_TYPE_RTK_FIXEDR", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSINPUTWrapper, "GPS_FIX_TYPE_STATIC", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSINPUTWrapper, "GPS_FIX_TYPE_PPP", {value: 8, writable: false, enumerable: true, configurable: true});
module.exports = GPSINPUTWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_NO_GPS",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_NO_FIX",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_2D_FIX",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_3D_FIX",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_DGPS",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_RTK_FLOATR",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_RTK_FIXEDR",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_STATIC",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "GPS_FIX_TYPE_PPP",
      "value": 8
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "fix_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ignore_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_week_ms",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_week",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lat",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lon",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "alt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hdop",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vdop",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vn",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ve",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vd",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "speed_accuracy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "horiz_accuracy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vert_accuracy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "satellites_visible",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "GPSINPUT",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GPSINPUT"
}
*/
