// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__CommandLong_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CommandLong_RequestRefStruct = StructType({
  broadcast: primitiveTypes.bool,
  command: primitiveTypes.uint16,
  confirmation: primitiveTypes.uint8,
  param1: primitiveTypes.float32,
  param2: primitiveTypes.float32,
  param3: primitiveTypes.float32,
  param4: primitiveTypes.float32,
  param5: primitiveTypes.float32,
  param6: primitiveTypes.float32,
  param7: primitiveTypes.float32,
});
const CommandLong_RequestRefArray = ArrayType(CommandLong_RequestRefStruct);
const CommandLong_RequestRefStructArray = StructType({
  data: CommandLong_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CommandLong_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CommandLong_RequestRefStruct(other._refObject.toObject());
      this._broadcastIntialized = true;
      this._commandIntialized = true;
      this._confirmationIntialized = true;
      this._param1Intialized = true;
      this._param2Intialized = true;
      this._param3Intialized = true;
      this._param4Intialized = true;
      this._param5Intialized = true;
      this._param6Intialized = true;
      this._param7Intialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CommandLong_RequestRefStruct();
    this._broadcastIntialized = false;
    this._commandIntialized = false;
    this._confirmationIntialized = false;
    this._param1Intialized = false;
    this._param2Intialized = false;
    this._param3Intialized = false;
    this._param4Intialized = false;
    this._param5Intialized = false;
    this._param6Intialized = false;
    this._param7Intialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CommandLong_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CommandLong_RequestArrayWrapper;
  }
  static get ArrayType() {
    return CommandLong_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return CommandLong_RequestRefStructArray
  }
  static get refObjectType() {
    return CommandLong_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._broadcastIntialized) {
        throw new TypeError('Invalid argument: broadcast in CommandLong_Request');
      }
      if (!this._commandIntialized) {
        throw new TypeError('Invalid argument: command in CommandLong_Request');
      }
      if (!this._confirmationIntialized) {
        throw new TypeError('Invalid argument: confirmation in CommandLong_Request');
      }
      if (!this._param1Intialized) {
        throw new TypeError('Invalid argument: param1 in CommandLong_Request');
      }
      if (!this._param2Intialized) {
        throw new TypeError('Invalid argument: param2 in CommandLong_Request');
      }
      if (!this._param3Intialized) {
        throw new TypeError('Invalid argument: param3 in CommandLong_Request');
      }
      if (!this._param4Intialized) {
        throw new TypeError('Invalid argument: param4 in CommandLong_Request');
      }
      if (!this._param5Intialized) {
        throw new TypeError('Invalid argument: param5 in CommandLong_Request');
      }
      if (!this._param6Intialized) {
        throw new TypeError('Invalid argument: param6 in CommandLong_Request');
      }
      if (!this._param7Intialized) {
        throw new TypeError('Invalid argument: param7 in CommandLong_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._broadcastIntialized = true;
    this._refObject.broadcast = refObject.broadcast;
    this._commandIntialized = true;
    this._refObject.command = refObject.command;
    this._confirmationIntialized = true;
    this._refObject.confirmation = refObject.confirmation;
    this._param1Intialized = true;
    this._refObject.param1 = refObject.param1;
    this._param2Intialized = true;
    this._refObject.param2 = refObject.param2;
    this._param3Intialized = true;
    this._refObject.param3 = refObject.param3;
    this._param4Intialized = true;
    this._refObject.param4 = refObject.param4;
    this._param5Intialized = true;
    this._refObject.param5 = refObject.param5;
    this._param6Intialized = true;
    this._refObject.param6 = refObject.param6;
    this._param7Intialized = true;
    this._refObject.param7 = refObject.param7;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    CommandLong_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'CommandLong_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get broadcast() {
    if (!this._broadcastIntialized) {
      return undefined;
    }
    return this._refObject.broadcast;
  }
  set broadcast(value) {
    this._broadcastIntialized = true;
    this._refObject.broadcast = value;
  }
  get command() {
    if (!this._commandIntialized) {
      return undefined;
    }
    return this._refObject.command;
  }
  set command(value) {
    this._commandIntialized = true;
    this._refObject.command = value;
  }
  get confirmation() {
    if (!this._confirmationIntialized) {
      return undefined;
    }
    return this._refObject.confirmation;
  }
  set confirmation(value) {
    this._confirmationIntialized = true;
    this._refObject.confirmation = value;
  }
  get param1() {
    if (!this._param1Intialized) {
      return undefined;
    }
    return this._refObject.param1;
  }
  set param1(value) {
    this._param1Intialized = true;
    this._refObject.param1 = value;
  }
  get param2() {
    if (!this._param2Intialized) {
      return undefined;
    }
    return this._refObject.param2;
  }
  set param2(value) {
    this._param2Intialized = true;
    this._refObject.param2 = value;
  }
  get param3() {
    if (!this._param3Intialized) {
      return undefined;
    }
    return this._refObject.param3;
  }
  set param3(value) {
    this._param3Intialized = true;
    this._refObject.param3 = value;
  }
  get param4() {
    if (!this._param4Intialized) {
      return undefined;
    }
    return this._refObject.param4;
  }
  set param4(value) {
    this._param4Intialized = true;
    this._refObject.param4 = value;
  }
  get param5() {
    if (!this._param5Intialized) {
      return undefined;
    }
    return this._refObject.param5;
  }
  set param5(value) {
    this._param5Intialized = true;
    this._refObject.param5 = value;
  }
  get param6() {
    if (!this._param6Intialized) {
      return undefined;
    }
    return this._refObject.param6;
  }
  set param6(value) {
    this._param6Intialized = true;
    this._refObject.param6 = value;
  }
  get param7() {
    if (!this._param7Intialized) {
      return undefined;
    }
    return this._refObject.param7;
  }
  set param7(value) {
    this._param7Intialized = true;
    this._refObject.param7 = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CommandLong_RequestRefStruct(refObject.toObject());
    this._broadcastIntialized = true;
    this._commandIntialized = true;
    this._confirmationIntialized = true;
    this._param1Intialized = true;
    this._param2Intialized = true;
    this._param3Intialized = true;
    this._param4Intialized = true;
    this._param5Intialized = true;
    this._param6Intialized = true;
    this._param7Intialized = true;
  }
  copy(other) {
    this._refObject = new CommandLong_RequestRefStruct(other._refObject.toObject());
    this._broadcastIntialized = true;
    this._commandIntialized = true;
    this._confirmationIntialized = true;
    this._param1Intialized = true;
    this._param2Intialized = true;
    this._param3Intialized = true;
    this._param4Intialized = true;
    this._param5Intialized = true;
    this._param6Intialized = true;
    this._param7Intialized = true;
  }
  static get classType() {
    return CommandLong_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"broadcast","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"command","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"confirmation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param5","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param6","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param7","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"CommandLong_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CommandLong_Request"};
  }
  hasMember(name) {
    let memberNames = ["broadcast","command","confirmation","param1","param2","param3","param4","param5","param6","param7"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CommandLong_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CommandLong_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CommandLong_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandLong_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandLong_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CommandLong_RequestRefArray(size);
    this._refObject = new CommandLong_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CommandLong_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CommandLong_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CommandLong_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CommandLong_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CommandLong_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CommandLong_RequestArrayWrapper;
  }
}
module.exports = CommandLong_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "broadcast",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "command",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "confirmation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param5",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param6",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param7",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "CommandLong_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CommandLong_Request"
}
*/
