// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__SetMode_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SetMode_RequestRefStruct = StructType({
  base_mode: primitiveTypes.uint8,
  custom_mode: primitiveTypes.string,
});
const SetMode_RequestRefArray = ArrayType(SetMode_RequestRefStruct);
const SetMode_RequestRefStructArray = StructType({
  data: SetMode_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetMode_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetMode_RequestRefStruct(other._refObject.toObject());
      this._base_modeIntialized = true;
      this._custom_modeIntialized = true;
      this._wrapperFields.custom_mode =  new StringWrapper(other._wrapperFields.custom_mode);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetMode_RequestRefStruct();
    this._base_modeIntialized = false;
    this._custom_modeIntialized = false;
    this._wrapperFields.custom_mode = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetMode_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetMode_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetMode_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetMode_RequestRefStructArray
  }
  static get refObjectType() {
    return SetMode_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._base_modeIntialized) {
        throw new TypeError('Invalid argument: base_mode in SetMode_Request');
      }
      if (!this._custom_modeIntialized) {
        throw new TypeError('Invalid argument: custom_mode in SetMode_Request');
      }
    }
    if (own) {
      this._wrapperFields.custom_mode.freeze(own, checkConsistency);
    }
    this._refObject.custom_mode = this._wrapperFields.custom_mode.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._base_modeIntialized = true;
    this._refObject.base_mode = refObject.base_mode;
    this._custom_modeIntialized = true;
    this._wrapperFields.custom_mode.data = refObject.custom_mode.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.custom_mode);
  }
  static destoryRawROS(msg) {
    SetMode_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'SetMode_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get base_mode() {
    if (!this._base_modeIntialized) {
      return undefined;
    }
    return this._refObject.base_mode;
  }
  set base_mode(value) {
    this._base_modeIntialized = true;
    this._refObject.base_mode = value;
  }
  get custom_mode() {
    if (!this._custom_modeIntialized) {
      return undefined;
    }
    return this._wrapperFields.custom_mode.data;
  }
  set custom_mode(value) {
    this._custom_modeIntialized = true;
    this._wrapperFields.custom_mode.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetMode_RequestRefStruct(refObject.toObject());
    this._base_modeIntialized = true;
    this._custom_modeIntialized = true;
    this._wrapperFields.custom_mode.copyRefObject(this._refObject.custom_mode);
  }
  copy(other) {
    this._refObject = new SetMode_RequestRefStruct(other._refObject.toObject());
    this._base_modeIntialized = true;
    this._custom_modeIntialized = true;
    this._wrapperFields.custom_mode.copy(other._wrapperFields.custom_mode);
  }
  static get classType() {
    return SetMode_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"MAV_MODE_PREFLIGHT","value":0},{"type":"uint8","name":"MAV_MODE_STABILIZE_DISARMED","value":80},{"type":"uint8","name":"MAV_MODE_STABILIZE_ARMED","value":208},{"type":"uint8","name":"MAV_MODE_MANUAL_DISARMED","value":64},{"type":"uint8","name":"MAV_MODE_MANUAL_ARMED","value":192},{"type":"uint8","name":"MAV_MODE_GUIDED_DISARMED","value":88},{"type":"uint8","name":"MAV_MODE_GUIDED_ARMED","value":216},{"type":"uint8","name":"MAV_MODE_AUTO_DISARMED","value":92},{"type":"uint8","name":"MAV_MODE_AUTO_ARMED","value":220},{"type":"uint8","name":"MAV_MODE_TEST_DISARMED","value":66},{"type":"uint8","name":"MAV_MODE_TEST_ARMED","value":194}],"fields":[{"name":"base_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"custom_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"SetMode_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetMode_Request"};
  }
  hasMember(name) {
    let memberNames = ["base_mode","custom_mode"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetMode_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetMode_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetMode_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMode_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMode_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetMode_RequestRefArray(size);
    this._refObject = new SetMode_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetMode_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetMode_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetMode_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetMode_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetMode_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetMode_RequestArrayWrapper;
  }
}
// Define constants (11 in total)
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_PREFLIGHT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_STABILIZE_DISARMED", {value: 80, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_STABILIZE_ARMED", {value: 208, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_MANUAL_DISARMED", {value: 64, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_MANUAL_ARMED", {value: 192, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_GUIDED_DISARMED", {value: 88, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_GUIDED_ARMED", {value: 216, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_AUTO_DISARMED", {value: 92, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_AUTO_ARMED", {value: 220, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_TEST_DISARMED", {value: 66, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMode_RequestWrapper, "MAV_MODE_TEST_ARMED", {value: 194, writable: false, enumerable: true, configurable: true});
module.exports = SetMode_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "MAV_MODE_PREFLIGHT",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "MAV_MODE_STABILIZE_DISARMED",
      "value": 80
    },
    {
      "type": "uint8",
      "name": "MAV_MODE_STABILIZE_ARMED",
      "value": 208
    },
    {
      "type": "uint8",
      "name": "MAV_MODE_MANUAL_DISARMED",
      "value": 64
    },
    {
      "type": "uint8",
      "name": "MAV_MODE_MANUAL_ARMED",
      "value": 192
    },
    {
      "type": "uint8",
      "name": "MAV_MODE_GUIDED_DISARMED",
      "value": 88
    },
    {
      "type": "uint8",
      "name": "MAV_MODE_GUIDED_ARMED",
      "value": 216
    },
    {
      "type": "uint8",
      "name": "MAV_MODE_AUTO_DISARMED",
      "value": 92
    },
    {
      "type": "uint8",
      "name": "MAV_MODE_AUTO_ARMED",
      "value": 220
    },
    {
      "type": "uint8",
      "name": "MAV_MODE_TEST_DISARMED",
      "value": 66
    },
    {
      "type": "uint8",
      "name": "MAV_MODE_TEST_ARMED",
      "value": 194
    }
  ],
  "fields": [
    {
      "name": "base_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "custom_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "SetMode_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetMode_Request"
}
*/
