// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__HilControls.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const HilControlsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  roll_ailerons: primitiveTypes.float32,
  pitch_elevator: primitiveTypes.float32,
  yaw_rudder: primitiveTypes.float32,
  throttle: primitiveTypes.float32,
  aux1: primitiveTypes.float32,
  aux2: primitiveTypes.float32,
  aux3: primitiveTypes.float32,
  aux4: primitiveTypes.float32,
  mode: primitiveTypes.uint8,
  nav_mode: primitiveTypes.uint8,
});
const HilControlsRefArray = ArrayType(HilControlsRefStruct);
const HilControlsRefStructArray = StructType({
  data: HilControlsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HilControlsWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HilControlsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._roll_aileronsIntialized = true;
      this._pitch_elevatorIntialized = true;
      this._yaw_rudderIntialized = true;
      this._throttleIntialized = true;
      this._aux1Intialized = true;
      this._aux2Intialized = true;
      this._aux3Intialized = true;
      this._aux4Intialized = true;
      this._modeIntialized = true;
      this._nav_modeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HilControlsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._roll_aileronsIntialized = false;
    this._pitch_elevatorIntialized = false;
    this._yaw_rudderIntialized = false;
    this._throttleIntialized = false;
    this._aux1Intialized = false;
    this._aux2Intialized = false;
    this._aux3Intialized = false;
    this._aux4Intialized = false;
    this._modeIntialized = false;
    this._nav_modeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new HilControlsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HilControlsArrayWrapper;
  }
  static get ArrayType() {
    return HilControlsArrayWrapper;
  }
  static get refObjectArrayType() {
    return HilControlsRefStructArray
  }
  static get refObjectType() {
    return HilControlsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._roll_aileronsIntialized) {
        throw new TypeError('Invalid argument: roll_ailerons in HilControls');
      }
      if (!this._pitch_elevatorIntialized) {
        throw new TypeError('Invalid argument: pitch_elevator in HilControls');
      }
      if (!this._yaw_rudderIntialized) {
        throw new TypeError('Invalid argument: yaw_rudder in HilControls');
      }
      if (!this._throttleIntialized) {
        throw new TypeError('Invalid argument: throttle in HilControls');
      }
      if (!this._aux1Intialized) {
        throw new TypeError('Invalid argument: aux1 in HilControls');
      }
      if (!this._aux2Intialized) {
        throw new TypeError('Invalid argument: aux2 in HilControls');
      }
      if (!this._aux3Intialized) {
        throw new TypeError('Invalid argument: aux3 in HilControls');
      }
      if (!this._aux4Intialized) {
        throw new TypeError('Invalid argument: aux4 in HilControls');
      }
      if (!this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in HilControls');
      }
      if (!this._nav_modeIntialized) {
        throw new TypeError('Invalid argument: nav_mode in HilControls');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._roll_aileronsIntialized = true;
    this._refObject.roll_ailerons = refObject.roll_ailerons;
    this._pitch_elevatorIntialized = true;
    this._refObject.pitch_elevator = refObject.pitch_elevator;
    this._yaw_rudderIntialized = true;
    this._refObject.yaw_rudder = refObject.yaw_rudder;
    this._throttleIntialized = true;
    this._refObject.throttle = refObject.throttle;
    this._aux1Intialized = true;
    this._refObject.aux1 = refObject.aux1;
    this._aux2Intialized = true;
    this._refObject.aux2 = refObject.aux2;
    this._aux3Intialized = true;
    this._refObject.aux3 = refObject.aux3;
    this._aux4Intialized = true;
    this._refObject.aux4 = refObject.aux4;
    this._modeIntialized = true;
    this._refObject.mode = refObject.mode;
    this._nav_modeIntialized = true;
    this._refObject.nav_mode = refObject.nav_mode;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    HilControlsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'HilControls'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get roll_ailerons() {
    if (!this._roll_aileronsIntialized) {
      return undefined;
    }
    return this._refObject.roll_ailerons;
  }
  set roll_ailerons(value) {
    this._roll_aileronsIntialized = true;
    this._refObject.roll_ailerons = value;
  }
  get pitch_elevator() {
    if (!this._pitch_elevatorIntialized) {
      return undefined;
    }
    return this._refObject.pitch_elevator;
  }
  set pitch_elevator(value) {
    this._pitch_elevatorIntialized = true;
    this._refObject.pitch_elevator = value;
  }
  get yaw_rudder() {
    if (!this._yaw_rudderIntialized) {
      return undefined;
    }
    return this._refObject.yaw_rudder;
  }
  set yaw_rudder(value) {
    this._yaw_rudderIntialized = true;
    this._refObject.yaw_rudder = value;
  }
  get throttle() {
    if (!this._throttleIntialized) {
      return undefined;
    }
    return this._refObject.throttle;
  }
  set throttle(value) {
    this._throttleIntialized = true;
    this._refObject.throttle = value;
  }
  get aux1() {
    if (!this._aux1Intialized) {
      return undefined;
    }
    return this._refObject.aux1;
  }
  set aux1(value) {
    this._aux1Intialized = true;
    this._refObject.aux1 = value;
  }
  get aux2() {
    if (!this._aux2Intialized) {
      return undefined;
    }
    return this._refObject.aux2;
  }
  set aux2(value) {
    this._aux2Intialized = true;
    this._refObject.aux2 = value;
  }
  get aux3() {
    if (!this._aux3Intialized) {
      return undefined;
    }
    return this._refObject.aux3;
  }
  set aux3(value) {
    this._aux3Intialized = true;
    this._refObject.aux3 = value;
  }
  get aux4() {
    if (!this._aux4Intialized) {
      return undefined;
    }
    return this._refObject.aux4;
  }
  set aux4(value) {
    this._aux4Intialized = true;
    this._refObject.aux4 = value;
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._refObject.mode;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._refObject.mode = value;
  }
  get nav_mode() {
    if (!this._nav_modeIntialized) {
      return undefined;
    }
    return this._refObject.nav_mode;
  }
  set nav_mode(value) {
    this._nav_modeIntialized = true;
    this._refObject.nav_mode = value;
  }
  copyRefObject(refObject) {
    this._refObject = new HilControlsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._roll_aileronsIntialized = true;
    this._pitch_elevatorIntialized = true;
    this._yaw_rudderIntialized = true;
    this._throttleIntialized = true;
    this._aux1Intialized = true;
    this._aux2Intialized = true;
    this._aux3Intialized = true;
    this._aux4Intialized = true;
    this._modeIntialized = true;
    this._nav_modeIntialized = true;
  }
  copy(other) {
    this._refObject = new HilControlsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._roll_aileronsIntialized = true;
    this._pitch_elevatorIntialized = true;
    this._yaw_rudderIntialized = true;
    this._throttleIntialized = true;
    this._aux1Intialized = true;
    this._aux2Intialized = true;
    this._aux3Intialized = true;
    this._aux4Intialized = true;
    this._modeIntialized = true;
    this._nav_modeIntialized = true;
  }
  static get classType() {
    return HilControlsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"roll_ailerons","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch_elevator","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_rudder","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"throttle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nav_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"HilControls","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HilControls"};
  }
  hasMember(name) {
    let memberNames = ["header","roll_ailerons","pitch_elevator","yaw_rudder","throttle","aux1","aux2","aux3","aux4","mode","nav_mode"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HilControlsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HilControlsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HilControlsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HilControlsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HilControlsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HilControlsRefArray(size);
    this._refObject = new HilControlsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HilControlsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HilControlsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HilControlsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HilControlsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HilControlsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HilControlsArrayWrapper;
  }
}
module.exports = HilControlsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "roll_ailerons",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch_elevator",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_rudder",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "throttle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "nav_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "HilControls",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HilControls"
}
*/
