// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__SetMavFrame_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SetMavFrame_RequestRefStruct = StructType({
  mav_frame: primitiveTypes.uint8,
});
const SetMavFrame_RequestRefArray = ArrayType(SetMavFrame_RequestRefStruct);
const SetMavFrame_RequestRefStructArray = StructType({
  data: SetMavFrame_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetMavFrame_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetMavFrame_RequestRefStruct(other._refObject.toObject());
      this._mav_frameIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetMavFrame_RequestRefStruct();
    this._mav_frameIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SetMavFrame_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetMavFrame_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetMavFrame_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetMavFrame_RequestRefStructArray
  }
  static get refObjectType() {
    return SetMavFrame_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._mav_frameIntialized) {
        throw new TypeError('Invalid argument: mav_frame in SetMavFrame_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._mav_frameIntialized = true;
    this._refObject.mav_frame = refObject.mav_frame;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SetMavFrame_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'SetMavFrame_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get mav_frame() {
    if (!this._mav_frameIntialized) {
      return undefined;
    }
    return this._refObject.mav_frame;
  }
  set mav_frame(value) {
    this._mav_frameIntialized = true;
    this._refObject.mav_frame = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetMavFrame_RequestRefStruct(refObject.toObject());
    this._mav_frameIntialized = true;
  }
  copy(other) {
    this._refObject = new SetMavFrame_RequestRefStruct(other._refObject.toObject());
    this._mav_frameIntialized = true;
  }
  static get classType() {
    return SetMavFrame_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"FRAME_GLOBAL","value":0},{"type":"uint8","name":"FRAME_LOCAL_NED","value":1},{"type":"uint8","name":"FRAME_MISSION","value":2},{"type":"uint8","name":"FRAME_GLOBAL_RELATIVE_ALT","value":3},{"type":"uint8","name":"FRAME_LOCAL_ENU","value":4},{"type":"uint8","name":"FRAME_GLOBAL_INT","value":5},{"type":"uint8","name":"FRAME_GLOBAL_RELATIVE_ALT_INT","value":6},{"type":"uint8","name":"FRAME_LOCAL_OFFSET_NED","value":7},{"type":"uint8","name":"FRAME_BODY_NED","value":8},{"type":"uint8","name":"FRAME_BODY_OFFSET_NED","value":9},{"type":"uint8","name":"FRAME_GLOBAL_TERRAIN_ALT","value":10},{"type":"uint8","name":"FRAME_GLOBAL_TERRAIN_ALT_INT","value":11},{"type":"uint8","name":"FRAME_BODY_FRD","value":12},{"type":"uint8","name":"FRAME_RESERVED_13","value":13},{"type":"uint8","name":"FRAME_RESERVED_14","value":14},{"type":"uint8","name":"FRAME_RESERVED_15","value":15},{"type":"uint8","name":"FRAME_RESERVED_16","value":16},{"type":"uint8","name":"FRAME_RESERVED_17","value":17},{"type":"uint8","name":"FRAME_RESERVED_18","value":18},{"type":"uint8","name":"FRAME_RESERVED_19","value":19},{"type":"uint8","name":"FRAME_LOCAL_FRD","value":20},{"type":"uint8","name":"FRAME_LOCAL_FLU","value":21}],"fields":[{"name":"mav_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"SetMavFrame_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetMavFrame_Request"};
  }
  hasMember(name) {
    let memberNames = ["mav_frame"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetMavFrame_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetMavFrame_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetMavFrame_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMavFrame_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMavFrame_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetMavFrame_RequestRefArray(size);
    this._refObject = new SetMavFrame_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetMavFrame_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetMavFrame_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetMavFrame_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetMavFrame_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetMavFrame_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetMavFrame_RequestArrayWrapper;
  }
}
// Define constants (22 in total)
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_GLOBAL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_LOCAL_NED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_MISSION", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_GLOBAL_RELATIVE_ALT", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_LOCAL_ENU", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_GLOBAL_INT", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_GLOBAL_RELATIVE_ALT_INT", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_LOCAL_OFFSET_NED", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_BODY_NED", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_BODY_OFFSET_NED", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_GLOBAL_TERRAIN_ALT", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_GLOBAL_TERRAIN_ALT_INT", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_BODY_FRD", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_RESERVED_13", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_RESERVED_14", {value: 14, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_RESERVED_15", {value: 15, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_RESERVED_16", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_RESERVED_17", {value: 17, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_RESERVED_18", {value: 18, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_RESERVED_19", {value: 19, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_LOCAL_FRD", {value: 20, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetMavFrame_RequestWrapper, "FRAME_LOCAL_FLU", {value: 21, writable: false, enumerable: true, configurable: true});
module.exports = SetMavFrame_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "FRAME_LOCAL_NED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "FRAME_MISSION",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_RELATIVE_ALT",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "FRAME_LOCAL_ENU",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_INT",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_RELATIVE_ALT_INT",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "FRAME_LOCAL_OFFSET_NED",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "FRAME_BODY_NED",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "FRAME_BODY_OFFSET_NED",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_TERRAIN_ALT",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_TERRAIN_ALT_INT",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "FRAME_BODY_FRD",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_13",
      "value": 13
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_14",
      "value": 14
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_15",
      "value": 15
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_16",
      "value": 16
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_17",
      "value": 17
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_18",
      "value": 18
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_19",
      "value": 19
    },
    {
      "type": "uint8",
      "name": "FRAME_LOCAL_FRD",
      "value": 20
    },
    {
      "type": "uint8",
      "name": "FRAME_LOCAL_FLU",
      "value": 21
    }
  ],
  "fields": [
    {
      "name": "mav_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "SetMavFrame_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetMavFrame_Request"
}
*/
