// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__Waypoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const WaypointRefStruct = StructType({
  frame: primitiveTypes.uint8,
  command: primitiveTypes.uint16,
  is_current: primitiveTypes.bool,
  autocontinue: primitiveTypes.bool,
  param1: primitiveTypes.float32,
  param2: primitiveTypes.float32,
  param3: primitiveTypes.float32,
  param4: primitiveTypes.float32,
  x_lat: primitiveTypes.float64,
  y_long: primitiveTypes.float64,
  z_alt: primitiveTypes.float64,
});
const WaypointRefArray = ArrayType(WaypointRefStruct);
const WaypointRefStructArray = StructType({
  data: WaypointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WaypointWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WaypointRefStruct(other._refObject.toObject());
      this._frameIntialized = true;
      this._commandIntialized = true;
      this._is_currentIntialized = true;
      this._autocontinueIntialized = true;
      this._param1Intialized = true;
      this._param2Intialized = true;
      this._param3Intialized = true;
      this._param4Intialized = true;
      this._x_latIntialized = true;
      this._y_longIntialized = true;
      this._z_altIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WaypointRefStruct();
    this._frameIntialized = false;
    this._commandIntialized = false;
    this._is_currentIntialized = false;
    this._autocontinueIntialized = false;
    this._param1Intialized = false;
    this._param2Intialized = false;
    this._param3Intialized = false;
    this._param4Intialized = false;
    this._x_latIntialized = false;
    this._y_longIntialized = false;
    this._z_altIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new WaypointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WaypointArrayWrapper;
  }
  static get ArrayType() {
    return WaypointArrayWrapper;
  }
  static get refObjectArrayType() {
    return WaypointRefStructArray
  }
  static get refObjectType() {
    return WaypointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._frameIntialized) {
        throw new TypeError('Invalid argument: frame in Waypoint');
      }
      if (!this._commandIntialized) {
        throw new TypeError('Invalid argument: command in Waypoint');
      }
      if (!this._is_currentIntialized) {
        throw new TypeError('Invalid argument: is_current in Waypoint');
      }
      if (!this._autocontinueIntialized) {
        throw new TypeError('Invalid argument: autocontinue in Waypoint');
      }
      if (!this._param1Intialized) {
        throw new TypeError('Invalid argument: param1 in Waypoint');
      }
      if (!this._param2Intialized) {
        throw new TypeError('Invalid argument: param2 in Waypoint');
      }
      if (!this._param3Intialized) {
        throw new TypeError('Invalid argument: param3 in Waypoint');
      }
      if (!this._param4Intialized) {
        throw new TypeError('Invalid argument: param4 in Waypoint');
      }
      if (!this._x_latIntialized) {
        throw new TypeError('Invalid argument: x_lat in Waypoint');
      }
      if (!this._y_longIntialized) {
        throw new TypeError('Invalid argument: y_long in Waypoint');
      }
      if (!this._z_altIntialized) {
        throw new TypeError('Invalid argument: z_alt in Waypoint');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._frameIntialized = true;
    this._refObject.frame = refObject.frame;
    this._commandIntialized = true;
    this._refObject.command = refObject.command;
    this._is_currentIntialized = true;
    this._refObject.is_current = refObject.is_current;
    this._autocontinueIntialized = true;
    this._refObject.autocontinue = refObject.autocontinue;
    this._param1Intialized = true;
    this._refObject.param1 = refObject.param1;
    this._param2Intialized = true;
    this._refObject.param2 = refObject.param2;
    this._param3Intialized = true;
    this._refObject.param3 = refObject.param3;
    this._param4Intialized = true;
    this._refObject.param4 = refObject.param4;
    this._x_latIntialized = true;
    this._refObject.x_lat = refObject.x_lat;
    this._y_longIntialized = true;
    this._refObject.y_long = refObject.y_long;
    this._z_altIntialized = true;
    this._refObject.z_alt = refObject.z_alt;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    WaypointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'Waypoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get frame() {
    if (!this._frameIntialized) {
      return undefined;
    }
    return this._refObject.frame;
  }
  set frame(value) {
    this._frameIntialized = true;
    this._refObject.frame = value;
  }
  get command() {
    if (!this._commandIntialized) {
      return undefined;
    }
    return this._refObject.command;
  }
  set command(value) {
    this._commandIntialized = true;
    this._refObject.command = value;
  }
  get is_current() {
    if (!this._is_currentIntialized) {
      return undefined;
    }
    return this._refObject.is_current;
  }
  set is_current(value) {
    this._is_currentIntialized = true;
    this._refObject.is_current = value;
  }
  get autocontinue() {
    if (!this._autocontinueIntialized) {
      return undefined;
    }
    return this._refObject.autocontinue;
  }
  set autocontinue(value) {
    this._autocontinueIntialized = true;
    this._refObject.autocontinue = value;
  }
  get param1() {
    if (!this._param1Intialized) {
      return undefined;
    }
    return this._refObject.param1;
  }
  set param1(value) {
    this._param1Intialized = true;
    this._refObject.param1 = value;
  }
  get param2() {
    if (!this._param2Intialized) {
      return undefined;
    }
    return this._refObject.param2;
  }
  set param2(value) {
    this._param2Intialized = true;
    this._refObject.param2 = value;
  }
  get param3() {
    if (!this._param3Intialized) {
      return undefined;
    }
    return this._refObject.param3;
  }
  set param3(value) {
    this._param3Intialized = true;
    this._refObject.param3 = value;
  }
  get param4() {
    if (!this._param4Intialized) {
      return undefined;
    }
    return this._refObject.param4;
  }
  set param4(value) {
    this._param4Intialized = true;
    this._refObject.param4 = value;
  }
  get x_lat() {
    if (!this._x_latIntialized) {
      return undefined;
    }
    return this._refObject.x_lat;
  }
  set x_lat(value) {
    this._x_latIntialized = true;
    this._refObject.x_lat = value;
  }
  get y_long() {
    if (!this._y_longIntialized) {
      return undefined;
    }
    return this._refObject.y_long;
  }
  set y_long(value) {
    this._y_longIntialized = true;
    this._refObject.y_long = value;
  }
  get z_alt() {
    if (!this._z_altIntialized) {
      return undefined;
    }
    return this._refObject.z_alt;
  }
  set z_alt(value) {
    this._z_altIntialized = true;
    this._refObject.z_alt = value;
  }
  copyRefObject(refObject) {
    this._refObject = new WaypointRefStruct(refObject.toObject());
    this._frameIntialized = true;
    this._commandIntialized = true;
    this._is_currentIntialized = true;
    this._autocontinueIntialized = true;
    this._param1Intialized = true;
    this._param2Intialized = true;
    this._param3Intialized = true;
    this._param4Intialized = true;
    this._x_latIntialized = true;
    this._y_longIntialized = true;
    this._z_altIntialized = true;
  }
  copy(other) {
    this._refObject = new WaypointRefStruct(other._refObject.toObject());
    this._frameIntialized = true;
    this._commandIntialized = true;
    this._is_currentIntialized = true;
    this._autocontinueIntialized = true;
    this._param1Intialized = true;
    this._param2Intialized = true;
    this._param3Intialized = true;
    this._param4Intialized = true;
    this._x_latIntialized = true;
    this._y_longIntialized = true;
    this._z_altIntialized = true;
  }
  static get classType() {
    return WaypointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"FRAME_GLOBAL","value":0},{"type":"uint8","name":"FRAME_LOCAL_NED","value":1},{"type":"uint8","name":"FRAME_MISSION","value":2},{"type":"uint8","name":"FRAME_GLOBAL_REL_ALT","value":3},{"type":"uint8","name":"FRAME_LOCAL_ENU","value":4},{"type":"uint8","name":"FRAME_GLOBAL_INT","value":5},{"type":"uint8","name":"FRAME_GLOBAL_RELATIVE_ALT_INT","value":6},{"type":"uint8","name":"FRAME_LOCAL_OFFSET_NED","value":7},{"type":"uint8","name":"FRAME_BODY_NED","value":8},{"type":"uint8","name":"FRAME_BODY_OFFSET_NED","value":9},{"type":"uint8","name":"FRAME_GLOBAL_TERRAIN_ALT","value":10},{"type":"uint8","name":"FRAME_GLOBAL_TERRAIN_ALT_INT","value":11},{"type":"uint8","name":"FRAME_BODY_FRD","value":12},{"type":"uint8","name":"FRAME_RESERVED_13","value":13},{"type":"uint8","name":"FRAME_RESERVED_14","value":14},{"type":"uint8","name":"FRAME_RESERVED_15","value":15},{"type":"uint8","name":"FRAME_RESERVED_16","value":16},{"type":"uint8","name":"FRAME_RESERVED_17","value":17},{"type":"uint8","name":"FRAME_RESERVED_18","value":18},{"type":"uint8","name":"FRAME_RESERVED_19","value":19},{"type":"uint8","name":"FRAME_LOCAL_FRD","value":20},{"type":"uint8","name":"FRAME_LOCAL_FLU","value":21}],"fields":[{"name":"frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"command","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_current","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"autocontinue","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x_lat","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y_long","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z_alt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"Waypoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Waypoint"};
  }
  hasMember(name) {
    let memberNames = ["frame","command","is_current","autocontinue","param1","param2","param3","param4","x_lat","y_long","z_alt"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WaypointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WaypointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WaypointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WaypointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WaypointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WaypointRefArray(size);
    this._refObject = new WaypointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WaypointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WaypointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WaypointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WaypointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WaypointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WaypointArrayWrapper;
  }
}
// Define constants (22 in total)
Object.defineProperty(WaypointWrapper, "FRAME_GLOBAL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_LOCAL_NED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_MISSION", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_GLOBAL_REL_ALT", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_LOCAL_ENU", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_GLOBAL_INT", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_GLOBAL_RELATIVE_ALT_INT", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_LOCAL_OFFSET_NED", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_BODY_NED", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_BODY_OFFSET_NED", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_GLOBAL_TERRAIN_ALT", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_GLOBAL_TERRAIN_ALT_INT", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_BODY_FRD", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_RESERVED_13", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_RESERVED_14", {value: 14, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_RESERVED_15", {value: 15, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_RESERVED_16", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_RESERVED_17", {value: 17, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_RESERVED_18", {value: 18, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_RESERVED_19", {value: 19, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_LOCAL_FRD", {value: 20, writable: false, enumerable: true, configurable: true});
Object.defineProperty(WaypointWrapper, "FRAME_LOCAL_FLU", {value: 21, writable: false, enumerable: true, configurable: true});
module.exports = WaypointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "FRAME_LOCAL_NED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "FRAME_MISSION",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_REL_ALT",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "FRAME_LOCAL_ENU",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_INT",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_RELATIVE_ALT_INT",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "FRAME_LOCAL_OFFSET_NED",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "FRAME_BODY_NED",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "FRAME_BODY_OFFSET_NED",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_TERRAIN_ALT",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "FRAME_GLOBAL_TERRAIN_ALT_INT",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "FRAME_BODY_FRD",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_13",
      "value": 13
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_14",
      "value": 14
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_15",
      "value": 15
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_16",
      "value": 16
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_17",
      "value": 17
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_18",
      "value": 18
    },
    {
      "type": "uint8",
      "name": "FRAME_RESERVED_19",
      "value": 19
    },
    {
      "type": "uint8",
      "name": "FRAME_LOCAL_FRD",
      "value": 20
    },
    {
      "type": "uint8",
      "name": "FRAME_LOCAL_FLU",
      "value": 21
    }
  ],
  "fields": [
    {
      "name": "frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "command",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "is_current",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "autocontinue",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x_lat",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y_long",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z_alt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "Waypoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Waypoint"
}
*/
