// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__Trajectory.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PositionTargetWrapper = require('../../generated/mavros_msgs/mavros_msgs__msg__PositionTarget.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const UInt16Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt16.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const TrajectoryRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  type: primitiveTypes.uint8,
  point_1: PositionTargetWrapper.refObjectType,
  point_2: PositionTargetWrapper.refObjectType,
  point_3: PositionTargetWrapper.refObjectType,
  point_4: PositionTargetWrapper.refObjectType,
  point_5: PositionTargetWrapper.refObjectType,
  point_valid: ArrayType(primitiveTypes.uint8, 5),
  command: ArrayType(primitiveTypes.uint16, 5),
  time_horizon: ArrayType(primitiveTypes.float32, 5),
});
const TrajectoryRefArray = ArrayType(TrajectoryRefStruct);
const TrajectoryRefStructArray = StructType({
  data: TrajectoryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TrajectoryWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TrajectoryRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._typeIntialized = true;
      this._wrapperFields.point_1 =  new PositionTargetWrapper(other._wrapperFields.point_1);
      this._wrapperFields.point_2 =  new PositionTargetWrapper(other._wrapperFields.point_2);
      this._wrapperFields.point_3 =  new PositionTargetWrapper(other._wrapperFields.point_3);
      this._wrapperFields.point_4 =  new PositionTargetWrapper(other._wrapperFields.point_4);
      this._wrapperFields.point_5 =  new PositionTargetWrapper(other._wrapperFields.point_5);
      this._wrapperFields.point_valid = UInt8Wrapper.createArray();
      this._wrapperFields.point_valid.copy(other._wrapperFields.point_valid);
      this._wrapperFields.command = UInt16Wrapper.createArray();
      this._wrapperFields.command.copy(other._wrapperFields.command);
      this._wrapperFields.time_horizon = Float32Wrapper.createArray();
      this._wrapperFields.time_horizon.copy(other._wrapperFields.time_horizon);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TrajectoryRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._typeIntialized = false;
    this._wrapperFields.point_1 = new PositionTargetWrapper();
    this._wrapperFields.point_2 = new PositionTargetWrapper();
    this._wrapperFields.point_3 = new PositionTargetWrapper();
    this._wrapperFields.point_4 = new PositionTargetWrapper();
    this._wrapperFields.point_5 = new PositionTargetWrapper();
    this._wrapperFields.point_valid = UInt8Wrapper.createArray();
    this._wrapperFields.command = UInt16Wrapper.createArray();
    this._wrapperFields.time_horizon = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new TrajectoryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TrajectoryArrayWrapper;
  }
  static get ArrayType() {
    return TrajectoryArrayWrapper;
  }
  static get refObjectArrayType() {
    return TrajectoryRefStructArray
  }
  static get refObjectType() {
    return TrajectoryRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in Trajectory');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.point_1.freeze(own, checkConsistency);
    this._refObject.point_1 = this._wrapperFields.point_1.refObject;
    this._wrapperFields.point_2.freeze(own, checkConsistency);
    this._refObject.point_2 = this._wrapperFields.point_2.refObject;
    this._wrapperFields.point_3.freeze(own, checkConsistency);
    this._refObject.point_3 = this._wrapperFields.point_3.refObject;
    this._wrapperFields.point_4.freeze(own, checkConsistency);
    this._refObject.point_4 = this._wrapperFields.point_4.refObject;
    this._wrapperFields.point_5.freeze(own, checkConsistency);
    this._refObject.point_5 = this._wrapperFields.point_5.refObject;
    for (let i = 0; i < this._wrapperFields.point_valid.data.length; i++) {
      this._refObject.point_valid[i] = this._wrapperFields.point_valid.data[i];
    }
    for (let i = 0; i < this._wrapperFields.command.data.length; i++) {
      this._refObject.command[i] = this._wrapperFields.command.data[i];
    }
    for (let i = 0; i < this._wrapperFields.time_horizon.data.length; i++) {
      this._refObject.time_horizon[i] = this._wrapperFields.time_horizon.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
    this._wrapperFields.point_1.copyRefObject(refObject.point_1);
    this._wrapperFields.point_2.copyRefObject(refObject.point_2);
    this._wrapperFields.point_3.copyRefObject(refObject.point_3);
    this._wrapperFields.point_4.copyRefObject(refObject.point_4);
    this._wrapperFields.point_5.copyRefObject(refObject.point_5);
    this._refObject.point_valid = refObject.point_valid;
    this._wrapperFields.point_valid.fill(refObject.point_valid.toArray());
    this._refObject.command = refObject.command;
    this._wrapperFields.command.fill(refObject.command.toArray());
    this._refObject.time_horizon = refObject.time_horizon;
    this._wrapperFields.time_horizon.fill(refObject.time_horizon.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    PositionTargetWrapper.freeStruct(refObject.point_1);
    PositionTargetWrapper.freeStruct(refObject.point_2);
    PositionTargetWrapper.freeStruct(refObject.point_3);
    PositionTargetWrapper.freeStruct(refObject.point_4);
    PositionTargetWrapper.freeStruct(refObject.point_5);
  }
  static destoryRawROS(msg) {
    TrajectoryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'Trajectory'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  get point_1() {
    return this._wrapperFields.point_1;
  }
  set point_1(value) {
    if (value instanceof PositionTargetWrapper) {
      this._wrapperFields.point_1.copy(value);
    } else {
      this._wrapperFields.point_1.copy(new PositionTargetWrapper(value));
    }
  }
  get point_2() {
    return this._wrapperFields.point_2;
  }
  set point_2(value) {
    if (value instanceof PositionTargetWrapper) {
      this._wrapperFields.point_2.copy(value);
    } else {
      this._wrapperFields.point_2.copy(new PositionTargetWrapper(value));
    }
  }
  get point_3() {
    return this._wrapperFields.point_3;
  }
  set point_3(value) {
    if (value instanceof PositionTargetWrapper) {
      this._wrapperFields.point_3.copy(value);
    } else {
      this._wrapperFields.point_3.copy(new PositionTargetWrapper(value));
    }
  }
  get point_4() {
    return this._wrapperFields.point_4;
  }
  set point_4(value) {
    if (value instanceof PositionTargetWrapper) {
      this._wrapperFields.point_4.copy(value);
    } else {
      this._wrapperFields.point_4.copy(new PositionTargetWrapper(value));
    }
  }
  get point_5() {
    return this._wrapperFields.point_5;
  }
  set point_5(value) {
    if (value instanceof PositionTargetWrapper) {
      this._wrapperFields.point_5.copy(value);
    } else {
      this._wrapperFields.point_5.copy(new PositionTargetWrapper(value));
    }
  }
  get point_valid() {
    return this._wrapperFields['point_valid'].data;
  }
  set point_valid(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields['point_valid'].fill(value);
  }
  get command() {
    return this._wrapperFields['command'].data;
  }
  set command(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields['command'].fill(value);
  }
  get time_horizon() {
    return this._wrapperFields['time_horizon'].data;
  }
  set time_horizon(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields['time_horizon'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new TrajectoryRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._typeIntialized = true;
    this._wrapperFields.point_1.copyRefObject(this._refObject.point_1);
    this._wrapperFields.point_2.copyRefObject(this._refObject.point_2);
    this._wrapperFields.point_3.copyRefObject(this._refObject.point_3);
    this._wrapperFields.point_4.copyRefObject(this._refObject.point_4);
    this._wrapperFields.point_5.copyRefObject(this._refObject.point_5);
    this._wrapperFields.point_valid.fill(refObject.point_valid.toArray());
    this._wrapperFields.command.fill(refObject.command.toArray());
    this._wrapperFields.time_horizon.fill(refObject.time_horizon.toArray());
  }
  copy(other) {
    this._refObject = new TrajectoryRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._typeIntialized = true;
    this._wrapperFields.point_1.copy(other._wrapperFields.point_1);
    this._wrapperFields.point_2.copy(other._wrapperFields.point_2);
    this._wrapperFields.point_3.copy(other._wrapperFields.point_3);
    this._wrapperFields.point_4.copy(other._wrapperFields.point_4);
    this._wrapperFields.point_5.copy(other._wrapperFields.point_5);
    this._wrapperFields.point_valid.copy(other._wrapperFields.point_valid);
    this._wrapperFields.command.copy(other._wrapperFields.command);
    this._wrapperFields.time_horizon.copy(other._wrapperFields.time_horizon);
  }
  static get classType() {
    return TrajectoryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"MAV_TRAJECTORY_REPRESENTATION_WAYPOINTS","value":0},{"type":"uint8","name":"MAV_TRAJECTORY_REPRESENTATION_BEZIER","value":1}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"point_1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"mavros_msgs","type":"PositionTarget","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"point_2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"mavros_msgs","type":"PositionTarget","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"point_3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"mavros_msgs","type":"PositionTarget","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"point_4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"mavros_msgs","type":"PositionTarget","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"point_5","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"mavros_msgs","type":"PositionTarget","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"point_valid","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"command","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_horizon","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"Trajectory","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Trajectory"};
  }
  hasMember(name) {
    let memberNames = ["header","type","point_1","point_2","point_3","point_4","point_5","point_valid","command","time_horizon"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TrajectoryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TrajectoryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TrajectoryWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectoryArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectoryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TrajectoryRefArray(size);
    this._refObject = new TrajectoryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TrajectoryWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TrajectoryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TrajectoryArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TrajectoryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TrajectoryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TrajectoryArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(TrajectoryWrapper, "MAV_TRAJECTORY_REPRESENTATION_WAYPOINTS", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(TrajectoryWrapper, "MAV_TRAJECTORY_REPRESENTATION_BEZIER", {value: 1, writable: false, enumerable: true, configurable: true});
module.exports = TrajectoryWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "MAV_TRAJECTORY_REPRESENTATION_WAYPOINTS",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "MAV_TRAJECTORY_REPRESENTATION_BEZIER",
      "value": 1
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "point_1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "mavros_msgs",
        "type": "PositionTarget",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "point_2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "mavros_msgs",
        "type": "PositionTarget",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "point_3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "mavros_msgs",
        "type": "PositionTarget",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "point_4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "mavros_msgs",
        "type": "PositionTarget",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "point_5",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "mavros_msgs",
        "type": "PositionTarget",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "point_valid",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "command",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_horizon",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "Trajectory",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Trajectory"
}
*/
