// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__CommandVtolTransition_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const CommandVtolTransition_RequestRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  state: primitiveTypes.uint8,
});
const CommandVtolTransition_RequestRefArray = ArrayType(CommandVtolTransition_RequestRefStruct);
const CommandVtolTransition_RequestRefStructArray = StructType({
  data: CommandVtolTransition_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CommandVtolTransition_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CommandVtolTransition_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._stateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CommandVtolTransition_RequestRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._stateIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CommandVtolTransition_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CommandVtolTransition_RequestArrayWrapper;
  }
  static get ArrayType() {
    return CommandVtolTransition_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return CommandVtolTransition_RequestRefStructArray
  }
  static get refObjectType() {
    return CommandVtolTransition_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._stateIntialized) {
        throw new TypeError('Invalid argument: state in CommandVtolTransition_Request');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._stateIntialized = true;
    this._refObject.state = refObject.state;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    CommandVtolTransition_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'CommandVtolTransition_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get state() {
    if (!this._stateIntialized) {
      return undefined;
    }
    return this._refObject.state;
  }
  set state(value) {
    this._stateIntialized = true;
    this._refObject.state = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CommandVtolTransition_RequestRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._stateIntialized = true;
  }
  copy(other) {
    this._refObject = new CommandVtolTransition_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._stateIntialized = true;
  }
  static get classType() {
    return CommandVtolTransition_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"STATE_MC","value":3},{"type":"uint8","name":"STATE_FW","value":4}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"CommandVtolTransition_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CommandVtolTransition_Request"};
  }
  hasMember(name) {
    let memberNames = ["header","state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CommandVtolTransition_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CommandVtolTransition_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CommandVtolTransition_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandVtolTransition_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandVtolTransition_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CommandVtolTransition_RequestRefArray(size);
    this._refObject = new CommandVtolTransition_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CommandVtolTransition_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CommandVtolTransition_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CommandVtolTransition_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CommandVtolTransition_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CommandVtolTransition_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CommandVtolTransition_RequestArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(CommandVtolTransition_RequestWrapper, "STATE_MC", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(CommandVtolTransition_RequestWrapper, "STATE_FW", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = CommandVtolTransition_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "STATE_MC",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "STATE_FW",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "CommandVtolTransition_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CommandVtolTransition_Request"
}
*/
