// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__EndpointDel_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const EndpointDel_RequestRefStruct = StructType({
  id: primitiveTypes.uint32,
  url: primitiveTypes.string,
  type: primitiveTypes.uint8,
});
const EndpointDel_RequestRefArray = ArrayType(EndpointDel_RequestRefStruct);
const EndpointDel_RequestRefStructArray = StructType({
  data: EndpointDel_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EndpointDel_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EndpointDel_RequestRefStruct(other._refObject.toObject());
      this._idIntialized = true;
      this._urlIntialized = true;
      this._wrapperFields.url =  new StringWrapper(other._wrapperFields.url);
      this._typeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EndpointDel_RequestRefStruct();
    this._idIntialized = false;
    this._urlIntialized = false;
    this._wrapperFields.url = new StringWrapper();
    this._typeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EndpointDel_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EndpointDel_RequestArrayWrapper;
  }
  static get ArrayType() {
    return EndpointDel_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return EndpointDel_RequestRefStructArray
  }
  static get refObjectType() {
    return EndpointDel_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in EndpointDel_Request');
      }
      if (!this._urlIntialized) {
        throw new TypeError('Invalid argument: url in EndpointDel_Request');
      }
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in EndpointDel_Request');
      }
    }
    if (own) {
      this._wrapperFields.url.freeze(own, checkConsistency);
    }
    this._refObject.url = this._wrapperFields.url.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._urlIntialized = true;
    this._wrapperFields.url.data = refObject.url.data;
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.url);
  }
  static destoryRawROS(msg) {
    EndpointDel_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'EndpointDel_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get url() {
    if (!this._urlIntialized) {
      return undefined;
    }
    return this._wrapperFields.url.data;
  }
  set url(value) {
    this._urlIntialized = true;
    this._wrapperFields.url.data = value;
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EndpointDel_RequestRefStruct(refObject.toObject());
    this._idIntialized = true;
    this._urlIntialized = true;
    this._wrapperFields.url.copyRefObject(this._refObject.url);
    this._typeIntialized = true;
  }
  copy(other) {
    this._refObject = new EndpointDel_RequestRefStruct(other._refObject.toObject());
    this._idIntialized = true;
    this._urlIntialized = true;
    this._wrapperFields.url.copy(other._wrapperFields.url);
    this._typeIntialized = true;
  }
  static get classType() {
    return EndpointDel_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"TYPE_FCU","value":0},{"type":"uint8","name":"TYPE_GCS","value":1},{"type":"uint8","name":"TYPE_UAS","value":2}],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"url","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"EndpointDel_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EndpointDel_Request"};
  }
  hasMember(name) {
    let memberNames = ["id","url","type"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EndpointDel_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EndpointDel_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EndpointDel_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EndpointDel_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EndpointDel_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EndpointDel_RequestRefArray(size);
    this._refObject = new EndpointDel_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EndpointDel_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EndpointDel_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EndpointDel_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EndpointDel_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EndpointDel_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EndpointDel_RequestArrayWrapper;
  }
}
// Define constants (3 in total)
Object.defineProperty(EndpointDel_RequestWrapper, "TYPE_FCU", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EndpointDel_RequestWrapper, "TYPE_GCS", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EndpointDel_RequestWrapper, "TYPE_UAS", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = EndpointDel_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "TYPE_FCU",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "TYPE_GCS",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "TYPE_UAS",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "url",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "EndpointDel_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EndpointDel_Request"
}
*/
