// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__ManualControl.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const ManualControlRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  x: primitiveTypes.float32,
  y: primitiveTypes.float32,
  z: primitiveTypes.float32,
  r: primitiveTypes.float32,
  buttons: primitiveTypes.uint16,
  buttons2: primitiveTypes.uint16,
  enabled_extensions: primitiveTypes.uint8,
  s: primitiveTypes.float32,
  t: primitiveTypes.float32,
  aux1: primitiveTypes.float32,
  aux2: primitiveTypes.float32,
  aux3: primitiveTypes.float32,
  aux4: primitiveTypes.float32,
  aux5: primitiveTypes.float32,
  aux6: primitiveTypes.float32,
});
const ManualControlRefArray = ArrayType(ManualControlRefStruct);
const ManualControlRefStructArray = StructType({
  data: ManualControlRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ManualControlWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ManualControlRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._xIntialized = true;
      this._yIntialized = true;
      this._zIntialized = true;
      this._rIntialized = true;
      this._buttonsIntialized = true;
      this._buttons2Intialized = true;
      this._enabled_extensionsIntialized = true;
      this._sIntialized = true;
      this._tIntialized = true;
      this._aux1Intialized = true;
      this._aux2Intialized = true;
      this._aux3Intialized = true;
      this._aux4Intialized = true;
      this._aux5Intialized = true;
      this._aux6Intialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ManualControlRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._xIntialized = false;
    this._yIntialized = false;
    this._zIntialized = false;
    this._rIntialized = false;
    this._buttonsIntialized = false;
    this._buttons2Intialized = false;
    this._enabled_extensionsIntialized = false;
    this._sIntialized = false;
    this._tIntialized = false;
    this._aux1Intialized = false;
    this._aux2Intialized = false;
    this._aux3Intialized = false;
    this._aux4Intialized = false;
    this._aux5Intialized = false;
    this._aux6Intialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ManualControlWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ManualControlArrayWrapper;
  }
  static get ArrayType() {
    return ManualControlArrayWrapper;
  }
  static get refObjectArrayType() {
    return ManualControlRefStructArray
  }
  static get refObjectType() {
    return ManualControlRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._xIntialized) {
        throw new TypeError('Invalid argument: x in ManualControl');
      }
      if (!this._yIntialized) {
        throw new TypeError('Invalid argument: y in ManualControl');
      }
      if (!this._zIntialized) {
        throw new TypeError('Invalid argument: z in ManualControl');
      }
      if (!this._rIntialized) {
        throw new TypeError('Invalid argument: r in ManualControl');
      }
      if (!this._buttonsIntialized) {
        throw new TypeError('Invalid argument: buttons in ManualControl');
      }
      if (!this._buttons2Intialized) {
        throw new TypeError('Invalid argument: buttons2 in ManualControl');
      }
      if (!this._enabled_extensionsIntialized) {
        throw new TypeError('Invalid argument: enabled_extensions in ManualControl');
      }
      if (!this._sIntialized) {
        throw new TypeError('Invalid argument: s in ManualControl');
      }
      if (!this._tIntialized) {
        throw new TypeError('Invalid argument: t in ManualControl');
      }
      if (!this._aux1Intialized) {
        throw new TypeError('Invalid argument: aux1 in ManualControl');
      }
      if (!this._aux2Intialized) {
        throw new TypeError('Invalid argument: aux2 in ManualControl');
      }
      if (!this._aux3Intialized) {
        throw new TypeError('Invalid argument: aux3 in ManualControl');
      }
      if (!this._aux4Intialized) {
        throw new TypeError('Invalid argument: aux4 in ManualControl');
      }
      if (!this._aux5Intialized) {
        throw new TypeError('Invalid argument: aux5 in ManualControl');
      }
      if (!this._aux6Intialized) {
        throw new TypeError('Invalid argument: aux6 in ManualControl');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._zIntialized = true;
    this._refObject.z = refObject.z;
    this._rIntialized = true;
    this._refObject.r = refObject.r;
    this._buttonsIntialized = true;
    this._refObject.buttons = refObject.buttons;
    this._buttons2Intialized = true;
    this._refObject.buttons2 = refObject.buttons2;
    this._enabled_extensionsIntialized = true;
    this._refObject.enabled_extensions = refObject.enabled_extensions;
    this._sIntialized = true;
    this._refObject.s = refObject.s;
    this._tIntialized = true;
    this._refObject.t = refObject.t;
    this._aux1Intialized = true;
    this._refObject.aux1 = refObject.aux1;
    this._aux2Intialized = true;
    this._refObject.aux2 = refObject.aux2;
    this._aux3Intialized = true;
    this._refObject.aux3 = refObject.aux3;
    this._aux4Intialized = true;
    this._refObject.aux4 = refObject.aux4;
    this._aux5Intialized = true;
    this._refObject.aux5 = refObject.aux5;
    this._aux6Intialized = true;
    this._refObject.aux6 = refObject.aux6;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    ManualControlWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'ManualControl'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get z() {
    if (!this._zIntialized) {
      return undefined;
    }
    return this._refObject.z;
  }
  set z(value) {
    this._zIntialized = true;
    this._refObject.z = value;
  }
  get r() {
    if (!this._rIntialized) {
      return undefined;
    }
    return this._refObject.r;
  }
  set r(value) {
    this._rIntialized = true;
    this._refObject.r = value;
  }
  get buttons() {
    if (!this._buttonsIntialized) {
      return undefined;
    }
    return this._refObject.buttons;
  }
  set buttons(value) {
    this._buttonsIntialized = true;
    this._refObject.buttons = value;
  }
  get buttons2() {
    if (!this._buttons2Intialized) {
      return undefined;
    }
    return this._refObject.buttons2;
  }
  set buttons2(value) {
    this._buttons2Intialized = true;
    this._refObject.buttons2 = value;
  }
  get enabled_extensions() {
    if (!this._enabled_extensionsIntialized) {
      return undefined;
    }
    return this._refObject.enabled_extensions;
  }
  set enabled_extensions(value) {
    this._enabled_extensionsIntialized = true;
    this._refObject.enabled_extensions = value;
  }
  get s() {
    if (!this._sIntialized) {
      return undefined;
    }
    return this._refObject.s;
  }
  set s(value) {
    this._sIntialized = true;
    this._refObject.s = value;
  }
  get t() {
    if (!this._tIntialized) {
      return undefined;
    }
    return this._refObject.t;
  }
  set t(value) {
    this._tIntialized = true;
    this._refObject.t = value;
  }
  get aux1() {
    if (!this._aux1Intialized) {
      return undefined;
    }
    return this._refObject.aux1;
  }
  set aux1(value) {
    this._aux1Intialized = true;
    this._refObject.aux1 = value;
  }
  get aux2() {
    if (!this._aux2Intialized) {
      return undefined;
    }
    return this._refObject.aux2;
  }
  set aux2(value) {
    this._aux2Intialized = true;
    this._refObject.aux2 = value;
  }
  get aux3() {
    if (!this._aux3Intialized) {
      return undefined;
    }
    return this._refObject.aux3;
  }
  set aux3(value) {
    this._aux3Intialized = true;
    this._refObject.aux3 = value;
  }
  get aux4() {
    if (!this._aux4Intialized) {
      return undefined;
    }
    return this._refObject.aux4;
  }
  set aux4(value) {
    this._aux4Intialized = true;
    this._refObject.aux4 = value;
  }
  get aux5() {
    if (!this._aux5Intialized) {
      return undefined;
    }
    return this._refObject.aux5;
  }
  set aux5(value) {
    this._aux5Intialized = true;
    this._refObject.aux5 = value;
  }
  get aux6() {
    if (!this._aux6Intialized) {
      return undefined;
    }
    return this._refObject.aux6;
  }
  set aux6(value) {
    this._aux6Intialized = true;
    this._refObject.aux6 = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ManualControlRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._rIntialized = true;
    this._buttonsIntialized = true;
    this._buttons2Intialized = true;
    this._enabled_extensionsIntialized = true;
    this._sIntialized = true;
    this._tIntialized = true;
    this._aux1Intialized = true;
    this._aux2Intialized = true;
    this._aux3Intialized = true;
    this._aux4Intialized = true;
    this._aux5Intialized = true;
    this._aux6Intialized = true;
  }
  copy(other) {
    this._refObject = new ManualControlRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._rIntialized = true;
    this._buttonsIntialized = true;
    this._buttons2Intialized = true;
    this._enabled_extensionsIntialized = true;
    this._sIntialized = true;
    this._tIntialized = true;
    this._aux1Intialized = true;
    this._aux2Intialized = true;
    this._aux3Intialized = true;
    this._aux4Intialized = true;
    this._aux5Intialized = true;
    this._aux6Intialized = true;
  }
  static get classType() {
    return ManualControlWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"r","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"buttons","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"buttons2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"enabled_extensions","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"s","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"t","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux5","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux6","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"ManualControl","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ManualControl"};
  }
  hasMember(name) {
    let memberNames = ["header","x","y","z","r","buttons","buttons2","enabled_extensions","s","t","aux1","aux2","aux3","aux4","aux5","aux6"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ManualControlArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ManualControlWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ManualControlWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ManualControlArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ManualControlArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ManualControlRefArray(size);
    this._refObject = new ManualControlRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ManualControlWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ManualControlArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ManualControlArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ManualControlWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ManualControlWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ManualControlArrayWrapper;
  }
}
module.exports = ManualControlWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "r",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "buttons",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "buttons2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "enabled_extensions",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "s",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "t",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux5",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux6",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "ManualControl",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ManualControl"
}
*/
