// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__OverrideRCIn.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt16Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt16.js');
const OverrideRCInRefStruct = StructType({
  channels: ArrayType(primitiveTypes.uint16, 18),
});
const OverrideRCInRefArray = ArrayType(OverrideRCInRefStruct);
const OverrideRCInRefStructArray = StructType({
  data: OverrideRCInRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OverrideRCInWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OverrideRCInRefStruct(other._refObject.toObject());
      this._wrapperFields.channels = UInt16Wrapper.createArray();
      this._wrapperFields.channels.copy(other._wrapperFields.channels);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OverrideRCInRefStruct();
    this._wrapperFields.channels = UInt16Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new OverrideRCInWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OverrideRCInArrayWrapper;
  }
  static get ArrayType() {
    return OverrideRCInArrayWrapper;
  }
  static get refObjectArrayType() {
    return OverrideRCInRefStructArray
  }
  static get refObjectType() {
    return OverrideRCInRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    for (let i = 0; i < this._wrapperFields.channels.data.length; i++) {
      this._refObject.channels[i] = this._wrapperFields.channels.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.channels = refObject.channels;
    this._wrapperFields.channels.fill(refObject.channels.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    OverrideRCInWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'OverrideRCIn'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get channels() {
    return this._wrapperFields['channels'].data;
  }
  set channels(value) {
    if (value.length !== 18) {
      throw new RangeError('The length of the array must be 18.');
    }
    this._wrapperFields['channels'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new OverrideRCInRefStruct(refObject.toObject());
    this._wrapperFields.channels.fill(refObject.channels.toArray());
  }
  copy(other) {
    this._refObject = new OverrideRCInRefStruct(other._refObject.toObject());
    this._wrapperFields.channels.copy(other._wrapperFields.channels);
  }
  static get classType() {
    return OverrideRCInWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint16","name":"CHAN_RELEASE","value":0},{"type":"uint16","name":"CHAN_NOCHANGE","value":65535}],"fields":[{"name":"channels","type":{"isArray":true,"arraySize":18,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"OverrideRCIn","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OverrideRCIn"};
  }
  hasMember(name) {
    let memberNames = ["channels"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OverrideRCInArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OverrideRCInWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OverrideRCInWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OverrideRCInArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OverrideRCInArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OverrideRCInRefArray(size);
    this._refObject = new OverrideRCInRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OverrideRCInWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OverrideRCInArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OverrideRCInArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OverrideRCInWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OverrideRCInWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OverrideRCInArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(OverrideRCInWrapper, "CHAN_RELEASE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OverrideRCInWrapper, "CHAN_NOCHANGE", {value: 65535, writable: false, enumerable: true, configurable: true});
module.exports = OverrideRCInWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint16",
      "name": "CHAN_RELEASE",
      "value": 0
    },
    {
      "type": "uint16",
      "name": "CHAN_NOCHANGE",
      "value": 65535
    }
  ],
  "fields": [
    {
      "name": "channels",
      "type": {
        "isArray": true,
        "arraySize": 18,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "OverrideRCIn",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OverrideRCIn"
}
*/
