// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__OpenDroneIDSelfID.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const OpenDroneIDSelfIDRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  id_or_mac: primitiveTypes.string,
  description_type: primitiveTypes.uint8,
  description: primitiveTypes.string,
});
const OpenDroneIDSelfIDRefArray = ArrayType(OpenDroneIDSelfIDRefStruct);
const OpenDroneIDSelfIDRefStructArray = StructType({
  data: OpenDroneIDSelfIDRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OpenDroneIDSelfIDWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OpenDroneIDSelfIDRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._id_or_macIntialized = true;
      this._wrapperFields.id_or_mac =  new StringWrapper(other._wrapperFields.id_or_mac);
      this._description_typeIntialized = true;
      this._descriptionIntialized = true;
      this._wrapperFields.description =  new StringWrapper(other._wrapperFields.description);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OpenDroneIDSelfIDRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._id_or_macIntialized = false;
    this._wrapperFields.id_or_mac = new StringWrapper();
    this._description_typeIntialized = false;
    this._descriptionIntialized = false;
    this._wrapperFields.description = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new OpenDroneIDSelfIDWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OpenDroneIDSelfIDArrayWrapper;
  }
  static get ArrayType() {
    return OpenDroneIDSelfIDArrayWrapper;
  }
  static get refObjectArrayType() {
    return OpenDroneIDSelfIDRefStructArray
  }
  static get refObjectType() {
    return OpenDroneIDSelfIDRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._id_or_macIntialized) {
        throw new TypeError('Invalid argument: id_or_mac in OpenDroneIDSelfID');
      }
      if (!this._description_typeIntialized) {
        throw new TypeError('Invalid argument: description_type in OpenDroneIDSelfID');
      }
      if (!this._descriptionIntialized) {
        throw new TypeError('Invalid argument: description in OpenDroneIDSelfID');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.id_or_mac.freeze(own, checkConsistency);
    }
    this._refObject.id_or_mac = this._wrapperFields.id_or_mac.refObject;
    if (own) {
      this._wrapperFields.description.freeze(own, checkConsistency);
    }
    this._refObject.description = this._wrapperFields.description.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.data = refObject.id_or_mac.data;
    this._description_typeIntialized = true;
    this._refObject.description_type = refObject.description_type;
    this._descriptionIntialized = true;
    this._wrapperFields.description.data = refObject.description.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.id_or_mac);
    StringWrapper.freeStruct(refObject.description);
  }
  static destoryRawROS(msg) {
    OpenDroneIDSelfIDWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'OpenDroneIDSelfID'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get id_or_mac() {
    if (!this._id_or_macIntialized) {
      return undefined;
    }
    return this._wrapperFields.id_or_mac.data;
  }
  set id_or_mac(value) {
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.data = value;
  }
  get description_type() {
    if (!this._description_typeIntialized) {
      return undefined;
    }
    return this._refObject.description_type;
  }
  set description_type(value) {
    this._description_typeIntialized = true;
    this._refObject.description_type = value;
  }
  get description() {
    if (!this._descriptionIntialized) {
      return undefined;
    }
    return this._wrapperFields.description.data;
  }
  set description(value) {
    this._descriptionIntialized = true;
    this._wrapperFields.description.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new OpenDroneIDSelfIDRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.copyRefObject(this._refObject.id_or_mac);
    this._description_typeIntialized = true;
    this._descriptionIntialized = true;
    this._wrapperFields.description.copyRefObject(this._refObject.description);
  }
  copy(other) {
    this._refObject = new OpenDroneIDSelfIDRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.copy(other._wrapperFields.id_or_mac);
    this._description_typeIntialized = true;
    this._descriptionIntialized = true;
    this._wrapperFields.description.copy(other._wrapperFields.description);
  }
  static get classType() {
    return OpenDroneIDSelfIDWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"DESC_TYPE_TEXT","value":0},{"type":"uint8","name":"DESC_TYPE_EMERGENCY","value":1},{"type":"uint8","name":"DESC_TYPE_EXTENDED_STATUS","value":2}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id_or_mac","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"description_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"description","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"OpenDroneIDSelfID","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OpenDroneIDSelfID"};
  }
  hasMember(name) {
    let memberNames = ["header","id_or_mac","description_type","description"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OpenDroneIDSelfIDArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OpenDroneIDSelfIDWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OpenDroneIDSelfIDWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpenDroneIDSelfIDArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpenDroneIDSelfIDArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OpenDroneIDSelfIDRefArray(size);
    this._refObject = new OpenDroneIDSelfIDRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OpenDroneIDSelfIDWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OpenDroneIDSelfIDArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OpenDroneIDSelfIDArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OpenDroneIDSelfIDWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OpenDroneIDSelfIDWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OpenDroneIDSelfIDArrayWrapper;
  }
}
// Define constants (3 in total)
Object.defineProperty(OpenDroneIDSelfIDWrapper, "DESC_TYPE_TEXT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDSelfIDWrapper, "DESC_TYPE_EMERGENCY", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDSelfIDWrapper, "DESC_TYPE_EXTENDED_STATUS", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = OpenDroneIDSelfIDWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "DESC_TYPE_TEXT",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "DESC_TYPE_EMERGENCY",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "DESC_TYPE_EXTENDED_STATUS",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "id_or_mac",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "description_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "description",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "OpenDroneIDSelfID",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OpenDroneIDSelfID"
}
*/
