// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__OpenDroneIDBasicID.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const OpenDroneIDBasicIDRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  id_or_mac: primitiveTypes.string,
  id_type: primitiveTypes.uint8,
  ua_type: primitiveTypes.uint8,
  uas_id: primitiveTypes.string,
});
const OpenDroneIDBasicIDRefArray = ArrayType(OpenDroneIDBasicIDRefStruct);
const OpenDroneIDBasicIDRefStructArray = StructType({
  data: OpenDroneIDBasicIDRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OpenDroneIDBasicIDWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OpenDroneIDBasicIDRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._id_or_macIntialized = true;
      this._wrapperFields.id_or_mac =  new StringWrapper(other._wrapperFields.id_or_mac);
      this._id_typeIntialized = true;
      this._ua_typeIntialized = true;
      this._uas_idIntialized = true;
      this._wrapperFields.uas_id =  new StringWrapper(other._wrapperFields.uas_id);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OpenDroneIDBasicIDRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._id_or_macIntialized = false;
    this._wrapperFields.id_or_mac = new StringWrapper();
    this._id_typeIntialized = false;
    this._ua_typeIntialized = false;
    this._uas_idIntialized = false;
    this._wrapperFields.uas_id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new OpenDroneIDBasicIDWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OpenDroneIDBasicIDArrayWrapper;
  }
  static get ArrayType() {
    return OpenDroneIDBasicIDArrayWrapper;
  }
  static get refObjectArrayType() {
    return OpenDroneIDBasicIDRefStructArray
  }
  static get refObjectType() {
    return OpenDroneIDBasicIDRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._id_or_macIntialized) {
        throw new TypeError('Invalid argument: id_or_mac in OpenDroneIDBasicID');
      }
      if (!this._id_typeIntialized) {
        throw new TypeError('Invalid argument: id_type in OpenDroneIDBasicID');
      }
      if (!this._ua_typeIntialized) {
        throw new TypeError('Invalid argument: ua_type in OpenDroneIDBasicID');
      }
      if (!this._uas_idIntialized) {
        throw new TypeError('Invalid argument: uas_id in OpenDroneIDBasicID');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.id_or_mac.freeze(own, checkConsistency);
    }
    this._refObject.id_or_mac = this._wrapperFields.id_or_mac.refObject;
    if (own) {
      this._wrapperFields.uas_id.freeze(own, checkConsistency);
    }
    this._refObject.uas_id = this._wrapperFields.uas_id.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.data = refObject.id_or_mac.data;
    this._id_typeIntialized = true;
    this._refObject.id_type = refObject.id_type;
    this._ua_typeIntialized = true;
    this._refObject.ua_type = refObject.ua_type;
    this._uas_idIntialized = true;
    this._wrapperFields.uas_id.data = refObject.uas_id.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.id_or_mac);
    StringWrapper.freeStruct(refObject.uas_id);
  }
  static destoryRawROS(msg) {
    OpenDroneIDBasicIDWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'OpenDroneIDBasicID'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get id_or_mac() {
    if (!this._id_or_macIntialized) {
      return undefined;
    }
    return this._wrapperFields.id_or_mac.data;
  }
  set id_or_mac(value) {
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.data = value;
  }
  get id_type() {
    if (!this._id_typeIntialized) {
      return undefined;
    }
    return this._refObject.id_type;
  }
  set id_type(value) {
    this._id_typeIntialized = true;
    this._refObject.id_type = value;
  }
  get ua_type() {
    if (!this._ua_typeIntialized) {
      return undefined;
    }
    return this._refObject.ua_type;
  }
  set ua_type(value) {
    this._ua_typeIntialized = true;
    this._refObject.ua_type = value;
  }
  get uas_id() {
    if (!this._uas_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.uas_id.data;
  }
  set uas_id(value) {
    this._uas_idIntialized = true;
    this._wrapperFields.uas_id.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new OpenDroneIDBasicIDRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.copyRefObject(this._refObject.id_or_mac);
    this._id_typeIntialized = true;
    this._ua_typeIntialized = true;
    this._uas_idIntialized = true;
    this._wrapperFields.uas_id.copyRefObject(this._refObject.uas_id);
  }
  copy(other) {
    this._refObject = new OpenDroneIDBasicIDRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._id_or_macIntialized = true;
    this._wrapperFields.id_or_mac.copy(other._wrapperFields.id_or_mac);
    this._id_typeIntialized = true;
    this._ua_typeIntialized = true;
    this._uas_idIntialized = true;
    this._wrapperFields.uas_id.copy(other._wrapperFields.uas_id);
  }
  static get classType() {
    return OpenDroneIDBasicIDWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ID_TYPE_NONE","value":0},{"type":"uint8","name":"ID_TYPE_SERIAL_NUMBER","value":1},{"type":"uint8","name":"ID_TYPE_CAA_REGISTRATION_ID","value":2},{"type":"uint8","name":"ID_TYPE_UTM_ASSIGNED_UUID","value":3},{"type":"uint8","name":"ID_TYPE_SPECIFIC_SESSION_ID","value":4},{"type":"uint8","name":"UA_TYPE_NONE","value":0},{"type":"uint8","name":"UA_TYPE_AEROPLANE","value":1},{"type":"uint8","name":"UA_TYPE_HELICOPTER_OR_MULTIROTOR","value":2},{"type":"uint8","name":"UA_TYPE_GYROPLANE","value":3},{"type":"uint8","name":"UA_TYPE_HYBRID_LIFT","value":4},{"type":"uint8","name":"UA_TYPE_ORNITHOPTER","value":5},{"type":"uint8","name":"UA_TYPE_GLIDER","value":6},{"type":"uint8","name":"UA_TYPE_KITE","value":7},{"type":"uint8","name":"UA_TYPE_FREE_BALLOON","value":8},{"type":"uint8","name":"UA_TYPE_CAPTIVE_BALLOON","value":9},{"type":"uint8","name":"UA_TYPE_AIRSHIP","value":10},{"type":"uint8","name":"UA_TYPE_FREE_FALL_PARACHUTE","value":11},{"type":"uint8","name":"UA_TYPE_ROCKET","value":12},{"type":"uint8","name":"UA_TYPE_TETHERED_POWERED_AIRCRAFT","value":13},{"type":"uint8","name":"UA_TYPE_GROUND_OBSTACLE","value":14},{"type":"uint8","name":"UA_TYPE_OTHER","value":15}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id_or_mac","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"id_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ua_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"uas_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"OpenDroneIDBasicID","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OpenDroneIDBasicID"};
  }
  hasMember(name) {
    let memberNames = ["header","id_or_mac","id_type","ua_type","uas_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OpenDroneIDBasicIDArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OpenDroneIDBasicIDWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OpenDroneIDBasicIDWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpenDroneIDBasicIDArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpenDroneIDBasicIDArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OpenDroneIDBasicIDRefArray(size);
    this._refObject = new OpenDroneIDBasicIDRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OpenDroneIDBasicIDWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OpenDroneIDBasicIDArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OpenDroneIDBasicIDArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OpenDroneIDBasicIDWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OpenDroneIDBasicIDWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OpenDroneIDBasicIDArrayWrapper;
  }
}
// Define constants (21 in total)
Object.defineProperty(OpenDroneIDBasicIDWrapper, "ID_TYPE_NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "ID_TYPE_SERIAL_NUMBER", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "ID_TYPE_CAA_REGISTRATION_ID", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "ID_TYPE_UTM_ASSIGNED_UUID", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "ID_TYPE_SPECIFIC_SESSION_ID", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_AEROPLANE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_HELICOPTER_OR_MULTIROTOR", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_GYROPLANE", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_HYBRID_LIFT", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_ORNITHOPTER", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_GLIDER", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_KITE", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_FREE_BALLOON", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_CAPTIVE_BALLOON", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_AIRSHIP", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_FREE_FALL_PARACHUTE", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_ROCKET", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_TETHERED_POWERED_AIRCRAFT", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_GROUND_OBSTACLE", {value: 14, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OpenDroneIDBasicIDWrapper, "UA_TYPE_OTHER", {value: 15, writable: false, enumerable: true, configurable: true});
module.exports = OpenDroneIDBasicIDWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ID_TYPE_NONE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ID_TYPE_SERIAL_NUMBER",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "ID_TYPE_CAA_REGISTRATION_ID",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "ID_TYPE_UTM_ASSIGNED_UUID",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "ID_TYPE_SPECIFIC_SESSION_ID",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_NONE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_AEROPLANE",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_HELICOPTER_OR_MULTIROTOR",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_GYROPLANE",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_HYBRID_LIFT",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_ORNITHOPTER",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_GLIDER",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_KITE",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_FREE_BALLOON",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_CAPTIVE_BALLOON",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_AIRSHIP",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_FREE_FALL_PARACHUTE",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_ROCKET",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_TETHERED_POWERED_AIRCRAFT",
      "value": 13
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_GROUND_OBSTACLE",
      "value": 14
    },
    {
      "type": "uint8",
      "name": "UA_TYPE_OTHER",
      "value": 15
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "id_or_mac",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "id_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ua_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "uas_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "OpenDroneIDBasicID",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OpenDroneIDBasicID"
}
*/
