// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__EndpointAdd_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const EndpointAdd_ResponseRefStruct = StructType({
  successful: primitiveTypes.bool,
  reason: primitiveTypes.string,
  id: primitiveTypes.uint32,
});
const EndpointAdd_ResponseRefArray = ArrayType(EndpointAdd_ResponseRefStruct);
const EndpointAdd_ResponseRefStructArray = StructType({
  data: EndpointAdd_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EndpointAdd_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EndpointAdd_ResponseRefStruct(other._refObject.toObject());
      this._successfulIntialized = true;
      this._reasonIntialized = true;
      this._wrapperFields.reason =  new StringWrapper(other._wrapperFields.reason);
      this._idIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EndpointAdd_ResponseRefStruct();
    this._successfulIntialized = false;
    this._reasonIntialized = false;
    this._wrapperFields.reason = new StringWrapper();
    this._idIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EndpointAdd_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EndpointAdd_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return EndpointAdd_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return EndpointAdd_ResponseRefStructArray
  }
  static get refObjectType() {
    return EndpointAdd_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successfulIntialized) {
        throw new TypeError('Invalid argument: successful in EndpointAdd_Response');
      }
      if (!this._reasonIntialized) {
        throw new TypeError('Invalid argument: reason in EndpointAdd_Response');
      }
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in EndpointAdd_Response');
      }
    }
    if (own) {
      this._wrapperFields.reason.freeze(own, checkConsistency);
    }
    this._refObject.reason = this._wrapperFields.reason.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successfulIntialized = true;
    this._refObject.successful = refObject.successful;
    this._reasonIntialized = true;
    this._wrapperFields.reason.data = refObject.reason.data;
    this._idIntialized = true;
    this._refObject.id = refObject.id;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.reason);
  }
  static destoryRawROS(msg) {
    EndpointAdd_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'EndpointAdd_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get successful() {
    if (!this._successfulIntialized) {
      return undefined;
    }
    return this._refObject.successful;
  }
  set successful(value) {
    this._successfulIntialized = true;
    this._refObject.successful = value;
  }
  get reason() {
    if (!this._reasonIntialized) {
      return undefined;
    }
    return this._wrapperFields.reason.data;
  }
  set reason(value) {
    this._reasonIntialized = true;
    this._wrapperFields.reason.data = value;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EndpointAdd_ResponseRefStruct(refObject.toObject());
    this._successfulIntialized = true;
    this._reasonIntialized = true;
    this._wrapperFields.reason.copyRefObject(this._refObject.reason);
    this._idIntialized = true;
  }
  copy(other) {
    this._refObject = new EndpointAdd_ResponseRefStruct(other._refObject.toObject());
    this._successfulIntialized = true;
    this._reasonIntialized = true;
    this._wrapperFields.reason.copy(other._wrapperFields.reason);
    this._idIntialized = true;
  }
  static get classType() {
    return EndpointAdd_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"successful","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reason","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"EndpointAdd_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EndpointAdd_Response"};
  }
  hasMember(name) {
    let memberNames = ["successful","reason","id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EndpointAdd_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EndpointAdd_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EndpointAdd_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EndpointAdd_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EndpointAdd_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EndpointAdd_ResponseRefArray(size);
    this._refObject = new EndpointAdd_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EndpointAdd_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EndpointAdd_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EndpointAdd_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EndpointAdd_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EndpointAdd_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EndpointAdd_ResponseArrayWrapper;
  }
}
module.exports = EndpointAdd_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "successful",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reason",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "EndpointAdd_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EndpointAdd_Response"
}
*/
