// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__LandingTarget.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const LandingTargetRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  target_num: primitiveTypes.uint8,
  frame: primitiveTypes.uint8,
  angle: ArrayType(primitiveTypes.float32, 2),
  distance: primitiveTypes.float32,
  size: ArrayType(primitiveTypes.float32, 2),
  pose: PoseWrapper.refObjectType,
  type: primitiveTypes.uint8,
});
const LandingTargetRefArray = ArrayType(LandingTargetRefStruct);
const LandingTargetRefStructArray = StructType({
  data: LandingTargetRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LandingTargetWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LandingTargetRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._target_numIntialized = true;
      this._frameIntialized = true;
      this._wrapperFields.angle = Float32Wrapper.createArray();
      this._wrapperFields.angle.copy(other._wrapperFields.angle);
      this._distanceIntialized = true;
      this._wrapperFields.size = Float32Wrapper.createArray();
      this._wrapperFields.size.copy(other._wrapperFields.size);
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
      this._typeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LandingTargetRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._target_numIntialized = false;
    this._frameIntialized = false;
    this._wrapperFields.angle = Float32Wrapper.createArray();
    this._distanceIntialized = false;
    this._wrapperFields.size = Float32Wrapper.createArray();
    this._wrapperFields.pose = new PoseWrapper();
    this._typeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LandingTargetWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LandingTargetArrayWrapper;
  }
  static get ArrayType() {
    return LandingTargetArrayWrapper;
  }
  static get refObjectArrayType() {
    return LandingTargetRefStructArray
  }
  static get refObjectType() {
    return LandingTargetRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._target_numIntialized) {
        throw new TypeError('Invalid argument: target_num in LandingTarget');
      }
      if (!this._frameIntialized) {
        throw new TypeError('Invalid argument: frame in LandingTarget');
      }
      if (!this._distanceIntialized) {
        throw new TypeError('Invalid argument: distance in LandingTarget');
      }
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in LandingTarget');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    for (let i = 0; i < this._wrapperFields.angle.data.length; i++) {
      this._refObject.angle[i] = this._wrapperFields.angle.data[i];
    }
    for (let i = 0; i < this._wrapperFields.size.data.length; i++) {
      this._refObject.size[i] = this._wrapperFields.size.data[i];
    }
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._target_numIntialized = true;
    this._refObject.target_num = refObject.target_num;
    this._frameIntialized = true;
    this._refObject.frame = refObject.frame;
    this._refObject.angle = refObject.angle;
    this._wrapperFields.angle.fill(refObject.angle.toArray());
    this._distanceIntialized = true;
    this._refObject.distance = refObject.distance;
    this._refObject.size = refObject.size;
    this._wrapperFields.size.fill(refObject.size.toArray());
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    PoseWrapper.freeStruct(refObject.pose);
  }
  static destoryRawROS(msg) {
    LandingTargetWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'LandingTarget'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get target_num() {
    if (!this._target_numIntialized) {
      return undefined;
    }
    return this._refObject.target_num;
  }
  set target_num(value) {
    this._target_numIntialized = true;
    this._refObject.target_num = value;
  }
  get frame() {
    if (!this._frameIntialized) {
      return undefined;
    }
    return this._refObject.frame;
  }
  set frame(value) {
    this._frameIntialized = true;
    this._refObject.frame = value;
  }
  get angle() {
    return this._wrapperFields['angle'].data;
  }
  set angle(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['angle'].fill(value);
  }
  get distance() {
    if (!this._distanceIntialized) {
      return undefined;
    }
    return this._refObject.distance;
  }
  set distance(value) {
    this._distanceIntialized = true;
    this._refObject.distance = value;
  }
  get size() {
    return this._wrapperFields['size'].data;
  }
  set size(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['size'].fill(value);
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LandingTargetRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._target_numIntialized = true;
    this._frameIntialized = true;
    this._wrapperFields.angle.fill(refObject.angle.toArray());
    this._distanceIntialized = true;
    this._wrapperFields.size.fill(refObject.size.toArray());
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._typeIntialized = true;
  }
  copy(other) {
    this._refObject = new LandingTargetRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._target_numIntialized = true;
    this._frameIntialized = true;
    this._wrapperFields.angle.copy(other._wrapperFields.angle);
    this._distanceIntialized = true;
    this._wrapperFields.size.copy(other._wrapperFields.size);
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._typeIntialized = true;
  }
  static get classType() {
    return LandingTargetWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"GLOBAL","value":0},{"type":"uint8","name":"LOCAL_NED","value":2},{"type":"uint8","name":"MISSION","value":3},{"type":"uint8","name":"GLOBAL_RELATIVE_ALT","value":4},{"type":"uint8","name":"LOCAL_ENU","value":5},{"type":"uint8","name":"GLOBAL_INT","value":6},{"type":"uint8","name":"GLOBAL_RELATIVE_ALT_INT","value":7},{"type":"uint8","name":"LOCAL_OFFSET_NED","value":8},{"type":"uint8","name":"BODY_NED","value":9},{"type":"uint8","name":"BODY_OFFSET_NED","value":10},{"type":"uint8","name":"GLOBAL_TERRAIN_ALT","value":11},{"type":"uint8","name":"GLOBAL_TERRAIN_ALT_INT","value":12},{"type":"uint8","name":"LIGHT_BEACON","value":0},{"type":"uint8","name":"RADIO_BEACON","value":1},{"type":"uint8","name":"VISION_FIDUCIAL","value":2},{"type":"uint8","name":"VISION_OTHER","value":3}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"target_num","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angle","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"distance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"LandingTarget","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LandingTarget"};
  }
  hasMember(name) {
    let memberNames = ["header","target_num","frame","angle","distance","size","pose","type"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LandingTargetArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LandingTargetWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LandingTargetWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandingTargetArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandingTargetArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LandingTargetRefArray(size);
    this._refObject = new LandingTargetRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LandingTargetWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LandingTargetArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LandingTargetArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LandingTargetWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LandingTargetWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LandingTargetArrayWrapper;
  }
}
// Define constants (16 in total)
Object.defineProperty(LandingTargetWrapper, "GLOBAL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "LOCAL_NED", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "MISSION", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "GLOBAL_RELATIVE_ALT", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "LOCAL_ENU", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "GLOBAL_INT", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "GLOBAL_RELATIVE_ALT_INT", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "LOCAL_OFFSET_NED", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "BODY_NED", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "BODY_OFFSET_NED", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "GLOBAL_TERRAIN_ALT", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "GLOBAL_TERRAIN_ALT_INT", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "LIGHT_BEACON", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "RADIO_BEACON", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "VISION_FIDUCIAL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingTargetWrapper, "VISION_OTHER", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = LandingTargetWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "GLOBAL",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "LOCAL_NED",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "MISSION",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "GLOBAL_RELATIVE_ALT",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "LOCAL_ENU",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "GLOBAL_INT",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "GLOBAL_RELATIVE_ALT_INT",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "LOCAL_OFFSET_NED",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "BODY_NED",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "BODY_OFFSET_NED",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "GLOBAL_TERRAIN_ALT",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "GLOBAL_TERRAIN_ALT_INT",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "LIGHT_BEACON",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "RADIO_BEACON",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "VISION_FIDUCIAL",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "VISION_OTHER",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "target_num",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angle",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "distance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "size",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "LandingTarget",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LandingTarget"
}
*/
