// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__GimbalManagerConfigure_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GimbalManagerConfigure_RequestRefStruct = StructType({
  sysid_primary: primitiveTypes.int16,
  compid_primary: primitiveTypes.int16,
  sysid_secondary: primitiveTypes.int16,
  compid_secondary: primitiveTypes.int16,
  gimbal_device_id: primitiveTypes.uint8,
});
const GimbalManagerConfigure_RequestRefArray = ArrayType(GimbalManagerConfigure_RequestRefStruct);
const GimbalManagerConfigure_RequestRefStructArray = StructType({
  data: GimbalManagerConfigure_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GimbalManagerConfigure_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GimbalManagerConfigure_RequestRefStruct(other._refObject.toObject());
      this._sysid_primaryIntialized = true;
      this._compid_primaryIntialized = true;
      this._sysid_secondaryIntialized = true;
      this._compid_secondaryIntialized = true;
      this._gimbal_device_idIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GimbalManagerConfigure_RequestRefStruct();
    this._sysid_primaryIntialized = false;
    this._compid_primaryIntialized = false;
    this._sysid_secondaryIntialized = false;
    this._compid_secondaryIntialized = false;
    this._gimbal_device_idIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GimbalManagerConfigure_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GimbalManagerConfigure_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GimbalManagerConfigure_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GimbalManagerConfigure_RequestRefStructArray
  }
  static get refObjectType() {
    return GimbalManagerConfigure_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._sysid_primaryIntialized) {
        throw new TypeError('Invalid argument: sysid_primary in GimbalManagerConfigure_Request');
      }
      if (!this._compid_primaryIntialized) {
        throw new TypeError('Invalid argument: compid_primary in GimbalManagerConfigure_Request');
      }
      if (!this._sysid_secondaryIntialized) {
        throw new TypeError('Invalid argument: sysid_secondary in GimbalManagerConfigure_Request');
      }
      if (!this._compid_secondaryIntialized) {
        throw new TypeError('Invalid argument: compid_secondary in GimbalManagerConfigure_Request');
      }
      if (!this._gimbal_device_idIntialized) {
        throw new TypeError('Invalid argument: gimbal_device_id in GimbalManagerConfigure_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._sysid_primaryIntialized = true;
    this._refObject.sysid_primary = refObject.sysid_primary;
    this._compid_primaryIntialized = true;
    this._refObject.compid_primary = refObject.compid_primary;
    this._sysid_secondaryIntialized = true;
    this._refObject.sysid_secondary = refObject.sysid_secondary;
    this._compid_secondaryIntialized = true;
    this._refObject.compid_secondary = refObject.compid_secondary;
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = refObject.gimbal_device_id;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GimbalManagerConfigure_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'GimbalManagerConfigure_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get sysid_primary() {
    if (!this._sysid_primaryIntialized) {
      return undefined;
    }
    return this._refObject.sysid_primary;
  }
  set sysid_primary(value) {
    this._sysid_primaryIntialized = true;
    this._refObject.sysid_primary = value;
  }
  get compid_primary() {
    if (!this._compid_primaryIntialized) {
      return undefined;
    }
    return this._refObject.compid_primary;
  }
  set compid_primary(value) {
    this._compid_primaryIntialized = true;
    this._refObject.compid_primary = value;
  }
  get sysid_secondary() {
    if (!this._sysid_secondaryIntialized) {
      return undefined;
    }
    return this._refObject.sysid_secondary;
  }
  set sysid_secondary(value) {
    this._sysid_secondaryIntialized = true;
    this._refObject.sysid_secondary = value;
  }
  get compid_secondary() {
    if (!this._compid_secondaryIntialized) {
      return undefined;
    }
    return this._refObject.compid_secondary;
  }
  set compid_secondary(value) {
    this._compid_secondaryIntialized = true;
    this._refObject.compid_secondary = value;
  }
  get gimbal_device_id() {
    if (!this._gimbal_device_idIntialized) {
      return undefined;
    }
    return this._refObject.gimbal_device_id;
  }
  set gimbal_device_id(value) {
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GimbalManagerConfigure_RequestRefStruct(refObject.toObject());
    this._sysid_primaryIntialized = true;
    this._compid_primaryIntialized = true;
    this._sysid_secondaryIntialized = true;
    this._compid_secondaryIntialized = true;
    this._gimbal_device_idIntialized = true;
  }
  copy(other) {
    this._refObject = new GimbalManagerConfigure_RequestRefStruct(other._refObject.toObject());
    this._sysid_primaryIntialized = true;
    this._compid_primaryIntialized = true;
    this._sysid_secondaryIntialized = true;
    this._compid_secondaryIntialized = true;
    this._gimbal_device_idIntialized = true;
  }
  static get classType() {
    return GimbalManagerConfigure_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"sysid_primary","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"compid_primary","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sysid_secondary","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"compid_secondary","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gimbal_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"GimbalManagerConfigure_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GimbalManagerConfigure_Request"};
  }
  hasMember(name) {
    let memberNames = ["sysid_primary","compid_primary","sysid_secondary","compid_secondary","gimbal_device_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GimbalManagerConfigure_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GimbalManagerConfigure_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GimbalManagerConfigure_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerConfigure_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerConfigure_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GimbalManagerConfigure_RequestRefArray(size);
    this._refObject = new GimbalManagerConfigure_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GimbalManagerConfigure_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GimbalManagerConfigure_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GimbalManagerConfigure_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GimbalManagerConfigure_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GimbalManagerConfigure_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GimbalManagerConfigure_RequestArrayWrapper;
  }
}
module.exports = GimbalManagerConfigure_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "sysid_primary",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "compid_primary",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sysid_secondary",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "compid_secondary",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gimbal_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "GimbalManagerConfigure_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GimbalManagerConfigure_Request"
}
*/
