// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__MountControl.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const MountControlRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  mode: primitiveTypes.uint8,
  pitch: primitiveTypes.float32,
  roll: primitiveTypes.float32,
  yaw: primitiveTypes.float32,
  altitude: primitiveTypes.float32,
  latitude: primitiveTypes.float32,
  longitude: primitiveTypes.float32,
});
const MountControlRefArray = ArrayType(MountControlRefStruct);
const MountControlRefStructArray = StructType({
  data: MountControlRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MountControlWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MountControlRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._modeIntialized = true;
      this._pitchIntialized = true;
      this._rollIntialized = true;
      this._yawIntialized = true;
      this._altitudeIntialized = true;
      this._latitudeIntialized = true;
      this._longitudeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MountControlRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._modeIntialized = false;
    this._pitchIntialized = false;
    this._rollIntialized = false;
    this._yawIntialized = false;
    this._altitudeIntialized = false;
    this._latitudeIntialized = false;
    this._longitudeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MountControlWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MountControlArrayWrapper;
  }
  static get ArrayType() {
    return MountControlArrayWrapper;
  }
  static get refObjectArrayType() {
    return MountControlRefStructArray
  }
  static get refObjectType() {
    return MountControlRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in MountControl');
      }
      if (!this._pitchIntialized) {
        throw new TypeError('Invalid argument: pitch in MountControl');
      }
      if (!this._rollIntialized) {
        throw new TypeError('Invalid argument: roll in MountControl');
      }
      if (!this._yawIntialized) {
        throw new TypeError('Invalid argument: yaw in MountControl');
      }
      if (!this._altitudeIntialized) {
        throw new TypeError('Invalid argument: altitude in MountControl');
      }
      if (!this._latitudeIntialized) {
        throw new TypeError('Invalid argument: latitude in MountControl');
      }
      if (!this._longitudeIntialized) {
        throw new TypeError('Invalid argument: longitude in MountControl');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._modeIntialized = true;
    this._refObject.mode = refObject.mode;
    this._pitchIntialized = true;
    this._refObject.pitch = refObject.pitch;
    this._rollIntialized = true;
    this._refObject.roll = refObject.roll;
    this._yawIntialized = true;
    this._refObject.yaw = refObject.yaw;
    this._altitudeIntialized = true;
    this._refObject.altitude = refObject.altitude;
    this._latitudeIntialized = true;
    this._refObject.latitude = refObject.latitude;
    this._longitudeIntialized = true;
    this._refObject.longitude = refObject.longitude;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    MountControlWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'MountControl'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._refObject.mode;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._refObject.mode = value;
  }
  get pitch() {
    if (!this._pitchIntialized) {
      return undefined;
    }
    return this._refObject.pitch;
  }
  set pitch(value) {
    this._pitchIntialized = true;
    this._refObject.pitch = value;
  }
  get roll() {
    if (!this._rollIntialized) {
      return undefined;
    }
    return this._refObject.roll;
  }
  set roll(value) {
    this._rollIntialized = true;
    this._refObject.roll = value;
  }
  get yaw() {
    if (!this._yawIntialized) {
      return undefined;
    }
    return this._refObject.yaw;
  }
  set yaw(value) {
    this._yawIntialized = true;
    this._refObject.yaw = value;
  }
  get altitude() {
    if (!this._altitudeIntialized) {
      return undefined;
    }
    return this._refObject.altitude;
  }
  set altitude(value) {
    this._altitudeIntialized = true;
    this._refObject.altitude = value;
  }
  get latitude() {
    if (!this._latitudeIntialized) {
      return undefined;
    }
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._latitudeIntialized = true;
    this._refObject.latitude = value;
  }
  get longitude() {
    if (!this._longitudeIntialized) {
      return undefined;
    }
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._longitudeIntialized = true;
    this._refObject.longitude = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MountControlRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._modeIntialized = true;
    this._pitchIntialized = true;
    this._rollIntialized = true;
    this._yawIntialized = true;
    this._altitudeIntialized = true;
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
  }
  copy(other) {
    this._refObject = new MountControlRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._modeIntialized = true;
    this._pitchIntialized = true;
    this._rollIntialized = true;
    this._yawIntialized = true;
    this._altitudeIntialized = true;
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
  }
  static get classType() {
    return MountControlWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"MAV_MOUNT_MODE_RETRACT","value":0},{"type":"uint8","name":"MAV_MOUNT_MODE_NEUTRAL","value":1},{"type":"uint8","name":"MAV_MOUNT_MODE_MAVLINK_TARGETING","value":2},{"type":"uint8","name":"MAV_MOUNT_MODE_RC_TARGETING","value":3},{"type":"uint8","name":"MAV_MOUNT_MODE_GPS_POINT","value":4}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"roll","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"MountControl","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MountControl"};
  }
  hasMember(name) {
    let memberNames = ["header","mode","pitch","roll","yaw","altitude","latitude","longitude"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MountControlArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MountControlWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MountControlWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MountControlArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MountControlArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MountControlRefArray(size);
    this._refObject = new MountControlRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MountControlWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MountControlArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MountControlArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MountControlWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MountControlWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MountControlArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(MountControlWrapper, "MAV_MOUNT_MODE_RETRACT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountControlWrapper, "MAV_MOUNT_MODE_NEUTRAL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountControlWrapper, "MAV_MOUNT_MODE_MAVLINK_TARGETING", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountControlWrapper, "MAV_MOUNT_MODE_RC_TARGETING", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MountControlWrapper, "MAV_MOUNT_MODE_GPS_POINT", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = MountControlWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "MAV_MOUNT_MODE_RETRACT",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "MAV_MOUNT_MODE_NEUTRAL",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "MAV_MOUNT_MODE_MAVLINK_TARGETING",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "MAV_MOUNT_MODE_RC_TARGETING",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "MAV_MOUNT_MODE_GPS_POINT",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "roll",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "MountControl",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MountControl"
}
*/
