// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__LogRequestData_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LogRequestData_RequestRefStruct = StructType({
  id: primitiveTypes.uint16,
  offset: primitiveTypes.uint32,
  count: primitiveTypes.uint32,
});
const LogRequestData_RequestRefArray = ArrayType(LogRequestData_RequestRefStruct);
const LogRequestData_RequestRefStructArray = StructType({
  data: LogRequestData_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LogRequestData_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LogRequestData_RequestRefStruct(other._refObject.toObject());
      this._idIntialized = true;
      this._offsetIntialized = true;
      this._countIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LogRequestData_RequestRefStruct();
    this._idIntialized = false;
    this._offsetIntialized = false;
    this._countIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LogRequestData_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LogRequestData_RequestArrayWrapper;
  }
  static get ArrayType() {
    return LogRequestData_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return LogRequestData_RequestRefStructArray
  }
  static get refObjectType() {
    return LogRequestData_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in LogRequestData_Request');
      }
      if (!this._offsetIntialized) {
        throw new TypeError('Invalid argument: offset in LogRequestData_Request');
      }
      if (!this._countIntialized) {
        throw new TypeError('Invalid argument: count in LogRequestData_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._offsetIntialized = true;
    this._refObject.offset = refObject.offset;
    this._countIntialized = true;
    this._refObject.count = refObject.count;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LogRequestData_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'LogRequestData_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get offset() {
    if (!this._offsetIntialized) {
      return undefined;
    }
    return this._refObject.offset;
  }
  set offset(value) {
    this._offsetIntialized = true;
    this._refObject.offset = value;
  }
  get count() {
    if (!this._countIntialized) {
      return undefined;
    }
    return this._refObject.count;
  }
  set count(value) {
    this._countIntialized = true;
    this._refObject.count = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LogRequestData_RequestRefStruct(refObject.toObject());
    this._idIntialized = true;
    this._offsetIntialized = true;
    this._countIntialized = true;
  }
  copy(other) {
    this._refObject = new LogRequestData_RequestRefStruct(other._refObject.toObject());
    this._idIntialized = true;
    this._offsetIntialized = true;
    this._countIntialized = true;
  }
  static get classType() {
    return LogRequestData_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"LogRequestData_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LogRequestData_Request"};
  }
  hasMember(name) {
    let memberNames = ["id","offset","count"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LogRequestData_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LogRequestData_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LogRequestData_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LogRequestData_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LogRequestData_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LogRequestData_RequestRefArray(size);
    this._refObject = new LogRequestData_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LogRequestData_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LogRequestData_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LogRequestData_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LogRequestData_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LogRequestData_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LogRequestData_RequestArrayWrapper;
  }
}
module.exports = LogRequestData_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "LogRequestData_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LogRequestData_Request"
}
*/
