// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__CamIMUStamp.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const CamIMUStampRefStruct = StructType({
  frame_stamp: TimeWrapper.refObjectType,
  frame_seq_id: primitiveTypes.int32,
});
const CamIMUStampRefArray = ArrayType(CamIMUStampRefStruct);
const CamIMUStampRefStructArray = StructType({
  data: CamIMUStampRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CamIMUStampWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CamIMUStampRefStruct(other._refObject.toObject());
      this._wrapperFields.frame_stamp =  new TimeWrapper(other._wrapperFields.frame_stamp);
      this._frame_seq_idIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CamIMUStampRefStruct();
    this._wrapperFields.frame_stamp = new TimeWrapper();
    this._frame_seq_idIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CamIMUStampWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CamIMUStampArrayWrapper;
  }
  static get ArrayType() {
    return CamIMUStampArrayWrapper;
  }
  static get refObjectArrayType() {
    return CamIMUStampRefStructArray
  }
  static get refObjectType() {
    return CamIMUStampRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._frame_seq_idIntialized) {
        throw new TypeError('Invalid argument: frame_seq_id in CamIMUStamp');
      }
    }
    this._wrapperFields.frame_stamp.freeze(own, checkConsistency);
    this._refObject.frame_stamp = this._wrapperFields.frame_stamp.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.frame_stamp.copyRefObject(refObject.frame_stamp);
    this._frame_seq_idIntialized = true;
    this._refObject.frame_seq_id = refObject.frame_seq_id;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.frame_stamp);
  }
  static destoryRawROS(msg) {
    CamIMUStampWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'CamIMUStamp'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get frame_stamp() {
    return this._wrapperFields.frame_stamp;
  }
  set frame_stamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.frame_stamp.copy(value);
    } else {
      this._wrapperFields.frame_stamp.copy(new TimeWrapper(value));
    }
  }
  get frame_seq_id() {
    if (!this._frame_seq_idIntialized) {
      return undefined;
    }
    return this._refObject.frame_seq_id;
  }
  set frame_seq_id(value) {
    this._frame_seq_idIntialized = true;
    this._refObject.frame_seq_id = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CamIMUStampRefStruct(refObject.toObject());
    this._wrapperFields.frame_stamp.copyRefObject(this._refObject.frame_stamp);
    this._frame_seq_idIntialized = true;
  }
  copy(other) {
    this._refObject = new CamIMUStampRefStruct(other._refObject.toObject());
    this._wrapperFields.frame_stamp.copy(other._wrapperFields.frame_stamp);
    this._frame_seq_idIntialized = true;
  }
  static get classType() {
    return CamIMUStampWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"frame_stamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"frame_seq_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"CamIMUStamp","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CamIMUStamp"};
  }
  hasMember(name) {
    let memberNames = ["frame_stamp","frame_seq_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CamIMUStampArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CamIMUStampWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CamIMUStampWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CamIMUStampArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CamIMUStampArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CamIMUStampRefArray(size);
    this._refObject = new CamIMUStampRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CamIMUStampWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CamIMUStampArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CamIMUStampArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CamIMUStampWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CamIMUStampWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CamIMUStampArrayWrapper;
  }
}
module.exports = CamIMUStampWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "frame_stamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "frame_seq_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "CamIMUStamp",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CamIMUStamp"
}
*/
