// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__CommandTriggerInterval_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CommandTriggerInterval_RequestRefStruct = StructType({
  cycle_time: primitiveTypes.float32,
  integration_time: primitiveTypes.float32,
});
const CommandTriggerInterval_RequestRefArray = ArrayType(CommandTriggerInterval_RequestRefStruct);
const CommandTriggerInterval_RequestRefStructArray = StructType({
  data: CommandTriggerInterval_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CommandTriggerInterval_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CommandTriggerInterval_RequestRefStruct(other._refObject.toObject());
      this._cycle_timeIntialized = true;
      this._integration_timeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CommandTriggerInterval_RequestRefStruct();
    this._cycle_timeIntialized = false;
    this._integration_timeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CommandTriggerInterval_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CommandTriggerInterval_RequestArrayWrapper;
  }
  static get ArrayType() {
    return CommandTriggerInterval_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return CommandTriggerInterval_RequestRefStructArray
  }
  static get refObjectType() {
    return CommandTriggerInterval_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._cycle_timeIntialized) {
        throw new TypeError('Invalid argument: cycle_time in CommandTriggerInterval_Request');
      }
      if (!this._integration_timeIntialized) {
        throw new TypeError('Invalid argument: integration_time in CommandTriggerInterval_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._cycle_timeIntialized = true;
    this._refObject.cycle_time = refObject.cycle_time;
    this._integration_timeIntialized = true;
    this._refObject.integration_time = refObject.integration_time;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    CommandTriggerInterval_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'CommandTriggerInterval_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get cycle_time() {
    if (!this._cycle_timeIntialized) {
      return undefined;
    }
    return this._refObject.cycle_time;
  }
  set cycle_time(value) {
    this._cycle_timeIntialized = true;
    this._refObject.cycle_time = value;
  }
  get integration_time() {
    if (!this._integration_timeIntialized) {
      return undefined;
    }
    return this._refObject.integration_time;
  }
  set integration_time(value) {
    this._integration_timeIntialized = true;
    this._refObject.integration_time = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CommandTriggerInterval_RequestRefStruct(refObject.toObject());
    this._cycle_timeIntialized = true;
    this._integration_timeIntialized = true;
  }
  copy(other) {
    this._refObject = new CommandTriggerInterval_RequestRefStruct(other._refObject.toObject());
    this._cycle_timeIntialized = true;
    this._integration_timeIntialized = true;
  }
  static get classType() {
    return CommandTriggerInterval_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"cycle_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"integration_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"CommandTriggerInterval_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CommandTriggerInterval_Request"};
  }
  hasMember(name) {
    let memberNames = ["cycle_time","integration_time"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CommandTriggerInterval_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CommandTriggerInterval_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CommandTriggerInterval_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandTriggerInterval_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CommandTriggerInterval_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CommandTriggerInterval_RequestRefArray(size);
    this._refObject = new CommandTriggerInterval_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CommandTriggerInterval_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CommandTriggerInterval_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CommandTriggerInterval_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CommandTriggerInterval_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CommandTriggerInterval_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CommandTriggerInterval_RequestArrayWrapper;
  }
}
module.exports = CommandTriggerInterval_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "cycle_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "integration_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "CommandTriggerInterval_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CommandTriggerInterval_Request"
}
*/
