// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__OpticalFlow.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const OpticalFlowRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  flow: Vector3Wrapper.refObjectType,
  flow_comp_m: Vector3Wrapper.refObjectType,
  quality: primitiveTypes.uint8,
  ground_distance: primitiveTypes.float32,
  flow_rate: Vector3Wrapper.refObjectType,
});
const OpticalFlowRefArray = ArrayType(OpticalFlowRefStruct);
const OpticalFlowRefStructArray = StructType({
  data: OpticalFlowRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OpticalFlowWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OpticalFlowRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.flow =  new Vector3Wrapper(other._wrapperFields.flow);
      this._wrapperFields.flow_comp_m =  new Vector3Wrapper(other._wrapperFields.flow_comp_m);
      this._qualityIntialized = true;
      this._ground_distanceIntialized = true;
      this._wrapperFields.flow_rate =  new Vector3Wrapper(other._wrapperFields.flow_rate);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OpticalFlowRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.flow = new Vector3Wrapper();
    this._wrapperFields.flow_comp_m = new Vector3Wrapper();
    this._qualityIntialized = false;
    this._ground_distanceIntialized = false;
    this._wrapperFields.flow_rate = new Vector3Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new OpticalFlowWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OpticalFlowArrayWrapper;
  }
  static get ArrayType() {
    return OpticalFlowArrayWrapper;
  }
  static get refObjectArrayType() {
    return OpticalFlowRefStructArray
  }
  static get refObjectType() {
    return OpticalFlowRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._qualityIntialized) {
        throw new TypeError('Invalid argument: quality in OpticalFlow');
      }
      if (!this._ground_distanceIntialized) {
        throw new TypeError('Invalid argument: ground_distance in OpticalFlow');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.flow.freeze(own, checkConsistency);
    this._refObject.flow = this._wrapperFields.flow.refObject;
    this._wrapperFields.flow_comp_m.freeze(own, checkConsistency);
    this._refObject.flow_comp_m = this._wrapperFields.flow_comp_m.refObject;
    this._wrapperFields.flow_rate.freeze(own, checkConsistency);
    this._refObject.flow_rate = this._wrapperFields.flow_rate.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.flow.copyRefObject(refObject.flow);
    this._wrapperFields.flow_comp_m.copyRefObject(refObject.flow_comp_m);
    this._qualityIntialized = true;
    this._refObject.quality = refObject.quality;
    this._ground_distanceIntialized = true;
    this._refObject.ground_distance = refObject.ground_distance;
    this._wrapperFields.flow_rate.copyRefObject(refObject.flow_rate);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    Vector3Wrapper.freeStruct(refObject.flow);
    Vector3Wrapper.freeStruct(refObject.flow_comp_m);
    Vector3Wrapper.freeStruct(refObject.flow_rate);
  }
  static destoryRawROS(msg) {
    OpticalFlowWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'OpticalFlow'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get flow() {
    return this._wrapperFields.flow;
  }
  set flow(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.flow.copy(value);
    } else {
      this._wrapperFields.flow.copy(new Vector3Wrapper(value));
    }
  }
  get flow_comp_m() {
    return this._wrapperFields.flow_comp_m;
  }
  set flow_comp_m(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.flow_comp_m.copy(value);
    } else {
      this._wrapperFields.flow_comp_m.copy(new Vector3Wrapper(value));
    }
  }
  get quality() {
    if (!this._qualityIntialized) {
      return undefined;
    }
    return this._refObject.quality;
  }
  set quality(value) {
    this._qualityIntialized = true;
    this._refObject.quality = value;
  }
  get ground_distance() {
    if (!this._ground_distanceIntialized) {
      return undefined;
    }
    return this._refObject.ground_distance;
  }
  set ground_distance(value) {
    this._ground_distanceIntialized = true;
    this._refObject.ground_distance = value;
  }
  get flow_rate() {
    return this._wrapperFields.flow_rate;
  }
  set flow_rate(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.flow_rate.copy(value);
    } else {
      this._wrapperFields.flow_rate.copy(new Vector3Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new OpticalFlowRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.flow.copyRefObject(this._refObject.flow);
    this._wrapperFields.flow_comp_m.copyRefObject(this._refObject.flow_comp_m);
    this._qualityIntialized = true;
    this._ground_distanceIntialized = true;
    this._wrapperFields.flow_rate.copyRefObject(this._refObject.flow_rate);
  }
  copy(other) {
    this._refObject = new OpticalFlowRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.flow.copy(other._wrapperFields.flow);
    this._wrapperFields.flow_comp_m.copy(other._wrapperFields.flow_comp_m);
    this._qualityIntialized = true;
    this._ground_distanceIntialized = true;
    this._wrapperFields.flow_rate.copy(other._wrapperFields.flow_rate);
  }
  static get classType() {
    return OpticalFlowWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"flow","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"flow_comp_m","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"quality","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ground_distance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flow_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"OpticalFlow","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OpticalFlow"};
  }
  hasMember(name) {
    let memberNames = ["header","flow","flow_comp_m","quality","ground_distance","flow_rate"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OpticalFlowArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OpticalFlowWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OpticalFlowWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpticalFlowArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpticalFlowArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OpticalFlowRefArray(size);
    this._refObject = new OpticalFlowRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OpticalFlowWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OpticalFlowArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OpticalFlowArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OpticalFlowWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OpticalFlowWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OpticalFlowArrayWrapper;
  }
}
module.exports = OpticalFlowWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "flow",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "flow_comp_m",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "quality",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ground_distance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flow_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "OpticalFlow",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OpticalFlow"
}
*/
