// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__FileChecksum_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const FileChecksum_RequestRefStruct = StructType({
  file_path: primitiveTypes.string,
});
const FileChecksum_RequestRefArray = ArrayType(FileChecksum_RequestRefStruct);
const FileChecksum_RequestRefStructArray = StructType({
  data: FileChecksum_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FileChecksum_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FileChecksum_RequestRefStruct(other._refObject.toObject());
      this._file_pathIntialized = true;
      this._wrapperFields.file_path =  new StringWrapper(other._wrapperFields.file_path);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FileChecksum_RequestRefStruct();
    this._file_pathIntialized = false;
    this._wrapperFields.file_path = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FileChecksum_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FileChecksum_RequestArrayWrapper;
  }
  static get ArrayType() {
    return FileChecksum_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return FileChecksum_RequestRefStructArray
  }
  static get refObjectType() {
    return FileChecksum_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._file_pathIntialized) {
        throw new TypeError('Invalid argument: file_path in FileChecksum_Request');
      }
    }
    if (own) {
      this._wrapperFields.file_path.freeze(own, checkConsistency);
    }
    this._refObject.file_path = this._wrapperFields.file_path.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._file_pathIntialized = true;
    this._wrapperFields.file_path.data = refObject.file_path.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.file_path);
  }
  static destoryRawROS(msg) {
    FileChecksum_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'FileChecksum_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get file_path() {
    if (!this._file_pathIntialized) {
      return undefined;
    }
    return this._wrapperFields.file_path.data;
  }
  set file_path(value) {
    this._file_pathIntialized = true;
    this._wrapperFields.file_path.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FileChecksum_RequestRefStruct(refObject.toObject());
    this._file_pathIntialized = true;
    this._wrapperFields.file_path.copyRefObject(this._refObject.file_path);
  }
  copy(other) {
    this._refObject = new FileChecksum_RequestRefStruct(other._refObject.toObject());
    this._file_pathIntialized = true;
    this._wrapperFields.file_path.copy(other._wrapperFields.file_path);
  }
  static get classType() {
    return FileChecksum_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"file_path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"FileChecksum_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FileChecksum_Request"};
  }
  hasMember(name) {
    let memberNames = ["file_path"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FileChecksum_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FileChecksum_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FileChecksum_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FileChecksum_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FileChecksum_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FileChecksum_RequestRefArray(size);
    this._refObject = new FileChecksum_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FileChecksum_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FileChecksum_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FileChecksum_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FileChecksum_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FileChecksum_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FileChecksum_RequestArrayWrapper;
  }
}
module.exports = FileChecksum_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "file_path",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "FileChecksum_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FileChecksum_Request"
}
*/
