// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__ADSBVehicle.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const ADSBVehicleRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  icao_address: primitiveTypes.uint32,
  callsign: primitiveTypes.string,
  latitude: primitiveTypes.float64,
  longitude: primitiveTypes.float64,
  altitude: primitiveTypes.float32,
  heading: primitiveTypes.float32,
  hor_velocity: primitiveTypes.float32,
  ver_velocity: primitiveTypes.float32,
  altitude_type: primitiveTypes.uint8,
  emitter_type: primitiveTypes.uint8,
  tslc: DurationWrapper.refObjectType,
  flags: primitiveTypes.uint16,
  squawk: primitiveTypes.uint16,
});
const ADSBVehicleRefArray = ArrayType(ADSBVehicleRefStruct);
const ADSBVehicleRefStructArray = StructType({
  data: ADSBVehicleRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ADSBVehicleWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ADSBVehicleRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._icao_addressIntialized = true;
      this._callsignIntialized = true;
      this._wrapperFields.callsign =  new StringWrapper(other._wrapperFields.callsign);
      this._latitudeIntialized = true;
      this._longitudeIntialized = true;
      this._altitudeIntialized = true;
      this._headingIntialized = true;
      this._hor_velocityIntialized = true;
      this._ver_velocityIntialized = true;
      this._altitude_typeIntialized = true;
      this._emitter_typeIntialized = true;
      this._wrapperFields.tslc =  new DurationWrapper(other._wrapperFields.tslc);
      this._flagsIntialized = true;
      this._squawkIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ADSBVehicleRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._icao_addressIntialized = false;
    this._callsignIntialized = false;
    this._wrapperFields.callsign = new StringWrapper();
    this._latitudeIntialized = false;
    this._longitudeIntialized = false;
    this._altitudeIntialized = false;
    this._headingIntialized = false;
    this._hor_velocityIntialized = false;
    this._ver_velocityIntialized = false;
    this._altitude_typeIntialized = false;
    this._emitter_typeIntialized = false;
    this._wrapperFields.tslc = new DurationWrapper();
    this._flagsIntialized = false;
    this._squawkIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ADSBVehicleWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ADSBVehicleArrayWrapper;
  }
  static get ArrayType() {
    return ADSBVehicleArrayWrapper;
  }
  static get refObjectArrayType() {
    return ADSBVehicleRefStructArray
  }
  static get refObjectType() {
    return ADSBVehicleRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._icao_addressIntialized) {
        throw new TypeError('Invalid argument: icao_address in ADSBVehicle');
      }
      if (!this._callsignIntialized) {
        throw new TypeError('Invalid argument: callsign in ADSBVehicle');
      }
      if (!this._latitudeIntialized) {
        throw new TypeError('Invalid argument: latitude in ADSBVehicle');
      }
      if (!this._longitudeIntialized) {
        throw new TypeError('Invalid argument: longitude in ADSBVehicle');
      }
      if (!this._altitudeIntialized) {
        throw new TypeError('Invalid argument: altitude in ADSBVehicle');
      }
      if (!this._headingIntialized) {
        throw new TypeError('Invalid argument: heading in ADSBVehicle');
      }
      if (!this._hor_velocityIntialized) {
        throw new TypeError('Invalid argument: hor_velocity in ADSBVehicle');
      }
      if (!this._ver_velocityIntialized) {
        throw new TypeError('Invalid argument: ver_velocity in ADSBVehicle');
      }
      if (!this._altitude_typeIntialized) {
        throw new TypeError('Invalid argument: altitude_type in ADSBVehicle');
      }
      if (!this._emitter_typeIntialized) {
        throw new TypeError('Invalid argument: emitter_type in ADSBVehicle');
      }
      if (!this._flagsIntialized) {
        throw new TypeError('Invalid argument: flags in ADSBVehicle');
      }
      if (!this._squawkIntialized) {
        throw new TypeError('Invalid argument: squawk in ADSBVehicle');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.callsign.freeze(own, checkConsistency);
    }
    this._refObject.callsign = this._wrapperFields.callsign.refObject;
    this._wrapperFields.tslc.freeze(own, checkConsistency);
    this._refObject.tslc = this._wrapperFields.tslc.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._icao_addressIntialized = true;
    this._refObject.icao_address = refObject.icao_address;
    this._callsignIntialized = true;
    this._wrapperFields.callsign.data = refObject.callsign.data;
    this._latitudeIntialized = true;
    this._refObject.latitude = refObject.latitude;
    this._longitudeIntialized = true;
    this._refObject.longitude = refObject.longitude;
    this._altitudeIntialized = true;
    this._refObject.altitude = refObject.altitude;
    this._headingIntialized = true;
    this._refObject.heading = refObject.heading;
    this._hor_velocityIntialized = true;
    this._refObject.hor_velocity = refObject.hor_velocity;
    this._ver_velocityIntialized = true;
    this._refObject.ver_velocity = refObject.ver_velocity;
    this._altitude_typeIntialized = true;
    this._refObject.altitude_type = refObject.altitude_type;
    this._emitter_typeIntialized = true;
    this._refObject.emitter_type = refObject.emitter_type;
    this._wrapperFields.tslc.copyRefObject(refObject.tslc);
    this._flagsIntialized = true;
    this._refObject.flags = refObject.flags;
    this._squawkIntialized = true;
    this._refObject.squawk = refObject.squawk;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.callsign);
    DurationWrapper.freeStruct(refObject.tslc);
  }
  static destoryRawROS(msg) {
    ADSBVehicleWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'ADSBVehicle'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get icao_address() {
    if (!this._icao_addressIntialized) {
      return undefined;
    }
    return this._refObject.icao_address;
  }
  set icao_address(value) {
    this._icao_addressIntialized = true;
    this._refObject.icao_address = value;
  }
  get callsign() {
    if (!this._callsignIntialized) {
      return undefined;
    }
    return this._wrapperFields.callsign.data;
  }
  set callsign(value) {
    this._callsignIntialized = true;
    this._wrapperFields.callsign.data = value;
  }
  get latitude() {
    if (!this._latitudeIntialized) {
      return undefined;
    }
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._latitudeIntialized = true;
    this._refObject.latitude = value;
  }
  get longitude() {
    if (!this._longitudeIntialized) {
      return undefined;
    }
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._longitudeIntialized = true;
    this._refObject.longitude = value;
  }
  get altitude() {
    if (!this._altitudeIntialized) {
      return undefined;
    }
    return this._refObject.altitude;
  }
  set altitude(value) {
    this._altitudeIntialized = true;
    this._refObject.altitude = value;
  }
  get heading() {
    if (!this._headingIntialized) {
      return undefined;
    }
    return this._refObject.heading;
  }
  set heading(value) {
    this._headingIntialized = true;
    this._refObject.heading = value;
  }
  get hor_velocity() {
    if (!this._hor_velocityIntialized) {
      return undefined;
    }
    return this._refObject.hor_velocity;
  }
  set hor_velocity(value) {
    this._hor_velocityIntialized = true;
    this._refObject.hor_velocity = value;
  }
  get ver_velocity() {
    if (!this._ver_velocityIntialized) {
      return undefined;
    }
    return this._refObject.ver_velocity;
  }
  set ver_velocity(value) {
    this._ver_velocityIntialized = true;
    this._refObject.ver_velocity = value;
  }
  get altitude_type() {
    if (!this._altitude_typeIntialized) {
      return undefined;
    }
    return this._refObject.altitude_type;
  }
  set altitude_type(value) {
    this._altitude_typeIntialized = true;
    this._refObject.altitude_type = value;
  }
  get emitter_type() {
    if (!this._emitter_typeIntialized) {
      return undefined;
    }
    return this._refObject.emitter_type;
  }
  set emitter_type(value) {
    this._emitter_typeIntialized = true;
    this._refObject.emitter_type = value;
  }
  get tslc() {
    return this._wrapperFields.tslc;
  }
  set tslc(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.tslc.copy(value);
    } else {
      this._wrapperFields.tslc.copy(new DurationWrapper(value));
    }
  }
  get flags() {
    if (!this._flagsIntialized) {
      return undefined;
    }
    return this._refObject.flags;
  }
  set flags(value) {
    this._flagsIntialized = true;
    this._refObject.flags = value;
  }
  get squawk() {
    if (!this._squawkIntialized) {
      return undefined;
    }
    return this._refObject.squawk;
  }
  set squawk(value) {
    this._squawkIntialized = true;
    this._refObject.squawk = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ADSBVehicleRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._icao_addressIntialized = true;
    this._callsignIntialized = true;
    this._wrapperFields.callsign.copyRefObject(this._refObject.callsign);
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._headingIntialized = true;
    this._hor_velocityIntialized = true;
    this._ver_velocityIntialized = true;
    this._altitude_typeIntialized = true;
    this._emitter_typeIntialized = true;
    this._wrapperFields.tslc.copyRefObject(this._refObject.tslc);
    this._flagsIntialized = true;
    this._squawkIntialized = true;
  }
  copy(other) {
    this._refObject = new ADSBVehicleRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._icao_addressIntialized = true;
    this._callsignIntialized = true;
    this._wrapperFields.callsign.copy(other._wrapperFields.callsign);
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._headingIntialized = true;
    this._hor_velocityIntialized = true;
    this._ver_velocityIntialized = true;
    this._altitude_typeIntialized = true;
    this._emitter_typeIntialized = true;
    this._wrapperFields.tslc.copy(other._wrapperFields.tslc);
    this._flagsIntialized = true;
    this._squawkIntialized = true;
  }
  static get classType() {
    return ADSBVehicleWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ALT_PRESSURE_QNH","value":0},{"type":"uint8","name":"ALT_GEOMETRIC","value":1},{"type":"uint8","name":"EMITTER_NO_INFO","value":0},{"type":"uint8","name":"EMITTER_LIGHT","value":1},{"type":"uint8","name":"EMITTER_SMALL","value":2},{"type":"uint8","name":"EMITTER_LARGE","value":3},{"type":"uint8","name":"EMITTER_HIGH_VORTEX_LARGE","value":4},{"type":"uint8","name":"EMITTER_HEAVY","value":5},{"type":"uint8","name":"EMITTER_HIGHLY_MANUV","value":6},{"type":"uint8","name":"EMITTER_ROTOCRAFT","value":7},{"type":"uint8","name":"EMITTER_UNASSIGNED","value":8},{"type":"uint8","name":"EMITTER_GLIDER","value":9},{"type":"uint8","name":"EMITTER_LIGHTER_AIR","value":10},{"type":"uint8","name":"EMITTER_PARACHUTE","value":11},{"type":"uint8","name":"EMITTER_ULTRA_LIGHT","value":12},{"type":"uint8","name":"EMITTER_UNASSIGNED2","value":13},{"type":"uint8","name":"EMITTER_UAV","value":14},{"type":"uint8","name":"EMITTER_SPACE","value":15},{"type":"uint8","name":"EMITTER_UNASSGINED3","value":16},{"type":"uint8","name":"EMITTER_EMERGENCY_SURFACE","value":17},{"type":"uint8","name":"EMITTER_SERVICE_SURFACE","value":18},{"type":"uint8","name":"EMITTER_POINT_OBSTACLE","value":19},{"type":"uint16","name":"FLAG_VALID_COORDS","value":1},{"type":"uint16","name":"FLAG_VALID_ALTITUDE","value":2},{"type":"uint16","name":"FLAG_VALID_HEADING","value":4},{"type":"uint16","name":"FLAG_VALID_VELOCITY","value":8},{"type":"uint16","name":"FLAG_VALID_CALLSIGN","value":16},{"type":"uint16","name":"FLAG_VALID_SQUAWK","value":32},{"type":"uint16","name":"FLAG_SIMULATED","value":64},{"type":"uint16","name":"FLAG_VERTICAL_VELOCITY_VALID","value":128},{"type":"uint16","name":"FLAG_BARO_VALID","value":256},{"type":"uint16","name":"FLAG_SOURCE_UAT","value":32768}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"icao_address","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"callsign","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"heading","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hor_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ver_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"emitter_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tslc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"squawk","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"ADSBVehicle","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ADSBVehicle"};
  }
  hasMember(name) {
    let memberNames = ["header","icao_address","callsign","latitude","longitude","altitude","heading","hor_velocity","ver_velocity","altitude_type","emitter_type","tslc","flags","squawk"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ADSBVehicleArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ADSBVehicleWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ADSBVehicleWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ADSBVehicleArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ADSBVehicleArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ADSBVehicleRefArray(size);
    this._refObject = new ADSBVehicleRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ADSBVehicleWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ADSBVehicleArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ADSBVehicleArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ADSBVehicleWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ADSBVehicleWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ADSBVehicleArrayWrapper;
  }
}
// Define constants (32 in total)
Object.defineProperty(ADSBVehicleWrapper, "ALT_PRESSURE_QNH", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "ALT_GEOMETRIC", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_NO_INFO", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_LIGHT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_SMALL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_LARGE", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_HIGH_VORTEX_LARGE", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_HEAVY", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_HIGHLY_MANUV", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_ROTOCRAFT", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_UNASSIGNED", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_GLIDER", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_LIGHTER_AIR", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_PARACHUTE", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_ULTRA_LIGHT", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_UNASSIGNED2", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_UAV", {value: 14, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_SPACE", {value: 15, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_UNASSGINED3", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_EMERGENCY_SURFACE", {value: 17, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_SERVICE_SURFACE", {value: 18, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "EMITTER_POINT_OBSTACLE", {value: 19, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "FLAG_VALID_COORDS", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "FLAG_VALID_ALTITUDE", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "FLAG_VALID_HEADING", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "FLAG_VALID_VELOCITY", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "FLAG_VALID_CALLSIGN", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "FLAG_VALID_SQUAWK", {value: 32, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "FLAG_SIMULATED", {value: 64, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "FLAG_VERTICAL_VELOCITY_VALID", {value: 128, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "FLAG_BARO_VALID", {value: 256, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ADSBVehicleWrapper, "FLAG_SOURCE_UAT", {value: 32768, writable: false, enumerable: true, configurable: true});
module.exports = ADSBVehicleWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ALT_PRESSURE_QNH",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ALT_GEOMETRIC",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "EMITTER_NO_INFO",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "EMITTER_LIGHT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "EMITTER_SMALL",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "EMITTER_LARGE",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "EMITTER_HIGH_VORTEX_LARGE",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "EMITTER_HEAVY",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "EMITTER_HIGHLY_MANUV",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "EMITTER_ROTOCRAFT",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "EMITTER_UNASSIGNED",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "EMITTER_GLIDER",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "EMITTER_LIGHTER_AIR",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "EMITTER_PARACHUTE",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "EMITTER_ULTRA_LIGHT",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "EMITTER_UNASSIGNED2",
      "value": 13
    },
    {
      "type": "uint8",
      "name": "EMITTER_UAV",
      "value": 14
    },
    {
      "type": "uint8",
      "name": "EMITTER_SPACE",
      "value": 15
    },
    {
      "type": "uint8",
      "name": "EMITTER_UNASSGINED3",
      "value": 16
    },
    {
      "type": "uint8",
      "name": "EMITTER_EMERGENCY_SURFACE",
      "value": 17
    },
    {
      "type": "uint8",
      "name": "EMITTER_SERVICE_SURFACE",
      "value": 18
    },
    {
      "type": "uint8",
      "name": "EMITTER_POINT_OBSTACLE",
      "value": 19
    },
    {
      "type": "uint16",
      "name": "FLAG_VALID_COORDS",
      "value": 1
    },
    {
      "type": "uint16",
      "name": "FLAG_VALID_ALTITUDE",
      "value": 2
    },
    {
      "type": "uint16",
      "name": "FLAG_VALID_HEADING",
      "value": 4
    },
    {
      "type": "uint16",
      "name": "FLAG_VALID_VELOCITY",
      "value": 8
    },
    {
      "type": "uint16",
      "name": "FLAG_VALID_CALLSIGN",
      "value": 16
    },
    {
      "type": "uint16",
      "name": "FLAG_VALID_SQUAWK",
      "value": 32
    },
    {
      "type": "uint16",
      "name": "FLAG_SIMULATED",
      "value": 64
    },
    {
      "type": "uint16",
      "name": "FLAG_VERTICAL_VELOCITY_VALID",
      "value": 128
    },
    {
      "type": "uint16",
      "name": "FLAG_BARO_VALID",
      "value": 256
    },
    {
      "type": "uint16",
      "name": "FLAG_SOURCE_UAT",
      "value": 32768
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "icao_address",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "callsign",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "heading",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hor_velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ver_velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "emitter_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tslc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "squawk",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "ADSBVehicle",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ADSBVehicle"
}
*/
