// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__OpenDroneIDSystemUpdate.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const OpenDroneIDSystemUpdateRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  operator_latitude: primitiveTypes.int32,
  operator_longitude: primitiveTypes.int32,
  operator_altitude_geo: primitiveTypes.float32,
});
const OpenDroneIDSystemUpdateRefArray = ArrayType(OpenDroneIDSystemUpdateRefStruct);
const OpenDroneIDSystemUpdateRefStructArray = StructType({
  data: OpenDroneIDSystemUpdateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OpenDroneIDSystemUpdateWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OpenDroneIDSystemUpdateRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._operator_latitudeIntialized = true;
      this._operator_longitudeIntialized = true;
      this._operator_altitude_geoIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OpenDroneIDSystemUpdateRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._operator_latitudeIntialized = false;
    this._operator_longitudeIntialized = false;
    this._operator_altitude_geoIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new OpenDroneIDSystemUpdateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OpenDroneIDSystemUpdateArrayWrapper;
  }
  static get ArrayType() {
    return OpenDroneIDSystemUpdateArrayWrapper;
  }
  static get refObjectArrayType() {
    return OpenDroneIDSystemUpdateRefStructArray
  }
  static get refObjectType() {
    return OpenDroneIDSystemUpdateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._operator_latitudeIntialized) {
        throw new TypeError('Invalid argument: operator_latitude in OpenDroneIDSystemUpdate');
      }
      if (!this._operator_longitudeIntialized) {
        throw new TypeError('Invalid argument: operator_longitude in OpenDroneIDSystemUpdate');
      }
      if (!this._operator_altitude_geoIntialized) {
        throw new TypeError('Invalid argument: operator_altitude_geo in OpenDroneIDSystemUpdate');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._operator_latitudeIntialized = true;
    this._refObject.operator_latitude = refObject.operator_latitude;
    this._operator_longitudeIntialized = true;
    this._refObject.operator_longitude = refObject.operator_longitude;
    this._operator_altitude_geoIntialized = true;
    this._refObject.operator_altitude_geo = refObject.operator_altitude_geo;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    OpenDroneIDSystemUpdateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'OpenDroneIDSystemUpdate'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get operator_latitude() {
    if (!this._operator_latitudeIntialized) {
      return undefined;
    }
    return this._refObject.operator_latitude;
  }
  set operator_latitude(value) {
    this._operator_latitudeIntialized = true;
    this._refObject.operator_latitude = value;
  }
  get operator_longitude() {
    if (!this._operator_longitudeIntialized) {
      return undefined;
    }
    return this._refObject.operator_longitude;
  }
  set operator_longitude(value) {
    this._operator_longitudeIntialized = true;
    this._refObject.operator_longitude = value;
  }
  get operator_altitude_geo() {
    if (!this._operator_altitude_geoIntialized) {
      return undefined;
    }
    return this._refObject.operator_altitude_geo;
  }
  set operator_altitude_geo(value) {
    this._operator_altitude_geoIntialized = true;
    this._refObject.operator_altitude_geo = value;
  }
  copyRefObject(refObject) {
    this._refObject = new OpenDroneIDSystemUpdateRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._operator_latitudeIntialized = true;
    this._operator_longitudeIntialized = true;
    this._operator_altitude_geoIntialized = true;
  }
  copy(other) {
    this._refObject = new OpenDroneIDSystemUpdateRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._operator_latitudeIntialized = true;
    this._operator_longitudeIntialized = true;
    this._operator_altitude_geoIntialized = true;
  }
  static get classType() {
    return OpenDroneIDSystemUpdateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"operator_latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"operator_longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"operator_altitude_geo","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"OpenDroneIDSystemUpdate","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OpenDroneIDSystemUpdate"};
  }
  hasMember(name) {
    let memberNames = ["header","operator_latitude","operator_longitude","operator_altitude_geo"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OpenDroneIDSystemUpdateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OpenDroneIDSystemUpdateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OpenDroneIDSystemUpdateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpenDroneIDSystemUpdateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OpenDroneIDSystemUpdateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OpenDroneIDSystemUpdateRefArray(size);
    this._refObject = new OpenDroneIDSystemUpdateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OpenDroneIDSystemUpdateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OpenDroneIDSystemUpdateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OpenDroneIDSystemUpdateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OpenDroneIDSystemUpdateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OpenDroneIDSystemUpdateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OpenDroneIDSystemUpdateArrayWrapper;
  }
}
module.exports = OpenDroneIDSystemUpdateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "operator_latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "operator_longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "operator_altitude_geo",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "OpenDroneIDSystemUpdate",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OpenDroneIDSystemUpdate"
}
*/
