// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__TimesyncStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const TimesyncStatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  remote_timestamp_ns: primitiveTypes.uint64,
  observed_offset_ns: primitiveTypes.int64,
  estimated_offset_ns: primitiveTypes.int64,
  round_trip_time_ms: primitiveTypes.float32,
});
const TimesyncStatusRefArray = ArrayType(TimesyncStatusRefStruct);
const TimesyncStatusRefStructArray = StructType({
  data: TimesyncStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TimesyncStatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TimesyncStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._remote_timestamp_nsIntialized = true;
      this._observed_offset_nsIntialized = true;
      this._estimated_offset_nsIntialized = true;
      this._round_trip_time_msIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TimesyncStatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._remote_timestamp_nsIntialized = false;
    this._observed_offset_nsIntialized = false;
    this._estimated_offset_nsIntialized = false;
    this._round_trip_time_msIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new TimesyncStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TimesyncStatusArrayWrapper;
  }
  static get ArrayType() {
    return TimesyncStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return TimesyncStatusRefStructArray
  }
  static get refObjectType() {
    return TimesyncStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._remote_timestamp_nsIntialized) {
        throw new TypeError('Invalid argument: remote_timestamp_ns in TimesyncStatus');
      }
      if (!this._observed_offset_nsIntialized) {
        throw new TypeError('Invalid argument: observed_offset_ns in TimesyncStatus');
      }
      if (!this._estimated_offset_nsIntialized) {
        throw new TypeError('Invalid argument: estimated_offset_ns in TimesyncStatus');
      }
      if (!this._round_trip_time_msIntialized) {
        throw new TypeError('Invalid argument: round_trip_time_ms in TimesyncStatus');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._remote_timestamp_nsIntialized = true;
    this._refObject.remote_timestamp_ns = refObject.remote_timestamp_ns;
    this._observed_offset_nsIntialized = true;
    this._refObject.observed_offset_ns = refObject.observed_offset_ns;
    this._estimated_offset_nsIntialized = true;
    this._refObject.estimated_offset_ns = refObject.estimated_offset_ns;
    this._round_trip_time_msIntialized = true;
    this._refObject.round_trip_time_ms = refObject.round_trip_time_ms;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    TimesyncStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'TimesyncStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get remote_timestamp_ns() {
    if (!this._remote_timestamp_nsIntialized) {
      return undefined;
    }
    return this._refObject.remote_timestamp_ns;
  }
  set remote_timestamp_ns(value) {
    this._remote_timestamp_nsIntialized = true;
    this._refObject.remote_timestamp_ns = value;
  }
  get observed_offset_ns() {
    if (!this._observed_offset_nsIntialized) {
      return undefined;
    }
    return this._refObject.observed_offset_ns;
  }
  set observed_offset_ns(value) {
    this._observed_offset_nsIntialized = true;
    this._refObject.observed_offset_ns = value;
  }
  get estimated_offset_ns() {
    if (!this._estimated_offset_nsIntialized) {
      return undefined;
    }
    return this._refObject.estimated_offset_ns;
  }
  set estimated_offset_ns(value) {
    this._estimated_offset_nsIntialized = true;
    this._refObject.estimated_offset_ns = value;
  }
  get round_trip_time_ms() {
    if (!this._round_trip_time_msIntialized) {
      return undefined;
    }
    return this._refObject.round_trip_time_ms;
  }
  set round_trip_time_ms(value) {
    this._round_trip_time_msIntialized = true;
    this._refObject.round_trip_time_ms = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TimesyncStatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._remote_timestamp_nsIntialized = true;
    this._observed_offset_nsIntialized = true;
    this._estimated_offset_nsIntialized = true;
    this._round_trip_time_msIntialized = true;
  }
  copy(other) {
    this._refObject = new TimesyncStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._remote_timestamp_nsIntialized = true;
    this._observed_offset_nsIntialized = true;
    this._estimated_offset_nsIntialized = true;
    this._round_trip_time_msIntialized = true;
  }
  static get classType() {
    return TimesyncStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"remote_timestamp_ns","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"observed_offset_ns","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"estimated_offset_ns","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"round_trip_time_ms","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"TimesyncStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TimesyncStatus"};
  }
  hasMember(name) {
    let memberNames = ["header","remote_timestamp_ns","observed_offset_ns","estimated_offset_ns","round_trip_time_ms"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TimesyncStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TimesyncStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TimesyncStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TimesyncStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TimesyncStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TimesyncStatusRefArray(size);
    this._refObject = new TimesyncStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TimesyncStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TimesyncStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TimesyncStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TimesyncStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TimesyncStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TimesyncStatusArrayWrapper;
  }
}
module.exports = TimesyncStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "remote_timestamp_ns",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "observed_offset_ns",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "estimated_offset_ns",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "round_trip_time_ms",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "TimesyncStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TimesyncStatus"
}
*/
