// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__ParamValue.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ParamValueRefStruct = StructType({
  integer: primitiveTypes.int64,
  real: primitiveTypes.float64,
});
const ParamValueRefArray = ArrayType(ParamValueRefStruct);
const ParamValueRefStructArray = StructType({
  data: ParamValueRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ParamValueWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ParamValueRefStruct(other._refObject.toObject());
      this._integerIntialized = true;
      this._realIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ParamValueRefStruct();
    this._integerIntialized = false;
    this._realIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ParamValueWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ParamValueArrayWrapper;
  }
  static get ArrayType() {
    return ParamValueArrayWrapper;
  }
  static get refObjectArrayType() {
    return ParamValueRefStructArray
  }
  static get refObjectType() {
    return ParamValueRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._integerIntialized) {
        throw new TypeError('Invalid argument: integer in ParamValue');
      }
      if (!this._realIntialized) {
        throw new TypeError('Invalid argument: real in ParamValue');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._integerIntialized = true;
    this._refObject.integer = refObject.integer;
    this._realIntialized = true;
    this._refObject.real = refObject.real;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ParamValueWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'ParamValue'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get integer() {
    if (!this._integerIntialized) {
      return undefined;
    }
    return this._refObject.integer;
  }
  set integer(value) {
    this._integerIntialized = true;
    this._refObject.integer = value;
  }
  get real() {
    if (!this._realIntialized) {
      return undefined;
    }
    return this._refObject.real;
  }
  set real(value) {
    this._realIntialized = true;
    this._refObject.real = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ParamValueRefStruct(refObject.toObject());
    this._integerIntialized = true;
    this._realIntialized = true;
  }
  copy(other) {
    this._refObject = new ParamValueRefStruct(other._refObject.toObject());
    this._integerIntialized = true;
    this._realIntialized = true;
  }
  static get classType() {
    return ParamValueWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"integer","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"real","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"ParamValue","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ParamValue"};
  }
  hasMember(name) {
    let memberNames = ["integer","real"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ParamValueArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ParamValueWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ParamValueWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParamValueArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParamValueArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ParamValueRefArray(size);
    this._refObject = new ParamValueRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ParamValueWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ParamValueArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ParamValueArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ParamValueWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ParamValueWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ParamValueArrayWrapper;
  }
}
module.exports = ParamValueWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "integer",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "real",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "ParamValue",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ParamValue"
}
*/
