// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__CameraImageCaptured.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const QuaternionWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Quaternion.js');
const GeoPointWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__GeoPoint.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const CameraImageCapturedRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  orientation: QuaternionWrapper.refObjectType,
  geo: GeoPointWrapper.refObjectType,
  relative_alt: primitiveTypes.float32,
  image_index: primitiveTypes.int32,
  capture_result: primitiveTypes.int8,
  file_url: primitiveTypes.string,
});
const CameraImageCapturedRefArray = ArrayType(CameraImageCapturedRefStruct);
const CameraImageCapturedRefStructArray = StructType({
  data: CameraImageCapturedRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CameraImageCapturedWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CameraImageCapturedRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.orientation =  new QuaternionWrapper(other._wrapperFields.orientation);
      this._wrapperFields.geo =  new GeoPointWrapper(other._wrapperFields.geo);
      this._relative_altIntialized = true;
      this._image_indexIntialized = true;
      this._capture_resultIntialized = true;
      this._file_urlIntialized = true;
      this._wrapperFields.file_url =  new StringWrapper(other._wrapperFields.file_url);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CameraImageCapturedRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.orientation = new QuaternionWrapper();
    this._wrapperFields.geo = new GeoPointWrapper();
    this._relative_altIntialized = false;
    this._image_indexIntialized = false;
    this._capture_resultIntialized = false;
    this._file_urlIntialized = false;
    this._wrapperFields.file_url = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CameraImageCapturedWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CameraImageCapturedArrayWrapper;
  }
  static get ArrayType() {
    return CameraImageCapturedArrayWrapper;
  }
  static get refObjectArrayType() {
    return CameraImageCapturedRefStructArray
  }
  static get refObjectType() {
    return CameraImageCapturedRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._relative_altIntialized) {
        throw new TypeError('Invalid argument: relative_alt in CameraImageCaptured');
      }
      if (!this._image_indexIntialized) {
        throw new TypeError('Invalid argument: image_index in CameraImageCaptured');
      }
      if (!this._capture_resultIntialized) {
        throw new TypeError('Invalid argument: capture_result in CameraImageCaptured');
      }
      if (!this._file_urlIntialized) {
        throw new TypeError('Invalid argument: file_url in CameraImageCaptured');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.orientation.freeze(own, checkConsistency);
    this._refObject.orientation = this._wrapperFields.orientation.refObject;
    this._wrapperFields.geo.freeze(own, checkConsistency);
    this._refObject.geo = this._wrapperFields.geo.refObject;
    if (own) {
      this._wrapperFields.file_url.freeze(own, checkConsistency);
    }
    this._refObject.file_url = this._wrapperFields.file_url.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.orientation.copyRefObject(refObject.orientation);
    this._wrapperFields.geo.copyRefObject(refObject.geo);
    this._relative_altIntialized = true;
    this._refObject.relative_alt = refObject.relative_alt;
    this._image_indexIntialized = true;
    this._refObject.image_index = refObject.image_index;
    this._capture_resultIntialized = true;
    this._refObject.capture_result = refObject.capture_result;
    this._file_urlIntialized = true;
    this._wrapperFields.file_url.data = refObject.file_url.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    QuaternionWrapper.freeStruct(refObject.orientation);
    GeoPointWrapper.freeStruct(refObject.geo);
    StringWrapper.freeStruct(refObject.file_url);
  }
  static destoryRawROS(msg) {
    CameraImageCapturedWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'CameraImageCaptured'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get orientation() {
    return this._wrapperFields.orientation;
  }
  set orientation(value) {
    if (value instanceof QuaternionWrapper) {
      this._wrapperFields.orientation.copy(value);
    } else {
      this._wrapperFields.orientation.copy(new QuaternionWrapper(value));
    }
  }
  get geo() {
    return this._wrapperFields.geo;
  }
  set geo(value) {
    if (value instanceof GeoPointWrapper) {
      this._wrapperFields.geo.copy(value);
    } else {
      this._wrapperFields.geo.copy(new GeoPointWrapper(value));
    }
  }
  get relative_alt() {
    if (!this._relative_altIntialized) {
      return undefined;
    }
    return this._refObject.relative_alt;
  }
  set relative_alt(value) {
    this._relative_altIntialized = true;
    this._refObject.relative_alt = value;
  }
  get image_index() {
    if (!this._image_indexIntialized) {
      return undefined;
    }
    return this._refObject.image_index;
  }
  set image_index(value) {
    this._image_indexIntialized = true;
    this._refObject.image_index = value;
  }
  get capture_result() {
    if (!this._capture_resultIntialized) {
      return undefined;
    }
    return this._refObject.capture_result;
  }
  set capture_result(value) {
    this._capture_resultIntialized = true;
    this._refObject.capture_result = value;
  }
  get file_url() {
    if (!this._file_urlIntialized) {
      return undefined;
    }
    return this._wrapperFields.file_url.data;
  }
  set file_url(value) {
    this._file_urlIntialized = true;
    this._wrapperFields.file_url.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CameraImageCapturedRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.orientation.copyRefObject(this._refObject.orientation);
    this._wrapperFields.geo.copyRefObject(this._refObject.geo);
    this._relative_altIntialized = true;
    this._image_indexIntialized = true;
    this._capture_resultIntialized = true;
    this._file_urlIntialized = true;
    this._wrapperFields.file_url.copyRefObject(this._refObject.file_url);
  }
  copy(other) {
    this._refObject = new CameraImageCapturedRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.orientation.copy(other._wrapperFields.orientation);
    this._wrapperFields.geo.copy(other._wrapperFields.geo);
    this._relative_altIntialized = true;
    this._image_indexIntialized = true;
    this._capture_resultIntialized = true;
    this._file_urlIntialized = true;
    this._wrapperFields.file_url.copy(other._wrapperFields.file_url);
  }
  static get classType() {
    return CameraImageCapturedWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"orientation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Quaternion","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"geo","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"GeoPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"relative_alt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"image_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"capture_result","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"file_url","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"CameraImageCaptured","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CameraImageCaptured"};
  }
  hasMember(name) {
    let memberNames = ["header","orientation","geo","relative_alt","image_index","capture_result","file_url"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CameraImageCapturedArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CameraImageCapturedWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CameraImageCapturedWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CameraImageCapturedArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CameraImageCapturedArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CameraImageCapturedRefArray(size);
    this._refObject = new CameraImageCapturedRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CameraImageCapturedWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CameraImageCapturedArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CameraImageCapturedArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CameraImageCapturedWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CameraImageCapturedWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CameraImageCapturedArrayWrapper;
  }
}
module.exports = CameraImageCapturedWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "orientation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Quaternion",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "geo",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "GeoPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "relative_alt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "image_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "capture_result",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "file_url",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "CameraImageCaptured",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CameraImageCaptured"
}
*/
