// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__ParamSet_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ParamValueWrapper = require('../../generated/mavros_msgs/mavros_msgs__msg__ParamValue.js');
const ParamSet_RequestRefStruct = StructType({
  param_id: primitiveTypes.string,
  value: ParamValueWrapper.refObjectType,
});
const ParamSet_RequestRefArray = ArrayType(ParamSet_RequestRefStruct);
const ParamSet_RequestRefStructArray = StructType({
  data: ParamSet_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ParamSet_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ParamSet_RequestRefStruct(other._refObject.toObject());
      this._param_idIntialized = true;
      this._wrapperFields.param_id =  new StringWrapper(other._wrapperFields.param_id);
      this._wrapperFields.value =  new ParamValueWrapper(other._wrapperFields.value);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ParamSet_RequestRefStruct();
    this._param_idIntialized = false;
    this._wrapperFields.param_id = new StringWrapper();
    this._wrapperFields.value = new ParamValueWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ParamSet_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ParamSet_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ParamSet_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ParamSet_RequestRefStructArray
  }
  static get refObjectType() {
    return ParamSet_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._param_idIntialized) {
        throw new TypeError('Invalid argument: param_id in ParamSet_Request');
      }
    }
    if (own) {
      this._wrapperFields.param_id.freeze(own, checkConsistency);
    }
    this._refObject.param_id = this._wrapperFields.param_id.refObject;
    this._wrapperFields.value.freeze(own, checkConsistency);
    this._refObject.value = this._wrapperFields.value.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._param_idIntialized = true;
    this._wrapperFields.param_id.data = refObject.param_id.data;
    this._wrapperFields.value.copyRefObject(refObject.value);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.param_id);
    ParamValueWrapper.freeStruct(refObject.value);
  }
  static destoryRawROS(msg) {
    ParamSet_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'ParamSet_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get param_id() {
    if (!this._param_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.param_id.data;
  }
  set param_id(value) {
    this._param_idIntialized = true;
    this._wrapperFields.param_id.data = value;
  }
  get value() {
    return this._wrapperFields.value;
  }
  set value(value) {
    if (value instanceof ParamValueWrapper) {
      this._wrapperFields.value.copy(value);
    } else {
      this._wrapperFields.value.copy(new ParamValueWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ParamSet_RequestRefStruct(refObject.toObject());
    this._param_idIntialized = true;
    this._wrapperFields.param_id.copyRefObject(this._refObject.param_id);
    this._wrapperFields.value.copyRefObject(this._refObject.value);
  }
  copy(other) {
    this._refObject = new ParamSet_RequestRefStruct(other._refObject.toObject());
    this._param_idIntialized = true;
    this._wrapperFields.param_id.copy(other._wrapperFields.param_id);
    this._wrapperFields.value.copy(other._wrapperFields.value);
  }
  static get classType() {
    return ParamSet_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"param_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"mavros_msgs","type":"ParamValue","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"ParamSet_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ParamSet_Request"};
  }
  hasMember(name) {
    let memberNames = ["param_id","value"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ParamSet_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ParamSet_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ParamSet_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParamSet_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParamSet_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ParamSet_RequestRefArray(size);
    this._refObject = new ParamSet_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ParamSet_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ParamSet_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ParamSet_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ParamSet_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ParamSet_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ParamSet_RequestArrayWrapper;
  }
}
module.exports = ParamSet_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "param_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "mavros_msgs",
        "type": "ParamValue",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "ParamSet_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ParamSet_Request"
}
*/
