// rclnodejs/generated/mavros_msgs/mavros_msgs__srv__FileRename_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const FileRename_RequestRefStruct = StructType({
  old_path: primitiveTypes.string,
  new_path: primitiveTypes.string,
});
const FileRename_RequestRefArray = ArrayType(FileRename_RequestRefStruct);
const FileRename_RequestRefStructArray = StructType({
  data: FileRename_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FileRename_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FileRename_RequestRefStruct(other._refObject.toObject());
      this._old_pathIntialized = true;
      this._wrapperFields.old_path =  new StringWrapper(other._wrapperFields.old_path);
      this._new_pathIntialized = true;
      this._wrapperFields.new_path =  new StringWrapper(other._wrapperFields.new_path);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FileRename_RequestRefStruct();
    this._old_pathIntialized = false;
    this._wrapperFields.old_path = new StringWrapper();
    this._new_pathIntialized = false;
    this._wrapperFields.new_path = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FileRename_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FileRename_RequestArrayWrapper;
  }
  static get ArrayType() {
    return FileRename_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return FileRename_RequestRefStructArray
  }
  static get refObjectType() {
    return FileRename_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._old_pathIntialized) {
        throw new TypeError('Invalid argument: old_path in FileRename_Request');
      }
      if (!this._new_pathIntialized) {
        throw new TypeError('Invalid argument: new_path in FileRename_Request');
      }
    }
    if (own) {
      this._wrapperFields.old_path.freeze(own, checkConsistency);
    }
    this._refObject.old_path = this._wrapperFields.old_path.refObject;
    if (own) {
      this._wrapperFields.new_path.freeze(own, checkConsistency);
    }
    this._refObject.new_path = this._wrapperFields.new_path.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._old_pathIntialized = true;
    this._wrapperFields.old_path.data = refObject.old_path.data;
    this._new_pathIntialized = true;
    this._wrapperFields.new_path.data = refObject.new_path.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.old_path);
    StringWrapper.freeStruct(refObject.new_path);
  }
  static destoryRawROS(msg) {
    FileRename_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'srv', interfaceName: 'FileRename_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get old_path() {
    if (!this._old_pathIntialized) {
      return undefined;
    }
    return this._wrapperFields.old_path.data;
  }
  set old_path(value) {
    this._old_pathIntialized = true;
    this._wrapperFields.old_path.data = value;
  }
  get new_path() {
    if (!this._new_pathIntialized) {
      return undefined;
    }
    return this._wrapperFields.new_path.data;
  }
  set new_path(value) {
    this._new_pathIntialized = true;
    this._wrapperFields.new_path.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FileRename_RequestRefStruct(refObject.toObject());
    this._old_pathIntialized = true;
    this._wrapperFields.old_path.copyRefObject(this._refObject.old_path);
    this._new_pathIntialized = true;
    this._wrapperFields.new_path.copyRefObject(this._refObject.new_path);
  }
  copy(other) {
    this._refObject = new FileRename_RequestRefStruct(other._refObject.toObject());
    this._old_pathIntialized = true;
    this._wrapperFields.old_path.copy(other._wrapperFields.old_path);
    this._new_pathIntialized = true;
    this._wrapperFields.new_path.copy(other._wrapperFields.new_path);
  }
  static get classType() {
    return FileRename_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"old_path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"new_path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"FileRename_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FileRename_Request"};
  }
  hasMember(name) {
    let memberNames = ["old_path","new_path"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FileRename_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FileRename_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FileRename_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FileRename_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FileRename_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FileRename_RequestRefArray(size);
    this._refObject = new FileRename_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FileRename_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FileRename_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FileRename_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FileRename_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FileRename_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FileRename_RequestArrayWrapper;
  }
}
module.exports = FileRename_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "old_path",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "new_path",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "FileRename_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FileRename_Request"
}
*/
