// rclnodejs/generated/mavros_msgs/mavros_msgs__msg__GimbalDeviceAttitudeStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const QuaternionWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Quaternion.js');
const GimbalDeviceAttitudeStatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  target_system: primitiveTypes.uint8,
  target_component: primitiveTypes.uint8,
  flags: primitiveTypes.uint16,
  q: QuaternionWrapper.refObjectType,
  angular_velocity_x: primitiveTypes.float32,
  angular_velocity_y: primitiveTypes.float32,
  angular_velocity_z: primitiveTypes.float32,
  failure_flags: primitiveTypes.uint32,
});
const GimbalDeviceAttitudeStatusRefArray = ArrayType(GimbalDeviceAttitudeStatusRefStruct);
const GimbalDeviceAttitudeStatusRefStructArray = StructType({
  data: GimbalDeviceAttitudeStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GimbalDeviceAttitudeStatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GimbalDeviceAttitudeStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._target_systemIntialized = true;
      this._target_componentIntialized = true;
      this._flagsIntialized = true;
      this._wrapperFields.q =  new QuaternionWrapper(other._wrapperFields.q);
      this._angular_velocity_xIntialized = true;
      this._angular_velocity_yIntialized = true;
      this._angular_velocity_zIntialized = true;
      this._failure_flagsIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GimbalDeviceAttitudeStatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._target_systemIntialized = false;
    this._target_componentIntialized = false;
    this._flagsIntialized = false;
    this._wrapperFields.q = new QuaternionWrapper();
    this._angular_velocity_xIntialized = false;
    this._angular_velocity_yIntialized = false;
    this._angular_velocity_zIntialized = false;
    this._failure_flagsIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GimbalDeviceAttitudeStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GimbalDeviceAttitudeStatusArrayWrapper;
  }
  static get ArrayType() {
    return GimbalDeviceAttitudeStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return GimbalDeviceAttitudeStatusRefStructArray
  }
  static get refObjectType() {
    return GimbalDeviceAttitudeStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._target_systemIntialized) {
        throw new TypeError('Invalid argument: target_system in GimbalDeviceAttitudeStatus');
      }
      if (!this._target_componentIntialized) {
        throw new TypeError('Invalid argument: target_component in GimbalDeviceAttitudeStatus');
      }
      if (!this._flagsIntialized) {
        throw new TypeError('Invalid argument: flags in GimbalDeviceAttitudeStatus');
      }
      if (!this._angular_velocity_xIntialized) {
        throw new TypeError('Invalid argument: angular_velocity_x in GimbalDeviceAttitudeStatus');
      }
      if (!this._angular_velocity_yIntialized) {
        throw new TypeError('Invalid argument: angular_velocity_y in GimbalDeviceAttitudeStatus');
      }
      if (!this._angular_velocity_zIntialized) {
        throw new TypeError('Invalid argument: angular_velocity_z in GimbalDeviceAttitudeStatus');
      }
      if (!this._failure_flagsIntialized) {
        throw new TypeError('Invalid argument: failure_flags in GimbalDeviceAttitudeStatus');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.q.freeze(own, checkConsistency);
    this._refObject.q = this._wrapperFields.q.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._target_systemIntialized = true;
    this._refObject.target_system = refObject.target_system;
    this._target_componentIntialized = true;
    this._refObject.target_component = refObject.target_component;
    this._flagsIntialized = true;
    this._refObject.flags = refObject.flags;
    this._wrapperFields.q.copyRefObject(refObject.q);
    this._angular_velocity_xIntialized = true;
    this._refObject.angular_velocity_x = refObject.angular_velocity_x;
    this._angular_velocity_yIntialized = true;
    this._refObject.angular_velocity_y = refObject.angular_velocity_y;
    this._angular_velocity_zIntialized = true;
    this._refObject.angular_velocity_z = refObject.angular_velocity_z;
    this._failure_flagsIntialized = true;
    this._refObject.failure_flags = refObject.failure_flags;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    QuaternionWrapper.freeStruct(refObject.q);
  }
  static destoryRawROS(msg) {
    GimbalDeviceAttitudeStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mavros_msgs', subFolder: 'msg', interfaceName: 'GimbalDeviceAttitudeStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get target_system() {
    if (!this._target_systemIntialized) {
      return undefined;
    }
    return this._refObject.target_system;
  }
  set target_system(value) {
    this._target_systemIntialized = true;
    this._refObject.target_system = value;
  }
  get target_component() {
    if (!this._target_componentIntialized) {
      return undefined;
    }
    return this._refObject.target_component;
  }
  set target_component(value) {
    this._target_componentIntialized = true;
    this._refObject.target_component = value;
  }
  get flags() {
    if (!this._flagsIntialized) {
      return undefined;
    }
    return this._refObject.flags;
  }
  set flags(value) {
    this._flagsIntialized = true;
    this._refObject.flags = value;
  }
  get q() {
    return this._wrapperFields.q;
  }
  set q(value) {
    if (value instanceof QuaternionWrapper) {
      this._wrapperFields.q.copy(value);
    } else {
      this._wrapperFields.q.copy(new QuaternionWrapper(value));
    }
  }
  get angular_velocity_x() {
    if (!this._angular_velocity_xIntialized) {
      return undefined;
    }
    return this._refObject.angular_velocity_x;
  }
  set angular_velocity_x(value) {
    this._angular_velocity_xIntialized = true;
    this._refObject.angular_velocity_x = value;
  }
  get angular_velocity_y() {
    if (!this._angular_velocity_yIntialized) {
      return undefined;
    }
    return this._refObject.angular_velocity_y;
  }
  set angular_velocity_y(value) {
    this._angular_velocity_yIntialized = true;
    this._refObject.angular_velocity_y = value;
  }
  get angular_velocity_z() {
    if (!this._angular_velocity_zIntialized) {
      return undefined;
    }
    return this._refObject.angular_velocity_z;
  }
  set angular_velocity_z(value) {
    this._angular_velocity_zIntialized = true;
    this._refObject.angular_velocity_z = value;
  }
  get failure_flags() {
    if (!this._failure_flagsIntialized) {
      return undefined;
    }
    return this._refObject.failure_flags;
  }
  set failure_flags(value) {
    this._failure_flagsIntialized = true;
    this._refObject.failure_flags = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GimbalDeviceAttitudeStatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._target_systemIntialized = true;
    this._target_componentIntialized = true;
    this._flagsIntialized = true;
    this._wrapperFields.q.copyRefObject(this._refObject.q);
    this._angular_velocity_xIntialized = true;
    this._angular_velocity_yIntialized = true;
    this._angular_velocity_zIntialized = true;
    this._failure_flagsIntialized = true;
  }
  copy(other) {
    this._refObject = new GimbalDeviceAttitudeStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._target_systemIntialized = true;
    this._target_componentIntialized = true;
    this._flagsIntialized = true;
    this._wrapperFields.q.copy(other._wrapperFields.q);
    this._angular_velocity_xIntialized = true;
    this._angular_velocity_yIntialized = true;
    this._angular_velocity_zIntialized = true;
    this._failure_flagsIntialized = true;
  }
  static get classType() {
    return GimbalDeviceAttitudeStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint16","name":"FLAGS_RETRACT","value":1},{"type":"uint16","name":"FLAGS_NEUTRAL","value":2},{"type":"uint16","name":"FLAGS_ROLL_LOCK","value":4},{"type":"uint16","name":"FLAGS_PITCH_LOCK","value":8},{"type":"uint16","name":"FLAGS_YAW_LOCK","value":16},{"type":"uint32","name":"ERROR_FLAGS_AT_ROLL_LIMIT","value":1},{"type":"uint32","name":"ERROR_FLAGS_AT_PITCH_LIMIT","value":2},{"type":"uint32","name":"ERROR_FLAGS_AT_YAW_LIMIT","value":4},{"type":"uint32","name":"ERROR_FLAGS_ENCODER_ERROR","value":8},{"type":"uint32","name":"ERROR_FLAGS_POWER_ERROR","value":16},{"type":"uint32","name":"ERROR_FLAGS_MOTOR_ERROR","value":32},{"type":"uint32","name":"ERROR_FLAGS_SOFTWARE_ERROR","value":64},{"type":"uint32","name":"ERROR_FLAGS_COMMS_ERROR","value":128},{"type":"uint32","name":"ERROR_FLAGS_CALIBRATION_RUNNING","value":256}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"target_system","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"target_component","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"q","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Quaternion","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"angular_velocity_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angular_velocity_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angular_velocity_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"failure_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mavros_msgs","type":"GimbalDeviceAttitudeStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GimbalDeviceAttitudeStatus"};
  }
  hasMember(name) {
    let memberNames = ["header","target_system","target_component","flags","q","angular_velocity_x","angular_velocity_y","angular_velocity_z","failure_flags"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GimbalDeviceAttitudeStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GimbalDeviceAttitudeStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GimbalDeviceAttitudeStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalDeviceAttitudeStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalDeviceAttitudeStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GimbalDeviceAttitudeStatusRefArray(size);
    this._refObject = new GimbalDeviceAttitudeStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GimbalDeviceAttitudeStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GimbalDeviceAttitudeStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GimbalDeviceAttitudeStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GimbalDeviceAttitudeStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GimbalDeviceAttitudeStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GimbalDeviceAttitudeStatusArrayWrapper;
  }
}
// Define constants (14 in total)
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "FLAGS_RETRACT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "FLAGS_NEUTRAL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "FLAGS_ROLL_LOCK", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "FLAGS_PITCH_LOCK", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "FLAGS_YAW_LOCK", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "ERROR_FLAGS_AT_ROLL_LIMIT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "ERROR_FLAGS_AT_PITCH_LIMIT", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "ERROR_FLAGS_AT_YAW_LIMIT", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "ERROR_FLAGS_ENCODER_ERROR", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "ERROR_FLAGS_POWER_ERROR", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "ERROR_FLAGS_MOTOR_ERROR", {value: 32, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "ERROR_FLAGS_SOFTWARE_ERROR", {value: 64, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "ERROR_FLAGS_COMMS_ERROR", {value: 128, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceAttitudeStatusWrapper, "ERROR_FLAGS_CALIBRATION_RUNNING", {value: 256, writable: false, enumerable: true, configurable: true});
module.exports = GimbalDeviceAttitudeStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint16",
      "name": "FLAGS_RETRACT",
      "value": 1
    },
    {
      "type": "uint16",
      "name": "FLAGS_NEUTRAL",
      "value": 2
    },
    {
      "type": "uint16",
      "name": "FLAGS_ROLL_LOCK",
      "value": 4
    },
    {
      "type": "uint16",
      "name": "FLAGS_PITCH_LOCK",
      "value": 8
    },
    {
      "type": "uint16",
      "name": "FLAGS_YAW_LOCK",
      "value": 16
    },
    {
      "type": "uint32",
      "name": "ERROR_FLAGS_AT_ROLL_LIMIT",
      "value": 1
    },
    {
      "type": "uint32",
      "name": "ERROR_FLAGS_AT_PITCH_LIMIT",
      "value": 2
    },
    {
      "type": "uint32",
      "name": "ERROR_FLAGS_AT_YAW_LIMIT",
      "value": 4
    },
    {
      "type": "uint32",
      "name": "ERROR_FLAGS_ENCODER_ERROR",
      "value": 8
    },
    {
      "type": "uint32",
      "name": "ERROR_FLAGS_POWER_ERROR",
      "value": 16
    },
    {
      "type": "uint32",
      "name": "ERROR_FLAGS_MOTOR_ERROR",
      "value": 32
    },
    {
      "type": "uint32",
      "name": "ERROR_FLAGS_SOFTWARE_ERROR",
      "value": 64
    },
    {
      "type": "uint32",
      "name": "ERROR_FLAGS_COMMS_ERROR",
      "value": 128
    },
    {
      "type": "uint32",
      "name": "ERROR_FLAGS_CALIBRATION_RUNNING",
      "value": 256
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "target_system",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "target_component",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "q",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Quaternion",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "angular_velocity_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angular_velocity_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angular_velocity_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "failure_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mavros_msgs",
    "type": "GimbalDeviceAttitudeStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GimbalDeviceAttitudeStatus"
}
*/
