// rclnodejs/generated/nmea_msgs/nmea_msgs__msg__Gpgsv.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GpgsvSatelliteWrapper = require('../../generated/nmea_msgs/nmea_msgs__msg__GpgsvSatellite.js');
const GpgsvRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  message_id: primitiveTypes.string,
  n_msgs: primitiveTypes.uint8,
  msg_number: primitiveTypes.uint8,
  n_satellites: primitiveTypes.uint8,
  satellites: GpgsvSatelliteWrapper.refObjectArrayType,
});
const GpgsvRefArray = ArrayType(GpgsvRefStruct);
const GpgsvRefStructArray = StructType({
  data: GpgsvRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GpgsvWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GpgsvRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._message_idIntialized = true;
      this._wrapperFields.message_id =  new StringWrapper(other._wrapperFields.message_id);
      this._n_msgsIntialized = true;
      this._msg_numberIntialized = true;
      this._n_satellitesIntialized = true;
      this._wrapperFields.satellites = GpgsvSatelliteWrapper.createArray();
      this._wrapperFields.satellites.copy(other._wrapperFields.satellites);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GpgsvRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._message_idIntialized = false;
    this._wrapperFields.message_id = new StringWrapper();
    this._n_msgsIntialized = false;
    this._msg_numberIntialized = false;
    this._n_satellitesIntialized = false;
    this._wrapperFields.satellites = GpgsvSatelliteWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GpgsvWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GpgsvArrayWrapper;
  }
  static get ArrayType() {
    return GpgsvArrayWrapper;
  }
  static get refObjectArrayType() {
    return GpgsvRefStructArray
  }
  static get refObjectType() {
    return GpgsvRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._message_idIntialized) {
        throw new TypeError('Invalid argument: message_id in Gpgsv');
      }
      if (!this._n_msgsIntialized) {
        throw new TypeError('Invalid argument: n_msgs in Gpgsv');
      }
      if (!this._msg_numberIntialized) {
        throw new TypeError('Invalid argument: msg_number in Gpgsv');
      }
      if (!this._n_satellitesIntialized) {
        throw new TypeError('Invalid argument: n_satellites in Gpgsv');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.message_id.freeze(own, checkConsistency);
    }
    this._refObject.message_id = this._wrapperFields.message_id.refObject;
    this._wrapperFields.satellites.freeze(own, checkConsistency);
    this._refObject.satellites = this._wrapperFields.satellites.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._message_idIntialized = true;
    this._wrapperFields.message_id.data = refObject.message_id.data;
    this._n_msgsIntialized = true;
    this._refObject.n_msgs = refObject.n_msgs;
    this._msg_numberIntialized = true;
    this._refObject.msg_number = refObject.msg_number;
    this._n_satellitesIntialized = true;
    this._refObject.n_satellites = refObject.n_satellites;
    this._wrapperFields.satellites.copyRefObject(refObject.satellites);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.message_id);
    if (refObject.satellites.size != 0) {
      GpgsvSatelliteWrapper.ArrayType.freeArray(refObject.satellites);
      if (GpgsvSatelliteWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.satellites, GpgsvSatelliteWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GpgsvWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nmea_msgs', subFolder: 'msg', interfaceName: 'Gpgsv'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get message_id() {
    if (!this._message_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.message_id.data;
  }
  set message_id(value) {
    this._message_idIntialized = true;
    this._wrapperFields.message_id.data = value;
  }
  get n_msgs() {
    if (!this._n_msgsIntialized) {
      return undefined;
    }
    return this._refObject.n_msgs;
  }
  set n_msgs(value) {
    this._n_msgsIntialized = true;
    this._refObject.n_msgs = value;
  }
  get msg_number() {
    if (!this._msg_numberIntialized) {
      return undefined;
    }
    return this._refObject.msg_number;
  }
  set msg_number(value) {
    this._msg_numberIntialized = true;
    this._refObject.msg_number = value;
  }
  get n_satellites() {
    if (!this._n_satellitesIntialized) {
      return undefined;
    }
    return this._refObject.n_satellites;
  }
  set n_satellites(value) {
    this._n_satellitesIntialized = true;
    this._refObject.n_satellites = value;
  }
  get satellites() {
    return this._wrapperFields.satellites;
  }
  set satellites(value) {
    this._wrapperFields.satellites.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GpgsvRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._message_idIntialized = true;
    this._wrapperFields.message_id.copyRefObject(this._refObject.message_id);
    this._n_msgsIntialized = true;
    this._msg_numberIntialized = true;
    this._n_satellitesIntialized = true;
    this._wrapperFields.satellites.copyRefObject(this._refObject.satellites);
  }
  copy(other) {
    this._refObject = new GpgsvRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._message_idIntialized = true;
    this._wrapperFields.message_id.copy(other._wrapperFields.message_id);
    this._n_msgsIntialized = true;
    this._msg_numberIntialized = true;
    this._n_satellitesIntialized = true;
    this._wrapperFields.satellites.copy(other._wrapperFields.satellites);
  }
  static get classType() {
    return GpgsvWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"message_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"n_msgs","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"msg_number","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"n_satellites","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"satellites","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"nmea_msgs","type":"GpgsvSatellite","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nmea_msgs","type":"Gpgsv","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Gpgsv"};
  }
  hasMember(name) {
    let memberNames = ["header","message_id","n_msgs","msg_number","n_satellites","satellites"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GpgsvArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GpgsvWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GpgsvWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GpgsvArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GpgsvArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GpgsvRefArray(size);
    this._refObject = new GpgsvRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GpgsvWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GpgsvArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GpgsvArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GpgsvWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GpgsvWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GpgsvArrayWrapper;
  }
}
module.exports = GpgsvWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "message_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "n_msgs",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "msg_number",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "n_satellites",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "satellites",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "nmea_msgs",
        "type": "GpgsvSatellite",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nmea_msgs",
    "type": "Gpgsv",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Gpgsv"
}
*/
