// rclnodejs/generated/nmea_msgs/nmea_msgs__msg__GpgsvSatellite.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GpgsvSatelliteRefStruct = StructType({
  prn: primitiveTypes.uint8,
  elevation: primitiveTypes.uint8,
  azimuth: primitiveTypes.uint16,
  snr: primitiveTypes.int8,
});
const GpgsvSatelliteRefArray = ArrayType(GpgsvSatelliteRefStruct);
const GpgsvSatelliteRefStructArray = StructType({
  data: GpgsvSatelliteRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GpgsvSatelliteWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GpgsvSatelliteRefStruct(other._refObject.toObject());
      this._prnIntialized = true;
      this._elevationIntialized = true;
      this._azimuthIntialized = true;
      this._snrIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GpgsvSatelliteRefStruct();
    this._prnIntialized = false;
    this._elevationIntialized = false;
    this._azimuthIntialized = false;
    this._snrIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GpgsvSatelliteWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GpgsvSatelliteArrayWrapper;
  }
  static get ArrayType() {
    return GpgsvSatelliteArrayWrapper;
  }
  static get refObjectArrayType() {
    return GpgsvSatelliteRefStructArray
  }
  static get refObjectType() {
    return GpgsvSatelliteRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._prnIntialized) {
        throw new TypeError('Invalid argument: prn in GpgsvSatellite');
      }
      if (!this._elevationIntialized) {
        throw new TypeError('Invalid argument: elevation in GpgsvSatellite');
      }
      if (!this._azimuthIntialized) {
        throw new TypeError('Invalid argument: azimuth in GpgsvSatellite');
      }
      if (!this._snrIntialized) {
        throw new TypeError('Invalid argument: snr in GpgsvSatellite');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._prnIntialized = true;
    this._refObject.prn = refObject.prn;
    this._elevationIntialized = true;
    this._refObject.elevation = refObject.elevation;
    this._azimuthIntialized = true;
    this._refObject.azimuth = refObject.azimuth;
    this._snrIntialized = true;
    this._refObject.snr = refObject.snr;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GpgsvSatelliteWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nmea_msgs', subFolder: 'msg', interfaceName: 'GpgsvSatellite'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get prn() {
    if (!this._prnIntialized) {
      return undefined;
    }
    return this._refObject.prn;
  }
  set prn(value) {
    this._prnIntialized = true;
    this._refObject.prn = value;
  }
  get elevation() {
    if (!this._elevationIntialized) {
      return undefined;
    }
    return this._refObject.elevation;
  }
  set elevation(value) {
    this._elevationIntialized = true;
    this._refObject.elevation = value;
  }
  get azimuth() {
    if (!this._azimuthIntialized) {
      return undefined;
    }
    return this._refObject.azimuth;
  }
  set azimuth(value) {
    this._azimuthIntialized = true;
    this._refObject.azimuth = value;
  }
  get snr() {
    if (!this._snrIntialized) {
      return undefined;
    }
    return this._refObject.snr;
  }
  set snr(value) {
    this._snrIntialized = true;
    this._refObject.snr = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GpgsvSatelliteRefStruct(refObject.toObject());
    this._prnIntialized = true;
    this._elevationIntialized = true;
    this._azimuthIntialized = true;
    this._snrIntialized = true;
  }
  copy(other) {
    this._refObject = new GpgsvSatelliteRefStruct(other._refObject.toObject());
    this._prnIntialized = true;
    this._elevationIntialized = true;
    this._azimuthIntialized = true;
    this._snrIntialized = true;
  }
  static get classType() {
    return GpgsvSatelliteWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"prn","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"elevation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"azimuth","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"snr","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nmea_msgs","type":"GpgsvSatellite","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GpgsvSatellite"};
  }
  hasMember(name) {
    let memberNames = ["prn","elevation","azimuth","snr"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GpgsvSatelliteArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GpgsvSatelliteWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GpgsvSatelliteWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GpgsvSatelliteArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GpgsvSatelliteArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GpgsvSatelliteRefArray(size);
    this._refObject = new GpgsvSatelliteRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GpgsvSatelliteWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GpgsvSatelliteArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GpgsvSatelliteArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GpgsvSatelliteWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GpgsvSatelliteWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GpgsvSatelliteArrayWrapper;
  }
}
module.exports = GpgsvSatelliteWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "prn",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "elevation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "azimuth",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "snr",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nmea_msgs",
    "type": "GpgsvSatellite",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GpgsvSatellite"
}
*/
