// rclnodejs/generated/can_msgs/can_msgs__msg__Frame.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const FrameRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  id: primitiveTypes.uint32,
  is_rtr: primitiveTypes.bool,
  is_extended: primitiveTypes.bool,
  is_error: primitiveTypes.bool,
  dlc: primitiveTypes.uint8,
  data: ArrayType(primitiveTypes.uint8, 8),
});
const FrameRefArray = ArrayType(FrameRefStruct);
const FrameRefStructArray = StructType({
  data: FrameRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FrameWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FrameRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._idIntialized = true;
      this._is_rtrIntialized = true;
      this._is_extendedIntialized = true;
      this._is_errorIntialized = true;
      this._dlcIntialized = true;
      this._wrapperFields.data = UInt8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FrameRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._idIntialized = false;
    this._is_rtrIntialized = false;
    this._is_extendedIntialized = false;
    this._is_errorIntialized = false;
    this._dlcIntialized = false;
    this._wrapperFields.data = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new FrameWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FrameArrayWrapper;
  }
  static get ArrayType() {
    return FrameArrayWrapper;
  }
  static get refObjectArrayType() {
    return FrameRefStructArray
  }
  static get refObjectType() {
    return FrameRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in Frame');
      }
      if (!this._is_rtrIntialized) {
        throw new TypeError('Invalid argument: is_rtr in Frame');
      }
      if (!this._is_extendedIntialized) {
        throw new TypeError('Invalid argument: is_extended in Frame');
      }
      if (!this._is_errorIntialized) {
        throw new TypeError('Invalid argument: is_error in Frame');
      }
      if (!this._dlcIntialized) {
        throw new TypeError('Invalid argument: dlc in Frame');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    for (let i = 0; i < this._wrapperFields.data.data.length; i++) {
      this._refObject.data[i] = this._wrapperFields.data.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._is_rtrIntialized = true;
    this._refObject.is_rtr = refObject.is_rtr;
    this._is_extendedIntialized = true;
    this._refObject.is_extended = refObject.is_extended;
    this._is_errorIntialized = true;
    this._refObject.is_error = refObject.is_error;
    this._dlcIntialized = true;
    this._refObject.dlc = refObject.dlc;
    this._refObject.data = refObject.data;
    this._wrapperFields.data.fill(refObject.data.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    FrameWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'can_msgs', subFolder: 'msg', interfaceName: 'Frame'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get is_rtr() {
    if (!this._is_rtrIntialized) {
      return undefined;
    }
    return this._refObject.is_rtr;
  }
  set is_rtr(value) {
    this._is_rtrIntialized = true;
    this._refObject.is_rtr = value;
  }
  get is_extended() {
    if (!this._is_extendedIntialized) {
      return undefined;
    }
    return this._refObject.is_extended;
  }
  set is_extended(value) {
    this._is_extendedIntialized = true;
    this._refObject.is_extended = value;
  }
  get is_error() {
    if (!this._is_errorIntialized) {
      return undefined;
    }
    return this._refObject.is_error;
  }
  set is_error(value) {
    this._is_errorIntialized = true;
    this._refObject.is_error = value;
  }
  get dlc() {
    if (!this._dlcIntialized) {
      return undefined;
    }
    return this._refObject.dlc;
  }
  set dlc(value) {
    this._dlcIntialized = true;
    this._refObject.dlc = value;
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    if (value.length !== 8) {
      throw new RangeError('The length of the array must be 8.');
    }
    this._wrapperFields['data'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new FrameRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._idIntialized = true;
    this._is_rtrIntialized = true;
    this._is_extendedIntialized = true;
    this._is_errorIntialized = true;
    this._dlcIntialized = true;
    this._wrapperFields.data.fill(refObject.data.toArray());
  }
  copy(other) {
    this._refObject = new FrameRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._idIntialized = true;
    this._is_rtrIntialized = true;
    this._is_extendedIntialized = true;
    this._is_errorIntialized = true;
    this._dlcIntialized = true;
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return FrameWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_rtr","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_extended","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dlc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":8,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"can_msgs","type":"Frame","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Frame"};
  }
  hasMember(name) {
    let memberNames = ["header","id","is_rtr","is_extended","is_error","dlc","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FrameArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FrameWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FrameWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FrameArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FrameArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FrameRefArray(size);
    this._refObject = new FrameRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FrameWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FrameArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FrameArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FrameWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FrameWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FrameArrayWrapper;
  }
}
module.exports = FrameWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "is_rtr",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "is_extended",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "is_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dlc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": 8,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "can_msgs",
    "type": "Frame",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Frame"
}
*/
